"use strict";(self.webpackChunkwebsite_2=self.webpackChunkwebsite_2||[]).push([[6484],{5680:(e,t,a)=>{a.d(t,{xA:()=>m,yg:()=>y});var n=a(6540);function i(e,t,a){return t in e?Object.defineProperty(e,t,{value:a,enumerable:!0,configurable:!0,writable:!0}):e[t]=a,e}function r(e,t){var a=Object.keys(e);if(Object.getOwnPropertySymbols){var n=Object.getOwnPropertySymbols(e);t&&(n=n.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),a.push.apply(a,n)}return a}function l(e){for(var t=1;t<arguments.length;t++){var a=null!=arguments[t]?arguments[t]:{};t%2?r(Object(a),!0).forEach((function(t){i(e,t,a[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(a)):r(Object(a)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(a,t))}))}return e}function p(e,t){if(null==e)return{};var a,n,i=function(e,t){if(null==e)return{};var a,n,i={},r=Object.keys(e);for(n=0;n<r.length;n++)a=r[n],t.indexOf(a)>=0||(i[a]=e[a]);return i}(e,t);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);for(n=0;n<r.length;n++)a=r[n],t.indexOf(a)>=0||Object.prototype.propertyIsEnumerable.call(e,a)&&(i[a]=e[a])}return i}var o=n.createContext({}),s=function(e){var t=n.useContext(o),a=t;return e&&(a="function"==typeof e?e(t):l(l({},t),e)),a},m=function(e){var t=s(e.components);return n.createElement(o.Provider,{value:t},e.children)},g="mdxType",d={inlineCode:"code",wrapper:function(e){var t=e.children;return n.createElement(n.Fragment,{},t)}},u=n.forwardRef((function(e,t){var a=e.components,i=e.mdxType,r=e.originalType,o=e.parentName,m=p(e,["components","mdxType","originalType","parentName"]),g=s(a),u=i,y=g["".concat(o,".").concat(u)]||g[u]||d[u]||r;return a?n.createElement(y,l(l({ref:t},m),{},{components:a})):n.createElement(y,l({ref:t},m))}));function y(e,t){var a=arguments,i=t&&t.mdxType;if("string"==typeof e||i){var r=a.length,l=new Array(r);l[0]=u;var p={};for(var o in t)hasOwnProperty.call(t,o)&&(p[o]=t[o]);p.originalType=e,p[g]="string"==typeof e?e:i,l[1]=p;for(var s=2;s<r;s++)l[s]=a[s];return n.createElement.apply(null,l)}return n.createElement.apply(null,a)}u.displayName="MDXCreateElement"},8384:(e,t,a)=>{a.r(t),a.d(t,{assets:()=>o,contentTitle:()=>l,default:()=>d,frontMatter:()=>r,metadata:()=>p,toc:()=>s});var n=a(8168),i=(a(6540),a(5680));const r={sidebar_position:1,slug:"/Pipelines/Smart-seq2_Multi_Sample_Pipeline/README"},l="Smart-seq2 Multi-Sample Overview",p={unversionedId:"Deprecated_Pipelines/Smart-seq2_Multi_Sample_Pipeline/README",id:"Deprecated_Pipelines/Smart-seq2_Multi_Sample_Pipeline/README",title:"Smart-seq2 Multi-Sample Overview",description:"9/12/2014",source:"@site/docs/Deprecated_Pipelines/Smart-seq2_Multi_Sample_Pipeline/README.md",sourceDirName:"Deprecated_Pipelines/Smart-seq2_Multi_Sample_Pipeline",slug:"/Pipelines/Smart-seq2_Multi_Sample_Pipeline/README",permalink:"/warp/docs/Pipelines/Smart-seq2_Multi_Sample_Pipeline/README",draft:!1,editUrl:"https://github.com/broadinstitute/warp/edit/develop/website/docs/Deprecated_Pipelines/Smart-seq2_Multi_Sample_Pipeline/README.md",tags:[],version:"current",lastUpdatedBy:"Nikelle Petrillo",lastUpdatedAt:1761573741,formattedLastUpdatedAt:"Oct 27, 2025",sidebarPosition:1,frontMatter:{sidebar_position:1,slug:"/Pipelines/Smart-seq2_Multi_Sample_Pipeline/README"},sidebar:"docsSidebar",previous:{title:"scATAC v1.3.2 Methods",permalink:"/warp/docs/Deprecated_Pipelines/Single_Cell_ATAC_Seq_Pipeline/scatac.methods"},next:{title:"Smart-seq2 Multi-Sample v2.2.21 Publication Methods",permalink:"/warp/docs/Deprecated_Pipelines/Smart-seq2_Multi_Sample_Pipeline/smart-seq2.methods"}},o={},s=[{value:"Introduction",id:"introduction",level:2},{value:"Inputs",id:"inputs",level:2},{value:"Sample and Reference Inputs",id:"sample-and-reference-inputs",level:3},{value:"Additional Input",id:"additional-input",level:3},{value:"Smart-seq2 Multi-Sample Task Summary",id:"smart-seq2-multi-sample-task-summary",level:3},{value:"Outputs",id:"outputs",level:3},{value:"Validation",id:"validation",level:2},{value:"Versioning",id:"versioning",level:2},{value:"Citing the Smart-seq2 Multi-Sample Pipeline",id:"citing-the-smart-seq2-multi-sample-pipeline",level:2},{value:"Consortia Support",id:"consortia-support",level:2},{value:"Have Suggestions?",id:"have-suggestions",level:2}],m={toc:s},g="wrapper";function d(e){let{components:t,...a}=e;return(0,i.yg)(g,(0,n.A)({},m,a,{components:t,mdxType:"MDXLayout"}),(0,i.yg)("h1",{id:"smart-seq2-multi-sample-overview"},"Smart-seq2 Multi-Sample Overview"),(0,i.yg)("admonition",{type:"warning"},(0,i.yg)("p",{parentName:"admonition"},"9/12/2014"),(0,i.yg)("p",{parentName:"admonition"},"We are deprecating the Smart-seq2 Multi-Sample Pipeline. Although the code will continue to be available, we are no longer supporting it. For an alternative, see the ",(0,i.yg)("a",{parentName:"p",href:"/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README"},"Smart-seq2 Single Nucleus Multi Sample workflow"),".")),(0,i.yg)("table",null,(0,i.yg)("thead",{parentName:"table"},(0,i.yg)("tr",{parentName:"thead"},(0,i.yg)("th",{parentName:"tr",align:"center"},"Pipeline Version"),(0,i.yg)("th",{parentName:"tr",align:"center"},"Date Updated"),(0,i.yg)("th",{parentName:"tr",align:"center"},"Documentation Author"),(0,i.yg)("th",{parentName:"tr",align:"center"},"Questions or Feedback"))),(0,i.yg)("tbody",{parentName:"table"},(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:"center"},(0,i.yg)("a",{parentName:"td",href:"https://github.com/broadinstitute/warp/releases"},"MultiSampleSmartSeq2_v2.2.21")),(0,i.yg)("td",{parentName:"tr",align:"center"},"December, 2023"),(0,i.yg)("td",{parentName:"tr",align:"center"},"Elizabeth Kiernan"),(0,i.yg)("td",{parentName:"tr",align:"center"},"Please ",(0,i.yg)("a",{parentName:"td",href:"https://github.com/broadinstitute/warp/issues"},"file an issue in WARP"),".")))),(0,i.yg)("h2",{id:"introduction"},"Introduction"),(0,i.yg)("p",null,"The Smart-seq2 Multi-Sample (Multi-SS2) Pipeline is a wrapper around the ",(0,i.yg)("a",{parentName:"p",href:"../Smart-seq2_Single_Sample_Pipeline/README"},"Smart-seq2 Single Sample")," pipeline. It is developed by the Data Coordination Platform of the Human Cell Atlas to process single-cell RNAseq (scRNAseq) data generated by Smart-seq2 assays. The workflow processes multiple cells by importing and running the ",(0,i.yg)("a",{parentName:"p",href:"https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_sample/SmartSeq2SingleSample.wdl"},"Smart-seq2 Single Sample workflow")," for each cell (sample) and then merging the resulting Loom matrix output into a single Loom matrix containing raw counts and TPMs."),(0,i.yg)("p",null,"Full details about the Smart-seq2 Pipeline can be read in the ",(0,i.yg)("a",{parentName:"p",href:"https://broadinstitute.github.io/warp/docs/Pipelines/Smart-seq2_Single_Sample_Pipeline/README"},"Smart-seq2 Single Sample Overview")," in GitHub."),(0,i.yg)("p",null,"The Multi-SS2 workflow can also be run in ",(0,i.yg)("a",{parentName:"p",href:"https://app.terra.bio"},"Terra"),", a cloud-based analysis platform. The Terra ",(0,i.yg)("a",{parentName:"p",href:"https://app.terra.bio/#workspaces/featured-workspaces-hca/HCA%20Smart-seq2%20Multi%20Sample%20Pipeline"},"Smart-seq2 public workspace")," contains the Smart-seq2 workflow, workflow configurations, required reference data and other inputs, and example testing data."),(0,i.yg)("admonition",{title:"Want to use the Multi-SS2 Pipeline for your publication?",type:"tip"},(0,i.yg)("p",{parentName:"admonition"},"Check out the ",(0,i.yg)("a",{parentName:"p",href:"/warp/docs/Deprecated_Pipelines/Smart-seq2_Multi_Sample_Pipeline/smart-seq2.methods"},"Smart-seq2 Publication Methods")," to get started!")),(0,i.yg)("h2",{id:"inputs"},"Inputs"),(0,i.yg)("p",null,"There are two example configuration (JSON) files available for testing the Multi-SS2 workflow. Both examples are also preloaded in the Terra ",(0,i.yg)("a",{parentName:"p",href:"https://app.terra.bio/#workspaces/featured-workspaces-hca/HCA%20Smart-seq2%20Multi%20Sample%20Pipeline"},"Smart-seq2 public workspace"),"."),(0,i.yg)("ul",null,(0,i.yg)("li",{parentName:"ul"},(0,i.yg)("a",{parentName:"li",href:"https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_multisample/human_single_example.json"},"human_single_example.json"),": Configurations for an example single-end human dataset consisting of two samples (cells)"),(0,i.yg)("li",{parentName:"ul"},(0,i.yg)("a",{parentName:"li",href:"https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_multisample/mouse_paired_example.json"},"mouse_paired_example.json"),": Configurations for an example paired-end mouse dataset consisting of two samples (cells)")),(0,i.yg)("h3",{id:"sample-and-reference-inputs"},"Sample and Reference Inputs"),(0,i.yg)("p",null,"The workflow\u2019s sample inputs are listed in the table below. Reference inputs are identical to those\nspecified in the ",(0,i.yg)("a",{parentName:"p",href:"../Smart-seq2_Single_Sample_Pipeline/README"},"Smart-seq2 Single Sample Overview"),"."),(0,i.yg)("p",null,"The workflow processes both single- and paired-end samples; however, these samples can not be mixed in the same run."),(0,i.yg)("table",null,(0,i.yg)("thead",{parentName:"table"},(0,i.yg)("tr",{parentName:"thead"},(0,i.yg)("th",{parentName:"tr",align:null},"Input name"),(0,i.yg)("th",{parentName:"tr",align:null},"Input Description"),(0,i.yg)("th",{parentName:"tr",align:null},"Input Type"))),(0,i.yg)("tbody",{parentName:"table"},(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},"fastq1_input_files"),(0,i.yg)("td",{parentName:"tr",align:null},"Cloud locations for each read1 file"),(0,i.yg)("td",{parentName:"tr",align:null},"Array of strings")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},"fastq2_input_files"),(0,i.yg)("td",{parentName:"tr",align:null},"Optional cloud locations for each read2 file if running paired-end samples"),(0,i.yg)("td",{parentName:"tr",align:null},"Array of strings")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},"input_ids"),(0,i.yg)("td",{parentName:"tr",align:null},"Unique identifiers or names for each cell; can be a UUID or human-readable name"),(0,i.yg)("td",{parentName:"tr",align:null},"Array of strings")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},"input_names"),(0,i.yg)("td",{parentName:"tr",align:null},"Optional unique identifiers/names to further describe each cell. If ",(0,i.yg)("inlineCode",{parentName:"td"},"input_id")," is a UUID, the ",(0,i.yg)("inlineCode",{parentName:"td"},"input_name")," could be used as a human-readable identifier"),(0,i.yg)("td",{parentName:"tr",align:null},"String")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},"batch_id"),(0,i.yg)("td",{parentName:"tr",align:null},"Identifier for the batch of multiple samples"),(0,i.yg)("td",{parentName:"tr",align:null},"String")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},"batch_name"),(0,i.yg)("td",{parentName:"tr",align:null},"Optional string to describe the batch or biological sample"),(0,i.yg)("td",{parentName:"tr",align:null},"String")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},"input_name_metadata_field"),(0,i.yg)("td",{parentName:"tr",align:null},"Optional input describing, when applicable, the metadata field containing the ",(0,i.yg)("inlineCode",{parentName:"td"},"input_names")),(0,i.yg)("td",{parentName:"tr",align:null},"String")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},"input_id_metadata_field"),(0,i.yg)("td",{parentName:"tr",align:null},"Optional string describing, when applicable, the metadata field containing the ",(0,i.yg)("inlineCode",{parentName:"td"},"input_ids")),(0,i.yg)("td",{parentName:"tr",align:null},"String")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},(0,i.yg)("inlineCode",{parentName:"td"},"project_id")),(0,i.yg)("td",{parentName:"tr",align:null},"Optional project identifier; usually a number"),(0,i.yg)("td",{parentName:"tr",align:null},"String")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},(0,i.yg)("inlineCode",{parentName:"td"},"project_name")),(0,i.yg)("td",{parentName:"tr",align:null},"Optional project identifier; usually a human-readable name"),(0,i.yg)("td",{parentName:"tr",align:null},"String")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},(0,i.yg)("inlineCode",{parentName:"td"},"library")),(0,i.yg)("td",{parentName:"tr",align:null},"Optional description of the sequencing method or approach"),(0,i.yg)("td",{parentName:"tr",align:null},"String")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},(0,i.yg)("inlineCode",{parentName:"td"},"organ")),(0,i.yg)("td",{parentName:"tr",align:null},"Optional description of the organ from which the cells were derived"),(0,i.yg)("td",{parentName:"tr",align:null},"String")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},(0,i.yg)("inlineCode",{parentName:"td"},"species")),(0,i.yg)("td",{parentName:"tr",align:null},"Optional description of the species from which the cells were derived"),(0,i.yg)("td",{parentName:"tr",align:null},"String")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},(0,i.yg)("inlineCode",{parentName:"td"},"paired-end")),(0,i.yg)("td",{parentName:"tr",align:null},"Boolean for whether samples are paired-end or not"),(0,i.yg)("td",{parentName:"tr",align:null},"Boolean")))),(0,i.yg)("h3",{id:"additional-input"},"Additional Input"),(0,i.yg)("p",null,'The reference inputs are identical to those specified in the "Additional Reference Inputs" section of the ',(0,i.yg)("a",{parentName:"p",href:"https://broadinstitute.github.io/warp/docs/Pipelines/Smart-seq2_Single_Sample_Pipeline/README#inputs"},"Smart-seq2 Single Sample Overview"),"."),(0,i.yg)("h3",{id:"smart-seq2-multi-sample-task-summary"},"Smart-seq2 Multi-Sample Task Summary"),(0,i.yg)("p",null,"The Multi-SS2 Pipeline calls two tasks:"),(0,i.yg)("p",null,"1) ",(0,i.yg)("a",{parentName:"p",href:"https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_sample/SmartSeq2SingleSample.wdl"},"SmartSeq2SingleSample"),": a task that runs the Smart-seq2 Single Sample workflow\n2) ",(0,i.yg)("a",{parentName:"p",href:"https://github.com/broadinstitute/warp/blob/master/tasks/wdl/LoomUtils.wdl"},"SmartSeq2PlateAggregation"),": the wrapper pipeline that aggregates the results"),(0,i.yg)("h3",{id:"outputs"},"Outputs"),(0,i.yg)("table",null,(0,i.yg)("thead",{parentName:"table"},(0,i.yg)("tr",{parentName:"thead"},(0,i.yg)("th",{parentName:"tr",align:null},"Output file name"),(0,i.yg)("th",{parentName:"tr",align:null},"Output Description"),(0,i.yg)("th",{parentName:"tr",align:null},"Output Type"))),(0,i.yg)("tbody",{parentName:"table"},(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},"bam_files"),(0,i.yg)("td",{parentName:"tr",align:null},"An array of genome-aligned BAM files (one for each sample) generated with HISAT2"),(0,i.yg)("td",{parentName:"tr",align:null},"Array")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},"bam_index_files"),(0,i.yg)("td",{parentName:"tr",align:null},"An array of BAM index files generated with HISAT2"),(0,i.yg)("td",{parentName:"tr",align:null},"Array")),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},"loom_output"),(0,i.yg)("td",{parentName:"tr",align:null},"A single Loom cell-by-gene matrix containing raw counts and TPMs for every cell"),(0,i.yg)("td",{parentName:"tr",align:null},"File")))),(0,i.yg)("p",null,"The final Loom matrix is an aggregate of all the individual Loom matrices generated using the ",(0,i.yg)("a",{parentName:"p",href:"https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_sample/SmartSeq2SingleSample.wdl"},"Smart-seq2 Single Sample workflow"),"."),(0,i.yg)("p",null,"The aggregated Loom filename contains the ",(0,i.yg)("inlineCode",{parentName:"p"},"batch_id")," prefix, which is the string specified in the input configuration. The ",(0,i.yg)("inlineCode",{parentName:"p"},"batch_id")," is also set as a global attribute in the Loom."),(0,i.yg)("p",null,"Both the individual sample Loom files and individual BAM files are described in the ",(0,i.yg)("a",{parentName:"p",href:"../Smart-seq2_Single_Sample_Pipeline/README"},"Smart-seq2 Single Sample Overview"),"."),(0,i.yg)("admonition",{title:"Zarr Array Deprecation Notice June 2020",type:"warning"},(0,i.yg)("p",{parentName:"admonition"},"Please note that we have deprecated the previously used Zarr array output. The pipeline now uses the Loom file format as the default output.")),(0,i.yg)("h2",{id:"validation"},"Validation"),(0,i.yg)("p",null,"The Multi-SS2 Pipeline has been validated for processing human and mouse, stranded or unstranded, paired- or single-end, and plate- or fluidigm-based Smart-seq2 datasets (see links to validation reports in the table below)."),(0,i.yg)("table",null,(0,i.yg)("thead",{parentName:"table"},(0,i.yg)("tr",{parentName:"thead"},(0,i.yg)("th",{parentName:"tr",align:null},"Workflow Configuration"),(0,i.yg)("th",{parentName:"tr",align:null},"Link to Report"))),(0,i.yg)("tbody",{parentName:"table"},(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},"Mouse paired-end"),(0,i.yg)("td",{parentName:"tr",align:null},(0,i.yg)("a",{parentName:"td",href:"https://docs.google.com/document/d/12zGTFROrcXEByt9z0h06qjSqb9vWutn28Tx6YiND1Ds/edit"},"Report"))),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},"Human and mouse single-end"),(0,i.yg)("td",{parentName:"tr",align:null},(0,i.yg)("a",{parentName:"td",href:"https://docs.google.com/document/d/1MonsTG8UnROHZ_XpulrSZNTxO988KEH6T6h45plFYQg/edit#heading=h.ixoqmhbabdvh"},"Report"))),(0,i.yg)("tr",{parentName:"tbody"},(0,i.yg)("td",{parentName:"tr",align:null},"Human stranded fluidigm"),(0,i.yg)("td",{parentName:"tr",align:null},(0,i.yg)("a",{parentName:"td",href:"https://docs.google.com/document/d/1FEg86Tlu657j9Kjw_v3keFQRXcBIs8gOqCwLbPSP-C0/edit#heading=h.wjr8otl7zg14"},"Report"))))),(0,i.yg)("h2",{id:"versioning"},"Versioning"),(0,i.yg)("p",null,"Release information for the Multi-SS2 Pipeline can be found in the ",(0,i.yg)("a",{parentName:"p",href:"https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_multisample/MultiSampleSmartSeq2.changelog.md"},"changelog"),". Please note that any major changes to the Smart-seq2 pipeline will be documented in the ",(0,i.yg)("a",{parentName:"p",href:"https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_sample/SmartSeq2SingleSample.changelog.md"},"Smart-seq2 Single Sample changelog"),"."),(0,i.yg)("h2",{id:"citing-the-smart-seq2-multi-sample-pipeline"},"Citing the Smart-seq2 Multi-Sample Pipeline"),(0,i.yg)("p",null,"If you use the Smart-seq2 Multi-Sample Pipeline in your research, please identify the pipeline in your methods section using the ",(0,i.yg)("a",{parentName:"p",href:"https://scicrunch.org/resources/data/record/nlx_144509-1/SCR_018920/resolver?q=SCR_018920&l=SCR_018920&i=rrid:scr_018920"},"Smart-seq2 Multi-Sample SciCrunch resource identifier"),"."),(0,i.yg)("ul",null,(0,i.yg)("li",{parentName:"ul"},"Ex: ",(0,i.yg)("em",{parentName:"li"},"Smart-seq2 Multi-Sample Pipeline (RRID:SCR_018920)"))),(0,i.yg)("p",null,"Please also consider citing our preprint:"),(0,i.yg)("p",null,"Degatano, K., Awdeh, A., Cox III, R.S., Dingman, W., Grant, G., Khajouei, F., Kiernan, E., Konwar, K., Mathews, K.L., Palis, K., et al. Warp Analysis Research Pipelines: Cloud-optimized workflows for biological data processing and reproducible analysis. Bioinformatics 2025; btaf494. ",(0,i.yg)("a",{parentName:"p",href:"https://doi.org/10.1093/bioinformatics/btaf494"},"https://doi.org/10.1093/bioinformatics/btaf494")),(0,i.yg)("h2",{id:"consortia-support"},"Consortia Support"),(0,i.yg)("p",null,"This pipeline is supported and used by the ",(0,i.yg)("a",{parentName:"p",href:"https://www.humancellatlas.org/"},"Human Cell Atlas")," (HCA) project. "),(0,i.yg)("p",null,"If your organization also uses this pipeline, we would love to list you! Please reach out to us by ",(0,i.yg)("a",{parentName:"p",href:"https://github.com/broadinstitute/warp/issues"},"filing an issue in WARP"),"."),(0,i.yg)("h2",{id:"have-suggestions"},"Have Suggestions?"),(0,i.yg)("p",null,"Please help us make our tools better by ",(0,i.yg)("a",{parentName:"p",href:"https://github.com/broadinstitute/warp/issues"},"filing an issue in WARP"),"; we welcome pipeline-related suggestions or questions."))}d.isMDXComponent=!0}}]);