"use strict";(self.webpackChunkwebsite_2=self.webpackChunkwebsite_2||[]).push([[2611],{5680:(e,t,a)=>{a.d(t,{xA:()=>p,yg:()=>y});var n=a(6540);function r(e,t,a){return t in e?Object.defineProperty(e,t,{value:a,enumerable:!0,configurable:!0,writable:!0}):e[t]=a,e}function i(e,t){var a=Object.keys(e);if(Object.getOwnPropertySymbols){var n=Object.getOwnPropertySymbols(e);t&&(n=n.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),a.push.apply(a,n)}return a}function l(e){for(var t=1;t<arguments.length;t++){var a=null!=arguments[t]?arguments[t]:{};t%2?i(Object(a),!0).forEach((function(t){r(e,t,a[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(a)):i(Object(a)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(a,t))}))}return e}function o(e,t){if(null==e)return{};var a,n,r=function(e,t){if(null==e)return{};var a,n,r={},i=Object.keys(e);for(n=0;n<i.length;n++)a=i[n],t.indexOf(a)>=0||(r[a]=e[a]);return r}(e,t);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(e);for(n=0;n<i.length;n++)a=i[n],t.indexOf(a)>=0||Object.prototype.propertyIsEnumerable.call(e,a)&&(r[a]=e[a])}return r}var s=n.createContext({}),c=function(e){var t=n.useContext(s),a=t;return e&&(a="function"==typeof e?e(t):l(l({},t),e)),a},p=function(e){var t=c(e.components);return n.createElement(s.Provider,{value:t},e.children)},g="mdxType",d={inlineCode:"code",wrapper:function(e){var t=e.children;return n.createElement(n.Fragment,{},t)}},m=n.forwardRef((function(e,t){var a=e.components,r=e.mdxType,i=e.originalType,s=e.parentName,p=o(e,["components","mdxType","originalType","parentName"]),g=c(a),m=r,y=g["".concat(s,".").concat(m)]||g[m]||d[m]||i;return a?n.createElement(y,l(l({ref:t},p),{},{components:a})):n.createElement(y,l({ref:t},p))}));function y(e,t){var a=arguments,r=t&&t.mdxType;if("string"==typeof e||r){var i=a.length,l=new Array(i);l[0]=m;var o={};for(var s in t)hasOwnProperty.call(t,s)&&(o[s]=t[s]);o.originalType=e,o[g]="string"==typeof e?e:r,l[1]=o;for(var c=2;c<i;c++)l[c]=a[c];return n.createElement.apply(null,l)}return n.createElement.apply(null,a)}m.displayName="MDXCreateElement"},8993:(e,t,a)=>{a.r(t),a.d(t,{assets:()=>s,contentTitle:()=>l,default:()=>d,frontMatter:()=>i,metadata:()=>o,toc:()=>c});var n=a(8168),r=(a(6540),a(5680));const i={sidebar_position:2},l="ATAC Library Metrics Overview",o={unversionedId:"Pipelines/ATAC/library-metrics",id:"Pipelines/ATAC/library-metrics",title:"ATAC Library Metrics Overview",description:"The ATAC pipeline uses SnapATAC2 to generate library-level metrics in CSV format.",source:"@site/docs/Pipelines/ATAC/library-metrics.md",sourceDirName:"Pipelines/ATAC",slug:"/Pipelines/ATAC/library-metrics",permalink:"/warp/docs/Pipelines/ATAC/library-metrics",draft:!1,editUrl:"https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/ATAC/library-metrics.md",tags:[],version:"current",lastUpdatedBy:"Nikelle Petrillo",lastUpdatedAt:1761573741,formattedLastUpdatedAt:"Oct 27, 2025",sidebarPosition:2,frontMatter:{sidebar_position:2},sidebar:"docsSidebar",previous:{title:"ATAC Count Matrix Overview",permalink:"/warp/docs/Pipelines/ATAC/count-matrix-overview"},next:{title:"ATAC v2.3.1 Methods",permalink:"/warp/docs/Pipelines/ATAC/atac.methods"}},s={},c=[],p={toc:c},g="wrapper";function d(e){let{components:t,...a}=e;return(0,r.yg)(g,(0,n.A)({},p,a,{components:t,mdxType:"MDXLayout"}),(0,r.yg)("h1",{id:"atac-library-metrics-overview"},"ATAC Library Metrics Overview"),(0,r.yg)("p",null,"The ",(0,r.yg)("a",{parentName:"p",href:"/warp/docs/Pipelines/ATAC/README"},"ATAC pipeline")," uses ",(0,r.yg)("a",{parentName:"p",href:"https://github.com/kaizhang/SnapATAC2"},"SnapATAC2")," to generate library-level metrics in CSV format.  "),(0,r.yg)("table",null,(0,r.yg)("thead",{parentName:"table"},(0,r.yg)("tr",{parentName:"thead"},(0,r.yg)("th",{parentName:"tr",align:null},"Metric"),(0,r.yg)("th",{parentName:"tr",align:null},"Description"))),(0,r.yg)("tbody",{parentName:"table"},(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"NHash_ID"),(0,r.yg)("td",{parentName:"tr",align:null},"A unique identifier used to track and reference the specific sample or dataset.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"sequenced_reads"),(0,r.yg)("td",{parentName:"tr",align:null},"The total number of reads generated from the sequencing process, which includes both reads that are mapped and unmapped.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"sequenced_read_pairs"),(0,r.yg)("td",{parentName:"tr",align:null},"The total number of read pairs (two reads per pair) generated from the sequencing process. This is typically half of the total sequenced reads if all reads are paired.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"fraction_valid_barcode"),(0,r.yg)("td",{parentName:"tr",align:null},"The fraction of reads that contain a valid barcode, indicating the proportion of reads that are correctly assigned to a specific cell or sample.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"fraction_Q30_bases_in_read_1"),(0,r.yg)("td",{parentName:"tr",align:null},"The proportion of bases in Read 1 that have a Phred quality score of 30 or higher, indicating high-confidence base calls.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"fraction_Q30_bases_in_read_2"),(0,r.yg)("td",{parentName:"tr",align:null},"The proportion of bases in Read 2 that have a Phred quality score of 30 or higher, indicating high-confidence base calls.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"number_of_cells"),(0,r.yg)("td",{parentName:"tr",align:null},"The estimated number of cells captured and sequenced in the experiment, based on the barcodes identified.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"mean_raw_read_pairs_per_cell"),(0,r.yg)("td",{parentName:"tr",align:null},"The average number of raw read pairs associated with each cell, providing an indication of the sequencing depth per cell.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"median_high-quality_fragments_per_cell"),(0,r.yg)("td",{parentName:"tr",align:null},"The median number of high-quality (e.g., confidently mapped) fragments associated with each cell, representing typical fragment quality across cells.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"fraction of high-quality fragments in cells"),(0,r.yg)("td",{parentName:"tr",align:null},"The fraction of high-quality fragments that are associated with identified cells, indicating the proportion of good-quality data that is cell-associated.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"fraction_of_transposition_events_in_peaks_in_cells"),(0,r.yg)("td",{parentName:"tr",align:null},"The fraction of transposition events within identified cells that occur within peaks, which are regions of accessible chromatin.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"fraction_duplicates"),(0,r.yg)("td",{parentName:"tr",align:null},"The fraction of sequenced fragments that are duplicates, which can result from PCR amplification or other factors, indicating the redundancy in the sequencing data.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"fraction_confidently_mapped"),(0,r.yg)("td",{parentName:"tr",align:null},"The fraction of sequenced fragments that are confidently mapped to the reference genome, indicating the proportion of reads that align well to the genome.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"fraction_unmapped"),(0,r.yg)("td",{parentName:"tr",align:null},"The fraction of sequenced fragments that could not be mapped to the reference genome, which can indicate sequencing errors, contamination, or regions not covered by the reference.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"fraction_nonnuclear"),(0,r.yg)("td",{parentName:"tr",align:null},"The fraction of sequenced fragments that are mapped to non-nuclear (e.g., mitochondrial or other organellar) DNA, providing insight into contamination or organellar activity.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"fraction_fragment_in_nucleosome_free_region"),(0,r.yg)("td",{parentName:"tr",align:null},"The fraction of sequenced fragments that map to nucleosome-free regions, which are indicative of accessible chromatin.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"fraction_fragment_flanking_single_nucleosome"),(0,r.yg)("td",{parentName:"tr",align:null},"The fraction of sequenced fragments that map to regions flanking single nucleosomes, indicating regions with partial chromatin accessibility.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"tss_enrichment_score"),(0,r.yg)("td",{parentName:"tr",align:null},"A measure of the enrichment of transposition events at transcription start sites (TSS), indicating the accessibility of promoters across the genome.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"fraction_of_high-quality_fragments_overlapping_TSS"),(0,r.yg)("td",{parentName:"tr",align:null},"The fraction of high-quality fragments that overlap transcription start sites (TSS), providing insight into promoter accessibility.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"Number_of_peaks"),(0,r.yg)("td",{parentName:"tr",align:null},"The total number of peaks, or regions of accessible chromatin, identified in the dataset, representing potential regulatory elements.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"fraction_of_genome_in_peaks"),(0,r.yg)("td",{parentName:"tr",align:null},"The fraction of the genome that is covered by identified peaks, indicating the extent of chromatin accessibility across the genome.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"fraction_of_high-quality_fragments_overlapping_peaks"),(0,r.yg)("td",{parentName:"tr",align:null},"The fraction of high-quality fragments that overlap with identified peaks, providing an indication of the efficiency of the assay in capturing accessible regions.")),(0,r.yg)("tr",{parentName:"tbody"},(0,r.yg)("td",{parentName:"tr",align:null},"atac_percent_target"),(0,r.yg)("td",{parentName:"tr",align:null},"Percent of cells recovered; value is calculated as estimated_cells/expected_cells.")))))}d.isMDXComponent=!0}}]);