"use strict";(self.webpackChunkwebsite_2=self.webpackChunkwebsite_2||[]).push([[6017],{5680:(t,e,a)=>{a.d(e,{xA:()=>p,yg:()=>N});var r=a(6540);function n(t,e,a){return e in t?Object.defineProperty(t,e,{value:a,enumerable:!0,configurable:!0,writable:!0}):t[e]=a,t}function i(t,e){var a=Object.keys(t);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(t);e&&(r=r.filter((function(e){return Object.getOwnPropertyDescriptor(t,e).enumerable}))),a.push.apply(a,r)}return a}function l(t){for(var e=1;e<arguments.length;e++){var a=null!=arguments[e]?arguments[e]:{};e%2?i(Object(a),!0).forEach((function(e){n(t,e,a[e])})):Object.getOwnPropertyDescriptors?Object.defineProperties(t,Object.getOwnPropertyDescriptors(a)):i(Object(a)).forEach((function(e){Object.defineProperty(t,e,Object.getOwnPropertyDescriptor(a,e))}))}return t}function g(t,e){if(null==t)return{};var a,r,n=function(t,e){if(null==t)return{};var a,r,n={},i=Object.keys(t);for(r=0;r<i.length;r++)a=i[r],e.indexOf(a)>=0||(n[a]=t[a]);return n}(t,e);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(t);for(r=0;r<i.length;r++)a=i[r],e.indexOf(a)>=0||Object.prototype.propertyIsEnumerable.call(t,a)&&(n[a]=t[a])}return n}var d=r.createContext({}),m=function(t){var e=r.useContext(d),a=e;return t&&(a="function"==typeof t?t(e):l(l({},e),t)),a},p=function(t){var e=m(t.components);return r.createElement(d.Provider,{value:e},t.children)},o="mdxType",s={inlineCode:"code",wrapper:function(t){var e=t.children;return r.createElement(r.Fragment,{},e)}},y=r.forwardRef((function(t,e){var a=t.components,n=t.mdxType,i=t.originalType,d=t.parentName,p=g(t,["components","mdxType","originalType","parentName"]),o=m(a),y=n,N=o["".concat(d,".").concat(y)]||o[y]||s[y]||i;return a?r.createElement(N,l(l({ref:e},p),{},{components:a})):r.createElement(N,l({ref:e},p))}));function N(t,e){var a=arguments,n=e&&e.mdxType;if("string"==typeof t||n){var i=a.length,l=new Array(i);l[0]=y;var g={};for(var d in e)hasOwnProperty.call(e,d)&&(g[d]=e[d]);g.originalType=t,g[o]="string"==typeof t?t:n,l[1]=g;for(var m=2;m<i;m++)l[m]=a[m];return r.createElement.apply(null,l)}return r.createElement.apply(null,a)}y.displayName="MDXCreateElement"},1846:(t,e,a)=>{a.r(e),a.d(e,{assets:()=>d,contentTitle:()=>l,default:()=>s,frontMatter:()=>i,metadata:()=>g,toc:()=>m});var r=a(9668),n=(a(6540),a(5680));const i={sidebar_position:3},l="Smart-seq2 Single Nucleus Multi-Sample Count Matrix Overview",g={unversionedId:"Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/count-matrix-overview",id:"Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/count-matrix-overview",title:"Smart-seq2 Single Nucleus Multi-Sample Count Matrix Overview",description:"The Smart-seq2 Single Nucleus Multi-Sample (Multi-snSS2) pipeline's default count matrix output is a Loom file, an HDF5 file generated using Loompy v.3.0.6. It contains the raw cell-by-gene intron and exon counts.",source:"@site/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/count-matrix-overview.md",sourceDirName:"Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline",slug:"/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/count-matrix-overview",permalink:"/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/count-matrix-overview",draft:!1,editUrl:"https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/count-matrix-overview.md",tags:[],version:"current",lastUpdatedBy:"Robert Sidney Cox",lastUpdatedAt:1733411364,formattedLastUpdatedAt:"Dec 5, 2024",sidebarPosition:3,frontMatter:{sidebar_position:3},sidebar:"docsSidebar",previous:{title:"Smart-seq2 Single Nucleus Multi-Sample v1.3.4 Publication Methods",permalink:"/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/multi_snss2.methods"},next:{title:"Consortia Data Processing",permalink:"/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/consortia-processing"}},d={},m=[{value:"Table 1. Global attributes",id:"table-1-global-attributes",level:2},{value:"Table 2. Column attributes (cell metrics)",id:"table-2-column-attributes-cell-metrics",level:2},{value:"Table 3. Row attributes (gene metrics)",id:"table-3-row-attributes-gene-metrics",level:2}],p={toc:m},o="wrapper";function s(t){let{components:e,...a}=t;return(0,n.yg)(o,(0,r.A)({},p,a,{components:e,mdxType:"MDXLayout"}),(0,n.yg)("h1",{id:"smart-seq2-single-nucleus-multi-sample-count-matrix-overview"},"Smart-seq2 Single Nucleus Multi-Sample Count Matrix Overview"),(0,n.yg)("p",null,"The Smart-seq2 Single Nucleus Multi-Sample (Multi-snSS2) pipeline's default count matrix output is a Loom file, an HDF5 file generated using ",(0,n.yg)("a",{parentName:"p",href:"http://loompy.org/"},"Loompy v.3.0.6"),". It contains the raw cell-by-gene intron and exon counts."),(0,n.yg)("p",null,"The matrix also contains multiple metrics for both individual cells (the columns of the matrix; ",(0,n.yg)("a",{parentName:"p",href:"#table-2-column-attributes-cell-metrics"},"Table 2"),") and individual genes (the rows of the matrix; ",(0,n.yg)("a",{parentName:"p",href:"#table-3-row-attributes-gene-metrics"},"Table 3"),")."),(0,n.yg)("p",null,"Additional details for each metric are provided in the JAVA source code for Picard's ",(0,n.yg)("a",{parentName:"p",href:"https://github.com/broadinstitute/picard/blob/4527d6de2f33f98f80e32e3acd32b09633529bd0/src/main/java/picard/analysis/AlignmentSummaryMetrics.java"},"AlignmentSummaryMetrics"),", ",(0,n.yg)("a",{parentName:"p",href:"https://github.com/broadinstitute/picard/blob/4527d6de2f33f98f80e32e3acd32b09633529bd0/src/main/java/picard/analysis/GcBiasSummaryMetrics.java"},"GcBiasSummaryMetrics"),", and ",(0,n.yg)("a",{parentName:"p",href:"https://github.com/broadinstitute/picard/blob/4527d6de2f33f98f80e32e3acd32b09633529bd0/src/main/java/picard/sam/DuplicationMetrics.java"},"DuplicationMetrics"),"."),(0,n.yg)("h2",{id:"table-1-global-attributes"},"Table 1. Global attributes"),(0,n.yg)("p",null,"The global attributes in the Loom apply to the whole file, not any specific part. "),(0,n.yg)("table",null,(0,n.yg)("thead",{parentName:"table"},(0,n.yg)("tr",{parentName:"thead"},(0,n.yg)("th",{parentName:"tr",align:"left"},"Attribute"),(0,n.yg)("th",{parentName:"tr",align:"left"},"Details"))),(0,n.yg)("tbody",{parentName:"table"},(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"CreationDate")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Date the Loom file was created.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"LOOM_SPEC_VERSION")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Loom file spec version used during creation of the Loom file.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"batch_id")),(0,n.yg)("td",{parentName:"tr",align:"left"},"The ",(0,n.yg)("inlineCode",{parentName:"td"},"batch_id")," provided to the pipeline as input.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"pipeline_version")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Version of the Multi-snSS2 pipeline used to generate the Loom file.")))),(0,n.yg)("h2",{id:"table-2-column-attributes-cell-metrics"},"Table 2. Column attributes (cell metrics)"),(0,n.yg)("p",null,"The cell metrics below are computed using ",(0,n.yg)("a",{parentName:"p",href:"https://broadinstitute.github.io/picard/"},"Picard"),", with the exception of ",(0,n.yg)("inlineCode",{parentName:"p"},"CellID"),", ",(0,n.yg)("inlineCode",{parentName:"p"},"cell_names"),", and ",(0,n.yg)("inlineCode",{parentName:"p"},"input_id")," which are provided to the pipeline as input."),(0,n.yg)("table",null,(0,n.yg)("thead",{parentName:"table"},(0,n.yg)("tr",{parentName:"thead"},(0,n.yg)("th",{parentName:"tr",align:"left"},"Cell Metrics"),(0,n.yg)("th",{parentName:"tr",align:"left"},"Tool"),(0,n.yg)("th",{parentName:"tr",align:"left"},"Details"))),(0,n.yg)("tbody",{parentName:"table"},(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"ACCUMULATION_LEVEL")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832753924763"},"CollectMultipleMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Level of metric accumulation; set to ",(0,n.yg)("inlineCode",{parentName:"td"},"ALL_READS")," using the ",(0,n.yg)("inlineCode",{parentName:"td"},"--METRIC_ACCUMULATION_LEVEL")," argument.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"ALIGNED_READS")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832711132315"},"CollectGcBiasMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Total number of aligned reads produced in a run.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"AT_DROPOUT")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832711132315"},"CollectGcBiasMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Percentage of misaligned reads with GC content below 50%.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"AVG_POS_3PRIME_SOFTCLIP_LENGTH.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Average length of soft-clipped bases at the 3' end of the first reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"AVG_POS_3PRIME_SOFTCLIP_LENGTH.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Average length of soft-clipped bases at the 3' end of all reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"AVG_POS_3PRIME_SOFTCLIP_LENGTH.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Average length of soft-clipped bases at the 3' end of the second reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"BAD_CYCLES.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of cycles with combined no-call and mismatch rates greater than or equal to 80% for the first reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"BAD_CYCLES.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of cycles with combined no-call and mismatch rates greater than or equal to 80% for all reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"BAD_CYCLES.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of cycles with combined no-call and mismatch rates greater than or equal to 80% for the second reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"CellID")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://github.com/broadinstitute/warp-tools"},"warp-tools")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Unique identifier for each cell provided to the pipeline as ",(0,n.yg)("inlineCode",{parentName:"td"},"input_ids"),"; identical to ",(0,n.yg)("inlineCode",{parentName:"td"},"cell_names")," and ",(0,n.yg)("inlineCode",{parentName:"td"},"input_id"),".")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"ESTIMATED_LIBRARY_SIZE")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832748517275"},"MarkDuplicates")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Estimated number of unique molecules in the library based on paired-end duplication.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"GC_DROPOUT")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832711132315"},"CollectGcBiasMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Percentage of misaligned reads with GC content above 50%.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"GC_NC_0_19")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832711132315"},"CollectGcBiasMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Normalized coverage over reads with GC content from 0 - 19%.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"GC_NC_20_39")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832711132315"},"CollectGcBiasMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Normalized coverage over reads with GC content from 20 - 39%.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"GC_NC_40_59")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832711132315"},"CollectGcBiasMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Normalized coverage over reads with GC content from 40 - 59%.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"GC_NC_60_79")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832711132315"},"CollectGcBiasMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Normalized coverage over reads with GC content from 60 - 79%.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"GC_NC_80_100")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832711132315"},"CollectGcBiasMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Normalized coverage over reads with GC content from 80 - 100%.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MAD_READ_LENGTH.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Median absolute deviation of the lengths of forward reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MAD_READ_LENGTH.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Median absolute deviation of the lengths of all reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MAD_READ_LENGTH.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Median absolute deviation of the lengths of reverse reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MAX_READ_LENGTH.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Maximum length of forward reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MAX_READ_LENGTH.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Maximum length of all reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MAX_READ_LENGTH.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Maximum length of reverse reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MEAN_READ_LENGTH.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Mean length of forward reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MEAN_READ_LENGTH.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Mean length of all reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MEAN_READ_LENGTH.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Mean length of reverse reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MEDIAN_READ_LENGTH.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Median length of forward reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MEDIAN_READ_LENGTH.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Median length of all reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MEDIAN_READ_LENGTH.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Median length of reverse reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MIN_READ_LENGTH.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Minimum length of forward reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MIN_READ_LENGTH.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Minimum length of all reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"MIN_READ_LENGTH.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Minimum length of reverse reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_ADAPTER.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of pass-filter forward reads that are unaligned or aligned with a mapping quality of 0 and match to a known adapter sequence from the start of the read.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_ADAPTER.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of all pass-filter reads that are unaligned or aligned with a mapping quality of 0 and match to a known adapter sequence from the start of the read.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_ADAPTER.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of pass-filter reverse reads that are unaligned or aligned with a mapping quality of 0 and match to a known adapter sequence from the start of the read.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_CHIMERAS.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of forward reads where the insert is larger than 100 kb or the ends of the pair map to different chromosomes.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_CHIMERAS.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of all reads where the insert is larger than 100 kb or the ends of the pair map to different chromosomes.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_CHIMERAS.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of reverse reads where the insert is larger than 100 kb or the ends of the pair map to different chromosomes.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_HARDCLIP.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of pass-filter bases that are hard-clipped from aligned, forward reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_HARDCLIP.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of pass-filter bases that are hard-clipped from all aligned reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_HARDCLIP.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of pass-filter bases that are hard-clipped from aligned, reverse reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_PF_READS.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of forward reads that pass vendor check (pass-filter).")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_PF_READS.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of reads that pass vendor check (pass-filter).")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_PF_READS.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of reverse reads that pass vendor check (pass-filter).")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_PF_READS_ALIGNED.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of pass-filter forward reads that are aligned.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_PF_READS_ALIGNED.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of all pass-filter reads that are aligned.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_PF_READS_ALIGNED.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of pass-filter reverse reads that are aligned.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_PF_READS_IMPROPER_PAIRS.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of forward reads not properly aligned in pairs.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_PF_READS_IMPROPER_PAIRS.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of reads not properly aligned in pairs.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_PF_READS_IMPROPER_PAIRS.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of reverse reads not properly aligned in pairs.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_READS_ALIGNED_IN_PAIRS.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of forward reads properly aligned in pairs.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_READS_ALIGNED_IN_PAIRS.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of reads properly aligned in pairs.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_READS_ALIGNED_IN_PAIRS.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of reverse reads properly aligned in pairs.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_SOFTCLIP.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of pass-filter bases that are soft-clipped from aligned forward reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_SOFTCLIP.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of pass-filter bases that are soft-clipped from all aligned reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PCT_SOFTCLIP.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of pass-filter bases that are soft-clipped from aligned reverse reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PERCENT_DUPLICATION")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832748517275"},"MarkDuplicates")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of mapped sequence marked as duplicate.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_ALIGNED_BASES.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Total number of aligned bases in pass-filter forward reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_ALIGNED_BASES.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Total number of aligned bases in all pass-filter reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_ALIGNED_BASES.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Total number of aligned bases in pass-filter reverse reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ALIGNED_BASES.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of bases aligned to the reference sequence in forward reads with high mapping quality.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ALIGNED_BASES.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of bases aligned to the reference sequence in all reads with high mapping quality.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ALIGNED_BASES.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of bases aligned to the reference sequence in reverse reads with high mapping quality.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ALIGNED_Q20_BASES.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Subset of ",(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ALIGNED_BASES.FIRST_OF_PAIR")," with a base call quality of at least 20.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ALIGNED_Q20_BASES.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Subset of ",(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ALIGNED_BASES.PAIR")," with a base call quality of at least 20.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ALIGNED_Q20_BASES.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Subset of ",(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ALIGNED_BASES.SECOND_OF_PAIR")," with a base call quality of at least 20.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ALIGNED_READS.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of pass-filter forward reads aligned with a mapping quality of at least 20.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ALIGNED_READS.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of all pass-filter reads aligned with a mapping quality of at least 20.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ALIGNED_READS.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of pass-filter reverse reads aligned with a mapping quality of at least 20.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ERROR_RATE.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of bases in pass-filter, high-quality forward reads that do not match the reference.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ERROR_RATE.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of bases in all pass-filter, high-quality reads that do not match the reference.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_ERROR_RATE.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Fraction of bases in pass-filter, high-quality reverse reads that do not match the reference.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_MEDIAN_MISMATCHES.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Median number of mismatches in high-quality forward reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_MEDIAN_MISMATCHES.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Median number of mismatches in all high-quality reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_HQ_MEDIAN_MISMATCHES.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Median number of mismatches in high-quality reverse reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_INDEL_RATE.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of insertion and deletion events per 100 aligned bases in forward reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_INDEL_RATE.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of insertion and deletion events per 100 aligned bases in all reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_INDEL_RATE.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of insertion and deletion events per 100 aligned bases in reverse reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_MISMATCH_RATE.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Rate of base mismatching for all aligned bases in forward reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_MISMATCH_RATE.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Rate of base mismatching for all aligned bases in all reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_MISMATCH_RATE.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Rate of base mismatching for all aligned bases in reverse reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_NOISE_READS.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of pass-filter forward reads marked as noise.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_NOISE_READS.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of all pass-filter reads marked as noise.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_NOISE_READS.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of pass-filter reverse reads marked as noise.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_READS.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of forward reads that pass vendor check (pass-filter).")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_READS.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of reads that pass vendor check (pass-filter).")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_READS.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of reverse reads that pass vendor check (pass-filter).")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_READS_ALIGNED.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of pass-filter forward reads that are aligned.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_READS_ALIGNED.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of pass-filter reads that are aligned.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_READS_ALIGNED.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of pass-filter reverse reads that are aligned.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_READS_IMPROPER_PAIRS.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of forward reads not properly aligned in pairs.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_READS_IMPROPER_PAIRS.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of reads not properly aligned in pairs.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"PF_READS_IMPROPER_PAIRS.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of reverse reads not properly aligned in pairs.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"READS_ALIGNED_IN_PAIRS.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of forward reads properly aligned in pairs.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"READS_ALIGNED_IN_PAIRS.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of reads properly aligned in pairs.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"READS_ALIGNED_IN_PAIRS.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of reverse reads properly aligned in pairs.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"READS_USED")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832711132315"},"CollectGcBiasMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"String describing whether duplicates are included in metrics produced by ",(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832711132315"},"CollectGcBiasMetrics"),"; the pipeline removes duplicates before metrics are calculated.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"READ_PAIRS_EXAMINED")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832748517275"},"MarkDuplicates")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of mapped read pairs examined by ",(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832748517275"},"MarkDuplicates"),".")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"READ_PAIR_DUPLICATES")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832748517275"},"MarkDuplicates")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of read pairs marked as duplicates.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"READ_PAIR_OPTICAL_DUPLICATES")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832748517275"},"MarkDuplicates")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of read pairs duplicates caused by optical duplication.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"SD_READ_LENGTH.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Standard deviation of forward read lengths.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"SD_READ_LENGTH.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Standard deviation of read lengths.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"SD_READ_LENGTH.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Standard deviation of reverse read lengths.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"SECONDARY_OR_SUPPLEMENTARY_RDS")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832748517275"},"MarkDuplicates")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of secondary or supplemetary reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"STRAND_BALANCE.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of pass-filter forward reads aligned divided by the total number of pass-filter reads aligned.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"STRAND_BALANCE.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Average strand balance of forward and reverse reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"STRAND_BALANCE.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of pass-filter reverse reads aligned divided by the total number of pass-filter reads aligned.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"TOTAL_CLUSTERS")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832711132315"},"CollectGcBiasMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Total number of reads after filtering used in GC bias calculation.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"TOTAL_READS.FIRST_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Total number of forward reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"TOTAL_READS.PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Total number of reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"TOTAL_READS.SECOND_OF_PAIR")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832683660955"},"CollectAlignmentSummaryMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Total number of reverse reads.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"UNMAPPED_READS")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832748517275"},"MarkDuplicates")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of unmapped reads examined by ",(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832748517275"},"MarkDuplicates"),".")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"UNPAIRED_READS_EXAMINED")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832748517275"},"MarkDuplicates")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of mapped reads without a mapped mate pair examined by ",(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832748517275"},"MarkDuplicates"),".")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"UNPAIRED_READ_DUPLICATES")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832748517275"},"MarkDuplicates")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Number of fragments marked as duplicates.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"WINDOW_SIZE")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832711132315"},"CollectGcBiasMetrics")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Genomic window size used in GC bias calculation.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"cell_names")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://github.com/broadinstitute/warp-tools"},"warp-tools")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Unique identifier for each cell provided to the pipeline as ",(0,n.yg)("inlineCode",{parentName:"td"},"input_ids"),"; identical to ",(0,n.yg)("inlineCode",{parentName:"td"},"Cell_ID")," and ",(0,n.yg)("inlineCode",{parentName:"td"},"input_id"),".")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("inlineCode",{parentName:"td"},"input_id")),(0,n.yg)("td",{parentName:"tr",align:"left"},(0,n.yg)("a",{parentName:"td",href:"https://github.com/broadinstitute/warp-tools"},"warp-tools")),(0,n.yg)("td",{parentName:"tr",align:"left"},"Unique identifier for each cell provided to the pipeline as ",(0,n.yg)("inlineCode",{parentName:"td"},"input_ids"),"; identical to ",(0,n.yg)("inlineCode",{parentName:"td"},"Cell_ID")," and ",(0,n.yg)("inlineCode",{parentName:"td"},"cell_names"),".")))),(0,n.yg)("h2",{id:"table-3-row-attributes-gene-metrics"},"Table 3. Row attributes (gene metrics)"),(0,n.yg)("table",null,(0,n.yg)("thead",{parentName:"table"},(0,n.yg)("tr",{parentName:"thead"},(0,n.yg)("th",{parentName:"tr",align:null},"Gene Metrics"),(0,n.yg)("th",{parentName:"tr",align:null},"Tool"),(0,n.yg)("th",{parentName:"tr",align:null},"Details"))),(0,n.yg)("tbody",{parentName:"table"},(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:null},(0,n.yg)("inlineCode",{parentName:"td"},"Gene")),(0,n.yg)("td",{parentName:"tr",align:null},(0,n.yg)("a",{parentName:"td",href:"https://www.gencodegenes.org/"},"GENCODE GTF")),(0,n.yg)("td",{parentName:"tr",align:null},"The unique ",(0,n.yg)("inlineCode",{parentName:"td"},"gene_ids")," provided in the ",(0,n.yg)("a",{parentName:"td",href:"https://www.gencodegenes.org/"},"GENCODE GTF"),"; identical to the ",(0,n.yg)("inlineCode",{parentName:"td"},"ensembl_ids")," attribute.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:null},(0,n.yg)("inlineCode",{parentName:"td"},"ensembl_ids")),(0,n.yg)("td",{parentName:"tr",align:null},(0,n.yg)("a",{parentName:"td",href:"https://www.gencodegenes.org/"},"GENCODE GTF")),(0,n.yg)("td",{parentName:"tr",align:null},"The unique ",(0,n.yg)("inlineCode",{parentName:"td"},"gene_ids")," provided in the ",(0,n.yg)("a",{parentName:"td",href:"https://www.gencodegenes.org/"},"GENCODE GTF"),"; identical to the ",(0,n.yg)("inlineCode",{parentName:"td"},"Gene")," attribute.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:null},(0,n.yg)("inlineCode",{parentName:"td"},"exon_lengths")),(0,n.yg)("td",{parentName:"tr",align:null},(0,n.yg)("a",{parentName:"td",href:"https://github.com/broadinstitute/warp-tools"},"warp-tools")),(0,n.yg)("td",{parentName:"tr",align:null},"The length in base pairs of the exons corresponding to this entity.")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:null},(0,n.yg)("inlineCode",{parentName:"td"},"gene_names")),(0,n.yg)("td",{parentName:"tr",align:null},(0,n.yg)("a",{parentName:"td",href:"https://www.gencodegenes.org/"},"GENCODE GTF")),(0,n.yg)("td",{parentName:"tr",align:null},"The unique ",(0,n.yg)("inlineCode",{parentName:"td"},"gene_name")," provided in the ",(0,n.yg)("a",{parentName:"td",href:"https://www.gencodegenes.org/"},"GENCODE GTF"),".")),(0,n.yg)("tr",{parentName:"tbody"},(0,n.yg)("td",{parentName:"tr",align:null},(0,n.yg)("inlineCode",{parentName:"td"},"intron_lengths")),(0,n.yg)("td",{parentName:"tr",align:null},(0,n.yg)("a",{parentName:"td",href:"https://github.com/broadinstitute/warp-tools"},"warp-tools")),(0,n.yg)("td",{parentName:"tr",align:null},"The length in base pairs of the introns corresponding to this entity.")))))}s.isMDXComponent=!0}}]);