# Smart-seq2 Single Nucleus Multi-Sample Overview

| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |
| :----: | :---: | :----: | :--------------: |
| [MultiSampleSmartSeq2SingleNuclei_v1.0.0.](https://github.com/broadinstitute/warp/releases) | May, 2021 | [Elizabeth Kiernan](mailto:ekiernan@broadinstitute.org) | Please file GitHub issues in WARP or contact [Kylee Degatano](mailto:kdegatano@broadinstitute.org) |

## Introduction

The Smart-seq2 Single Nucleus Multi-Sample (Multi-snSS2) Pipeline is a wrapper around the [Smart-seq2 Single Nucleus](../Smart-seq2_Single_Nucleus_Pipeline/) pipeline. It is developed in collaboration with the [BRAIN Initiative Cell Census Network](https://biccn.org/) to process single-nucleus RNAseq (snRNAseq) data generated by [Smart-seq2 assays](https://www.nature.com/articles/nmeth.2639). The pipeline's workflow is written in WDL, is freely available in the [WARP repository](https://github.com/broadinstitute/warp/blob/develop/pipelines/skylab/smartseq2_single_nucleus_multisample/MultiSampleSmartSeq2SingleNucleus.wdl) on GitHub, and can be run by any compliant WDL runner (e.g. [Crowmell](https://github.com/broadinstitute/cromwell)).  

The workflow processes multiple cells by importing and running the [Smart-seq2 Single Nucleus workflow](https://github.com/broadinstitute/warp/blob/develop/pipelines/skylab/smartseq2_single_nucleus/SmartSeq2SingleNucleus.wdl) for each cell and then merging the resulting count matrices (Loom format) into a single matrix that includes exonic and intronic counts, as well as quality metrics.

Full details about the Smart-seq2 pipeline can be read in the [Smart-seq2 Single Sample Overivew](../Smart-seq2_Single_Nucleus_Pipeline/) in WARP documentation.

## Inputs

An [example configuration file (JSON)](https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/smartseq2_single_nucleus_multisample/human_single_example.json) is available to test the pipeline.


### Sample and Reference Inputs

The workflow sample and reference inputs are identical to those specified in the [Smart-seq2 Single Nucleus Overview](../Smart-seq2_Single_Nucleus_Pipeline/) with the exception of the additional inputs specified in the table below. 

| Input name | Input Description | Input Type |
| --- | --- | --- |
| fastq1_input_files | Cloud locations for each read1 file | Array of strings | 
| fastq2_input_files | Optional cloud locations for each read2 file  |Array of strings |
| input_ids | Unique identifiers or names for each cell; can be a UUID or human-readable name | Array of strings |
| input_names | Optional unique identifiers/names to further describe each cell. If `input_id` is a UUID, the `input_name` could be used as a human-readable identifier | String |
| batch_id | Identifier for the batch of multiple samples | String |
| batch_name | Optional string to describe the batch or biological sample | String |
| input_name_metadata_field | Optional input describing, when applicable, the metadata field containing the `input_names` | String |
| input_id_metadata_field | Optional string describing, when applicable, the metadata field containing the `input_ids` | String |
| `project_id` | Optional project identifier; usually a number | String |
| `project_name` | Optional project identifier; usually a human-readable name | String |
| `library` | Optional description of the sequencing method or approach | String |
| `organ` | Optional description of the organ from which the cells were derived | String |
| `species` | Optional description of the species from which the cells were derived | String |


### Smart-seq2 Multi-Sample Task Summary

| Task name and link to WDL | Description | Software | Tools |
| --- | --- | --- | --- |
| [CheckInputs.checkInputArrays](https://github.com/broadinstitute/warp/blob/develop/tasks/skylab/CheckInputs.wdl) | Checks the inputs and initiates the per cell processing | Bash | NA | 
| [single_nucleus_run.SmartSeq2SingleNucleus](https://github.com/broadinstitute/warp/blob/develop/pipelines/skylab/smartseq2_single_nucleus/SmartSeq2SingleNucleus.wdl) | Calls the Smart-seq2 Single Nucleus workflow to perform alignment, deduplication, quality metric calculating, and exon and intron counting. Produces a final matrix (Loom format) with counts and metrics | See the [Smart-seq2 Single Nucleus Overview](../Smart-seq2_Single_Nucleus_Pipeline/) for details about task software | See the [Smart-seq2 Single Nucleus Overview](../Smart-seq2_Single_Nucleus_Pipeline/) for details about task tools |
| [LoomUtils.AggregateSmartSeq2Loom](https://github.com/broadinstitute/warp/blob/develop/tasks/skylab/LoomUtils.wdl) | Aggregates the matrix files (Loom format) for each sample to produce one final Loom output | Python 3 | Custom script: [ss2_loom_merge.py](https://github.com/broadinstitute/warp/blob/develop/dockers/skylab/loom-output/ss2_loom_merge.py) | 


### Outputs

The following table lists the outputs of the Multi-snSS2 workflow.

| Output file name | Output Description | Output Type |
| --- | --- | --- |
| bam_files | Array of genome-aligned BAM files (one for each cell) generated with Star  | Array [BAM]|
| loom_output | Cell-by-gene matrix in Loom format containing intronic and exonic counts for every cell | File |
| exon_intron_count_files | Array of TXT files (one per cell) that contain intronic and exonic counts | Array [TXT]| 

The final count matrix (Loom) is an aggregate of all the individual matrix files generated using the [Smart-seq2 Single Nucleus workflow](https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/smartseq2_single_nucleus/SmartSeq2SingleNucleus.wdl). 

The aggregated matrix filename contains the `batch_id` prefix, which is the string specified in the input configuration. The `batch_id` is also set as a global attribute in the Loom.

The individual sample Loom files are described in the [Smart-seq2 Single Nucleus Overview](../Smart-seq2_Single_Nucleus_Pipeline/).


## Validation
The Multi-snSS2 pipeline was scientifically validated by the BRAIN Initiatives Cell Census Network (BICCN) 2.0 Whole Mouse Brain Working Group. 

## Versioning

Release information for the Smart-seq2 Single Nucleus Multi-Sample Pipeline can be found in the [changelog](https://github.com/broadinstitute/warp/blob/develop/pipelines/skylab/smartseq2_single_nucleus_multisample/MultiSampleSmartSeq2SingleNucleus.changelog.md). Please note that any major changes to the Smart-seq2 Single Nucleus workflow will be documented in the [Smart-seq2 Single Nucleus changelog](https://github.com/broadinstitute/warp/blob/develop/pipelines/skylab/smartseq2_single_nucleus/SmartSeq2SingleNucleus.changelog.md).

## Have Suggestions?
Help us make our tools better by contacting [Kylee Degatano](mailto:kdegatano@broadinstitute.org) for pipeline-related suggestions or questions.

