<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.5">
<link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Blog Atom Feed">
<link rel="preconnect" href="https://www.google-analytics.com">
<script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-181048107-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview")</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><title data-react-helmet="true">Imputation Overview | WARP</title><meta data-react-helmet="true" name="twitter:card" content="summary_large_image"><meta data-react-helmet="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/Imputation_Pipeline/README"><meta data-react-helmet="true" name="docsearch:language" content="en"><meta data-react-helmet="true" name="docsearch:version" content="current"><meta data-react-helmet="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="Imputation Overview | WARP"><meta data-react-helmet="true" name="description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><meta data-react-helmet="true" property="og:description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><link data-react-helmet="true" rel="shortcut icon" href="/warp/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/Imputation_Pipeline/README"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Imputation_Pipeline/README" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Imputation_Pipeline/README" hreflang="x-default"><link data-react-helmet="true" rel="preconnect" href="https://BH4D9OD16A-dsn.algolia.net" crossorigin="anonymous"><link rel="stylesheet" href="/warp/assets/css/styles.57c5b218.css">
<link rel="preload" href="/warp/assets/js/runtime~main.190adbdd.js" as="script">
<link rel="preload" href="/warp/assets/js/main.fa639502.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#" class="skipToContent_1oUP">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle toggle_3Zt9 react-toggle--disabled"><div class="react-toggle-track" role="button" tabindex="-1"><div class="react-toggle-track-check"><span class="toggle_71bT">üåú</span></div><div class="react-toggle-track-x"><span class="toggle_71bT">üåû</span></div><div class="react-toggle-thumb"></div></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div><div class="searchBox_1Doo"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docs-wrapper docs-doc-page"><div class="docPage_31aa"><button class="clean-btn backToTopButton_35hR" type="button"><svg viewBox="0 0 24 24" width="28"><path d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z" fill="currentColor"></path></svg></button><aside class="docSidebarContainer_3Kbt"><div class="sidebar_15mo"><nav class="menu thin-scrollbar menu_Bmed menuWithAnnouncementBar_2WvA"><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">About WARP</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#">WARP Pipelines</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">CEMBA</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Exome Germline Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">GDC Whole Genome Somatic Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Illumina Genotyping Array</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#" tabindex="0">Imputation</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/Imputation_Pipeline/README">Imputation Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Imputation_Pipeline/references_overview">Creating the Imputation Pipeline&#x27;s Modified 1000 Genomes Reference</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Optimus</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">RNA with UMIs</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Single Cell ATAC</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Slide-seq</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Multi-Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Single Nucleus Multi-Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Ultima Genomics Whole Genome Germline</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Whole Genome Germline Single Sample</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">WDL Best Practices</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">Contribution Guide</a></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_1CGd"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_3E-R"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></aside><main class="docMainContainer_3ufF"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_3FnS"><div class="docItemContainer_33ec"><article><div class="tocCollapsible_1PrD tocMobile_3Hoh"><button type="button" class="clean-btn tocCollapsibleButton_2O1e">On this page</button></div><div class="markdown"><header><h1>Imputation Overview</h1></header><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases?q=Imputation_v1.0.0&amp;expanded=true" target="_blank" rel="noopener noreferrer">Imputation_v1.0.0</a></td><td align="center">August, 2021</td><td align="center"><a href="mailto:ekiernan@broadinstitute.org" target="_blank" rel="noopener noreferrer">Elizabeth Kiernan</a></td><td align="center">Please file GitHub issues in warp or contact <a href="mailto:warp-pipelines-help@broadinstitute.org" target="_blank" rel="noopener noreferrer">the WARP team</a></td></tr></tbody></table><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="introduction-to-the-imputation-pipeline"></a>Introduction to the Imputation pipeline<a class="hash-link" href="#introduction-to-the-imputation-pipeline" title="Direct link to heading">#</a></h2><p>The Imputation pipeline imputes missing genotypes from either a multi-sample VCF or an array of single sample VCFs using a large genomic reference panel. It is based on the <a href="https://imputationserver.readthedocs.io/en/latest/pipeline/" target="_blank" rel="noopener noreferrer">Michigan Imputation Server pipeline</a>. Overall, the pipeline filters, phases, and performs imputation on a multi-sample VCF. It outputs the imputed VCF along with key imputation metrics.</p><p><img src="/warp/assets/images/imputation-fa25d2e8cbb65cb07b1da17932ca6be7.png"></p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="set-up"></a>Set-up<a class="hash-link" href="#set-up" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="workflow-installation-and-requirements"></a>Workflow installation and requirements<a class="hash-link" href="#workflow-installation-and-requirements" title="Direct link to heading">#</a></h3><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/broad/arrays/imputation/Imputation.wdl" target="_blank" rel="noopener noreferrer">Imputation workflow</a> is written in the Workflow Description Language (WDL) and can be deployed using a WDL-compatible execution engine like  <a href="https://github.com/broadinstitute/cromwell" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH compliant, flexible workflow management system that supports multiple computing platforms.</p><p>To identify the latest workflow version and release notes, please see the Imputation workflow <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/broad/arrays/imputation/Imputation.changelog.md" target="_blank" rel="noopener noreferrer">changelog</a>. </p><p>The latest release of the workflow, example data, and dependencies are available from the WARP releases page. To discover and search releases, use the WARP command-line tool <a href="https://github.com/broadinstitute/warp/tree/develop/wreleaser" target="_blank" rel="noopener noreferrer">Wreleaser</a>.</p><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Try the Imputation pipeline in Terra</h5></div><div class="admonition-content"><p>You can run the pipeline in the <a href="https://app.terra.bio/#workspaces/warp-pipelines/Imputation" target="_blank" rel="noopener noreferrer">Imputation workspace</a> on <a href="https://app.terra.bio" target="_blank" rel="noopener noreferrer">Terra</a>, a cloud-optimized scalable bioinformatics platform. The workspace contains a preconfigured workflow, example inputs, instructions, and cost-estimates. </p></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="input-descriptions"></a>Input descriptions<a class="hash-link" href="#input-descriptions" title="Direct link to heading">#</a></h3><p>The table below describes each of the Imputation pipeline inputs. The workflow requires <strong>either</strong> a multi-sample VCF or an array of single sample VCFs. These samples must be from the same species and genotyping chip.</p><p><strong>You must have two or more samples to run the pipeline.</strong> </p><p>However, the pipeline is cost-optimized for between 100 and 1,000 samples. After 1,000 samples, the cost per sample no longer decreases (see the <a href="#price-estimates">Price estimates</a> section). </p><p>For examples of how to specify each input in a configuration file, as well as cloud locations for different example input files, see the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/broad/arrays/imputation/example_inputs.json" target="_blank" rel="noopener noreferrer">example input configuration file (JSON)</a>.</p><table><thead><tr><th>Input name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>ChunkLength</td><td>Size of chunks; default set to 25 MB.</td><td>Int</td></tr><tr><td>chunkOverlaps</td><td>Padding adding to the beginning and end of each chunk to reduce edge effects; default set 5 MB.</td><td>Int</td></tr><tr><td>multi_sample_vcf</td><td>Merged VCF containing multiple samples; can also use an array of individual VCFs.</td><td>File</td></tr><tr><td>multi_sample_vcf_index</td><td>Merged index for the merged VCF; can also use an array of index files if using an array of VCFs.</td><td>Index</td></tr><tr><td>single_sample_vcfs</td><td>Array of VCFs, one for each sample; can be used in lieu of a merged VCF containing all samples.</td><td>Array of files</td></tr><tr><td>single_sample_vcf_indices</td><td>Array of indices, one for each sample; can be used in lieu of a merged index for a multi-sample VCF.</td><td>Array of index files</td></tr><tr><td>perform_extra_qc_steps</td><td>Boolean to indicate if additional QC steps should be performed before imputing; when true, sites with call rates below 95% or low Hardy Weinberg Equilibrium (HWE) p-value are removed before imputation. Default is set to false.</td><td>Boolean</td></tr><tr><td>optional_qc_max_missing</td><td>Optional float used for the additional QC steps that sets a max threshold for the maximum rate of missing data allowed for sites; default set to 0.05.</td><td>Float</td></tr><tr><td>optional_qc_hwe</td><td>Optional HWE p-value when performing additional QC steps; default set to 0.000001.</td><td>Float</td></tr><tr><td>ref_dict</td><td>Reference dictionary.</td><td>File</td></tr><tr><td>contigs</td><td>Array of strings defining which contigs (chromsomes) should be used for the reference panel.</td><td>Array of strings</td></tr><tr><td>reference_panel_path</td><td>Path to the cloud storage containing the reference panel files for all contigs.</td><td>String</td></tr><tr><td>genetics_maps_eagle</td><td>Genetic map file for phasing.</td><td>File</td></tr><tr><td>output_callset_name</td><td>Output callset name.</td><td>String</td></tr><tr><td>split_output_to_single_sample</td><td>Boolean to split out the final combined VCF to individual sample VCFs; set to false by default.</td><td>Boolean</td></tr><tr><td>merge_ssvcf_mem_gb</td><td>Memory allocation for MergeSingleSampleVcfs (in GB).</td><td>Int</td></tr><tr><td>frac_well_imputed_threshold</td><td>Threshold for the fraction of well-imputed sites; default set to 0.9.</td><td>Float</td></tr><tr><td>chunks_fail_threshold</td><td>Maximum threshold for the number of chunks allowed to fail; default set to 1.</td><td>Float</td></tr><tr><td>vcf_suffix</td><td>File extension used for the VCF in the reference panel.</td><td>String</td></tr><tr><td>vcf_index_suffix</td><td>File extension used for the VCF index in the reference panel.</td><td>String</td></tr><tr><td>bcf_suffix</td><td>File extension used for the BCF in the reference panel.</td><td>String</td></tr><tr><td>bcf_index_suffix</td><td>File extension used for the BCF index in the reference panel.</td><td>String</td></tr><tr><td>m3vcf_suffix</td><td>File extension used for the M3VCF in the reference panel.</td><td>String</td></tr></tbody></table><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="imputation-reference-panel"></a>Imputation reference panel<a class="hash-link" href="#imputation-reference-panel" title="Direct link to heading">#</a></h3><p>The Imputation workflow&#x27;s reference panel files are hosted in a <a href="https://console.cloud.google.com/storage/browser/gcp-public-data--broad-references/hg19/v0/1000G_reference_panel;tab=objects?prefix=&amp;forceOnObjectsSortingFiltering=false" target="_blank" rel="noopener noreferrer">public Google Bucket</a>. For the cloud-path (URI) to the files, see the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/broad/arrays/imputation/example_inputs.json" target="_blank" rel="noopener noreferrer">example input configuration</a>.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="generation-of-the-modified-1000-genomes-reference"></a>Generation of the modified 1000 Genomes reference<a class="hash-link" href="#generation-of-the-modified-1000-genomes-reference" title="Direct link to heading">#</a></h4><p>Initial tests of the Imputation workflow followed by assessments of polygenic risk score revealed that disease risk scores were lower when computed from imputed array data as opposed to whole-genome sequencing data. This was found to be due to incorrectly genotyped sites in the 1000G reference panel. As a result, the 1000G reference files were modified for the Imputation pipeline as described in the <a href="/warp/docs/Pipelines/Imputation_Pipeline/references_overview">references overview</a>. You can view the original, unmodified 1000G VCFs <a href="https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/" target="_blank" rel="noopener noreferrer">here</a>.</p><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>X-chromosome not imputed</h5></div><div class="admonition-content"><p>Currently, the pipeline does not perform imputation on the X-chromosome and no reference panels are needed for the X-chromosome. Any sites identified on the X-chromosome after array analysis are merged back into the VCF after the imputation steps. </p></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="workflow-tasks-and-tools"></a>Workflow tasks and tools<a class="hash-link" href="#workflow-tasks-and-tools" title="Direct link to heading">#</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/broad/arrays/imputation/Imputation.wdl" target="_blank" rel="noopener noreferrer">Imputation workflow</a> imports a series of tasks from the ImputationTasks WDL, which is hosted in the Broad <a href="https://github.com/broadinstitute/warp/tree/develop/tasks/broad" target="_blank" rel="noopener noreferrer">tasks library</a>. The table below describes each workflow task, including the task name, tools, relevant software and non-default parameters. </p><table><thead><tr><th>Task name (alias) in WDL</th><th>Tool</th><th>Software</th><th>Description</th></tr></thead><tbody><tr><td>MergeSingleSampleVcfs</td><td>merge</td><td><a href="http://samtools.github.io/bcftools/bcftools.html" target="_blank" rel="noopener noreferrer">bcftools</a></td><td>If an array of single sample VCFs are pipeline input, the task merges them into a single VCF.</td></tr><tr><td>SetIDs (SetIdsVcfToImpute)</td><td>annotate</td><td><a href="http://samtools.github.io/bcftools/bcftools.html" target="_blank" rel="noopener noreferrer">bcftools</a>, bash</td><td>Adds variant IDs to the combined input VCF to create a new VCF. Sorts the alleles for a given variant ID so that REF:ALT is lexicographically consistent across IDs.</td></tr><tr><td>ExtractIDs (ExtractIdsVcfToImpute)</td><td>query</td><td><a href="http://samtools.github.io/bcftools/bcftools.html" target="_blank" rel="noopener noreferrer">bcftools</a></td><td>Extracts the variant IDs from the SortIDs output VCF to a new ‚Äú.ids‚Äù file so that any missing variants can be added back to the final VCF after imputation.</td></tr><tr><td>CountSamples</td><td>query</td><td><a href="http://samtools.github.io/bcftools/bcftools.html" target="_blank" rel="noopener noreferrer">bcftools</a></td><td>Uses the merged input VCF file to count the number of samples and output a TXT file containing the count.</td></tr><tr><td>CalculateChromsomeLength</td><td>grep</td><td>bash</td><td>Reads chromosome lengths from the reference dictionary and uses these to generate chunk intervals for the GenerateChunk task.</td></tr><tr><td>GenerateChunk</td><td>SelectVariants</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Performs site filtering by selecting SNPs only and excluding InDels, removing duplicate sites from the VCF, selecting biallelic variants, excluding symbolic/mixed variants, and removing sites with a maximum fraction of samples with no-call genotypes greater than 0.1. Also subsets to only a specified chunk of the genome.</td></tr><tr><td>OptionalQCSites</td><td>---</td><td><a href="http://vcftools.sourceforge.net/" target="_blank" rel="noopener noreferrer">vcftools</a>, <a href="http://samtools.github.io/bcftools/bcftools.html" target="_blank" rel="noopener noreferrer">bcftools</a></td><td>If the boolean extra_qc_steps is true, performs additional QC steps; excludes sites with more than 95% missing data and assesses sites for Hardy Weinberg Equilibrium, excluding any site with a p-value less than 0.000001.</td></tr><tr><td>CountVariantsInChunks</td><td>CountVariants</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Counts variants in the filtered VCF file; Returns the number of chunks in the array and in the reference file.</td></tr><tr><td>CheckChunks</td><td>convert, index</td><td><a href="http://samtools.github.io/bcftools/bcftools.html" target="_blank" rel="noopener noreferrer">bcftools</a></td><td>Confirms that there are no chunks where less than 3 sites or less than 50% of the sites in the array are also in the reference panel; if valid, creates a new VCF output.</td></tr><tr><td>PhaseVariantsEagle</td><td>eagle</td><td><a href="https://alkesgroup.broadinstitute.org/Eagle/Eagle_manual.html" target="_blank" rel="noopener noreferrer">Eagle2</a></td><td>Performs phasing on the filtered, validated VCF using the phased reference panel; allows for REF/ALT swaps</td></tr><tr><td>Minimac4</td><td>Minimac4</td><td><a href="https://genome.sph.umich.edu/wiki/Minimac4_Documentation" target="_blank" rel="noopener noreferrer">minimac4</a>, <a href="/warp/docs/Pipelines/Imputation_Pipeline/[bcftools](http://samtools.github.io/bcftools/bcftools.html)">bcftools</a></td><td>Performs imputation on the prephased VCF; parameterized to include variants that were genotyped but NOT in the reference panel and to specify a minRatio of 0.00001.</td></tr><tr><td>AggregateImputationQCMetrics</td><td>---</td><td>R</td><td>Uses an R script to take calculate metrics from minimac4 output info file, including total sites, total sites with variants, and sites with an <a href="https://genome.sph.umich.edu/wiki/Minimac3_Info_File" target="_blank" rel="noopener noreferrer">R2 metric</a> of 0.3 (total_sites_r2_gt_0.3); adds the metrics to a new TSV output.</td></tr><tr><td>UpdateHeader</td><td>UpdateVCFSequenceDictionary</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Updates the header of the imputed VCF; adds contig lengths</td></tr><tr><td>SeparateMultiallelics</td><td>norm</td><td><a href="http://samtools.github.io/bcftools/bcftools.html" target="_blank" rel="noopener noreferrer">bcftools</a></td><td>Splits multiallelic sites in the imputed VCF into biallelic records.</td></tr><tr><td>RemoveSymbolicAlleles</td><td>SelectVariants</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Removes SYMBOLIC alleles from the output VCF of the SeparateMultiallelics.</td></tr><tr><td>SetIds</td><td>annotate, index</td><td><a href="http://samtools.github.io/bcftools/bcftools.html" target="_blank" rel="noopener noreferrer">bcftools</a></td><td>Sorts the alleles in the variant ID from the RemoveSymbolicAllele output VCF so that REF:ALT is lexicographically consistent across IDs.</td></tr><tr><td>GatherVcfs</td><td>GatherVCFs</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Gathers the array of imputed VCFs and merges them into one VCF output.</td></tr><tr><td>ExtractIDs</td><td>query</td><td><a href="http://samtools.github.io/bcftools/bcftools.html" target="_blank" rel="noopener noreferrer">bcftools</a></td><td>Extracts the variant IDs from the imputed VCF.</td></tr><tr><td>FindSitesUniqueToFileTwoOnly</td><td>---</td><td>Ubuntu</td><td>Uses the IDs extracted from imputed VCF and those extracted from original VCF to identify missing variant sites from the original VCF; outputs the IDs to a file.</td></tr><tr><td>SelectVariantsByIds</td><td>SelectVariants</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Selects from the original input VCF any sites which were not included in the imputed VCF.</td></tr><tr><td>RemoveAnnotations</td><td>annotate</td><td><a href="http://samtools.github.io/bcftools/bcftools.html" target="_blank" rel="noopener noreferrer">bcftools</a></td><td>Removes the FORMAT and INFO annotations from the new VCF generated by the SelectVariantsbyIds task that contains the missing variants.</td></tr><tr><td>InterleaveVariants</td><td>MergeVCFs</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Combines the missing variants from the original VCF and the imputed variants into a new VCF.</td></tr><tr><td>MergeImputationQCMetrics</td><td>---</td><td>R</td><td>Uses an R script to calculate the fraction of well-imputed sites and outputs them to a TXT file; the fraction of &quot;well-imputed&quot; sites is based on the minimac reported R2 metric, with R2&gt;0.3 being &quot;well-imputed.&quot; Since homomorphic sites lead to an R2 value of 0, we report the fraction of sites with any variation which are well-imputed in addition to the fraction of total sites.</td></tr><tr><td>StoreChunksInfo</td><td>---</td><td>R</td><td>Uses an R script to record the coordinates of each imputation chunk, number of sites in the original array, and number of sites in the original array which are also in the reference panel, for each imputation chunk.</td></tr><tr><td>SplitMultiSampleVcf</td><td>split</td><td><a href="http://samtools.github.io/bcftools/bcftools.html" target="_blank" rel="noopener noreferrer">bcftools</a></td><td>If boolean is set to true, will split the interleave variants VCF into single sample VCFs.</td></tr></tbody></table><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="workflow-outputs"></a>Workflow outputs<a class="hash-link" href="#workflow-outputs" title="Direct link to heading">#</a></h2><p>The table below summarizes the workflow outputs. If running the workflow on Cromwell, these outputs are found in the task execution directory.</p><table><thead><tr><th>Output name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>imputed_single_sample_vcfs</td><td>Array of imputed single sample VCFs from the SplitMultiSampleVcf task.</td><td>Array</td></tr><tr><td>imputed_single_sample_vcf_indices</td><td>Array of indices for the imputed VCFs from the SplitMultiSampleVcf task</td><td>Array</td></tr><tr><td>imputed_multisample_vcf</td><td>VCF from the InterleaveVariants task; contains imputed variants as well as missing variants from the input VCF.</td><td>VCF</td></tr><tr><td>imputed_multisample_vcf_index</td><td>Index file for VCF from the InterleaveVariants task.</td><td>Index</td></tr><tr><td>aggregated_imputation_metrics</td><td>Aggregated QC metrics from the MergeImputationQcMetrics task; reports the fraction of sites well-imputed and outputs to TXT file; fraction of &quot;well-imputed&quot; is based on the minimac reported R2 metric, with R2&gt;0.3 being &quot;well-imputed.&quot; Since homomorphic sites lead to an R2 value of 0, we report the fraction of sites with any variation which are well-imputed in addition to the fraction of total sites.</td><td>TXT</td></tr><tr><td>chunks_info</td><td>TSV from StoreChunksInfo task; contains the chunk intervals as well as the number of variants in the array.</td><td>TSV</td></tr><tr><td>failed_chunks</td><td>File with the failed chunks from the StoreChunksInfo task.</td><td>File</td></tr><tr><td>n_failed_chunks</td><td>File with the number of failed chunks from the StoreChunksInfo task.</td><td>File</td></tr></tbody></table><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="important-notes"></a>Important notes<a class="hash-link" href="#important-notes" title="Direct link to heading">#</a></h2><ul><li>Runtime parameters are optimized for Broad&#x27;s Google Cloud Platform implementation.</li></ul><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="price-estimates"></a>Price estimates<a class="hash-link" href="#price-estimates" title="Direct link to heading">#</a></h2><p>The pipeline is cost-optimized for between 100 and 1,000 samples, where the cost per sample continues to decrease until 1,000 samples are run. Cost estimates per sample are provided below:</p><table><thead><tr><th>Cohort size ( # samples)</th><th>Cost per sample ($)</th></tr></thead><tbody><tr><td>1</td><td>8</td></tr><tr><td>10</td><td>0.8</td></tr><tr><td>100</td><td>0.11</td></tr><tr><td>1000</td><td>0.024</td></tr><tr><td>13.5 K</td><td>0.025</td></tr></tbody></table><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="contact-us"></a>Contact us<a class="hash-link" href="#contact-us" title="Direct link to heading">#</a></h2><p>Help us make our tools better by contacting <a href="mailto:warp-pipelines-help@broadinstitute.org" target="_blank" rel="noopener noreferrer">the WARP team</a> for pipeline-related suggestions or questions.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="licensing"></a>Licensing<a class="hash-link" href="#licensing" title="Direct link to heading">#</a></h2><p>Copyright Broad Institute, 2020 | BSD-3</p><p>The workflow script is released under the <strong>WDL open source code license (BSD-3)</strong> (full license text at <a href="https://github.com/broadinstitute/warp/blob/master/LICENSE" target="_blank" rel="noopener noreferrer">https://github.com/broadinstitute/warp/blob/master/LICENSE</a>). However, please note that the programs it calls may be subject to different licenses. Users are responsible for checking that they are authorized to run all programs before running this script.</p><ul><li><a href="https://github.com/broadinstitute/gatk/blob/master/LICENSE.TXT" target="_blank" rel="noopener noreferrer">GATK</a></li><li><a href="https://github.com/broadinstitute/picard/blob/master/LICENSE.txt" target="_blank" rel="noopener noreferrer">Picard</a></li><li><a href="https://alkesgroup.broadinstitute.org/Eagle/#x1-340007" target="_blank" rel="noopener noreferrer">Eagle2</a></li><li><a href="https://github.com/statgen/Minimac4/blob/master/LICENSE" target="_blank" rel="noopener noreferrer">minimac4</a></li><li><a href="https://github.com/samtools/bcftools/blob/develop/LICENSE" target="_blank" rel="noopener noreferrer">bcftools</a>
-<a href="http://vcftools.sourceforge.net/license.html" target="_blank" rel="noopener noreferrer">vcftools</a></li></ul></div><footer class="docusaurus-mt-lg"><div class="row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/Imputation_Pipeline/README.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_2_ui" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_13-_">Last updated on <b><time datetime="2023-03-21T17:26:48.000Z">3/21/2023</time></b> by <b>Kaylee Mathews</b></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/Illumina_genotyping_array_spec"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">¬´ VCF Overview: Illumina Genotyping Array</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/warp/docs/Pipelines/Imputation_Pipeline/references_overview"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Creating the Imputation Pipeline&#x27;s Modified 1000 Genomes Reference ¬ª</div></a></div></nav></div></div><div class="col col--3"><div class="tableOfContents_35-E thin-scrollbar"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-imputation-pipeline" class="table-of-contents__link">Introduction to the Imputation pipeline</a></li><li><a href="#set-up" class="table-of-contents__link">Set-up</a><ul><li><a href="#workflow-installation-and-requirements" class="table-of-contents__link">Workflow installation and requirements</a></li><li><a href="#input-descriptions" class="table-of-contents__link">Input descriptions</a></li><li><a href="#imputation-reference-panel" class="table-of-contents__link">Imputation reference panel</a></li></ul></li><li><a href="#workflow-tasks-and-tools" class="table-of-contents__link">Workflow tasks and tools</a></li><li><a href="#workflow-outputs" class="table-of-contents__link">Workflow outputs</a></li><li><a href="#important-notes" class="table-of-contents__link">Important notes</a></li><li><a href="#price-estimates" class="table-of-contents__link">Price estimates</a></li><li><a href="#contact-us" class="table-of-contents__link">Contact us</a></li><li><a href="#licensing" class="table-of-contents__link">Licensing</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item"><span>GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_3J9K"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></span></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright ¬© 2023 Copyright ¬© Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.190adbdd.js"></script>
<script src="/warp/assets/js/main.fa639502.js"></script>
</body>
</html>