<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Deprecated_Pipelines/Smart-seq2_Single_Sample_Pipeline/README">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">Smart-seq2 Single Sample Overview | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/Smart-seq2_Single_Sample_Pipeline/README"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Smart-seq2 Single Sample Overview | WARP"><meta data-rh="true" name="description" content="9/12/2014"><meta data-rh="true" property="og:description" content="9/12/2014"><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/Smart-seq2_Single_Sample_Pipeline/README"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Smart-seq2_Single_Sample_Pipeline/README" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Smart-seq2_Single_Sample_Pipeline/README" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.91c3cfb7.js" as="script">
<link rel="preload" href="/warp/assets/js/main.affc3f00.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">CEMBA</a><button aria-label="Toggle the collapsible sidebar category &#x27;CEMBA&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/README">GDC Whole Genome Somatic Single Sample</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Single_Cell_ATAC_Seq_Pipeline/README">Single Cell ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Cell ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Multi_Sample_Pipeline/README">Smart-seq2 Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible menu__list-item-collapsible--active"><a class="menu__link menu__link--sublist menu__link--active" aria-current="page" aria-expanded="true" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Sample_Pipeline/README">Smart-seq2 Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Deprecated_Pipelines/Smart-seq2_Single_Sample_Pipeline/Loom_schema">What&#x27;s in the Smart-seq2 Single Sample Pipeline Loom File?</a></li></ul></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">Deprecated Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Smart-seq2 Single Sample</span><meta itemprop="position" content="2"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Smart-seq2 Single Sample Overview</h1><div class="theme-admonition theme-admonition-warning alert alert--danger admonition_LlT9"><div class="admonitionHeading_tbUL"><span class="admonitionIcon_kALy"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M5.05.31c.81 2.17.41 3.38-.52 4.31C3.55 5.67 1.98 6.45.9 7.98c-1.45 2.05-1.7 6.53 3.53 7.7-2.2-1.16-2.67-4.52-.3-6.61-.61 2.03.53 3.33 1.94 2.86 1.39-.47 2.3.53 2.27 1.67-.02.78-.31 1.44-1.13 1.81 3.42-.59 4.78-3.42 4.78-5.56 0-2.84-2.53-3.22-1.25-5.61-1.52.13-2.03 1.13-1.89 2.75.09 1.08-1.02 1.8-1.86 1.33-.67-.41-.66-1.19-.06-1.78C8.18 5.31 8.68 2.45 5.05.32L5.03.3l.02.01z"></path></svg></span>danger</div><div class="admonitionContent_S0QG"><p>9/12/2014</p><p>We are deprecating the Smart-seq2 Single Sample Pipeline. Although the code will continue to be available, we are no longer supporting it. For an alternative, see the <a href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README">Smart-seq2 Single Nucleus Multi Sample workflow</a>.</p></div></div><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">smartseq2_v5.1.20</a></td><td align="center">February, 2024</td><td align="center">Elizabeth Kiernan</td><td align="center">Please <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">file an issue in WARP</a>.</td></tr></tbody></table><p><img loading="lazy" src="/warp/assets/images/smartseq_image-4574ae53d1ef76e202eac615f71940a8.png" width="1854" height="618" class="img_ev3q"></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="introduction-to-the-smart-seq2-single-sample-pipeline">Introduction to the Smart-seq2 Single Sample Pipeline<a href="#introduction-to-the-smart-seq2-single-sample-pipeline" class="hash-link" aria-label="Direct link to Introduction to the Smart-seq2 Single Sample Pipeline" title="Direct link to Introduction to the Smart-seq2 Single Sample Pipeline">​</a></h2><p>The Smart-seq2 Single Sample Pipeline (SS2) is designed by the <a href="https://data.humancellatlas.org/about" target="_blank" rel="noopener noreferrer">Data Coordination Platform</a> of the <a href="https://www.humancellatlas.org/" target="_blank" rel="noopener noreferrer">Human Cell Atlas</a> to process single-cell RNAseq (scRNAseq) data generated by <a href="https://www.nature.com/articles/nmeth.2639" target="_blank" rel="noopener noreferrer">Smart-seq2 assays</a>. The pipeline&#x27;s workflow is written in WDL, is freely available on <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_sample/SmartSeq2SingleSample.wdl" target="_blank" rel="noopener noreferrer">GitHub</a>, and can be run by any compliant WDL runner (e.g. <a href="https://github.com/broadinstitute/cromwell" target="_blank" rel="noopener noreferrer">Crowmell</a>). The pipeline is designed to process scRNA-seq data from an individual cell (to process multiple cells, see the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_multisample/MultiSampleSmartSeq2.wdl" target="_blank" rel="noopener noreferrer">Smart-seq2_Multi Sample</a> pipeline).</p><p>The SS2 workflow is divided into two parts which run in parallel. In part one, the workflow aligns reads to the genome and performs quality control on genome-aligned BAMs. In part two, it aligns reads to the transcriptome and generates gene expression estimates from transcriptome-aligned BAMs. The pipeline returns reads and expression estimates in BAM, and read counts and QC metrics in Loom and CSV file format.</p><p>The pipeline has been validated to support <a href="https://docs.google.com/document/d/1MonsTG8UnROHZ_XpulrSZNTxO988KEH6T6h45plFYQg/edit#heading=h.ixoqmhbabdvh" target="_blank" rel="noopener noreferrer">human</a> and <a href="https://docs.google.com/document/d/12zGTFROrcXEByt9z0h06qjSqb9vWutn28Tx6YiND1Ds/edit" target="_blank" rel="noopener noreferrer">mouse</a>, stranded or unstranded, paired- or single-end, and plate- or fluidigm-based Smart-seq2 datasets. Read more in the <a href="#validation">validation section</a>.</p><div class="theme-admonition theme-admonition-tip alert alert--success admonition_LlT9"><div class="admonitionHeading_tbUL"><span class="admonitionIcon_kALy"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Want to use the Smart-seq2 pipeline for your publication?</div><div class="admonitionContent_S0QG"><p>Check out the <a href="/warp/docs/Deprecated_Pipelines/Smart-seq2_Multi_Sample_Pipeline/smart-seq2.methods">Smart-seq2 Publication Methods</a> to get started!</p></div></div><h2 class="anchor anchorWithStickyNavbar_LWe7" id="quick-start-table">Quick Start Table<a href="#quick-start-table" class="hash-link" aria-label="Direct link to Quick Start Table" title="Direct link to Quick Start Table">​</a></h2><table><thead><tr><th>Pipeline Features</th><th>Description</th><th>Source</th></tr></thead><tbody><tr><td>Assay Type</td><td>Smart-seq2 Single Sample</td><td><a href="https://www.nature.com/articles/nprot.2014.006" target="_blank" rel="noopener noreferrer">Smart-seq2</a></td></tr><tr><td>Overall Workflow</td><td>Quality control and transcriptome quantification</td><td>Code available from <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_sample/SmartSeq2SingleSample.wdl" target="_blank" rel="noopener noreferrer">GitHub</a></td></tr><tr><td>Workflow Language</td><td>WDL</td><td><a href="https://github.com/openwdl/wdl" target="_blank" rel="noopener noreferrer">openWDL</a></td></tr><tr><td>Genomic Reference Sequence (for validation)</td><td>GRCh38 human genome primary sequence and M21 (GRCm38.p6) mouse genome primary sequence</td><td>GENCODE <a href="https://www.gencodegenes.org/human/release_27.html" target="_blank" rel="noopener noreferrer">human reference files</a> and <a href="https://www.gencodegenes.org/mouse/release_M21.html" target="_blank" rel="noopener noreferrer">mouse reference files</a></td></tr><tr><td>Transcriptomic Reference Annotation (for validation)</td><td>V27 GENCODE human transcriptome and M21 mouse transcriptome</td><td>GENCODE <a href="ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_27/gencode.v27.annotation.gtf.gz" target="_blank" rel="noopener noreferrer">human GTF</a> and <a href="ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M21/gencode.vM21.annotation.gff3.gz" target="_blank" rel="noopener noreferrer">mouse GTF</a></td></tr><tr><td>Aligner</td><td>HISAT2 (v.2.1.0)</td><td><a href="https://www.nature.com/articles/s41587-019-0201-4" target="_blank" rel="noopener noreferrer">Kim, et al.,2019</a></td></tr><tr><td>QC Metrics</td><td>Picard (v.2.26.10)</td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Broad Institute</a></td></tr><tr><td>Transcript Quantification</td><td>Utilities for processing large-scale single cell datasets</td><td><a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-323" target="_blank" rel="noopener noreferrer">RSEM v.1.3.0</a></td></tr><tr><td>Data Input File Format</td><td>File format in which sequencing data is provided</td><td><a href="https://academic.oup.com/nar/article/38/6/1767/3112533" target="_blank" rel="noopener noreferrer">FASTQ</a></td></tr><tr><td>Data Output File Formats</td><td>File formats in which Smart-seq2 output is provided</td><td><a href="http://samtools.github.io/hts-specs/" target="_blank" rel="noopener noreferrer">BAM</a>, Loom (generated with <a href="http://loompy.org/" target="_blank" rel="noopener noreferrer">Loompy v.3.0.6)</a>, CSV (QC metrics and counts)</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="set-up">Set-Up<a href="#set-up" class="hash-link" aria-label="Direct link to Set-Up" title="Direct link to Set-Up">​</a></h2><h3 class="anchor anchorWithStickyNavbar_LWe7" id="smart-seq2-single-sample-installation-and-requirements">Smart-seq2 Single Sample Installation and Requirements<a href="#smart-seq2-single-sample-installation-and-requirements" class="hash-link" aria-label="Direct link to Smart-seq2 Single Sample Installation and Requirements" title="Direct link to Smart-seq2 Single Sample Installation and Requirements">​</a></h3><p>The SS2 workflow code can be downloaded by cloning the GitHub <a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer">WARP repository</a>. For the latest release of Smart-seq2, please see the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_sample/SmartSeq2SingleSample.changelog.md" target="_blank" rel="noopener noreferrer">Smart-seq2 Single Sample changelog</a>.</p><p>The workflow is deployed using <a href="https://github.com/broadinstitute/cromwell" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH compliant, flexible workflow management system that supports multiple computing platforms.</p><div class="theme-admonition theme-admonition-tip alert alert--success admonition_LlT9"><div class="admonitionHeading_tbUL"><span class="admonitionIcon_kALy"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Try the SS2 workflow in Terra</div><div class="admonitionContent_S0QG"><p>You can run the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_multisample/MultiSampleSmartSeq2.wdl" target="_blank" rel="noopener noreferrer">Smart-seq2 Multi-Sample workflow</a> (a wrapper for the Single Sample workflow described in this document) in <a href="https://app.terra.bio" target="_blank" rel="noopener noreferrer">Terra</a>, a cloud-based analysis platform. The Terra <a href="https://app.terra.bio/#workspaces/featured-workspaces-hca/HCA%20Smart-seq2%20Multi%20Sample%20Pipeline" target="_blank" rel="noopener noreferrer">Smart-seq2 public workspace</a> is preloaded with the Smart-seq2 workflow, example testing data, and all the required reference data.</p></div></div><h3 class="anchor anchorWithStickyNavbar_LWe7" id="inputs">Inputs<a href="#inputs" class="hash-link" aria-label="Direct link to Inputs" title="Direct link to Inputs">​</a></h3><p>There are two example configuration (JSON) files available to test the SS2 workflow.</p><ul><li><a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_sample/human_single_example.json" target="_blank" rel="noopener noreferrer">human_single_example.json</a>: Configurations for an example single-end human dataset</li><li><a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_sample/mouse_paired_example.json" target="_blank" rel="noopener noreferrer">mouse_paired_example.json</a>: Configurations for an example mouse paired-end dataset</li></ul><h4 class="anchor anchorWithStickyNavbar_LWe7" id="sample-data-input">Sample Data Input<a href="#sample-data-input" class="hash-link" aria-label="Direct link to Sample Data Input" title="Direct link to Sample Data Input">​</a></h4><p>The pipeline is designed for both single- and paired-end reads in the form of FASTQ files. The workflow accepts two FASTQ files for paired-end experiments and one FASTQ file single-end experiments. It processes one sample (cell).</p><ul><li><code>fastq1</code>: forward reads for sample with paired-end sequencing (or reads for sample with single-end sequencing)</li><li><code>fastq2</code>: reverse reads for sample with paired-end sequencing (not applicable for samples with single-end sequencing)</li></ul><p>The workflow will use modified tasks depending if a sample is paired-end or single-end. This is designated with a boolean, as detailed in the following Reference Inputs section.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="additional-reference-inputs">Additional Reference Inputs<a href="#additional-reference-inputs" class="hash-link" aria-label="Direct link to Additional Reference Inputs" title="Direct link to Additional Reference Inputs">​</a></h4><p>The Smart-seq2 Single Sample workflow requires multiple reference indexes. Information on how these indexes are created is found in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/build_indices/README.md" target="_blank" rel="noopener noreferrer">BuildIndices workflow overview</a> documentation. The table below describes the references and each tool that uses the reference as input, when applicable.</p><table><thead><tr><th>Workflow Step</th><th>Reference name</th><th>Reference Description</th><th>Tool</th></tr></thead><tbody><tr><td>All</td><td>input_id</td><td>String containing unique identifier or name for the cell. It can be a human-readable descriptor (i.e. &quot;cell1&quot;) or a UUID</td><td>NA</td></tr><tr><td>All</td><td>input_name</td><td>Optional string that can be used as a name/identifier for the cell</td><td>NA</td></tr><tr><td>All</td><td>output_name</td><td>Output name, can include path</td><td>NA</td></tr><tr><td>All</td><td>paired_end</td><td>A boolean describing if sample is paired-end</td><td>NA</td></tr><tr><td>All</td><td>input_id_metadata_field</td><td>Optional string that describes, when applicable, the metadata field used for <code>input_id</code></td><td>NA</td></tr><tr><td>All</td><td>input_name_metadata_field</td><td>Optional string that describes, when applicable, the metadata field used for the <code>input_name</code></td><td>NA</td></tr><tr><td><strong>Genomic alignment with HISAT2</strong></td><td>hisat2_ref_index</td><td>HISAT2 reference index file in tarball</td><td>HISAT2</td></tr><tr><td></td><td>hisat2_ref_name</td><td>HISAT2 reference index name</td><td>HISAT2</td></tr><tr><td><strong>Picard-generated quality control metrics</strong></td><td>genome_ref_fasta</td><td>Genome reference in fasta format</td><td>Picard</td></tr><tr><td></td><td>gene_ref_flat</td><td><a href="https://software.broadinstitute.org/software/igv/genePred" target="_blank" rel="noopener noreferrer">RefFlat</a> file containing the location of RNA transcripts, exon start sites, etc.</td><td>Picard</td></tr><tr><td></td><td>rrna_intervals</td><td>Ribosomal RNA intervals file</td><td>Picard</td></tr><tr><td></td><td>stranded</td><td>Library strand information for HISAT2; example values include FR (read corresponds to transcript), RF(read corresponds to reverse complement of transcript), or NONE</td><td>Picard</td></tr><tr><td><strong>Transcriptomic alignment with HISAT2</strong></td><td>hisat2_ref_trans_index</td><td>HISAT2 transcriptome index file in tarball</td><td>HISAT2</td></tr><tr><td></td><td>hisat2_ref_trans_name</td><td>HISAT2 transcriptome index file name</td><td>HISAT2</td></tr><tr><td><strong>Gene expression quantification with RSEM</strong></td><td>rsem_ref_index</td><td>RSEM reference index file in tarball</td><td>RSEM</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="running-smart-seq2">Running Smart-seq2<a href="#running-smart-seq2" class="hash-link" aria-label="Direct link to Running Smart-seq2" title="Direct link to Running Smart-seq2">​</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/smartseq2_single_sample/SmartSeq2SingleSample.wdl" target="_blank" rel="noopener noreferrer">SmartSeq2SingleSample.wdl</a> is in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_sample" target="_blank" rel="noopener noreferrer">pipelines/smartseq2_single_sample folder</a> of the WARP repository and implements the workflow by importing individual tasks (written in WDL script) from the WARP <a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/" target="_blank" rel="noopener noreferrer">tasks folder</a>.</p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="smart-seq2-workflow-summary">Smart-seq2 Workflow Summary<a href="#smart-seq2-workflow-summary" class="hash-link" aria-label="Direct link to Smart-seq2 Workflow Summary" title="Direct link to Smart-seq2 Workflow Summary">​</a></h3><p>Overall, the workflow is divided into two parts that are completed after an initial input validation step. Each workflow part comprises tasks (WDL scripts) that are summarized below and are available in the WARP <a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/" target="_blank" rel="noopener noreferrer">tasks folder</a> in GitHub.</p><p><strong>Part 1: Quality Control Tasks</strong></p><ol><li>Aligns reads to the genome with HISAT2 v.2.1.0</li><li>Calculates summary metrics from an aligned BAM using Picard v.2.26.10</li></ol><p><strong>Part 2: Transcriptome Quantification Tasks</strong></p><ol><li>Aligns reads to the transcriptome with HISAT v.2.1.0</li><li>Quantifies gene expression using RSEM v.1.3.0</li></ol><p>The table below provides links to each of the Smart-seq2 Single Sample pipeline&#x27;s task WDL scripts, as well as documentation for each task&#x27;s software tools.</p><p>If you are looking for the parameters used for each task/tool, click on the task link in the table below and see the <code>command {}</code> section of the task WDL. The task&#x27;s Docker image is also specified in the task WDL in the <code># runtime values</code> section as <code>String docker =</code>.</p><table><thead><tr><th>Task</th><th>Tool</th></tr></thead><tbody><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/HISAT2.wdl" target="_blank" rel="noopener noreferrer">HISAT2.HISATPairedEnd</a></td><td><a href="https://www.nature.com/articles/s41587-019-0201-4" target="_blank" rel="noopener noreferrer">HISAT2 software</a></td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/HISAT2.wdl" target="_blank" rel="noopener noreferrer">HISAT2.HISATSingleEnd</a></td><td><a href="https://www.nature.com/articles/s41587-019-0201-4" target="_blank" rel="noopener noreferrer">HISAT2 software</a></td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/Picard.wdl" target="_blank" rel="noopener noreferrer">Picard.CollectMultipleMetrics</a></td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard software</a></td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/Picard.wdl" target="_blank" rel="noopener noreferrer">Picard.CollectRnaMetrics</a></td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard software</a></td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/Picard.wdl" target="_blank" rel="noopener noreferrer">Picard.CollectDuplicationMetrics</a></td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard software</a></td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/HISAT2.wdl" target="_blank" rel="noopener noreferrer">HISAT2.HISAT2RSEM</a></td><td><a href="https://www.nature.com/articles/s41587-019-0201-4" target="_blank" rel="noopener noreferrer">HISAT2 software</a></td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/HISAT2.wdl" target="_blank" rel="noopener noreferrer">HISAT2.HISAT2RSEMSingleEnd</a></td><td><a href="https://www.nature.com/articles/s41587-019-0201-4" target="_blank" rel="noopener noreferrer">HISAT2 software</a></td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/RSEM.wdl" target="_blank" rel="noopener noreferrer">RSEM.RSEMExpression</a></td><td><a href="https://deweylab.github.io/RSEM/README.html#de" target="_blank" rel="noopener noreferrer">RSEM software</a></td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/GroupMetricsOutputs.wdl" target="_blank" rel="noopener noreferrer">GroupMetricsOutputs.GroupQCOutputs</a></td><td><a href="https://sctools.readthedocs.io/en/latest/readme.html" target="_blank" rel="noopener noreferrer">sctools software</a></td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/LoomUtils.wdl" target="_blank" rel="noopener noreferrer">LoomUtils.SmartSeq2LoomOutput</a></td><td><a href="https://www.python.org/download/releases/3.0/" target="_blank" rel="noopener noreferrer">python3 software</a></td></tr></tbody></table><h4 class="anchor anchorWithStickyNavbar_LWe7" id="part-1-quality-control-tasks">Part 1: Quality Control Tasks<a href="#part-1-quality-control-tasks" class="hash-link" aria-label="Direct link to Part 1: Quality Control Tasks" title="Direct link to Part 1: Quality Control Tasks">​</a></h4><h5 class="anchor anchorWithStickyNavbar_LWe7" id="11-align-reads-to-the-genome-using-hisat2">1.1 Align reads to the genome using HISAT2<a href="#11-align-reads-to-the-genome-using-hisat2" class="hash-link" aria-label="Direct link to 1.1 Align reads to the genome using HISAT2" title="Direct link to 1.1 Align reads to the genome using HISAT2">​</a></h5><p>HISAT2 is a fast, cost-efficient alignment tool that can determine the presence of non-transcript sequences and true transcript sequences, taking into account the presence of single-nucleotide polymorphisms (<a href="https://www.nature.com/articles/s41587-019-0201-4" target="_blank" rel="noopener noreferrer">Kim et al.,2019</a>). The Smart-seq2 Single Sample workflow uses the <a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/HISAT2.wdl" target="_blank" rel="noopener noreferrer">HISAT2 task</a> to call HISAT2 and perform graph-based alignment of paired- or single-end reads (in the form of FASTQ files) to a reference genome. This task requires a reference index which can be built using the <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/wdl/build_indices" target="_blank" rel="noopener noreferrer">BuildIndices.wdl</a> documentation. The outputs of the task include a genome-aligned BAM file, a BAM index, and an alignment log file.</p><h5 class="anchor anchorWithStickyNavbar_LWe7" id="12-calculate-summary-metrics-using-picard">1.2 Calculate summary metrics using Picard<a href="#12-calculate-summary-metrics-using-picard" class="hash-link" aria-label="Direct link to 1.2 Calculate summary metrics using Picard" title="Direct link to 1.2 Calculate summary metrics using Picard">​</a></h5><p><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a> is a suite of command line tools used for manipulating high-throughput sequencing data. The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/Picard.wdl" target="_blank" rel="noopener noreferrer">Picard task</a> uses Picard tools to calculate quality control measurements on the HISAT2 genome-aligned BAM file. The task requires a reference genome fasta, a RefFlat gene annotation file, and an RNA intervals file (see the <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/wdl/build_indices" target="_blank" rel="noopener noreferrer">BuildIndices workflow</a> for creating references). The outputs of the task are text and PDF files for each metric.</p><p>The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/Picard.wdl" target="_blank" rel="noopener noreferrer">Picard task</a> generates QC metrics by using three sub-tasks:</p><ul><li><p>CollectMultipleMetrics: calls the  <a href="https://software.broadinstitute.org/gatk/documentation/tooldocs/4.2.6.1/picard_analysis_CollectMultipleMetrics.php" target="_blank" rel="noopener noreferrer">CollectMultipleMetrics</a> tool which uses the aligned BAM file and reference genome fasta to collect metrics on <a href="http://broadinstitute.github.io/picard/picard-metric-definitions.html#AlignmentSummaryMetrics" target="_blank" rel="noopener noreferrer">alignment</a>, <a href="http://broadinstitute.github.io/picard/picard-metric-definitions.html#InsertSizeMetrics" target="_blank" rel="noopener noreferrer">insert size</a>, <a href="https://broadinstitute.github.io/picard/command-line-overview.html#CollectGcBiasMetrics" target="_blank" rel="noopener noreferrer">GC bias</a>, <a href="http://broadinstitute.github.io/picard/picard-metric-definitions.html#BaseDistributionByCycleMetrics" target="_blank" rel="noopener noreferrer">base distribution by cycle</a>, <a href="https://broadinstitute.github.io/picard/command-line-overview.html#QualityScoreDistribution" target="_blank" rel="noopener noreferrer">quality score distribution</a>, <a href="https://broadinstitute.github.io/picard/command-line-overview.html#MeanQualityByCycle" target="_blank" rel="noopener noreferrer">quality distribution by cycle</a>, <a href="http://broadinstitute.github.io/picard/picard-metric-definitions.html#ErrorSummaryMetrics" target="_blank" rel="noopener noreferrer">sequencing artifacts</a>, and <a href="http://broadinstitute.github.io/picard/picard-metric-definitions.html#CollectQualityYieldMetrics.QualityYieldMetrics" target="_blank" rel="noopener noreferrer">quality yield</a>.</p></li><li><p>CollectRnaMetrics: calls the <a href="https://software.broadinstitute.org/gatk/documentation/tooldocs/4.2.6.1/picard_analysis_CollectRnaSeqMetrics.php" target="_blank" rel="noopener noreferrer">CollectRnaSeqMetrics</a> tool which uses the aligned BAM, a RefFlat genome annotation file, and a ribosomal intervals file to produce RNA alignment metrics (metric descriptions are found in the <a href="http://broadinstitute.github.io/picard/picard-metric-definitions.html#RnaSeqMetrics" target="_blank" rel="noopener noreferrer">Picard Metrics Dictionary</a>).</p></li><li><p>CollectDuplicationMetrics: calls the <a href="https://software.broadinstitute.org/gatk/documentation/tooldocs/4.2.6.1/picard_sam_markduplicates_MarkDuplicates.php" target="_blank" rel="noopener noreferrer">MarkDuplicates</a> tool which uses the aligned BAM to identify duplicate reads (output metrics are listed in the <a href="http://broadinstitute.github.io/picard/picard-metric-definitions.html#DuplicationMetrics" target="_blank" rel="noopener noreferrer">Picard Metrics Dictionary</a>).</p></li></ul><h4 class="anchor anchorWithStickyNavbar_LWe7" id="part-2-transcriptome-quantification-tasks">Part 2: Transcriptome Quantification Tasks<a href="#part-2-transcriptome-quantification-tasks" class="hash-link" aria-label="Direct link to Part 2: Transcriptome Quantification Tasks" title="Direct link to Part 2: Transcriptome Quantification Tasks">​</a></h4><h5 class="anchor anchorWithStickyNavbar_LWe7" id="21-align-reads-to-the-transcriptome-using-hisat2">2.1 Align reads to the transcriptome using HISAT2<a href="#21-align-reads-to-the-transcriptome-using-hisat2" class="hash-link" aria-label="Direct link to 2.1 Align reads to the transcriptome using HISAT2" title="Direct link to 2.1 Align reads to the transcriptome using HISAT2">​</a></h5><p>The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/HISAT2.wdl" target="_blank" rel="noopener noreferrer">HISAT2.HISAT2RSEM task</a> uses HISAT2 to align reads to the reference transcriptome. The task requires the hisat2_ref_trans_index file and the sample FASTQ files as input. The output of this task is a transcriptome-aligned BAM file and an alignment log file.</p><h5 class="anchor anchorWithStickyNavbar_LWe7" id="22-quantify-gene-expression-using-rsem">2.2 Quantify gene expression using RSEM<a href="#22-quantify-gene-expression-using-rsem" class="hash-link" aria-label="Direct link to 2.2 Quantify gene expression using RSEM" title="Direct link to 2.2 Quantify gene expression using RSEM">​</a></h5><p><a href="https://deweylab.github.io/RSEM/README.html#de" target="_blank" rel="noopener noreferrer">RSEM</a> is a software package for quantifying transcript abundances (<a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-323" target="_blank" rel="noopener noreferrer">Li and Dewey, 2011</a>). The Smart-seq2 Single Sample workflow uses the <a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/RSEM.wdl" target="_blank" rel="noopener noreferrer">RSEM.RSEMExpression task</a> to calculate expression estimates from a transcriptome-aligned BAM file using the rsem_ref_index file for reference input. The RSEM tool <a href="http://deweylab.biostat.wisc.edu/rsem/rsem-calculate-expression.html" target="_blank" rel="noopener noreferrer">rsem-calculate-expression</a> is used to estimate gene/isoform expression.</p><p>The RSEM task returns the following output files:</p><ul><li>rsem_gene: gene level expression estimates in FPKM, TPM, and counts (&quot;expected_count&quot;)</li><li>rsem_isoform: isoform level expression estimates in FPKM, TPM, and counts (&quot;expected_count&quot;)</li><li>rsem_time: the time consumed by aligning reads</li><li>rsem_cnt: alignment statistics</li><li>rsem_model: RNA-seq model parameters</li><li>rsem_theta: fraction of reads resulting from background</li></ul><p>Only the rsem_gene (TPM and expected_count), rsem_isoform (TPM and expected_count), and rsem_cnt files are used for the final outputs of the Smart-seq2 Single Sample workflow.</p><p>After the Smart-seq2 workflow generates HISAT2, Picard and RSEM metrics, the <a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/GroupMetricsOutputs.wdl" target="_blank" rel="noopener noreferrer">GroupMetricsOutputs task</a> combines most metrics into a &quot;group_results&quot; CSV file array. The workflow uses this array in downstream Loom file generation. The following HISAT2, <a href="http://broadinstitute.github.io/picard/picard-metric-definitions.html#" target="_blank" rel="noopener noreferrer">Picard</a> and <a href="http://deweylab.biostat.wisc.edu/rsem/rsem-calculate-expression.html" target="_blank" rel="noopener noreferrer">RSEM</a> outputs are inputs for this task:</p><ul><li>HISAT2_log_file</li><li>base_call_dist_metrics</li><li>gc_bias_detail_metrics</li><li>pre_adapter_details_metrics</li><li>pre_adapter_summary_metrics</li><li>bait_bias_detail_metrics</li><li>error_summary_metrics</li><li>alignment_summary_metrics</li><li>dedup_metrics</li><li>rna_metrics</li><li>gc_bias_summary_metrics</li><li>HISAT2RSEM_log_file</li><li>RSEMExpression.rsem_cnt</li></ul><h3 class="anchor anchorWithStickyNavbar_LWe7" id="outputs">Outputs<a href="#outputs" class="hash-link" aria-label="Direct link to Outputs" title="Direct link to Outputs">​</a></h3><p>The table below details the final outputs of the SS2 workflow.</p><table><thead><tr><th>Output Name</th><th>Output Description</th><th>Output Format</th></tr></thead><tbody><tr><td>Pipeline version</td><td>Version of the processing pipeline run on this data</td><td>string</td></tr><tr><td>aligned_bam</td><td>HISAT2 out BAM</td><td>BAM</td></tr><tr><td>bam_index</td><td>HISAT2 BAM index</td><td>BAM</td></tr><tr><td>insert_size_metrics</td><td>Picard insert size metrics</td><td>txt</td></tr><tr><td>quality_distribution_metrics</td><td>Picard quality distribution metrics</td><td>txt</td></tr><tr><td>quality_by_cycle_metrics</td><td>Picard quality by cycle metrics</td><td>txt</td></tr><tr><td>bait_bias_summary_metrics</td><td>Picard bait bias summary metrics</td><td>txt</td></tr><tr><td>rna_metrics</td><td>Picard RNA metrics</td><td>txt</td></tr><tr><td>group_results</td><td>Array of Picard metric files generated by the GroupQCOutputs task</td><td>CSV</td></tr><tr><td>rsem_gene_results</td><td>RSEM file containing gene-level expression estimates</td><td>tab-delimited</td></tr><tr><td>rsem_isoform_results</td><td>RSEM file containing isoform-level expression estimates</td><td>tab delimited</td></tr><tr><td>loom_output_files</td><td>Loom file containing RSEM TPM and expected_counts and metrics</td><td>Loom</td></tr></tbody></table><p>The final count matrix is in Loom format and includes only the group_results CSV files and the rsem_gene_results (TPM and expected_count). Note that the TPMs are contained in the Loom “matrix&quot;, whereas expected_count is contained in the “layers/estimated_counts&quot;.</p><div class="theme-admonition theme-admonition-warning alert alert--danger admonition_LlT9"><div class="admonitionHeading_tbUL"><span class="admonitionIcon_kALy"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M5.05.31c.81 2.17.41 3.38-.52 4.31C3.55 5.67 1.98 6.45.9 7.98c-1.45 2.05-1.7 6.53 3.53 7.7-2.2-1.16-2.67-4.52-.3-6.61-.61 2.03.53 3.33 1.94 2.86 1.39-.47 2.3.53 2.27 1.67-.02.78-.31 1.44-1.13 1.81 3.42-.59 4.78-3.42 4.78-5.56 0-2.84-2.53-3.22-1.25-5.61-1.52.13-2.03 1.13-1.89 2.75.09 1.08-1.02 1.8-1.86 1.33-.67-.41-.66-1.19-.06-1.78C8.18 5.31 8.68 2.45 5.05.32L5.03.3l.02.01z"></path></svg></span>Zarr Array Deprecation Notice June 2020</div><div class="admonitionContent_S0QG"><p>Please note that we have deprecated the previously used Zarr array output. The pipeline now uses the Loom file format as the default output.</p></div></div><h2 class="anchor anchorWithStickyNavbar_LWe7" id="validation">Validation<a href="#validation" class="hash-link" aria-label="Direct link to Validation" title="Direct link to Validation">​</a></h2><p>The SS2 pipeline has been validated for processing human and mouse, stranded or unstranded, paired- or single-end, and plate- or fluidigm-based Smart-seq2 datasets (see links to validation reports in the table below).</p><table><thead><tr><th>Workflow Configuration</th><th>Link to Report</th></tr></thead><tbody><tr><td>Mouse paired-end</td><td><a href="https://docs.google.com/document/d/12zGTFROrcXEByt9z0h06qjSqb9vWutn28Tx6YiND1Ds/edit" target="_blank" rel="noopener noreferrer">Report</a></td></tr><tr><td>Human and mouse single-end</td><td><a href="https://docs.google.com/document/d/1MonsTG8UnROHZ_XpulrSZNTxO988KEH6T6h45plFYQg/edit#heading=h.ixoqmhbabdvh" target="_blank" rel="noopener noreferrer">Report</a></td></tr><tr><td>Human stranded fluidigm</td><td><a href="https://docs.google.com/document/d/1FEg86Tlu657j9Kjw_v3keFQRXcBIs8gOqCwLbPSP-C0/edit#heading=h.wjr8otl7zg14" target="_blank" rel="noopener noreferrer">Report</a></td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="versioning">Versioning<a href="#versioning" class="hash-link" aria-label="Direct link to Versioning" title="Direct link to Versioning">​</a></h2><p>All SS2 release notes are documented in the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/smartseq2_single_sample/SmartSeq2SingleSample.changelog.md" target="_blank" rel="noopener noreferrer">Smartseq2 Single Sample changelog</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="citing-the-smart-seq2-single-sample-pipeline">Citing the Smart-seq2 Single Sample Pipeline<a href="#citing-the-smart-seq2-single-sample-pipeline" class="hash-link" aria-label="Direct link to Citing the Smart-seq2 Single Sample Pipeline" title="Direct link to Citing the Smart-seq2 Single Sample Pipeline">​</a></h2><p>If you use the Smart-seq2 Single Sample Pipeline in your research, please identify the pipeline in your methods section using the <a href="https://scicrunch.org/resources/data/record/nlx_144509-1/SCR_021228/resolver?q=SCR_021228&amp;l=SCR_021228&amp;i=rrid:scr_021228" target="_blank" rel="noopener noreferrer">Smart-seq2 Single Sample SciCrunch resource identifier</a>.</p><ul><li>Ex: <em>Smart-seq2 Single Sample Pipeline (RRID:SCR_021228)</em></li></ul><p>Please also consider citing our preprint:</p><p>Degatano, K., Awdeh, A., Cox III, R.S., Dingman, W., Grant, G., Khajouei, F., Kiernan, E., Konwar, K., Mathews, K.L., Palis, K., et al. Warp Analysis Research Pipelines: Cloud-optimized workflows for biological data processing and reproducible analysis. Bioinformatics 2025; btaf494. <a href="https://doi.org/10.1093/bioinformatics/btaf494" target="_blank" rel="noopener noreferrer">https://doi.org/10.1093/bioinformatics/btaf494</a></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="consortia-support">Consortia Support<a href="#consortia-support" class="hash-link" aria-label="Direct link to Consortia Support" title="Direct link to Consortia Support">​</a></h2><p>This pipeline is supported and used by the <a href="https://www.humancellatlas.org/" target="_blank" rel="noopener noreferrer">Human Cell Atlas</a> (HCA) project. </p><p>If your organization also uses this pipeline, we would love to list you! Please <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="have-suggestions">Have Suggestions?<a href="#have-suggestions" class="hash-link" aria-label="Direct link to Have Suggestions?" title="Direct link to Have Suggestions?">​</a></h2><p>Please help us make our tools better by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>; we welcome pipeline-related suggestions or questions.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Deprecated_Pipelines/Smart-seq2_Single_Sample_Pipeline/README.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2025-10-27T14:02:21.000Z">Oct 27, 2025</time></b> by <b>Nikelle Petrillo</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/Deprecated_Pipelines/Smart-seq2_Multi_Sample_Pipeline/Loom_schema"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">What&#x27;s in the Smart-seq2 Multi Sample Pipeline Loom File?</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/Deprecated_Pipelines/Smart-seq2_Single_Sample_Pipeline/Loom_schema"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">What&#x27;s in the Smart-seq2 Single Sample Pipeline Loom File?</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-smart-seq2-single-sample-pipeline" class="table-of-contents__link toc-highlight">Introduction to the Smart-seq2 Single Sample Pipeline</a></li><li><a href="#quick-start-table" class="table-of-contents__link toc-highlight">Quick Start Table</a></li><li><a href="#set-up" class="table-of-contents__link toc-highlight">Set-Up</a><ul><li><a href="#smart-seq2-single-sample-installation-and-requirements" class="table-of-contents__link toc-highlight">Smart-seq2 Single Sample Installation and Requirements</a></li><li><a href="#inputs" class="table-of-contents__link toc-highlight">Inputs</a></li></ul></li><li><a href="#running-smart-seq2" class="table-of-contents__link toc-highlight">Running Smart-seq2</a><ul><li><a href="#smart-seq2-workflow-summary" class="table-of-contents__link toc-highlight">Smart-seq2 Workflow Summary</a></li><li><a href="#outputs" class="table-of-contents__link toc-highlight">Outputs</a></li></ul></li><li><a href="#validation" class="table-of-contents__link toc-highlight">Validation</a></li><li><a href="#versioning" class="table-of-contents__link toc-highlight">Versioning</a></li><li><a href="#citing-the-smart-seq2-single-sample-pipeline" class="table-of-contents__link toc-highlight">Citing the Smart-seq2 Single Sample Pipeline</a></li><li><a href="#consortia-support" class="table-of-contents__link toc-highlight">Consortia Support</a></li><li><a href="#have-suggestions" class="table-of-contents__link toc-highlight">Have Suggestions?</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.91c3cfb7.js"></script>
<script src="/warp/assets/js/main.affc3f00.js"></script>
</body>
</html>