<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Deprecated_Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/README">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">Genomic Data Commons (GDC) Whole Genome Somatic Single Sample Overview | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/README"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Genomic Data Commons (GDC) Whole Genome Somatic Single Sample Overview | WARP"><meta data-rh="true" name="description" content="9/12/2014"><meta data-rh="true" property="og:description" content="9/12/2014"><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/README"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/README" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/README" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.59de646f.js" as="script">
<link rel="preload" href="/warp/assets/js/main.435ed8cb.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">CEMBA</a><button aria-label="Toggle the collapsible sidebar category &#x27;CEMBA&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/README">GDC Whole Genome Somatic Single Sample</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Single_Cell_ATAC_Seq_Pipeline/README">Single Cell ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Cell ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Multi_Sample_Pipeline/README">Smart-seq2 Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Sample_Pipeline/README">Smart-seq2 Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">Deprecated Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">GDC Whole Genome Somatic Single Sample</span><meta itemprop="position" content="2"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Genomic Data Commons (GDC) Whole Genome Somatic Single Sample Overview</h1><div class="theme-admonition theme-admonition-warning alert alert--danger admonition_LlT9"><div class="admonitionHeading_tbUL"><span class="admonitionIcon_kALy"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M5.05.31c.81 2.17.41 3.38-.52 4.31C3.55 5.67 1.98 6.45.9 7.98c-1.45 2.05-1.7 6.53 3.53 7.7-2.2-1.16-2.67-4.52-.3-6.61-.61 2.03.53 3.33 1.94 2.86 1.39-.47 2.3.53 2.27 1.67-.02.78-.31 1.44-1.13 1.81 3.42-.59 4.78-3.42 4.78-5.56 0-2.84-2.53-3.22-1.25-5.61-1.52.13-2.03 1.13-1.89 2.75.09 1.08-1.02 1.8-1.86 1.33-.67-.41-.66-1.19-.06-1.78C8.18 5.31 8.68 2.45 5.05.32L5.03.3l.02.01z"></path></svg></span>danger</div><div class="admonitionContent_S0QG"><p>9/12/2014</p><p>We are deprecating the Genomic Data Commons Whole Genome Somatic Single Sample Pipeline. Although the code will continue to be available, we are no longer supporting it.</p></div></div><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">GDCWholeGenomeSomaticSingleSample_v1.3.1</a></td><td align="center">January, 2024</td><td align="center">Elizabeth Kiernan</td><td align="center">Please <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">file an issue in WARP</a>.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="introduction-to-the-gdc-whole-genome-somatic-single-sample-pipeline">Introduction to the GDC Whole Genome Somatic Single Sample pipeline<a href="#introduction-to-the-gdc-whole-genome-somatic-single-sample-pipeline" class="hash-link" aria-label="Direct link to Introduction to the GDC Whole Genome Somatic Single Sample pipeline" title="Direct link to Introduction to the GDC Whole Genome Somatic Single Sample pipeline">​</a></h2><p>The GDC Whole Genome Somatic Single Sample (abbreviated GDC here) pipeline is the alignment and preprocessing workflow for genomic data designed for the National Cancer Institute&#x27;s <a href="https://gdc.cancer.gov/about-gdc" target="_blank" rel="noopener noreferrer">Genomic Data Commons</a>.</p><p>A high-level overview of the pipeline in addition to tool parameters are available on the <a href="https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/DNA_Seq_Variant_Calling_Pipeline/" target="_blank" rel="noopener noreferrer">GDC Documentation site</a>.</p><p>Overall, the pipeline converts reads (CRAM or BAM) to FASTQ and (re)aligns them to the latest human reference genome (see the <a href="#gdc-reference-genome">GDC Reference Genome</a> section below). Each read group is aligned separately. Read group alignments that belong to a single sample are then merged and duplicate reads are flagged for downstream variant calling.</p><p><img loading="lazy" src="/warp/assets/images/GDC-f632671b22c77921a8a18af81f65b521.png" width="1752" height="775" class="img_ev3q"></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="set-up">Set-up<a href="#set-up" class="hash-link" aria-label="Direct link to Set-up" title="Direct link to Set-up">​</a></h2><h3 class="anchor anchorWithStickyNavbar_LWe7" id="workflow-installation-and-requirements">Workflow installation and requirements<a href="#workflow-installation-and-requirements" class="hash-link" aria-label="Direct link to Workflow installation and requirements" title="Direct link to Workflow installation and requirements">​</a></h3><p>The workflow is written in the Workflow Description Language WDL and can be downloaded by cloning the <a href="https://github.com/broadinstitute/warp/tree/master" target="_blank" rel="noopener noreferrer">warp repository</a> in GitHub. The workflow can be deployed using <a href="https://github.com/broadinstitute/cromwell" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH compliant, flexible workflow management system that supports multiple computing platforms.</p><p>For the latest workflow version and release notes, please see the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/dna_seq/somatic/single_sample/wgs/gdc_genome/GDCWholeGenomeSomaticSingleSample.changelog.md" target="_blank" rel="noopener noreferrer">changelog</a>.</p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="software-version-requirements">Software version requirements<a href="#software-version-requirements" class="hash-link" aria-label="Direct link to Software version requirements" title="Direct link to Software version requirements">​</a></h3><ul><li>GATK 4.5.0.0</li><li>Picard 2.26.10</li><li>Samtools 1.11</li><li>Python 3.0</li><li>Cromwell version support<ul><li>Successfully tested on v52</li><li>Does not work on versions &lt; v23 due to output syntax</li></ul></li><li>Papi version support<ul><li>Successfully tested on Papi v2</li></ul></li></ul><h3 class="anchor anchorWithStickyNavbar_LWe7" id="input-descriptions">Input descriptions<a href="#input-descriptions" class="hash-link" aria-label="Direct link to Input descriptions" title="Direct link to Input descriptions">​</a></h3><p>The table below describes each of the GDC pipeline inputs. The workflow requires a <strong>single</strong> aligned CRAM or BAM file, or a <strong>single</strong> unmapped BAM (uBAM) as input, and is set up to run on samples with reads greater than 75 bp.</p><p>For examples of how to specify each input as well as cloud locations for different example input files, see the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/dna_seq/somatic/single_sample/wgs/gdc_genome/input_files/GDCWholeGenomeSomaticSingleSample.inputs.json" target="_blank" rel="noopener noreferrer">input configuration file (JSON)</a>.</p><table><thead><tr><th>Input name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>input_cram (optional)</td><td>A single mapped CRAM file; alternatively, input can be a mapped BAM (<code>input_bam</code>) or unmapped BAM (uBAM; <code>ubam</code>). If using a CRAM file aligned to a reference different than the reference chosen for the GDC workflow (the <code>ref_fasta</code> input), then you must specify both the <code>cram_ref_fasta</code> and the <code>cram_ref_fasta_index</code> inputs.</td><td>File</td></tr><tr><td>input_bam (optional)</td><td>A single mapped BAM file; alternatively input can be a CRAM (<code>input_cram</code>) or uBAM (<code>ubam</code>).</td><td>File</td></tr><tr><td>cram_ref_fasta (optional)</td><td>The reference file that was used to align the CRAM input (if used); if unspecified, the workflow will use the <code>ref_fasta</code> input file by default.</td><td>File</td></tr><tr><td>cram_ref_fasta_index (optional)</td><td>CRAM reference FASTA index for the <code>cram_ref_fata</code> if CRAM is used as workflow input.</td><td>File</td></tr><tr><td>output_map (optional)</td><td>Tab-separated file containing two columns: a list of all the read group IDs found in the input_cram (or input_bam) and a list of the desired name of the uBAMs generated for read group.</td><td>File</td></tr><tr><td>unmapped_bam_suffix (optional)</td><td>Optional string used to name the output uBAM file.</td><td>String</td></tr><tr><td>ubam (optional)</td><td>A single uBAM file; alternatively, input can be a mapped BAM (<code>input_bam</code>) or CRAM (<code>input_cram</code>).</td><td>File</td></tr><tr><td>contamination_vcf</td><td>VCF file of common variant sites that is used for the check_contamination task.</td><td>File</td></tr><tr><td>contamination_vcf_index</td><td>Index file for the <code>contamination_vcf</code> input.</td><td>File</td></tr><tr><td>dbsnp_vcf</td><td>VCF file of known variation sites that can be used to exclude these sites from the analysis; used for the gatk_baserecalibrator task.</td><td>File</td></tr><tr><td>dbsnp_vcf_index</td><td>Index file for the <code>dbsnp_vcf</code> input.</td><td>File</td></tr><tr><td>ref_fasta</td><td>Reference FASTA used to convert an input CRAM to BAM; if CRAM is not used, this file does not need to be specified.</td><td>File</td></tr><tr><td>ref_fai</td><td>Reference FASTA index for the <code>ref_fasta</code> input.</td><td>File</td></tr><tr><td>ref_dict</td><td>BWA reference dictionary used for alignment.</td><td>File</td></tr><tr><td>ref_amb</td><td>BWA reference file used for alignment.</td><td>File</td></tr><tr><td>ref_ann</td><td>BWA reference file used for alignment.</td><td>File</td></tr><tr><td>ref_bwt</td><td>BWA reference file used for alignment.</td><td>File</td></tr><tr><td>ref_pac</td><td>BWA reference file used for alignment.</td><td>File</td></tr><tr><td>ref_sa</td><td>BWA reference file used for alignment.</td><td>File</td></tr></tbody></table><h3 class="anchor anchorWithStickyNavbar_LWe7" id="gdc-reference-genome">GDC reference genome<a href="#gdc-reference-genome" class="hash-link" aria-label="Direct link to GDC reference genome" title="Direct link to GDC reference genome">​</a></h3><p>The GDC uses the human reference genome GRCh38.d1.vd1 for all data processing. Unlike the GRCh38 reference used by WARP pipelines for production, the GDC reference includes decoy viral sequences for ten types of human viral genomes. You can learn more about the reference from the <a href="https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/DNA_Seq_Variant_Calling_Pipeline/#reference-genome" target="_blank" rel="noopener noreferrer">GDC documentation</a>.</p><p>The reference files required for the GDC workflow are hosted in a <a href="https://console.cloud.google.com/storage/browser/gcp-public-data--broad-references/hg38/v0/gdc" target="_blank" rel="noopener noreferrer">public Google Bucket</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="workflow-tasks-and-tools">Workflow tasks and tools<a href="#workflow-tasks-and-tools" class="hash-link" aria-label="Direct link to Workflow tasks and tools" title="Direct link to Workflow tasks and tools">​</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/dna_seq/somatic/single_sample/wgs/gdc_genome/GDCWholeGenomeSomaticSingleSample.wdl" target="_blank" rel="noopener noreferrer">workflow</a> imports a series of tasks either from the Workflow script or the Broad <a href="https://github.com/broadinstitute/warp/tree/master/tasks/broad" target="_blank" rel="noopener noreferrer">tasks library</a>.</p><table><thead><tr><th>Task name in WDL</th><th>Tool</th><th>Software</th><th>Description</th></tr></thead><tbody><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/reprocessing/cram_to_unmapped_bams/CramToUnmappedBams.wdl" target="_blank" rel="noopener noreferrer">CramToUnmappedBams</a></td><td>view, index, RevertSam, ValidateSamFile, SortSam</td><td><a href="http://www.htslib.org/terms/" target="_blank" rel="noopener noreferrer">SamTools</a> and <a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a></td><td>If a CRAM file is used as input, the task converts to uBAM and generates an output map that is then used to split the uBAM by readgroup. The resulting BAM is sorted by query name using Picard.</td></tr><tr><td>bam_readgroup_to_contents</td><td>view</td><td><a href="http://www.htslib.org" target="_blank" rel="noopener noreferrer">Samtools</a></td><td>Extracts all the readgroups from the BAM header and returns a WDL array where each row is a readgroup.</td></tr><tr><td>biobambam_bamtofastq</td><td>bamtofastq</td><td><a href="https://gitlab.com/german.tischler/biobambam2" target="_blank" rel="noopener noreferrer">biobambam2</a></td><td>Converts the uBAMs to FASTQ.</td></tr><tr><td>emit_pe_records/emit_se_records</td><td>---</td><td>---</td><td>Associates the fasta file(s) generated by biobambam_bamtofastq with their respective readgroup; creates an array of structs to be used as input for alignment.</td></tr><tr><td>bwa_pe/ bwa_se</td><td>BWA mem, view</td><td><a href="http://bio-bwa.sourceforge.net/" target="_blank" rel="noopener noreferrer">BWA</a> and <a href="http://www.htslib.org" target="_blank" rel="noopener noreferrer">Samtools</a></td><td>Aligns FASTQ reads to the reference genome, generating an aligned BAM file.</td></tr><tr><td>picard_markduplicates</td><td><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-" target="_blank" rel="noopener noreferrer">MarkDuplicates</a></td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a></td><td>Runs MarkDuplicates in silent mode to locate and tag duplicate reads. Outputs a tagged BAM and metrics file.</td></tr><tr><td>sort_and_index_markdup_bam</td><td>sort</td><td><a href="http://www.htslib.org" target="_blank" rel="noopener noreferrer">Samtools</a></td><td>Sorts and indexes the BAM file; outputs the sorted BAM and index file.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/CheckContaminationSomatic.wdl" target="_blank" rel="noopener noreferrer">check_contamination</a></td><td><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037593451" target="_blank" rel="noopener noreferrer">GetPileUpSummaries</a> and <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360036888972" target="_blank" rel="noopener noreferrer">CalculateContamination</a></td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Summarizes counts of reads that support reference, alternate and other alleles for given sites and then calculates the fraction of reads coming from cross-sample contamination.</td></tr><tr><td>gatk_baserecalibrator</td><td><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360036898312" target="_blank" rel="noopener noreferrer">BaseRecalibrator</a></td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Generates recalibration table for Base Quality Score Recalibration (BQSR).</td></tr><tr><td>gatk_applybqsr</td><td><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037055712" target="_blank" rel="noopener noreferrer">ApplyBQSR</a></td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Recalibrates the base qualities of the input reads based on the recalibration table produced by the BaseRecalibrator tool and outputs a recalibrated BAM file; uses the emit_original_quals parameter to write the original base qualities under the BAM OQ tag.</td></tr><tr><td>collect_insert_size_metrics</td><td><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037055772" target="_blank" rel="noopener noreferrer">CollectInsertSizeMetrics</a></td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a></td><td>Generates metrics about insert size distribution in the form of both a histogram and table (txt file).</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="workflow-outputs">Workflow outputs<a href="#workflow-outputs" class="hash-link" aria-label="Direct link to Workflow outputs" title="Direct link to Workflow outputs">​</a></h2><p>The following table describes the workflow outputs. If running the workflow using Cromwell, these outputs will automatically be placed in the respective task execution directory.</p><p>Alternatively, Cromwell allows you to specify an output directory using an options.json, as described in Cromwell&#x27;s <a href="https://cromwell.readthedocs.io/en/stable/wf_options/Overview/" target="_blank" rel="noopener noreferrer">Workflow Options Overview</a>(see the section on <a href="https://cromwell.readthedocs.io/en/stable/wf_options/Overview/#output-copying" target="_blank" rel="noopener noreferrer">Output Copying</a>).</p><table><thead><tr><th>Output name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>validation_report (optional)</td><td>Samtools validation report(s); only returned if an aligned CRAM or BAM file is used as workflow input</td><td>TXT</td></tr><tr><td>unmapped_bams (optional)</td><td>uBAM file only returned if an aligned CRAM or BAM is used as workflow input.</td><td>BAM</td></tr><tr><td>bam</td><td>Base recalibrated BAM file.</td><td>BAM</td></tr><tr><td>bai</td><td>Index file for the BAM.</td><td>BAI</td></tr><tr><td>md_metrics</td><td>Picard MarkDuplicates metrics.</td><td>TXT</td></tr><tr><td>insert_size_metrics</td><td>Picard insert size metrics.</td><td>TXT</td></tr><tr><td>insert_size_histogram_pdf</td><td>Histogram representation of insert size metrics.</td><td>PDF</td></tr><tr><td>contamination</td><td>File containing a value indicating the fraction contamination.</td><td>TXT</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="important-notes">Important notes<a href="#important-notes" class="hash-link" aria-label="Direct link to Important notes" title="Direct link to Important notes">​</a></h2><ul><li>Runtime parameters are optimized for Broad&#x27;s Google Cloud Platform implementation.</li><li>Please visit the <a href="https://gatk.broadinstitute.org/hc/en-us/categories/360002310591" target="_blank" rel="noopener noreferrer">GATK Technical Documentation</a> site for further documentation on GATK-related workflows and tools.</li></ul><h2 class="anchor anchorWithStickyNavbar_LWe7" id="citing-the-gdc-pipeline">Citing the GDC Pipeline<a href="#citing-the-gdc-pipeline" class="hash-link" aria-label="Direct link to Citing the GDC Pipeline" title="Direct link to Citing the GDC Pipeline">​</a></h2><p>If you use the GDC Pipeline in your research, please consider citing our preprint:</p><p>Degatano, K., Awdeh, A., Cox III, R.S., Dingman, W., Grant, G., Khajouei, F., Kiernan, E., Konwar, K., Mathews, K.L., Palis, K., et al. Warp Analysis Research Pipelines: Cloud-optimized workflows for biological data processing and reproducible analysis. Bioinformatics 2025; btaf494. <a href="https://doi.org/10.1093/bioinformatics/btaf494" target="_blank" rel="noopener noreferrer">https://doi.org/10.1093/bioinformatics/btaf494</a></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="contact-us">Contact us<a href="#contact-us" class="hash-link" aria-label="Direct link to Contact us" title="Direct link to Contact us">​</a></h2><p>Please help us make our tools better by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>; we welcome pipeline-related suggestions or questions.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="licensing">Licensing<a href="#licensing" class="hash-link" aria-label="Direct link to Licensing" title="Direct link to Licensing">​</a></h2><p>Copyright Broad Institute, 2020 | BSD-3</p><p>The workflow script is released under the <strong>WDL open source code license (BSD-3)</strong> (full license text at <a href="https://github.com/broadinstitute/warp/blob/master/LICENSE" target="_blank" rel="noopener noreferrer">https://github.com/broadinstitute/warp/blob/master/LICENSE</a>). However, please note that the programs it calls may be subject to different licenses. Users are responsible for checking that they are authorized to run all programs before running this script.</p><ul><li><a href="https://software.broadinstitute.org/gatk/download/licensing.php" target="_blank" rel="noopener noreferrer">GATK</a></li><li><a href="http://bio-bwa.sourceforge.net/bwa.shtml#13" target="_blank" rel="noopener noreferrer">BWA</a></li><li><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a></li><li><a href="http://www.htslib.org/terms/" target="_blank" rel="noopener noreferrer">Samtools</a></li></ul></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Deprecated_Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/README.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2026-02-19T15:27:23.000Z">Feb 19, 2026</time></b> by <b>Nikelle Petrillo</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/Deprecated_Pipelines/CEMBA_MethylC_Seq_Pipeline/CEMBA.methods"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">CEMBA_v1.1.6 Publication Methods</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/Pipelines/Single_Cell_ATAC_Seq_Pipeline/README"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">scATAC Overview</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-gdc-whole-genome-somatic-single-sample-pipeline" class="table-of-contents__link toc-highlight">Introduction to the GDC Whole Genome Somatic Single Sample pipeline</a></li><li><a href="#set-up" class="table-of-contents__link toc-highlight">Set-up</a><ul><li><a href="#workflow-installation-and-requirements" class="table-of-contents__link toc-highlight">Workflow installation and requirements</a></li><li><a href="#software-version-requirements" class="table-of-contents__link toc-highlight">Software version requirements</a></li><li><a href="#input-descriptions" class="table-of-contents__link toc-highlight">Input descriptions</a></li><li><a href="#gdc-reference-genome" class="table-of-contents__link toc-highlight">GDC reference genome</a></li></ul></li><li><a href="#workflow-tasks-and-tools" class="table-of-contents__link toc-highlight">Workflow tasks and tools</a></li><li><a href="#workflow-outputs" class="table-of-contents__link toc-highlight">Workflow outputs</a></li><li><a href="#important-notes" class="table-of-contents__link toc-highlight">Important notes</a></li><li><a href="#citing-the-gdc-pipeline" class="table-of-contents__link toc-highlight">Citing the GDC Pipeline</a></li><li><a href="#contact-us" class="table-of-contents__link toc-highlight">Contact us</a></li><li><a href="#licensing" class="table-of-contents__link toc-highlight">Licensing</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2026 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.59de646f.js"></script>
<script src="/warp/assets/js/main.435ed8cb.js"></script>
</body>
</html>