<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.5">
<link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Blog Atom Feed">
<link rel="preconnect" href="https://www.google-analytics.com">
<script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-181048107-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview")</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><title data-react-helmet="true">Consortia Data Processing | WARP</title><meta data-react-helmet="true" name="twitter:card" content="summary_large_image"><meta data-react-helmet="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing"><meta data-react-helmet="true" name="docsearch:language" content="en"><meta data-react-helmet="true" name="docsearch:version" content="current"><meta data-react-helmet="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="Consortia Data Processing | WARP"><meta data-react-helmet="true" name="description" content="Human Cell Atlas Data Coordination Platform Matrix Processing"><meta data-react-helmet="true" property="og:description" content="Human Cell Atlas Data Coordination Platform Matrix Processing"><link data-react-helmet="true" rel="shortcut icon" href="/warp/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing" hreflang="x-default"><link data-react-helmet="true" rel="preconnect" href="https://BH4D9OD16A-dsn.algolia.net" crossorigin="anonymous"><link rel="stylesheet" href="/warp/assets/css/styles.57c5b218.css">
<link rel="preload" href="/warp/assets/js/runtime~main.73aa2fb7.js" as="script">
<link rel="preload" href="/warp/assets/js/main.fa639502.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#" class="skipToContent_1oUP">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle toggle_3Zt9 react-toggle--disabled"><div class="react-toggle-track" role="button" tabindex="-1"><div class="react-toggle-track-check"><span class="toggle_71bT">ðŸŒœ</span></div><div class="react-toggle-track-x"><span class="toggle_71bT">ðŸŒž</span></div><div class="react-toggle-thumb"></div></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div><div class="searchBox_1Doo"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docs-wrapper docs-doc-page"><div class="docPage_31aa"><button class="clean-btn backToTopButton_35hR" type="button"><svg viewBox="0 0 24 24" width="28"><path d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z" fill="currentColor"></path></svg></button><aside class="docSidebarContainer_3Kbt"><div class="sidebar_15mo"><nav class="menu thin-scrollbar menu_Bmed menuWithAnnouncementBar_2WvA"><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">About WARP</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#">WARP Pipelines</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">CEMBA</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Exome Germline Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">GDC Whole Genome Somatic Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Illumina Genotyping Array</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Imputation</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#" tabindex="0">Optimus</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/optimus.methods">Optimus v5.3.0 Methods</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/Bam_tags">What tags are included in an Optimus BAM file?</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a></li><li class="menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing">Consortia Data Processing</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">RNA with UMIs</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Single Cell ATAC</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">SlideSeq</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Multi-Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Single Nucleus Multi-Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Ultima Genomics Whole Genome Germline</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Whole Genome Germline Single Sample</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">WDL Best Practices</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">Contribution Guide</a></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_1CGd"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_3E-R"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></aside><main class="docMainContainer_3ufF"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_3FnS"><div class="docItemContainer_33ec"><article><div class="tocCollapsible_1PrD tocMobile_3Hoh"><button type="button" class="clean-btn tocCollapsibleButton_2O1e">On this page</button></div><div class="markdown"><header><h1>Consortia Data Processing</h1></header><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="human-cell-atlas-data-coordination-platform-matrix-processing"></a>Human Cell Atlas Data Coordination Platform Matrix Processing<a class="hash-link" href="#human-cell-atlas-data-coordination-platform-matrix-processing" title="Direct link to heading">#</a></h2><p>Optimus supports data processing for the Human Cell Atlas (HCA) Data Coordination Platform (DCP). Learn more about the DCP at the <a href="https://data.humancellatlas.org/" target="_blank" rel="noopener noreferrer">HCA Data Portal</a>).</p><p>All DCP Projects processed with Optimus have matrices containing the standard metrics and counts detailed in the <a href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a>, but also have additional post-processing to combine project data and incorporate DCP-curated metadata.</p><div class="admonition admonition-warning alert alert--danger"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M5.05.31c.81 2.17.41 3.38-.52 4.31C3.55 5.67 1.98 6.45.9 7.98c-1.45 2.05-1.7 6.53 3.53 7.7-2.2-1.16-2.67-4.52-.3-6.61-.61 2.03.53 3.33 1.94 2.86 1.39-.47 2.3.53 2.27 1.67-.02.78-.31 1.44-1.13 1.81 3.42-.59 4.78-3.42 4.78-5.56 0-2.84-2.53-3.22-1.25-5.61-1.52.13-2.03 1.13-1.89 2.75.09 1.08-1.02 1.8-1.86 1.33-.67-.41-.66-1.19-.06-1.78C8.18 5.31 8.68 2.45 5.05.32L5.03.3l.02.01z"></path></svg></span>warning</h5></div><div class="admonition-content"><p>This section details matrices produced for the Human Cell Atlas (HCA) <a href="https://data.humancellatlas.org/" target="_blank" rel="noopener noreferrer">Data Coordination Platform (DCP)2.0</a>, which includes matrices processed with Optimus v4.1.7 and later. The DCP is currently reprocessing data generated with earlier Optimus versions and will deprecate previous matrices once reprocessing is complete.</p></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="dcp-project-matrix-post-processing"></a>DCP project matrix post-processing<a class="hash-link" href="#dcp-project-matrix-post-processing" title="Direct link to heading">#</a></h3><p>DCP project matrices combine project data from individual library preparations that contain the same species, developmental age, and sequencing technology. </p><p>Key differences between the standard Optimus Loom matrix described in the <a href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a> and the DCP Project matrix include: </p><ul><li>Combined project data in the DCP matrix</li><li>Filtering: to reduce file size, DCP project matrices are in sparse format and minimally filtered so that <strong>only cells with 100 molecules or more</strong> are retained.</li><li>DCP-curated metadata in the Loom global attributes: see table below.</li><li>input_ids in the global attributes: all input_ids representing each library preparation in the matrix are added as a comma-separated string.</li></ul><p>DCP project matrices contain DCP-curated metadata in the Loom global attributes which may be useful when exploring the data and linking it back to the Project metadata. </p><p>Read more about each metadata field in the DCP <a href="https://data.humancellatlas.org/metadata" target="_blank" rel="noopener noreferrer">Metadata Dictionary</a>.</p><table><thead><tr><th>Metadata Attribute Name in Count Matrix</th><th>Metadata Description</th></tr></thead><tbody><tr><td><code>donor_organism.genus_species</code></td><td>species information; human or mouse</td></tr><tr><td><code>library_preparation_protocol.library_construction_approach</code></td><td>technology used for library preparation, i.e 10x or SS2</td></tr><tr><td><code>specimen_from_organism.organ</code></td><td>organ</td></tr><tr><td><code>project.project_core.project_name</code></td><td>project name</td></tr><tr><td><code>project.provenance.document_id</code></td><td>project id</td></tr><tr><td><code>input_id</code></td><td>metadata values for  <code>sequencing_process.provenance.document_id</code>; unique ID to demarcate the library prep</td></tr><tr><td><code>input_name</code></td><td>metadata values for <code>sequencing_input.biomaterial_core.biomaterial_id</code>; unique ID for the biomaterial</td></tr><tr><td><code>input_id_metadata_field</code></td><td>string describing the DCP-curated metadata field used for input_id: <code>sequencing_process.provenance.document_id</code></td></tr><tr><td><code>input_name_metadata_field</code></td><td>string describing the DCP-curated metadata field used for input_name: <code>sequencing_input.biomaterial_core.biomaterial_id</code></td></tr></tbody></table><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="post-processing-steps"></a>Post-processing steps<a class="hash-link" href="#post-processing-steps" title="Direct link to heading">#</a></h3><p>To create the DCP project matrices, Loom outputs from individual 10x library preparations, each with their own <code>input_id</code>, are combined into a single Loom file.</p><p><img src="/warp/assets/images/post_processing-256dd7dbd7f417796924b928983dadc3.png"></p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="making-project-matrix-cell-barcodes-unique"></a>Making project matrix cell barcodes unique<a class="hash-link" href="#making-project-matrix-cell-barcodes-unique" title="Direct link to heading">#</a></h3><p>Since DCP project matrices often contain combined data from multiple library preparations, the project matrix cell barcodes are modified so that they are unique for each library preparation, allowing the barcodes to be used by downstream community tools like Cumulus and Seurat.</p><p>In the standard Optimus matrix, cell barcodes are listed in both the Loom columns <code>CellID</code> and the <code>cell_names</code>.</p><p>For DCP projects, however, <strong>the cell barcodes in the  <code>cell_names</code> column are modified</strong> so that each cell barcode belonging to an individual library preparation is unique. This is done by adding a numerical suffix to the barcodes that corresponds to the <code>input_id</code> for the library preparation from which the cell barcodes came. </p><p>The <code>input_ids</code> are listed in the matrix global attributes. The order of the input_ids serves as an index for the cell barcode suffix.</p><p><img src="/warp/assets/images/unique_barcodes-9a129aad53dcd952b00b44ad268b51f3.png"></p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="example-cell-barcodes-from-a-dcp-project-matrix"></a>Example cell barcodes from a DCP project matrix<a class="hash-link" href="#example-cell-barcodes-from-a-dcp-project-matrix" title="Direct link to heading">#</a></h4><p>Let&#x27;s use loompy to look at the global attribute <code>input_id</code> for a DCP project matrix (loom format):</p><div class="codeBlockContainer_K1bP"><div class="codeBlockContent_hGly python"><pre tabindex="0" class="prism-code language-python codeBlock_23N8 thin-scrollbar" style="color:#bfc7d5;background-color:#292d3e"><code class="codeBlockLines_39YC"><span class="token-line" style="color:#bfc7d5"><span class="token operator" style="color:rgb(137, 221, 255)">&gt;&gt;</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> ds</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token plain">attrs</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token plain">input_id</span></span><span class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;166c1b1a-ad9c-4476-a4ec-8b52eb5032c7, 22b7da3d-a301-433e-99e1-e67266c1ee8b, 337a48c5-e363-45aa-886f-ccd4425edc2b, 40630e8b-c3a3-4813-b1e4-b156637c5cc3, 58d703d1-d366-42d0-af44-a3bb836838a5, 70c8d647-7984-4d03-912a-f2437aa1ba4f, 7c86cf30-4284-4a0d-817f-6047560c05c3, 8ef7aca4-be00-4c03-8576-1b2eff4ce7af, ae0cfa6e-e7cb-4a88-9f89-1c44abaa2291, cbd23025-b1bf-4e9e-a297-ddab4a217b76, df049da4-3d20-4da7-a1d7-7d6e8f7740ff, e17bf5ea-788b-4756-a008-a07aec091e10&#x27;</span><span class="token plain"></span></span><span class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;&gt;</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> </span></span></code></pre><button type="button" aria-label="Copy code to clipboard" class="copyButton_Ue-o clean-btn">Copy</button></div></div><p>Notice the attribute&#x27;s value is a string of comma- <strong>and space-</strong> separated UUIDs.</p><p>Each of these UUIDs represents one library preparation. This matrix contains data from 12 library preparations total. </p><p>Now let&#x27;s look at the <code>cell_names</code> column attribute which contains the unique cell barcodes:</p><div class="codeBlockContainer_K1bP"><div class="codeBlockContent_hGly python"><pre tabindex="0" class="prism-code language-python codeBlock_23N8 thin-scrollbar" style="color:#bfc7d5;background-color:#292d3e"><code class="codeBlockLines_39YC"><span class="token-line" style="color:#bfc7d5"><span class="token operator" style="color:rgb(137, 221, 255)">&gt;&gt;</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> ds</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token plain">ca</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token plain">cell_names</span></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">array</span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token punctuation" style="color:rgb(199, 146, 234)">[</span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;GGACAAGAGTGCGTGA-0&#x27;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"> </span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;GATCGATCACCAGGTC-0&#x27;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"> </span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;AGCGGTCAGGGCTTGA-0&#x27;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"></span></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">       </span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"> </span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;GTACGTAAGCTATGCT-11&#x27;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"> </span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;CAGAATCTCTGAGTGT-11&#x27;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"></span></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">       </span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;AACACGTAGTGTTTGC-11&#x27;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">]</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"> dtype</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token builtin" style="color:rgb(130, 170, 255)">object</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token plain"></span></span><span class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;&gt;</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> </span></span></code></pre><button type="button" aria-label="Copy code to clipboard" class="copyButton_Ue-o clean-btn">Copy</button></div></div><p>The suffix appended to the barcodes in the <code>cell_names</code> column is the index for the <code>input_id</code> UUID to which the cell barcodes belong. </p><p>For example, cell barcodes with a &quot;-0&quot; suffix belong to the library preparation represented by the first UUID, <code>166c1b1a-ad9c-4476-a4ec-8b52eb5032c7</code>, whereas cell barcodes with a &quot;-11&quot; suffix represent the 10th UUID, <code>cbd23025-b1bf-4e9e-a297-ddab4a217b76</code>.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="mapping-dcp-project-matrix-data-to-the-metadata-manifest"></a>Mapping DCP project matrix data to the metadata manifest<a class="hash-link" href="#mapping-dcp-project-matrix-data-to-the-metadata-manifest" title="Direct link to heading">#</a></h3><p>While the project matrices contain some project metadata (listed in the table above), there is additionally useful metadata in the project&#x27;s metadata manifest, a TSV file containing all of a project&#x27;s metadata, including donor and disease state information.</p><p>In addition to the global attribute <code>input_id</code>, each project matrix has an <code>input_id</code> column that can be useful for mapping matrix data back to the DCP metadata manifest. </p><p>The values listed in the <code>input_id</code> column match the library preparation UUID in the metadata manifest column <code>sequencing_process.provenance.document_id</code>. </p><p>Read more about the metadata manifest in the DCP <a href="https://data.humancellatlas.org/guides" target="_blank" rel="noopener noreferrer">Exploring Projects guide</a>.</p><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Explore HCA Project matrices in Terra</h5></div><div class="admonition-content"><p>HCA matrices produced with Optimus are compatible with multiple downstream community analysis tools. For a tutorial on using the Optimus matrix with <a href="https://satijalab.org/seurat/index.html" target="_blank" rel="noopener noreferrer">Seurat</a>, <a href="https://scanpy.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Scanpy</a>, <a href="https://cumulus.readthedocs.io/en/latest/index.html" target="_blank" rel="noopener noreferrer">Cumulus</a>, or <a href="https://pegasus.readthedocs.io/en/stable/#" target="_blank" rel="noopener noreferrer">Pegasus</a>, see the public <a href="https://app.terra.bio/#workspaces/featured-workspaces-hca/Intro-to-HCA-data-on-Terra" target="_blank" rel="noopener noreferrer">Intro-to-HCA-data-on-Terra workspace</a> (login required) and its accompanying <a href="https://support.terra.bio/hc/en-us/articles/360060041772" target="_blank" rel="noopener noreferrer">step-by-step guide</a>.</p></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="mapping-dcp-project-matrix-data-to-the-contributor-matrices"></a>Mapping DCP project matrix data to the contributor matrices<a class="hash-link" href="#mapping-dcp-project-matrix-data-to-the-contributor-matrices" title="Direct link to heading">#</a></h3><p>Contributor matrices contain data analyzed and provided by the original project contributors. While they vary in format and content from project to project, they often include cell type annotations and additional metadata such as donor information and cell barcodes. </p><p><em>If the contributor matrix contains donor metadata that matches a field in the project metadata manifest</em>, the matrix can be linked to the DCP-generated project matrix in a two-step process.</p><ul><li><p>First, map the DCP matrix to the metadata manifest using the Loom&#x27;s <code>input_id</code> <strong>column</strong>; this column contains the same library preparation/donor IDs as the project metadata manifest&#x27;s <code>sequencing_process.provenance.document_id</code> column. </p></li><li><p>Second, map the contributor matrix to the metadata manifest using the contributor matrix column that matches the metadata manifest. </p></li></ul><p>Contributor matrices might contain a column for cell barcodes for each library/preparation donor. These barcodes should match the non-unique barcodes listed in the DCP project matrix, with the exception of cells that might have been filtered out of the Loom matrix due to low UMIs. The Loom&#x27;s non-unique barcodes are listed in the <code>CellID</code> column. </p><p><img src="/warp/assets/images/mapping_matrices-431db5e1dfb615be0c1741fbb863465e.png"></p><p>For example code showing how to link a contributor matrix to a DCP project matrix, see the <a target="_blank" href="/warp/assets/files/Matrix_matching-2c4edc3d0b1881f352a39bd44e22513d.ipynb">Matrix_matching Jupyter Notebook</a>.</p><p>If you have any questions related to the contributor matrix and content, reach out to the individual project contributors listed on the Project page.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="brain-initiative-cell-census-network-processing"></a>Brain Initiative Cell Census Network Processing<a class="hash-link" href="#brain-initiative-cell-census-network-processing" title="Direct link to heading">#</a></h2><p>The Optimus pipeline supports data processing for the <a href="https://biccn.org/" target="_blank" rel="noopener noreferrer">BRAIN Initiative Cell Census Network (BICCN)</a>. An overview of the BICCN pipeline resources is available on the BICCN&#x27;s <a href="https://biccn.org/tools/biccn-pipelines" target="_blank" rel="noopener noreferrer">Pipelines page</a>.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="optimus-reference-files-for-biccn-data-processing"></a>Optimus reference files for BICCN data processing<a class="hash-link" href="#optimus-reference-files-for-biccn-data-processing" title="Direct link to heading">#</a></h3><p>The BICCN 2.0 Whole Mouse Brain Working Group uses the Ensembl GRCm38 reference for alignment and a modified GTF for gene annotation (see table below). All Optimus pipeline reference inputs were created with the <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/skylab/build_indices" target="_blank" rel="noopener noreferrer">BuildIndices workflow</a>.</p><p>BICCN processes single-nucleus data, which is enriched in pre-mRNAs containing introns. To account for this, the BuildIndices workflow uses the <code>BuildStarSingleNucleus</code> task to add intron annotations to the GTF with a custom <a href="https://github.com/broadinstitute/warp/blob/master/dockers/skylab/snss2-build-indices/add-introns-to-gtf.py" target="_blank" rel="noopener noreferrer">python script</a>. The GTF contains all annotations for any <code>gene_id</code> that has at least one transcript. This reduces the number of genes in the GTF to ~32,000. </p><p>All reference files are available in a public Google bucket (see table below) and are accompanied by a README that details reference provenance (gs://gcp-public-data--broad-references/mm10/v0/README_mm10_singlecell_gencode.txt). </p><table><thead><tr><th>Optimus reference input name</th><th>Google bucket URI</th><th>Reference source</th><th>Description</th></tr></thead><tbody><tr><td><code>annotations_gtf</code></td><td>gs://gcp-public-data--broad-references/mm10/v0/single_nucleus/modified_gencode.vM23.primary_assembly.annotation.gtf</td><td><a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/gencode.vM23.annotation.gtf.gzf" target="_blank" rel="noopener noreferrer">https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/gencode.vM23.annotation.gtf.gzf</a></td><td>Modified GENCODE GTF including intron annotations that can be used for intron counting with featureCounts.</td></tr><tr><td><code>ref_genome_fasta</code></td><td>gs://gcp-public-data--broad-references/mm10/v0/single_nucleus/modified_mm10.primary_assembly.genome.fa</td><td><a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/GRCm38.p6.genome.fa.gz" target="_blank" rel="noopener noreferrer">https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/GRCm38.p6.genome.fa.gz</a></td><td>FASTA filed used to create the STAR reference files.</td></tr><tr><td><code>tar_star_reference</code></td><td>gs://gcp-public-data--broad-references/mm10/v0/single_nucleus/star/modified_star_2.7.9a_primary_gencode_mouse_vM23.tar</td><td>NA â€” built with the BuildIndices workflow.</td><td>Reference files used for alignment with STAR.</td></tr><tr><td><code>whitelist</code></td><td>gs://broad-gotc-test-storage/Optimus/truth/plumbing/master/inputs/737K-august-2016.txt (for v2 chemistry) and gs://broad-gotc-test-storage/Optimus/truth/plumbing/master/inputs/3M-february-2018.txt (for v3 chemisty)</td><td>See <a href="https://kb.10xgenomics.com/hc/en-us/articles/115004506263-What-is-a-barcode-whitelist" target="_blank" rel="noopener noreferrer">10x barcode descriptions</a>.</td><td>List of barcode sequences included in 10x library preparations. Each barcode demarcates an individual cell.</td></tr></tbody></table></div><footer class="docusaurus-mt-lg"><div class="row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/Optimus_Pipeline/consortia-processing.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_2_ui" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_13-_">Last updated on <b><time datetime="2023-02-28T21:53:29.000Z">2/28/2023</time></b> by <b>Nikelle Petrillo</b></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Â« Optimus Count Matrix Overview</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">RNA with UMIs Overview Â»</div></a></div></nav></div></div><div class="col col--3"><div class="tableOfContents_35-E thin-scrollbar"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#human-cell-atlas-data-coordination-platform-matrix-processing" class="table-of-contents__link">Human Cell Atlas Data Coordination Platform Matrix Processing</a><ul><li><a href="#dcp-project-matrix-post-processing" class="table-of-contents__link">DCP project matrix post-processing</a></li><li><a href="#post-processing-steps" class="table-of-contents__link">Post-processing steps</a></li><li><a href="#making-project-matrix-cell-barcodes-unique" class="table-of-contents__link">Making project matrix cell barcodes unique</a></li><li><a href="#mapping-dcp-project-matrix-data-to-the-metadata-manifest" class="table-of-contents__link">Mapping DCP project matrix data to the metadata manifest</a></li><li><a href="#mapping-dcp-project-matrix-data-to-the-contributor-matrices" class="table-of-contents__link">Mapping DCP project matrix data to the contributor matrices</a></li></ul></li><li><a href="#brain-initiative-cell-census-network-processing" class="table-of-contents__link">Brain Initiative Cell Census Network Processing</a><ul><li><a href="#optimus-reference-files-for-biccn-data-processing" class="table-of-contents__link">Optimus reference files for BICCN data processing</a></li></ul></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item"><span>GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_3J9K"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></span></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright Â© 2023 Copyright Â© Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.73aa2fb7.js"></script>
<script src="/warp/assets/js/main.fa639502.js"></script>
</body>
</html>