# What tags are included in an Optimus BAM file?
The Optimus Pipeline outputs a barcoded BAM file of aligned reads. There are multiple tags within the BAM file, including standard tags from [10X genomics](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/bam) and [Sequence Alignment Map (SAM) files](https://samtools.github.io/hts-specs/SAMtags.pdf). The table below details the tags used by the Optimus Pipeline and the relevant sources/tools from which the pipeline obtains the tags.

|Optimus Pipeline BAM Tag | Details | Source
|---| --- | --- |
| AS | Alignment score generated by aligner | [SAM](https://samtools.github.io/hts-specs/SAMtags.pdf) |
| CB | Chromium cellular barcode sequence that is error-corrected and confirmed against a list of known-good barcode sequences | [10X Genomics](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/bam) |
| CR | Chromium cellular barcode sequence as reported by the sequencer | [10X Genomics](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/bam) |
| CY | Chromium cellular barcode read quality. Phred scores as reported by sequencer | [10X Genomics](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/bam) |
| GE | Gene exon | [Drop-seq Tools](https://github.com/broadinstitute/Drop-seq) |
| GS | Reserved for backwards compatibility reasons | [SAM](https://samtools.github.io/hts-specs/SAMtags.pdf) |
| HI | Query hit index | [SAM](https://samtools.github.io/hts-specs/SAMtags.pdf) |
| MD | String for mismatching indexes | [SAM](https://samtools.github.io/hts-specs/SAMtags.pdf) |
| NH | Number of reported alignments that contain the query in the current record | [SAM](https://samtools.github.io/hts-specs/SAMtags.pdf) |
| NM | Edit difference to the reference | [SAM](https://samtools.github.io/hts-specs/SAMtags.pdf) |
| RG | Read group | [SAM](https://samtools.github.io/hts-specs/SAMtags.pdf) |
| UB | Chromium molecular barcode sequence that is error-corrected among other molecular barcodes with the same cellular barcode and gene alignment | [10X Genomics](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/bam) |
| UG | Read group ID | [Umi-tools](https://github.com/CGATOxford/UMI-tools) |
| UR | Chromium molecular barcode sequence as reported by the sequencer | [10X Genomics](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/bam) |
| UY | Chromium molecular barcode read quality. Phred scores as reported by sequencer | [10X Genomics](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/bam) |
| XF | Extra alignment flags | [10X Genomics](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/bam) | 
| jI | [jI:B:I,Start1,End1,Start2,End2,...] Start and End of introns for all junctions (1-based) | [STAR](https://github.com/alexdobin/STAR) |
| jM | [jM:B:c,M1,M2,...] Intron motifs for all junctions (i.e. N in CIGAR)| [STAR](https://github.com/alexdobin/STAR) |
| nM | The number of mismatches per (paired) alignment | [STAR](https://github.com/alexdobin/STAR) |


