<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Pipelines/PairedTag_Pipeline/README">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">Paired-Tag Overview | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/PairedTag_Pipeline/README"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Paired-Tag Overview | WARP"><meta data-rh="true" name="description" content="|                          Pipeline Version                           | Date Updated | Documentation Author | Questions or Feedback |"><meta data-rh="true" property="og:description" content="|                          Pipeline Version                           | Date Updated | Documentation Author | Questions or Feedback |"><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/PairedTag_Pipeline/README"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/PairedTag_Pipeline/README" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/PairedTag_Pipeline/README" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.91c3cfb7.js" as="script">
<link rel="preload" href="/warp/assets/js/main.affc3f00.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/ATAC/README">ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/BuildIndices_Pipeline/README">BuildIndices</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README">Exome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Exome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/README">Illumina Genotyping Array</a><button aria-label="Toggle the collapsible sidebar category &#x27;Illumina Genotyping Array&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/ImputationBeagle_Pipeline/README">ImputationBeagle</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Imputation_Pipeline/README">Imputation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Imputation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/JointGenotyping_Pipeline/README">JointGenotyping</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Multiome_Pipeline/README">Multiome scATAC and GEX</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus</a><button aria-label="Toggle the collapsible sidebar category &#x27;Optimus&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/PairedTag_Pipeline/README">Paired-Tag</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README">RNA with UMIs</a><button aria-label="Toggle the collapsible sidebar category &#x27;RNA with UMIs&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/SlideSeq_Pipeline/README">Slide-seq</a><button aria-label="Toggle the collapsible sidebar category &#x27;Slide-seq&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/SlideTags_Pipeline/README">Slide-tags</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README">Smart-seq2 Single Nucleus Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Nucleus Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/snm3C/README">Single Nucleus Methyl-Seq and Chromatin Capture</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Nucleus Methyl-Seq and Chromatin Capture&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Ultima_Genomics_Whole_Genome_Germline_Pipeline/README">Ultima Genomics Whole Genome Germline</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/README">Whole Genome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Whole Genome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">WARP Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Paired-Tag</span><meta itemprop="position" content="2"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Paired-Tag Overview</h1><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">PairedTag_v1.10.0</a></td><td align="center">January, 2025</td><td align="center">WARP Pipelines</td><td align="center">Please <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">file an issue in WARP</a>.</td></tr></tbody></table><p><img loading="lazy" alt="pairedtag_diagram" src="/warp/assets/images/pairedtag_diagram-73f6a873b7625ba8545a4b5cf52d657c.png" width="1400" height="2019" class="img_ev3q"></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="introduction-to-the-paired-tag-workflow">Introduction to the Paired-Tag workflow<a href="#introduction-to-the-paired-tag-workflow" class="hash-link" aria-label="Direct link to Introduction to the Paired-Tag workflow" title="Direct link to Introduction to the Paired-Tag workflow">​</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/paired_tag/PairedTag.wdl" target="_blank" rel="noopener noreferrer">Paired-Tag workflow</a> is an open-source, cloud-optimized pipeline developed in collaboration with the <a href="https://biccn.org/" target="_blank" rel="noopener noreferrer">BRAIN Initiative Cell Census Network</a> (BICCN) and the BRAIN Initiative Cell Atlas Network (BICAN). It supports the processing of 3&#x27; single-nucleus histone modification data (generated with the <a href="https://www.nature.com/articles/s41594-023-01060-1" target="_blank" rel="noopener noreferrer">paired-tag protocol</a>]) and 10x gene expression (GEX) data generated with the <a href="https://www.10xgenomics.com/products/single-cell-multiome-atac-plus-gene-expression" target="_blank" rel="noopener noreferrer">10x Chromium Multiome assay</a>.</p><p>The workflow is a wrapper WDL script that calls two subworkflows: the Optimus workflow for single-nucleus GEX data and the ATAC workflow for single-nucleus histone modification data.</p><p>The <a href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus workflow</a> (GEX) corrects cell barcodes (CBs) and Unique Molecular Identifiers (UMIs), aligns reads to the genome, calculates per-barcode and per-gene quality metrics, and produces a raw cell-by-gene count matrix.</p><p>The <a href="/warp/docs/Pipelines/ATAC/README">ATAC workflow</a> (histone modification) performs demultiplexing for samples that have a preindex barcode, corrects CBs, aligns reads to the genome, calculates per-barcode quality metrics, and produces a fragment file.</p><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/paired_tag/PairedTag.wdl" target="_blank" rel="noopener noreferrer">wrapper WDL</a> is available in the <a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer">WARP repository</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="quickstart-table">Quickstart table<a href="#quickstart-table" class="hash-link" aria-label="Direct link to Quickstart table" title="Direct link to Quickstart table">​</a></h2><p>The following table provides a quick glance at the Paired-Tag pipeline features:</p><table><thead><tr><th>Pipeline features</th><th>Description</th><th>Source</th></tr></thead><tbody><tr><td>Assay type</td><td>Droplet Paired-Tag (parallel analysis of individual cells for RNA expression and DNA from targeted tagmentation by sequencing)</td><td><a href="https://www.nature.com/articles/s41594-023-01060-1" target="_blank" rel="noopener noreferrer">Xie et al. 2023</a></td></tr><tr><td>Overall workflow</td><td>Barcode correction, read alignment, gene and fragment quantification</td><td>Code available on <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/paired_tag/PairedTag.wdl" target="_blank" rel="noopener noreferrer">GitHub</a></td></tr><tr><td>Workflow language</td><td>WDL 1.0</td><td><a href="https://github.com/openwdl/wdl" target="_blank" rel="noopener noreferrer">openWDL</a></td></tr><tr><td>Genomic Reference Sequence</td><td>GRCh38 human genome primary sequence</td><td>GENCODE <a href="https://www.gencodegenes.org/human/release_43.html" target="_blank" rel="noopener noreferrer">human reference files</a></td></tr><tr><td>Gene annotation reference (GTF)</td><td>Reference containing gene annotations</td><td><a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_43/gencode.v43.annotation.gtf.gz" target="_blank" rel="noopener noreferrer">GENCODE human GTF</a></td></tr><tr><td>Aligners</td><td>STARsolo (GEX), BWA-mem2 (ATAC)</td><td><a href="https://www.biorxiv.org/content/10.1101/2021.05.05.442755v1" target="_blank" rel="noopener noreferrer">Kaminow et al. 2021</a>, <a href="https://ieeexplore.ieee.org/document/8820962" target="_blank" rel="noopener noreferrer">Vasimuddin et al. 2019</a></td></tr><tr><td>Transcript and fragment quantification</td><td>STARsolo (GEX), SnapATAC2 (ATAC)</td><td><a href="https://www.biorxiv.org/content/10.1101/2021.05.05.442755v1" target="_blank" rel="noopener noreferrer">Kaminow et al. 2021</a>, <a href="https://kzhang.org/SnapATAC2/" target="_blank" rel="noopener noreferrer">SnapATAC2</a></td></tr><tr><td>Data input file format</td><td>File format in which sequencing data is provided</td><td><a href="https://academic.oup.com/nar/article/38/6/1767/3112533" target="_blank" rel="noopener noreferrer">FASTQ</a></td></tr><tr><td>Data output file format</td><td>File formats in which Multiome output is provided</td><td><a href="http://samtools.github.io/hts-specs/" target="_blank" rel="noopener noreferrer">BAM</a> and <a href="https://anndata.readthedocs.io/en/latest/" target="_blank" rel="noopener noreferrer">h5ad</a></td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="set-up">Set-up<a href="#set-up" class="hash-link" aria-label="Direct link to Set-up" title="Direct link to Set-up">​</a></h2><h3 class="anchor anchorWithStickyNavbar_LWe7" id="paired-tag-installation">Paired-Tag installation<a href="#paired-tag-installation" class="hash-link" aria-label="Direct link to Paired-Tag installation" title="Direct link to Paired-Tag installation">​</a></h3><p>To download the latest Paired-Tag release, see the release tags prefixed with &quot;PairedTag&quot; on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a>. All Paired-Tag pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/paired_tag/PairedTag.changelog.md" target="_blank" rel="noopener noreferrer">Paired-Tag changelog</a>. </p><p>To search releases of this and other pipelines, use the WARP command-line tool <a href="https://github.com/broadinstitute/warp/tree/master/wreleaser" target="_blank" rel="noopener noreferrer">Wreleaser</a>.</p><p>If you’re running a Paired-Tag workflow version prior to the latest release, the accompanying documentation for that release may be downloaded with the source code on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a> (see the folder <code>website/docs/Pipelines/PairedTag_Pipeline</code>).</p><p>The Paired-Tag pipeline can be deployed using <a href="https://cromwell.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH-compliant, flexible workflow management system that supports multiple computing platforms. The workflow can also be run in <a href="https://app.terra.bio" target="_blank" rel="noopener noreferrer">Terra</a>, a cloud-based analysis platform.</p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="inputs">Inputs<a href="#inputs" class="hash-link" aria-label="Direct link to Inputs" title="Direct link to Inputs">​</a></h3><p>The Paired-Tag workflow inputs are specified in JSON configuration files. Example configuration files can be found in the <a href="https://github.com/broadinstitute/warp/tree/develop/pipelines/wdl/paired_tag/test_inputs" target="_blank" rel="noopener noreferrer"><code>test_inputs</code></a> folder in the WARP repository.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="input-descriptions">Input descriptions<a href="#input-descriptions" class="hash-link" aria-label="Direct link to Input descriptions" title="Direct link to Input descriptions">​</a></h4><table><thead><tr><th>Parameter name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>input_id</td><td>Unique identifier describing the biological sample or replicate that corresponds with the FASTQ files; can be a human-readable name or UUID.</td><td>String</td></tr><tr><td>gex_nhash_id</td><td>Optional identifier that can be used to demarcate the gene expression library aliquot or sample.</td><td></td></tr><tr><td>atac_nhash_id</td><td>Optional identifier that can be used to demarcate the ATAC library aliquot or sample.</td><td></td></tr><tr><td>counting_mode</td><td>Optional string that determines whether the Optimus (GEX) pipeline should be run in single-cell mode (sc_rna) or single-nucleus mode (sn_rna); default is &quot;sn_rna&quot;.</td><td>String</td></tr><tr><td>gex_r1_fastq</td><td>Array of read 1 FASTQ files representing a single GEX 10x library.</td><td>Array<!-- -->[File]</td></tr><tr><td>gex_r2_fastq</td><td>Array of read 2 FASTQ files representing a single GEX 10x library.</td><td>Array<!-- -->[File]</td></tr><tr><td>gex_i1_fastq</td><td>Optional array of index FASTQ files representing a single GEX 10x library; multiplexed samples are not currently supported, but the file may be passed to the pipeline.</td><td>Array<!-- -->[File]</td></tr><tr><td>tar_star_reference</td><td>TAR file containing a species-specific reference genome and GTF for Optimus (GEX) pipeline.</td><td>File</td></tr><tr><td>annotations_gtf</td><td>GTF file containing gene annotations used for GEX cell metric calculation and ATAC fragment metrics; must match the GTF used to build the STAR aligner.</td><td>File</td></tr><tr><td>mt_genes</td><td>Optional file for the Optimus (GEX) pipeline containing mitochondrial gene names used for metric calculation; default assumes &#x27;mt&#x27; prefix in GTF (case insensitive).</td><td>File</td></tr><tr><td>tenx_chemistry_version</td><td>Optional integer for the Optimus (GEX) pipeline specifying the 10x version chemistry the data was generated with; validated by examination of the first read 1 FASTQ file read structure; default is &quot;3&quot;.</td><td>Integer</td></tr><tr><td>emptydrops_lower</td><td><strong>Not used for single-nucleus data.</strong> Optional threshold for UMIs for the Optimus (GEX) pipeline that empty drops tool should consider for determining cell; data below threshold is not removed; default is &quot;100&quot;.</td><td>Integer</td></tr><tr><td>force_no_check</td><td>Optional boolean for the Optimus (GEX) pipeline indicating if the pipeline should perform checks; default is &quot;false&quot;.</td><td>Boolean</td></tr><tr><td>ignore_r1_read_length</td><td>Optional boolean for the Optimus (GEX) pipeline indicating if the pipeline should ignore barcode chemistry check; if &quot;true&quot;, the workflow will not ensure the <code>10x_chemistry_version</code> input matches the chemistry in the read 1 FASTQ; default is &quot;false&quot;.</td><td>Boolean</td></tr><tr><td>star_strand_mode</td><td>Optional string for the Optimus (GEX) pipeline for performing STARsolo alignment on forward stranded, reverse stranded, or unstranded data; default is &quot;Forward&quot;.</td><td>String</td></tr><tr><td>count_exons</td><td>Optional boolean for the Optimus (GEX) pipeline indicating if the workflow should calculate exon counts <strong>when in single-nucleus (sn_rna) mode</strong>; if &quot;true&quot; in sc_rna mode, the workflow will return an error; default is &quot;false&quot;.</td><td>Boolean</td></tr><tr><td>gex_whitelist</td><td>Optional file containing the list of valid barcodes for 10x multiome GEX data; default is &quot;gs://gcp-public-data--broad-references/RNA/resources/arc-v1/737K-arc-v1_gex.txt&quot;.</td><td>File</td></tr><tr><td>soloMultiMappers</td><td>Optional string describing whether or not the Optimus (GEX) pipeline should run STARsolo with the <code>--soloMultiMappers</code> flag; default is &quot;EM&quot;.</td><td>String</td></tr><tr><td>atac_r1_fastq</td><td>Array of read 1 paired-end FASTQ files representing a single paired-tag DNA library.</td><td>Array<!-- -->[File]</td></tr><tr><td>atac_r2_fastq</td><td>Array of barcodes FASTQ files representing a single paired-tag DNA library.</td><td>Array<!-- -->[File]</td></tr><tr><td>atac_r3_fastq</td><td>Array of read 2 paired-end FASTQ files representing a single paired-tag DNA library.</td><td>Array<!-- -->[File]</td></tr><tr><td>tar_bwa_reference</td><td>TAR file containing the reference index files for BWA-mem alignment for the ATAC pipeline.</td><td>File</td></tr><tr><td>chrom_sizes</td><td>File containing the genome chromosome sizes; used to calculate ATAC fragment file metrics.</td><td>File</td></tr><tr><td>adapter_seq_read1</td><td>Optional string describing the adapter sequence for ATAC read 1 paired-end reads to be used during adapter trimming with Cutadapt; default is &quot;GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG&quot;.</td><td>String</td></tr><tr><td>adapter_seq_read3</td><td>Optional string describing the adapter sequence for ATAC read 2 paired-end reads to be used during adapter trimming with Cutadapt; default is &quot;TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG&quot;.</td><td>String</td></tr><tr><td>atac_whitelist</td><td>Optional file containing the list of valid barcodes for 10x multiome ATAC adata; default is &quot;gs://gcp-public-data--broad-references/RNA/resources/arc-v1/737K-arc-v1_atac.txt&quot;.</td><td>File</td></tr><tr><td>preindex</td><td>Optional boolean for the ATAC workflow; if “true”, the pipeline will run the ATAC workflow with a preindexing task necessary for processing of droplet-based Paired-Tag data where sample barcodes from read2 are combined with cell barcodes into the BB tag of the output BAM file; if “false”, the pipeline will run the ATAC workflow without preindexing and cell barcodes are stored in the CB tag of the output BAM file; default is “true”.</td><td>Boolean</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="paired-tag-tasks-and-tools">Paired-Tag tasks and tools<a href="#paired-tag-tasks-and-tools" class="hash-link" aria-label="Direct link to Paired-Tag tasks and tools" title="Direct link to Paired-Tag tasks and tools">​</a></h2><p>The Paired-Tag workflow calls two WARP subworkflows and an additional task which are described briefly in the table below. For more details on each subworkflow and task, see the documentation and WDL scripts linked in the table.</p><table><thead><tr><th>Subworkflow/Task</th><th>Software</th><th>Description</th></tr></thead><tbody><tr><td>Optimus (<a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/optimus/Optimus.wdl" target="_blank" rel="noopener noreferrer">WDL</a> and <a href="/warp/docs/Pipelines/Optimus_Pipeline/README">documentation</a>)</td><td>fastqprocess, STARsolo, Emptydrops</td><td>Workflow used to analyze 10x single-cell GEX data.</td></tr><tr><td>PairedTagDemultiplex as demultiplex (<a href="https://github.com/broadinstitute/warp/blob/develop/tasks/wdl/PairedTagUtils.wdl" target="_blank" rel="noopener noreferrer">WDL</a>)</td><td>UPStools</td><td>Task used to check the length of the read2 FASTQ (should be either 27 or 24 bp). If <code>preindex</code> is set to true, the task will perform demultiplexing of the 3-bp sample barcode from the read2 ATAC fastq files and stores it in the readname. It will then perform barcode orientation checking. The ATAC workflow will then add a combined 3 bp sample barcode and cellular barcode to the BB tag of the BAM. If <code>preindex</code> is false and then length is 27 bp, the task will perform trimming and subsequent barcode orientation checking.</td></tr><tr><td>ATAC (<a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/atac/atac.wdl" target="_blank" rel="noopener noreferrer">WDL</a> and <a href="/warp/docs/Pipelines/ATAC/README">documentation</a>)</td><td>fastqprocess, bwa-mem, SnapATAC2</td><td>Workflow used to analyze single-nucleus paired-tag DNA (histone modifications) data.</td></tr><tr><td>ParseBarcodes as ParseBarcodes (<a href="https://github.com/broadinstitute/warp/blob/develop/tasks/wdl/PairedTagUtils.wdl" target="_blank" rel="noopener noreferrer">WDL</a>)</td><td>python3</td><td>Task used to parse and split the cell barcodes and sample barcodes from the combined index in the h5ad and fragment files when <code>preindex</code> is set to true.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="outputs">Outputs<a href="#outputs" class="hash-link" aria-label="Direct link to Outputs" title="Direct link to Outputs">​</a></h2><table><thead><tr><th>Output variable name</th><th>Filename, if applicable</th><th>Output format and description</th></tr></thead><tbody><tr><td>pairedtag_pipeline_version_out</td><td>N/A</td><td>String describing the version of the Paired-Tag pipeline used.</td></tr><tr><td>bam_aligned_output_atac</td><td><code>&lt;input_id&gt;_atac.bam</code> or if <code>preindex</code> = true, <code>&lt;input_id&gt;_atac.bam.BB.bam</code></td><td>BAM file containing aligned reads from ATAC workflow; contains sample and cell barcodes stored in the BB tag if <code>preindex</code> is “true”.</td></tr><tr><td>fragment_file_atac</td><td><code>&lt;input_id&gt;_atac.fragments.sorted.tsv.gz</code></td><td>Bgzipped TSV file containing fragment start and stop coordinates per barcode. The columns are &quot;Chromosome&quot;, &quot;Start&quot;, &quot;Stop&quot;, &quot;Barcode&quot;, and &quot;Number of reads&quot;. When <code>preindex</code> = true, the file has additional columns include &quot;Sample Barcode&quot;, &quot;Cell Barcode&quot;, and &quot;Duplicates&quot; (which indicates if a cell barcode matches more than one sample barcode).</td></tr><tr><td>snap_metrics_atac</td><td><code>&lt;input_id&gt;_atac.metrics.h5ad</code></td><td>h5ad (Anndata) file containing per-barcode metrics from SnapATAC2. See the <a href="/warp/docs/Pipelines/ATAC/count-matrix-overview">ATAC Count Matrix Overview</a> for more details. If the preindex option is used, the h5ad.obs will contain 3 extra columns: preindex (the sample barcode), CB (cell barcodes), and duplicates (indicates with a 1 if the cell barcode matches more than preindex, otherwise it is 0).</td></tr><tr><td>genomic_reference_version_gex</td><td><code>&lt;reference_version&gt;.txt</code></td><td>File containing the Genome build, source and GTF annotation version.</td></tr><tr><td>bam_gex</td><td><code>&lt;input_id&gt;_gex.bam</code></td><td>BAM file containing aligned reads from Optimus workflow.</td></tr><tr><td>matrix_gex</td><td><code>&lt;input_id&gt;_gex_sparse_counts.npz</code></td><td>NPZ file containing raw gene by cell counts.</td></tr><tr><td>matrix_row_index_gex</td><td><code>&lt;input_id&gt;_gex_sparse_counts_row_index.npy</code></td><td>NPY file containing the row indices.</td></tr><tr><td>matrix_col_index_gex</td><td><code>&lt;input_id&gt;_gex_sparse_counts_col_index.npy</code></td><td>NPY file containing the column indices.</td></tr><tr><td>cell_metrics_gex</td><td><code>&lt;input_id&gt;_gex.cell_metrics.csv.gz</code></td><td>CSV file containing the per-cell (barcode) metrics.</td></tr><tr><td>gene_metrics_gex</td><td><code>&lt;input_id&gt;_gex.gene_metrics.csv.gz</code></td><td>CSV file containing the per-gene metrics.</td></tr><tr><td>cell_calls_gex</td><td><code>&lt;input_id&gt;_gex.emptyDrops</code></td><td>TSV file containing the EmptyDrops results when the Optimus workflow is run in sc_rna mode.</td></tr><tr><td>h5ad_output_file_gex</td><td><code>&lt;input_id&gt;_gex.h5ad</code></td><td>h5ad (Anndata) file containing the raw cell-by-gene count matrix, gene metrics, cell metrics, and global attributes. See the <a href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a> for more details.</td></tr><tr><td>library_metrics</td><td><code>&lt;input_id&gt;_gex_&lt;gex_nhash_id&gt;_library_metrics.csv</code></td><td>Optional CSV file containing all library-level metrics calculated with STARsolo for gene expression data.</td></tr><tr><td>atac_library_final</td><td><code>&lt;input_id&gt;_atac_&lt;atac_nhash_id&gt;_library_metrics</code></td><td>CSV file containing all the library-level metrics calucalted by SnapATAC2.</td></tr><tr><td>cloud_provider</td><td>N/A</td><td>String describing the cloud provider that should be used to run the workflow; value should be &quot;gcp&quot; or &quot;azure&quot;.</td></tr><tr><td>multimappers_EM_matrix</td><td><code>UniqueAndMult-EM.mtx</code></td><td>Optional output produced when <code>soloMultiMappers</code> is &quot;EM&quot; (default); see STARsolo <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md#multi-gene-reads" target="_blank" rel="noopener noreferrer">documentation</a> for more information.</td></tr><tr><td>multimappers_Uniform_matrix</td><td><code>UniqueAndMult-Uniform.mtx</code></td><td>Optional output produced when <code>soloMultiMappers</code> is &quot;Uniform&quot;; see STARsolo <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md#multi-gene-reads" target="_blank" rel="noopener noreferrer">documentation</a> for more information.</td></tr><tr><td>multimappers_Rescue_matrix</td><td><code>UniqueAndMult-Rescue.mtx</code></td><td>Optional output produced when <code>soloMultiMappers</code> is &quot;Rescue&quot;; see STARsolo <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md#multi-gene-reads" target="_blank" rel="noopener noreferrer">documentation</a> for more information.</td></tr><tr><td>multimappers_PropUnique_matrix</td><td><code>UniqueAndMult-PropUnique.mtx</code></td><td>Optional output produced when <code>soloMultiMappers</code> is &quot;PropUnique&quot;; see STARsolo <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md#multi-gene-reads" target="_blank" rel="noopener noreferrer">documentation</a> for more information.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="versioning-and-testing">Versioning and testing<a href="#versioning-and-testing" class="hash-link" aria-label="Direct link to Versioning and testing" title="Direct link to Versioning and testing">​</a></h2><p>All Paired-Tag pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/paired_tag/PairedTag.wdl" target="_blank" rel="noopener noreferrer">Paired-Tag changelog</a> and tested using <a href="https://github.com/broadinstitute/warp/tree/develop/pipelines/wdl/paired_tag/test_inputs" target="_blank" rel="noopener noreferrer">plumbing and scientific test data</a>. To learn more about WARP pipeline testing, see <a href="https://broadinstitute.github.io/warp/docs/About_WARP/TestingPipelines" target="_blank" rel="noopener noreferrer">Testing Pipelines</a>. Note that paired-tag tests are still in development.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="citing-the-paired-tag-pipeline">Citing the Paired-Tag Pipeline<a href="#citing-the-paired-tag-pipeline" class="hash-link" aria-label="Direct link to Citing the Paired-Tag Pipeline" title="Direct link to Citing the Paired-Tag Pipeline">​</a></h2><p>If you use the Paired-Tag Pipeline in your research, please identify the pipeline in your methods section using the <a href="https://scicrunch.org/resources/data/record/nlx_144509-1/SCR_025042/resolver?q=paired_tag&amp;l=paired_tag&amp;i=rrid:scr_025042" target="_blank" rel="noopener noreferrer">Paired-Tag SciCrunch resource identifier</a>.</p><ul><li>Ex: <em>Paired-Tag Pipeline (RRID:SCR_025041)</em></li></ul><p>Please also consider citing our preprint:</p><p>Degatano, K., Awdeh, A., Cox III, R.S., Dingman, W., Grant, G., Khajouei, F., Kiernan, E., Konwar, K., Mathews, K.L., Palis, K., et al. Warp Analysis Research Pipelines: Cloud-optimized workflows for biological data processing and reproducible analysis. Bioinformatics 2025; btaf494. <a href="https://doi.org/10.1093/bioinformatics/btaf494" target="_blank" rel="noopener noreferrer">https://doi.org/10.1093/bioinformatics/btaf494</a></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="consortia-support">Consortia support<a href="#consortia-support" class="hash-link" aria-label="Direct link to Consortia support" title="Direct link to Consortia support">​</a></h2><p>This pipeline is supported by the <a href="https://braininitiative.nih.gov/" target="_blank" rel="noopener noreferrer">BRAIN Initiative</a> (BICCN and BICAN). </p><p>If your organization also uses this pipeline, we would like to list you! Please reach out to us by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="acknowledgements">Acknowledgements<a href="#acknowledgements" class="hash-link" aria-label="Direct link to Acknowledgements" title="Direct link to Acknowledgements">​</a></h2><p>We are immensely grateful to the members of the BRAIN Initiative (BICAN Sequencing Working Group) and SCORCH for their invaluable and exceptional contributions to this pipeline. Our heartfelt appreciation goes to Dr. Bing Ren&#x27;s lab, Yang Xie, and Lei Chang for their unwavering dedication and remarkable efforts.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="feedback">Feedback<a href="#feedback" class="hash-link" aria-label="Direct link to Feedback" title="Direct link to Feedback">​</a></h2><p>Please help us make our tools better by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>; we welcome pipeline-related suggestions or questions.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/PairedTag_Pipeline/README.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2025-10-27T14:02:21.000Z">Oct 27, 2025</time></b> by <b>Nikelle Petrillo</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/Pipelines/Optimus_Pipeline/starsolo-metrics"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">STAR Aligner Metrics</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">RNA with UMIs Overview</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-paired-tag-workflow" class="table-of-contents__link toc-highlight">Introduction to the Paired-Tag workflow</a></li><li><a href="#quickstart-table" class="table-of-contents__link toc-highlight">Quickstart table</a></li><li><a href="#set-up" class="table-of-contents__link toc-highlight">Set-up</a><ul><li><a href="#paired-tag-installation" class="table-of-contents__link toc-highlight">Paired-Tag installation</a></li><li><a href="#inputs" class="table-of-contents__link toc-highlight">Inputs</a></li></ul></li><li><a href="#paired-tag-tasks-and-tools" class="table-of-contents__link toc-highlight">Paired-Tag tasks and tools</a></li><li><a href="#outputs" class="table-of-contents__link toc-highlight">Outputs</a></li><li><a href="#versioning-and-testing" class="table-of-contents__link toc-highlight">Versioning and testing</a></li><li><a href="#citing-the-paired-tag-pipeline" class="table-of-contents__link toc-highlight">Citing the Paired-Tag Pipeline</a></li><li><a href="#consortia-support" class="table-of-contents__link toc-highlight">Consortia support</a></li><li><a href="#acknowledgements" class="table-of-contents__link toc-highlight">Acknowledgements</a></li><li><a href="#feedback" class="table-of-contents__link toc-highlight">Feedback</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.91c3cfb7.js"></script>
<script src="/warp/assets/js/main.affc3f00.js"></script>
</body>
</html>