<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">Smart-seq2 Single Nucleus Multi-Sample Overview | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Smart-seq2 Single Nucleus Multi-Sample Overview | WARP"><meta data-rh="true" name="description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><meta data-rh="true" property="og:description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.fefd6ae4.js" as="script">
<link rel="preload" href="/warp/assets/js/main.435ed8cb.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/ATAC/README">ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/BuildIndices_Pipeline/README">BuildIndices</a><button aria-label="Toggle the collapsible sidebar category &#x27;BuildIndices&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README">Exome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Exome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/README">Illumina Genotyping Array</a><button aria-label="Toggle the collapsible sidebar category &#x27;Illumina Genotyping Array&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/ImputationBeagle_Pipeline/README">ImputationBeagle</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Imputation_Pipeline/README">Imputation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Imputation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/JointGenotyping_Pipeline/README">JointGenotyping</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Multiome_Pipeline/README">Multiome scATAC and GEX</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus</a><button aria-label="Toggle the collapsible sidebar category &#x27;Optimus&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/PairedTag_Pipeline/README">Paired-Tag</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README">RNA with UMIs</a><button aria-label="Toggle the collapsible sidebar category &#x27;RNA with UMIs&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/SlideSeq_Pipeline/README">Slide-seq</a><button aria-label="Toggle the collapsible sidebar category &#x27;Slide-seq&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/SlideTags_Pipeline/README">Slide-tags</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible menu__list-item-collapsible--active"><a class="menu__link menu__link--sublist menu__link--active" aria-current="page" aria-expanded="true" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README">Smart-seq2 Single Nucleus Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Nucleus Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/multi_snss2.methods">Smart-seq2 Single Nucleus Multi-Sample v1.3.4 Publication Methods</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/count-matrix-overview">Smart-seq2 Single Nucleus Multi-Sample Count Matrix Overview</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/consortia-processing">Consortia Data Processing</a></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/snm3C/README">Single Nucleus Methyl-Seq and Chromatin Capture</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Nucleus Methyl-Seq and Chromatin Capture&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Ultima_Genomics_Whole_Genome_Germline_Pipeline/README">Ultima Genomics Whole Genome Germline</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/README">Whole Genome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Whole Genome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">WARP Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Smart-seq2 Single Nucleus Multi-Sample</span><meta itemprop="position" content="2"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Smart-seq2 Single Nucleus Multi-Sample Overview</h1><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">MultiSampleSmartSeq2SingleNuclei_v2.0.1</a></td><td align="center">September, 2024</td><td align="center">Elizabeth Kiernan</td><td align="center">Please <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">file an issue in WARP</a>.</td></tr></tbody></table><p><img loading="lazy" src="/warp/assets/images/snSS2-49fa29b5c051bd785b5af73abd3bfd35.png" width="884" height="1246" class="img_ev3q"></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="introduction-to-the-smart-seq2-single-nucleus-multi-sample-pipeline">Introduction to the Smart-seq2 Single Nucleus Multi-Sample pipeline<a href="#introduction-to-the-smart-seq2-single-nucleus-multi-sample-pipeline" class="hash-link" aria-label="Direct link to Introduction to the Smart-seq2 Single Nucleus Multi-Sample pipeline" title="Direct link to Introduction to the Smart-seq2 Single Nucleus Multi-Sample pipeline">​</a></h2><p>The Smart-seq2 Single Nucleus Multi-Sample (Multi-snSS2) pipeline was developed in collaboration with the <a href="https://biccn.org/" target="_blank" rel="noopener noreferrer">BRAIN Initiative Cell Census Network</a> (BICCN) to process single-nucleus RNAseq (snRNAseq) data generated by <a href="https://www.nature.com/articles/nmeth.2639" target="_blank" rel="noopener noreferrer">Smart-seq2 assays</a>. The pipeline&#x27;s workflow is written in WDL, is freely available in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_nucleus_multisample/MultiSampleSmartSeq2SingleNucleus.wdl" target="_blank" rel="noopener noreferrer">WARP repository</a> on GitHub, and can be run by any compliant WDL runner (e.g. <a href="https://github.com/broadinstitute/cromwell" target="_blank" rel="noopener noreferrer">Crowmell</a>). </p><p>The pipeline is designed to process snRNA-seq data from multiple cells. Overall, the workflow trims paired-end FASTQ files, aligns reads to the genome using a modified GTF, <a href="#6-creating-the-h5ad-cell-by-gene-matrix">counts intronic and exonic reads</a>, and calculates quality control metrics.</p><p>The pipeline has been scientifically validated by the BRAIN Institute. Read more in the <a href="#validation">validation section</a>.</p><div class="theme-admonition theme-admonition-tip alert alert--success admonition_LlT9"><div class="admonitionHeading_tbUL"><span class="admonitionIcon_kALy"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Try the Multi-snSS2 workflow in Terra</div><div class="admonitionContent_S0QG"><p>You can run the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_nucleus_multisample/MultiSampleSmartSeq2SingleNucleus.wdl" target="_blank" rel="noopener noreferrer">Smart-seq2 Single Nucleus Multi-Sample workflow</a> in <a href="https://app.terra.bio" target="_blank" rel="noopener noreferrer">Terra</a>, a cloud-based analysis platform. The Terra <a href="https://app.terra.bio/#workspaces/warp-pipelines/Smart-seq2_Single_Nucleus_Muti-Sample" target="_blank" rel="noopener noreferrer">Smart-seq2 Single Nucleus Multi-Sample public workspace</a> is preloaded with the Multi-snSS2 workflow, example testing data, and all the required reference data.</p></div></div><h2 class="anchor anchorWithStickyNavbar_LWe7" id="quick-start-table">Quick start table<a href="#quick-start-table" class="hash-link" aria-label="Direct link to Quick start table" title="Direct link to Quick start table">​</a></h2><table><thead><tr><th>Pipeline features</th><th>Description</th><th>Source</th></tr></thead><tbody><tr><td>Assay type</td><td>Smart-seq2 Single Nucleus</td><td><a href="https://www.nature.com/articles/nprot.2014.006" target="_blank" rel="noopener noreferrer">Smart-seq2</a></td></tr><tr><td>Overall workflow</td><td>Quality control and transcriptome quantification.</td><td>Code available from the <a href="https://github.com/broadinstitute/warp/tree/develop/pipelines/wdl/smartseq2_single_nucleus/SmartSeq2SingleNucleus.wdl" target="_blank" rel="noopener noreferrer">WARP repository</a> in GitHub</td></tr><tr><td>Workflow language</td><td>WDL</td><td><a href="https://github.com/openwdl/wdl" target="_blank" rel="noopener noreferrer">openWDL</a></td></tr><tr><td>Genomic reference sequence (for validation)</td><td>GRCm38 mouse genome primary sequence.</td><td>GENCODE GRCm38 <a href="http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/GRCm38.primary_assembly.genome.fa.gz" target="_blank" rel="noopener noreferrer">mouse reference files</a></td></tr><tr><td>Transcriptomic reference annotation (for validation)</td><td>Modified M23 GTF built with the  <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/wdl/build_indices/BuildIndices.wdl" target="_blank" rel="noopener noreferrer">BuildIndices workflow</a>.</td><td>GENCODE <a href="https://www.gencodegenes.org/mouse/release_M23.html" target="_blank" rel="noopener noreferrer">M23 GTF</a>; modified version available <a href="gs://gcp-public-data--broad-references/mm10/v0/single_nucleus/modified_gencode.vM23.primary_assembly.annotation.gtf" target="_blank" rel="noopener noreferrer">Broad references</a></td></tr><tr><td>Aligner</td><td>STAR</td><td><a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">STAR</a></td></tr><tr><td>QC metrics</td><td>Picard</td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Broad Institute</a></td></tr><tr><td>Transcript quantification</td><td>featureCounts (utilities for counting reads to genomic features).</td><td><a href="http://subread.sourceforge.net/" target="_blank" rel="noopener noreferrer">featureCounts</a></td></tr><tr><td>Data input file format</td><td>File format in which sequencing data is provided.</td><td><a href="https://academic.oup.com/nar/article/38/6/1767/3112533" target="_blank" rel="noopener noreferrer">FASTQ</a></td></tr><tr><td>Data output file formats</td><td>File formats in which Smart-seq2 output is provided.</td><td><a href="http://samtools.github.io/hts-specs/" target="_blank" rel="noopener noreferrer">BAM</a>, Loom (counts and metrics; generated with <a href="http://loompy.org/" target="_blank" rel="noopener noreferrer">Loompy v.3.0.6)</a>, TSV (counts)</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="set-up">Set-Up<a href="#set-up" class="hash-link" aria-label="Direct link to Set-Up" title="Direct link to Set-Up">​</a></h2><h3 class="anchor anchorWithStickyNavbar_LWe7" id="installation">Installation<a href="#installation" class="hash-link" aria-label="Direct link to Installation" title="Direct link to Installation">​</a></h3><p>To download the latest release of the Multi-snSS2 pipeline along with example data and dependencies, see the release tags prefixed with &quot;MultiSampleSmartSeq2SingleNucleus&quot; on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a>. All releases of the Multi-snSS2 pipeline are documented in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_nucleus_multisample/MultiSampleSmartSeq2SingleNucleus.changelog.md" target="_blank" rel="noopener noreferrer">Multi-snSS2 changelog</a>.</p><p>To search releases of this and other pipelines, use the WARP command-line tool <a href="https://github.com/broadinstitute/warp/tree/master/wreleaser" target="_blank" rel="noopener noreferrer">Wreleaser</a>.</p><p>The workflow is deployed using <a href="https://github.com/broadinstitute/cromwell" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH compliant, flexible workflow management system that supports multiple computing platforms.</p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="inputs">Inputs<a href="#inputs" class="hash-link" aria-label="Direct link to Inputs" title="Direct link to Inputs">​</a></h3><p>There is an <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_nucleus_multisample/mouse_example.json" target="_blank" rel="noopener noreferrer">example configuration (JSON) file</a> available that you can use to test the Multi-snSS2 workflow. It points to publicly available reference files and sample paired-end FASTQs.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="sample-data-and-reference-inputs">Sample data and reference inputs<a href="#sample-data-and-reference-inputs" class="hash-link" aria-label="Direct link to Sample data and reference inputs" title="Direct link to Sample data and reference inputs">​</a></h4><p>The table below details the Multi-snSS2 inputs. The pipeline is designed to take in an array of paired-end reads in the form of two FASTQ files per cell. </p><ul><li>The example mouse reference inputs are created using the <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/wdl/build_indices" target="_blank" rel="noopener noreferrer">BuildIndices Pipeline</a>.</li><li>The pipeline modifies a given GTF downloaded from GENCODE to only include biotypes that are listed in a tab separated file (<a href="https://github.com/broadinstitute/warp-tools/blob/develop/3rd-party-tools/build-indices/Biotypes.tsv" target="_blank" rel="noopener noreferrer">biotypes.tsv</a>).</li><li>The example references do not include the pseudogene biotype. Learn more about Ensembl biotypes in the <a href="https://m.ensembl.org/info/genome/genebuild/biotypes.html" target="_blank" rel="noopener noreferrer">Ensembl overview</a>.</li><li>To enable intron counting, the workflow calls a python script, <a href="https://github.com/broadinstitute/warp-tools/blob/develop/3rd-party-tools/build-indices/add-introns-to-gtf.py" target="_blank" rel="noopener noreferrer"><code>add-introns-to-gtf.py</code></a>, to create a custom GTF with intron annotations. Introns are considered any part of a contig that is not exonic nor intergenic. </li></ul><table><thead><tr><th>Input Name</th><th>Input Description</th><th>Input Format</th></tr></thead><tbody><tr><td>genome_ref_fasta</td><td>FASTA file used for metric collection by <a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a>.</td><td>File</td></tr><tr><td>tar_star_reference</td><td>TAR file containing genome indices used for the <a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">STAR aligner</a>.</td><td>File</td></tr><tr><td>annotations_gtf</td><td>Custom GTF file containing annotations for exon and intron tagging; must match <code>tar_star_reference</code>.</td><td>File</td></tr><tr><td>adapter_list</td><td>FASTA file containing adapter sequences used in the library preparation (i.e. Illumina adapters for Illumina sequencing).</td><td>File</td></tr><tr><td>input_ids</td><td>Unique identifiers or names for each cell; can be a UUID or human-readable name.</td><td>Array of strings</td></tr><tr><td>input_names</td><td>Optional unique identifiers/names to further describe each cell. If <code>input_ids</code> is a UUID, the <code>input_names</code> could be used as a human-readable identifier.</td><td>Array of strings</td></tr><tr><td>fastq1_input_files</td><td>Cloud path to FASTQ files containing forward paired-end sequencing reads for each cell (sample); order must match the order in input_id.</td><td>Array of strings</td></tr><tr><td>fastq2_input_files</td><td>Cloud path to FASTQ files containing reverse paired-end sequencing reads for each cell (sample); order must match the order in input_id.</td><td>Array of strings</td></tr><tr><td>batch_id</td><td>Identifier for the batch of multiple samples.</td><td>String</td></tr><tr><td>batch_name</td><td>Optional string to describe the batch or biological sample.</td><td>String</td></tr><tr><td>project_id</td><td>Optional project identifier; usually a number.</td><td>Array of strings</td></tr><tr><td>project_name</td><td>Optional project identifier; usually a human-readable name.</td><td>Array of strings</td></tr><tr><td>library</td><td>Optional description of the sequencing method or approach.</td><td>Array of strings</td></tr><tr><td>organ</td><td>Optional description of the organ from which the cells were derived.</td><td>Array of strings</td></tr><tr><td>species</td><td>Optional description of the species from which the cells were derived.</td><td>Array of strings</td></tr><tr><td>input_name_metadata_field</td><td>Optional input describing, when applicable, the metadata field containing the <code>input_names</code>.</td><td>String</td></tr><tr><td>input_id_metadata_field</td><td>Optional string describing, when applicable, the metadata field containing the <code>input_ids</code>.</td><td>String</td></tr><tr><td>cloud_provider</td><td>String describing the cloud provider that should be used to run the workflow; value should be &quot;gcp&quot; or &quot;azure&quot;.</td><td>String</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="multi-snss2-tasks-and-tools">Multi-snSS2 tasks and tools<a href="#multi-snss2-tasks-and-tools" class="hash-link" aria-label="Direct link to Multi-snSS2 tasks and tools" title="Direct link to Multi-snSS2 tasks and tools">​</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_nucleus_multisample/MultiSampleSmartSeq2SingleNucleus.wdl" target="_blank" rel="noopener noreferrer">Multi-snSS2 workflow</a> imports individual &quot;tasks&quot; (written in WDL script) from the WARP <a href="https://github.com/broadinstitute/warp/tree/master/tasks/wdl/" target="_blank" rel="noopener noreferrer">tasks folder</a>.</p><p>Overall, the Multi-snSS2 workflow:</p><ol><li>Trims adapters.</li><li>Aligns reads.</li><li>Removes duplicate reads.</li><li>Calculates metrics.</li><li>Quantifies gene counts.</li><li>Merges exon counts, intron counts, and metrics into a h5ad-formatted matrix.</li></ol><p>The tools each task employs in the Multi-snSS2 workflow are detailed in the table below. </p><p>To see specific tool parameters, select the task WDL link in the table; then view the <code>command {}</code> section of the task WDL script. To view or use the exact tool software, see the task&#x27;s Docker image which is specified in the task WDL <code># runtime values</code> section as <code>String docker =</code>.</p><table><thead><tr><th>Task name and WDL link</th><th>Tool</th><th>Software</th><th>Description</th></tr></thead><tbody><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/CheckInputs.wdl" target="_blank" rel="noopener noreferrer">CheckInputs.checkInputArrays</a></td><td>---</td><td>Bash</td><td>Checks the inputs and initiates the per cell processing.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/tree/master/tasks/wdl/StarAlign.wdl" target="_blank" rel="noopener noreferrer">StarAlign.STARGenomeRefVersion</a></td><td>---</td><td>Bash</td><td>Reads the <code>tar_star_reference</code> file to obtain the genomic reference source, build version, and annotation version.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/tree/master/tasks/wdl/TrimAdapters.wdl" target="_blank" rel="noopener noreferrer">TrimAdapters.TrimAdapters</a></td><td><a href="https://github.com/ExpressionAnalysis/ea-utils/tree/master/clipper" target="_blank" rel="noopener noreferrer">fastq-mcf</a></td><td><a href="https://github.com/ExpressionAnalysis/ea-utils" target="_blank" rel="noopener noreferrer">ea-utils</a></td><td>Trims adapter sequences from the FASTQ inputs</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/tree/master/tasks/wdl/StarAlign.wdl" target="_blank" rel="noopener noreferrer">StarAlign.StarAlignFastqMultisample</a></td><td>STAR</td><td><a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">STAR</a></td><td>Aligns reads to the genome.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/tree/master/tasks/wdl/Picard.wdl" target="_blank" rel="noopener noreferrer">Picard.RemoveDuplicatesFromBam</a></td><td>MarkDuplicates, AddOrReplaceReadGroups</td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a></td><td>Removes duplicate reads, producing a new BAM output; adds regroups to deduplicated BAM.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/tree/master/tasks/wdl/Picard.wdl" target="_blank" rel="noopener noreferrer">Picard.CollectMultipleMetricsMultiSample</a></td><td>CollectMultipleMetrics</td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a></td><td>Collects QC metrics on the deduplicated BAM files.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/tree/master/tasks/wdl/FeatureCounts.wdl" target="_blank" rel="noopener noreferrer">CountAlignments.CountAlignments</a></td><td>FeatureCounts</td><td><a href="http://subread.sourceforge.net/" target="_blank" rel="noopener noreferrer">Subread</a>, Python 3</td><td>Uses a custom GTF with featureCounts and Python to mark introns, create a BAM that has alignments spanning intron-exon junctions removed, and counts exons using the custom BAM and by excluding intron tags.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/H5adUtils.wdl" target="_blank" rel="noopener noreferrer">H5adUtils.SingleNucleusSmartSeq2H5adOutput</a></td><td>Custom script: <a href="https://github.com/broadinstitute/warp-tools/blob/develop/tools/scripts/create_h5ad_snss2.py" target="_blank" rel="noopener noreferrer">create_h5ad_snss2.py</a></td><td>Python 3</td><td>Creates the matrix files (h5ad format) for each sample.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/wdl/H5adUtils.wdl" target="_blank" rel="noopener noreferrer">H5adUtils.AggregateSmartSeq2H5ad</a></td><td>Custom script: <a href="https://github.com/broadinstitute/warp-tools/blob/develop/tools/scripts/ss2_h5ad_merge.py" target="_blank" rel="noopener noreferrer">ss2_h5ad_merge.py</a></td><td>Python 3</td><td>Aggregates the matrix files (h5ad format) for each sample to produce one final h5ad output.</td></tr></tbody></table><h4 class="anchor anchorWithStickyNavbar_LWe7" id="1-trimming-adapters">1. Trimming adapters<a href="#1-trimming-adapters" class="hash-link" aria-label="Direct link to 1. Trimming adapters" title="Direct link to 1. Trimming adapters">​</a></h4><p>The TrimAdapters task uses the adapter list reference file to run the <a href="https://github.com/ExpressionAnalysis/ea-utils/tree/master/clipper" target="_blank" rel="noopener noreferrer">fastq-mcf</a> tool. This tool identifies the adapters in the input FASTQ files and performs clipping by using a subsampling parameter of 200,000 reads. The task outputs the trimmed FASTQ files which are then used for alignment. </p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="2-aligning-reads">2. Aligning reads<a href="#2-aligning-reads" class="hash-link" aria-label="Direct link to 2. Aligning reads" title="Direct link to 2. Aligning reads">​</a></h4><p>The StarAlignFastq task runs the STAR aligner on the trimmed FASTQ files. The STAR quantMode parameter is set to GeneCounts, which counts the number of reads per gene while mapping. The task outputs a coordinate-sorted aligned BAM file. </p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="3-removing-duplicate-reads">3. Removing duplicate reads<a href="#3-removing-duplicate-reads" class="hash-link" aria-label="Direct link to 3. Removing duplicate reads" title="Direct link to 3. Removing duplicate reads">​</a></h4><p>The RemoveDuplicatesFromBam task removes multi-mapped reads, optical duplicates, and PCR duplicates from the aligned BAM. It then adds readgroup information and creates a new, coordinate-sorted aligned BAM output.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="4-collecting-metrics">4. Collecting metrics<a href="#4-collecting-metrics" class="hash-link" aria-label="Direct link to 4. Collecting metrics" title="Direct link to 4. Collecting metrics">​</a></h4><p>The CollectMultipleMetrics task uses the Picard tool CollectMultipleMetrics to perform QC on the deduplicated BAM file. These metrics are copied to the final cell-by-gene matrix output (h5ad file). A detailed list of these metrics can be found in the <a href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/count-matrix-overview">Multi-snSS2 Count Matrix Overview</a>.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="5-counting-genes">5. Counting genes<a href="#5-counting-genes" class="hash-link" aria-label="Direct link to 5. Counting genes" title="Direct link to 5. Counting genes">​</a></h4><p>The CountAlignments task uses the featureCounts package to count introns and exons. First, the featureCounts tool counts intronic alignments in the deduplicated BAM using a custom GTF with annotated introns. The tool flags intronic alignments if they overlap an annotated intron by a minimum of 3 bp. </p><p>Next, following pipeline processes established by the BICCN Whole Mouse Brain Working Group, a custom Python script (“<a href="https://github.com/broadinstitute/warp/tree/master/dockers/skylab/featureCounts/remove-reads-on-junctions.py" target="_blank" rel="noopener noreferrer">remove-reads-on-junctions.py</a>”) removes all alignments in the deduplicated BAM that cross only one intron-exon junction and produces an intermediate BAM file for exon counting. This removes a negligible amount of putative intronic alignments that did not meet the 3 bp intron overlap criteria. </p><p>Lastly, featureCounts uses the intermediate BAM with junctions removed to count exons. The final outputs of this step include a cell-by-gene matrix of intronic counts, a cell-by-gene matrix of exonic counts, and summary metrics for the different count types.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="6-creating-the-h5ad-cell-by-gene-matrix">6. Creating the h5ad cell by gene matrix<a href="#6-creating-the-h5ad-cell-by-gene-matrix" class="hash-link" aria-label="Direct link to 6. Creating the h5ad cell by gene matrix" title="Direct link to 6. Creating the h5ad cell by gene matrix">​</a></h4><p>The H5adUtils task combines the Picard metrics (alignment_summary_metrics, deduplication metrics, and the G/C bias summary metrics) with the featureCount exon and intron counts to create an h5ad formatted cell-by-gene count matrix. </p><p>Read full details for all the metrics in the <a href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/count-matrix-overview">Multi-snSS2 Count Matrix Overview</a>.</p><p>The cell-by-gene matrix can be examined using <a href="https://anndata.readthedocs.io/en/latest/" target="_blank" rel="noopener noreferrer">anndata software</a>. Exonic counts are stored in the main h5ad matrix which is unnamed by default. They are the default return value of the <code>anndata.read_h5ad()</code> command. Intronic counts are stored in the h5ad as an additional layer which is named <code>intron_counts</code>.</p><p>You can also access exonic and intronic counts using anndatas&#x27;s <code>layers()</code> method. For example, <code>anndata.layers[“”]</code> will return the exonic counts from the output h5ad file. Similarly, <code>anndata.connect.layers[“intron_counts”]</code> will return the intronic counts from the output h5ad. </p><p>Whole gene counts (which include both intronic and exonic counts) can be accessed by adding the intronic and exonic counts together. </p><p>Below is example anndata code for accessing the h5ad&#x27;s exonic, intronic, and whole gene counts.</p><div class="language-python codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#bfc7d5;--prism-background-color:#292d3e"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-python codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#bfc7d5"><span class="token keyword" style="font-style:italic">import</span><span class="token plain"> anndata</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">adata </span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token plain"> anndata</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token plain">read_h5ad</span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token string" style="color:rgb(195, 232, 141)">&quot;/PATH/TO/File.h5ad&quot;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token plain"></span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">count_exons </span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token plain"> adata</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token plain">X </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic">#geneXcell table for the exonic read counts</span><span class="token plain"></span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">count_introns </span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token plain"> adata</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token plain">layers</span><span class="token punctuation" style="color:rgb(199, 146, 234)">[</span><span class="token string" style="color:rgb(195, 232, 141)">&quot;intron_counts&quot;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">]</span><span class="token plain"> </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic">#geneXcell table for the intronic read counts</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>If you would like to get the counts for both introns and exons, you can sum the counts together.</p><p>To read more about the h5ad file format and use of layers, see the <a href="https://anndata.readthedocs.io/en/latest/" target="_blank" rel="noopener noreferrer">h5ad documentation</a>.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="7-outputs">7. Outputs<a href="#7-outputs" class="hash-link" aria-label="Direct link to 7. Outputs" title="Direct link to 7. Outputs">​</a></h4><p>The table below details the final outputs of the Multi-snSS2 workflow.  </p><table><thead><tr><th>Output variable name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>h5ad_output</td><td>Cell-by-gene count matrix that includes the raw exon counts (in matrix), intron counts (in matrix layer), cell metrics (column attributes) and gene IDs (row attributes).</td><td>h5ad</td></tr><tr><td>exon_intron_count_files</td><td>Array of TXT files (one per cell) that contain intronic and exonic counts.</td><td>Array <!-- -->[TXT]</td></tr><tr><td>bam_files</td><td>Array of genome-aligned BAM files (one for each cell) generated with STAR.</td><td>Array <!-- -->[BAM]</td></tr><tr><td>pipeline_version_out</td><td>Version of the processing pipeline run on this data.</td><td>String</td></tr></tbody></table><p>The h5ad matrix is the default output. See the <a href="https://github.com/broadinstitute/warp-tools/blob/develop/tools/scripts/create_h5ad_snss2.py" target="_blank" rel="noopener noreferrer">create_h5ad_snss2.py</a> script for the detailed code. This matrix contains the count matrices, as well as the gene and cell metrics detailed in the <a href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/count-matrix-overview">Multi-snSS2 Count Matrix Overview</a>.</p><p>To facilitate downstream analysis, the output h5ad file contains both gene names and gene IDs.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="validation">Validation<a href="#validation" class="hash-link" aria-label="Direct link to Validation" title="Direct link to Validation">​</a></h2><p>The Multi-snSS2 pipeline was scientifically validated by the BRAIN Initiatives Cell Census Network (BICCN) 2.0 Whole Mouse Brain Working Group. </p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="versioning">Versioning<a href="#versioning" class="hash-link" aria-label="Direct link to Versioning" title="Direct link to Versioning">​</a></h2><p>All Multi-snSS2 release notes are documented in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/smartseq2_single_nucleus_multisample/MultiSampleSmartSeq2SingleNucleus.changelog.md" target="_blank" rel="noopener noreferrer">Multi-snSS2 changelog</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="citing-the-multi-snss2-pipeline">Citing the Multi-snSS2 Pipeline<a href="#citing-the-multi-snss2-pipeline" class="hash-link" aria-label="Direct link to Citing the Multi-snSS2 Pipeline" title="Direct link to Citing the Multi-snSS2 Pipeline">​</a></h2><p>If you use the Multi-snSS2 Pipeline in your research, please identify the pipeline in your methods section using the <a href="https://scicrunch.org/resources/data/record/nlx_144509-1/SCR_021312/resolver?q=SCR_021312&amp;l=SCR_021312&amp;i=rrid:scr_021312" target="_blank" rel="noopener noreferrer">Multi-snSS2 SciCrunch resource identifier</a>.</p><ul><li>Ex: <em>Smart-seq2 Single Nucleus Multi-Sample Pipeline (RRID:SCR_021312)</em></li></ul><p>Please also consider citing our preprint:</p><p>Degatano, K., Awdeh, A., Cox III, R.S., Dingman, W., Grant, G., Khajouei, F., Kiernan, E., Konwar, K., Mathews, K.L., Palis, K., et al. Warp Analysis Research Pipelines: Cloud-optimized workflows for biological data processing and reproducible analysis. Bioinformatics 2025; btaf494. <a href="https://doi.org/10.1093/bioinformatics/btaf494" target="_blank" rel="noopener noreferrer">https://doi.org/10.1093/bioinformatics/btaf494</a></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="consortia-support">Consortia Support<a href="#consortia-support" class="hash-link" aria-label="Direct link to Consortia Support" title="Direct link to Consortia Support">​</a></h2><p>This pipeline is supported and used by the <a href="https://biccn.org/" target="_blank" rel="noopener noreferrer">BRAIN Initiative Cell Census Network</a> (BICCN). </p><p>Each consortia may use slightly different reference files for data analysis or have different post-processing steps. Learn more by reading the <a href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/consortia-processing">Consortia Processing</a> overview.</p><p>If your organization also uses this pipeline, we would love to list you! Please reach out to us by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="feedback">Feedback<a href="#feedback" class="hash-link" aria-label="Direct link to Feedback" title="Direct link to Feedback">​</a></h2><p>Please help us make our tools better by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>; we welcome pipeline-related suggestions or questions.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2026-02-11T21:48:59.000Z">Feb 11, 2026</time></b> by <b>Robert Sidney Cox</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/Pipelines/SlideTags_Pipeline/README"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Slide-tags Overview</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/multi_snss2.methods"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Smart-seq2 Single Nucleus Multi-Sample v1.3.4 Publication Methods</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-smart-seq2-single-nucleus-multi-sample-pipeline" class="table-of-contents__link toc-highlight">Introduction to the Smart-seq2 Single Nucleus Multi-Sample pipeline</a></li><li><a href="#quick-start-table" class="table-of-contents__link toc-highlight">Quick start table</a></li><li><a href="#set-up" class="table-of-contents__link toc-highlight">Set-Up</a><ul><li><a href="#installation" class="table-of-contents__link toc-highlight">Installation</a></li><li><a href="#inputs" class="table-of-contents__link toc-highlight">Inputs</a></li></ul></li><li><a href="#multi-snss2-tasks-and-tools" class="table-of-contents__link toc-highlight">Multi-snSS2 tasks and tools</a></li><li><a href="#validation" class="table-of-contents__link toc-highlight">Validation</a></li><li><a href="#versioning" class="table-of-contents__link toc-highlight">Versioning</a></li><li><a href="#citing-the-multi-snss2-pipeline" class="table-of-contents__link toc-highlight">Citing the Multi-snSS2 Pipeline</a></li><li><a href="#consortia-support" class="table-of-contents__link toc-highlight">Consortia Support</a></li><li><a href="#feedback" class="table-of-contents__link toc-highlight">Feedback</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2026 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.fefd6ae4.js"></script>
<script src="/warp/assets/js/main.435ed8cb.js"></script>
</body>
</html>