---
sidebar_position: 4
---

# Optimus Count Matrix Overview

The Optimus pipeline's default count matrix output is a Loom file, an HDF5 file generated using [Loompy v.3.0.6](http://loompy.org/). 

It contains the raw, but UMI-corrected cell by gene counts, which vary depending on the workflow's `counting_mode` and `count_exons` parameters. If running single-cell data (`counting_mode` is sc_rna), the counts will include only exonic gene counts. If running single-nucleus data (`counting_mode` is sn_rna), the counts will be whole transcript. Additionally, if `count_exons` is set to true in sn_rna mode, the Loom will contain the whole transcript counts as well as an additional layer with exonic counts.

You can determine which type of counts are in the loom by looking at the global attribute `expression_data_type` (see Table 1 below).


The matrix also contains multiple metrics for both individual cells (the columns of the matrix; [Table 2](#table-2-column-attributes-cell-metrics)) and individual genes (the rows of the matrix; [Table 3](#table-3-row-attributes-gene-metrics)). 



:::tip Additional Matrix Processing for Consortia
Loom files generated by Optimus for consortia, such as the Human Cell Atlas (HCA) or the BRAIN Initiative Cell Census Network (BICCN), may have additional processing steps. Read the [Consortia Processing Overview](consortia-processing.md#hca-data-coordination-platform-matrix-processing) for details on consortia-specific matrix changes.
:::

## Table 1. Global attributes

The global attributes in the Loom apply to the whole file, not any specific part. 

| Attribute | Details |
| :-- | :-- |
| optimus_output_schema_version | String with the loom file spec version |
| expression_data_type | String describing if the pipeline counts exonic or whole transcript (exonic and intronic) reads. For the single-cell mode (counting_mode = sc_rna), the value will be "exonic"; for the single-nucleus mode (counting_mode = sn_rna), the value will be "whole_transcript" |
| input_id | The sample or cell id listed in the pipeline configuration file. This can be any string, but we recommend it be consistent with any sample metadata. |
| input_name | Optional string that can be used to further describe the input |
| input_id_metadata_field | Optional string that describes, when applicable, the metadata field containing the `input_id` |
| input_name_metadata_field | Optional string that describes, when applicable, the metadata field containing the `input_name` |
| pipeline_version | String describing the Optimus version |

## Table 2. Column attributes (cell metrics)

| Cell Metrics | Program |Details             |
|:---|:---:|:--------------------|
|`CellID` | [warp-tools](https://github.com/broadinstitute/warp-tools) | The unique identifier for each cell based on cell barcodes; identical to `cell_names`. |
|`cell_names` |  [warp-tools](https://github.com/broadinstitute/warp-tools) | The unique identifier for each cell based on cell barcodes; identical to `CellID`. |
| `input_id` | Provided as pipeline input | The sample or cell id listed in the pipeline configuration file. This can be any string, but we recommend it be consistent with any sample metadata. |
|`n_reads`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of reads associated with this entity. n_reads, like all metrics, are calculated from the Optimus output BAM. Prior to alignment with STARsolo, reads are checked against the whitelist (1 hamming distance). These CB-corrected reads are the input to the STAR aligner. Then, the reads also get CB correction during STAR. For this reason, almost all reads in the aligned BAM have a CB tag and UB tag. Therefore, n_reads represents CB corrected reads, not all reads in the input FASTQ files. |
|`noise_reads`|[warp-tools](https://github.com/broadinstitute/warp-tools)| Number of reads that are categorized by 10x Genomics Cell Ranger as "noise". Refers to long polymers, or reads with high numbers of N (ambiguous) nucleotides. |
|`perfect_molecule_barcodes`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of reads with molecule barcodes that have no errors. |
|`n_mitochondrial_genes`| [warp-tools](https://github.com/broadinstitute/warp-tools)| The number of mitochondrial genes detected by this cell. |
|`n_mitochondrial_molecules`| [warp-tools](https://github.com/broadinstitute/warp-tools)| The number of molecules from mitochondrial genes detected for this cell. |
|`pct_mitochondrial_molecules`| [warp-tools](https://github.com/broadinstitute/warp-tools)| The percentage of molecules from mitochondrial genes detected for this cell. |
|`reads_mapped_uniquely`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of reads mapped to a single unambiguous location in the genome. |
|`reads_mapped_multiple`|[warp-tools](https://github.com/broadinstitute/warp-tools)|The number of reads mapped to multiple genomic positions with equal confidence. |
|`spliced_reads`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of reads that overlap splicing junctions. |
|`antisense_reads`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of reads that are mapped to the antisense strand instead of the transcribed strand. |
|`molecule_barcode_fraction_bases_above_30_mean`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The average fraction of bases in molecule barcodes that receive quality scores greater than 30 across the reads of this entity. |
|`molecule_barcode_fraction_bases_above_30_variance`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The variance in the fraction of bases in molecule barcodes that receive quality scores greater than 30 across the reads of this entity. |
|`genomic_reads_fraction_bases_quality_above_30_mean`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The average fraction of bases in the genomic read that receive quality scores greater than 30 across the reads of this entity (included for 10x Cell Ranger count comparison). |
|`genomic_reads_fraction_bases_quality_above_30_variance`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The variance in the fraction of bases in the genomic read that receive quality scores greater than 30 across the reads of this entity (included for 10x Cell Ranger count comparison). |
|`genomic_read_quality_mean`|[warp-tools](https://github.com/broadinstitute/warp-tools)| Average quality of Illumina base calls in the genomic reads corresponding to this entity. |
|`genomic_read_quality_variance`|[warp-tools](https://github.com/broadinstitute/warp-tools)|Variance in quality of Illumina base calls in the genomic reads corresponding to this entity. |
|`n_molecules`|[warp-tools](https://github.com/broadinstitute/warp-tools)| Number of molecules corresponding to this entity (only reflects reads with CB and UB tags). |
|`n_fragments`|[warp-tools](https://github.com/broadinstitute/warp-tools)| Number of fragments corresponding to this entity. See README.md for the definition of a Fragment. |
|`reads_per_fragment`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The average number of reads associated with each fragment in this entity. |
|`fragments_per_molecule`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The average number of fragments associated with each molecule in this entity. |
|`fragments_with_single_read_evidence`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of fragments associated with this entity that are observed by only one read. |
|`molecules_with_single_read_evidence`|[warp-tools](https://github.com/broadinstitute/warp-tools)|The number of molecules associated with this entity that are observed by only one read. |
|`perfect_cell_barcodes`|[warp-tools](https://github.com/broadinstitute/warp-tools)|The number of reads whose cell barcodes contain no error. |
|`reads_mapped_too_many_loci`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of reads that were mapped to too many loci across the genome and as a consequence, are reported unmapped by the aligner. |
|`cell_barcode_fraction_bases_above_30_variance`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The variance of the fraction of Illumina base calls for the cell barcode sequence that are greater than 30, across molecules. |
|`cell_barcode_fraction_bases_above_30_mean`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The average fraction of Illumina base calls for the cell barcode sequences that are greater than 30, across molecules. |
|`n_genes`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of genes detected by this cell. |
|`genes_detected_multiple_observations`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of genes that are observed by more than one read in this cell. |
| `emptydrops_FDR` | [dropletUtils](https://bioconductor.org/packages/release/bioc/html/DropletUtils.html) | False Discovery Rate (FDR) for being a non-empty droplet; single-cell data will read "NA" if task is unable to detect knee point inflection. Column is not included for data run in the sn_rna mode |
| `emptydrops_IsCell` | [dropletUtils](https://bioconductor.org/packages/release/bioc/html/DropletUtils.html) | Binarized call of cell/background based on predefined FDR cutoff; single-cell data will read "NA" if task is unable to detect knee point inflection. Column is not included for data run in the sn_rna mode |
| `emptydrops_Limited` | [dropletUtils](https://bioconductor.org/packages/release/bioc/html/DropletUtils.html) | Indicates whether a lower p-value could be obtained by increasing the number of iterations; single-cell data will read "NA" if task is unable to detect knee point inflection. Column is not included for data run in the sn_rna mode  |
|`emptydrops_LogProb` | [dropletUtils](https://bioconductor.org/packages/release/bioc/html/DropletUtils.html) | The log-probability of observing the barcode’s count vector under the null model; single-cell data will read "NA" if the task is unable to detect knee point inflection. Column is not included for data run in the sn_rna mode  |
| `emptydrops_PValue` | [dropletUtils](https://bioconductor.org/packages/release/bioc/html/DropletUtils.html) | Numeric, the Monte Carlo p-value against the null model; single-cell data will read "NA" if task is unable to detect knee point inflection. Column is not included for data run in the sn_rna mode  |
| `emptydrops_Total` | [dropletUtils](https://bioconductor.org/packages/release/bioc/html/DropletUtils.html) | Numeric, the total read counts for each barcode; single-cell data will read "NA" if task is unable to detect knee point inflection. Column is not included for data run in the sn_rna mode  |

## Table 3. Row Attributes (Gene Metrics)

| Gene Metrics                  | Program            |Details                 |
|-------------------------------|--------------------|------------------------|
|`ensembl_ids` | [GENCODE GTF](https://www.gencodegenes.org/) | The gene_id listed in the GENCODE GTF. |
| `Gene` | [GENCODE GTF](https://www.gencodegenes.org/) | The unique gene_name provided in the GENCODE GTF; identical to the `gene_names` attribute. |
|`gene_names` | [GENCODE GTF](https://www.gencodegenes.org/) | The unique gene_name provided in the GENCODE GTF; identical to the `Gene` attribute. |
|`n_reads`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of reads associated with this entity. |
|`noise_reads`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of reads that are categorized by 10x Genomics Cell Ranger as "noise". Refers to long polymers, or reads with high numbers of N (ambiguous) nucleotides. |
|`perfect_molecule_barcodes`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of reads with molecule barcodes that have no errors. |
|`reads_mapped_uniquely`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of reads mapped to a single unambiguous location in the genome. |
|`reads_mapped_multiple`|[warp-tools](https://github.com/broadinstitute/warp-tools)|The number of reads mapped to multiple genomic positions with equal confidence. |
|`spliced_reads`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of reads that overlap splicing junctions. |
|`antisense_reads`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of reads that are mapped to the antisense strand instead of the transcribed strand. |
|`molecule_barcode_fraction_bases_above_30_mean`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The average fraction of bases in molecule barcodes that receive quality scores greater than 30 across the reads of this entity. |
|`molecule_barcode_fraction_bases_above_30_variance`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The variance in the fraction of bases in molecule barcodes that receive quality scores greater than 30 across the reads of this entity. |
|`genomic_reads_fraction_bases_quality_above_30_mean`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The average fraction of bases in the genomic read that receive quality scores greater than 30 across the reads of this entity (included for 10x Cell Ranger count comparison). |
|`genomic_reads_fraction_bases_quality_above_30_variance`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The variance in the fraction of bases in the genomic read that receive quality scores greater than 30 across the reads of this entity (included for 10x Cell Ranger count comparison). |
|`genomic_read_quality_mean`|[warp-tools](https://github.com/broadinstitute/warp-tools)| Average quality of Illumina base calls in the genomic reads corresponding to this entity. |
|`genomic_read_quality_variance`|[warp-tools](https://github.com/broadinstitute/warp-tools)| Variance in quality of Illumina base calls in the genomic reads corresponding to this entity. |
|`n_molecules`|[warp-tools](https://github.com/broadinstitute/warp-tools)| Number of molecules corresponding to this entity (only reflects reads with CB and UB tags). |
|`n_fragments`|[warp-tools](https://github.com/broadinstitute/warp-tools)| Number of fragments corresponding to this entity. See README.md for the definition of a Fragment. |
|`reads_per_molecule`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The average number of reads associated with each molecule in this entity. |
|`reads_per_fragment`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The average number of reads associated with each fragment in this entity. |
|`fragments_per_molecule`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The average number of fragments associated with each molecule in this entity. |
|`fragments_with_single_read_evidence`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of fragments associated with this entity that are observed by only one read. |
|`molecules_with_single_read_evidence`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of molecules associated with this entity that are observed by only one read. |
|`number_cells_detected_multiple`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of cells which observe more than one read of this gene. |
|`number_cells_expressing`|[warp-tools](https://github.com/broadinstitute/warp-tools)| The number of cells that detect this gene. |
