<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Pipelines/BuildIndices_Pipeline/README">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">BuildIndices Overview | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/BuildIndices_Pipeline/README"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="BuildIndices Overview | WARP"><meta data-rh="true" name="description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><meta data-rh="true" property="og:description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/BuildIndices_Pipeline/README"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/BuildIndices_Pipeline/README" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/BuildIndices_Pipeline/README" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.f756d810.js" as="script">
<link rel="preload" href="/warp/assets/js/main.affc3f00.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/ATAC/README">ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/BuildIndices_Pipeline/README">BuildIndices</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README">Exome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Exome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/README">Illumina Genotyping Array</a><button aria-label="Toggle the collapsible sidebar category &#x27;Illumina Genotyping Array&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/ImputationBeagle_Pipeline/README">ImputationBeagle</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Imputation_Pipeline/README">Imputation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Imputation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/JointGenotyping_Pipeline/README">JointGenotyping</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Multiome_Pipeline/README">Multiome scATAC and GEX</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus</a><button aria-label="Toggle the collapsible sidebar category &#x27;Optimus&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/PairedTag_Pipeline/README">Paired-Tag</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README">RNA with UMIs</a><button aria-label="Toggle the collapsible sidebar category &#x27;RNA with UMIs&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/SlideSeq_Pipeline/README">Slide-seq</a><button aria-label="Toggle the collapsible sidebar category &#x27;Slide-seq&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/SlideTags_Pipeline/README">Slide-tags</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README">Smart-seq2 Single Nucleus Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Nucleus Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/snm3C/README">Single Nucleus Methyl-Seq and Chromatin Capture</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Nucleus Methyl-Seq and Chromatin Capture&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Ultima_Genomics_Whole_Genome_Germline_Pipeline/README">Ultima Genomics Whole Genome Germline</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/README">Whole Genome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Whole Genome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">WARP Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">BuildIndices</span><meta itemprop="position" content="2"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>BuildIndices Overview</h1><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">BuildIndices_v4.0.0</a></td><td align="center">January, 2025</td><td align="center">WARP Pipelines</td><td align="center">Please <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">file an issue in WARP</a>.</td></tr></tbody></table><p><img loading="lazy" alt="BuildIndices_diagram" src="/warp/assets/images/buildindices_diagram-df706c5e53b274ea9d451efd93ce1c0e.png" width="1047" height="1048" class="img_ev3q"></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="introduction-to-the-buildindices-workflow">Introduction to the BuildIndices workflow<a href="#introduction-to-the-buildindices-workflow" class="hash-link" aria-label="Direct link to Introduction to the BuildIndices workflow" title="Direct link to Introduction to the BuildIndices workflow">​</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/build_indices/BuildIndices.wdl" target="_blank" rel="noopener noreferrer">BuildIndices workflow</a> is an open-source, cloud-optimized pipeline developed in collaboration with the <a href="https://biccn.org/" target="_blank" rel="noopener noreferrer">BRAIN Initiative Cell Census Network</a> (BICCN) and the BRAIN Initiative Cell Atlas Network (BICAN). </p><p>Overall, the workflow filters GTF files for selected gene biotypes, calculates chromosome sizes, and builds reference bundles with required files for <a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">STAR</a> and <a href="https://github.com/bwa-mem2/bwa-mem2" target="_blank" rel="noopener noreferrer">bwa-mem2</a> aligners.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="quickstart-table">Quickstart table<a href="#quickstart-table" class="hash-link" aria-label="Direct link to Quickstart table" title="Direct link to Quickstart table">​</a></h2><p>The following table provides a quick glance at the BuildIndices pipeline features:</p><table><thead><tr><th>Pipeline features</th><th>Description</th><th>Source</th></tr></thead><tbody><tr><td>Overall workflow</td><td>Reference bundle creation for STAR and bwa-mem2 aligners</td><td>Code available on <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/build_indices/BuildIndices.wdl" target="_blank" rel="noopener noreferrer">GitHub</a></td></tr><tr><td>Workflow language</td><td>WDL 1.0</td><td><a href="https://github.com/openwdl/wdl" target="_blank" rel="noopener noreferrer">openWDL</a></td></tr><tr><td>Genomic Reference Sequence</td><td>GRCh38 human genome primary sequence, M32 (GRCm39) mouse genome primary sequence, and release 103 (GCF_003339765.1) macaque genome primary sequence</td><td>GENCODE <a href="https://www.gencodegenes.org/human/release_43.html" target="_blank" rel="noopener noreferrer">human reference files</a>, GENCODE <a href="https://www.gencodegenes.org/mouse/release_M32.html" target="_blank" rel="noopener noreferrer">mouse reference files</a>, and NCBI <a href="https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_003339765.1/" target="_blank" rel="noopener noreferrer">macaque reference files</a></td></tr><tr><td>Gene annotation reference (GTF)</td><td>Reference containing gene annotations</td><td>GENCODE <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_43/gencode.v43.annotation.gtf.gz" target="_blank" rel="noopener noreferrer">human GTF</a>, GENCODE <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M32/gencode.vM32.primary_assembly.annotation.gtf.gz" target="_blank" rel="noopener noreferrer">mouse GTF</a>, and NCBI <a href="https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_003339765.1/" target="_blank" rel="noopener noreferrer">macaque GTF</a></td></tr><tr><td>Reference builders</td><td>STAR, bwa-mem2</td><td><a href="https://pubmed.ncbi.nlm.nih.gov/23104886/" target="_blank" rel="noopener noreferrer">Dobin et al. 2013</a>, <a href="https://ieeexplore.ieee.org/document/8820962" target="_blank" rel="noopener noreferrer">Vasimuddin et al. 2019</a></td></tr><tr><td>Data input file format</td><td>File format in which reference files are provided</td><td>FASTA, GTF, TSV</td></tr><tr><td>Data output file format</td><td>File formats in which BuildIndices output is provided</td><td>GTF, TAR, TXT</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="set-up">Set-up<a href="#set-up" class="hash-link" aria-label="Direct link to Set-up" title="Direct link to Set-up">​</a></h2><h3 class="anchor anchorWithStickyNavbar_LWe7" id="buildindices-installation">BuildIndices installation<a href="#buildindices-installation" class="hash-link" aria-label="Direct link to BuildIndices installation" title="Direct link to BuildIndices installation">​</a></h3><p>To download the latest BuildIndices release, see the release tags prefixed with &quot;BuildIndices&quot; on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a>. All BuildIndices pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/build_indices/BuildIndices.changelog.md" target="_blank" rel="noopener noreferrer">BuildIndices changelog</a>. </p><p>To search releases of this and other pipelines, use the WARP command-line tool <a href="https://github.com/broadinstitute/warp/tree/master/wreleaser" target="_blank" rel="noopener noreferrer">Wreleaser</a>.</p><p>If you’re running a BuildIndices workflow version prior to the latest release, the accompanying documentation for that release may be downloaded with the source code on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a> (see the folder <code>website/docs/Pipelines/BuildIndices_Pipeline</code>).</p><p>The BuildIndices pipeline can be deployed using <a href="https://cromwell.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH-compliant, flexible workflow management system that supports multiple computing platforms. The workflow can also be run in <a href="https://app.terra.bio" target="_blank" rel="noopener noreferrer">Terra</a>, a cloud-based analysis platform.</p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="inputs">Inputs<a href="#inputs" class="hash-link" aria-label="Direct link to Inputs" title="Direct link to Inputs">​</a></h3><p>The BuildIndices workflow inputs are specified in JSON configuration files. Configuration files for <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/build_indices/Macaque.json" target="_blank" rel="noopener noreferrer">macaque</a> and <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/build_indices/Mouse.json" target="_blank" rel="noopener noreferrer">mouse</a> references can be found in the WARP repository.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="input-descriptions">Input descriptions<a href="#input-descriptions" class="hash-link" aria-label="Direct link to Input descriptions" title="Direct link to Input descriptions">​</a></h4><p>The table below describes the input variables for the BuildIndices workflow. </p><div class="theme-admonition theme-admonition-tip alert alert--success admonition_LlT9"><div class="admonitionHeading_tbUL"><span class="admonitionIcon_kALy"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>tip</div><div class="admonitionContent_S0QG"><p>Marmoset scripts expect a custom-modified input Marmoset GTF file and FASTA file. These inputs and accompanying README are located in a <a href="https://drive.google.com/drive/folders/15JcUhwOqkJwTVS8BOlA0yIdjh4RwJOdz" target="_blank" rel="noopener noreferrer">public Google Drive</a> maintained by Mike Debardine from the BICAN consortium.</p></div></div><table><thead><tr><th>Parameter name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>genome_source</td><td>Describes the source of the reference genome listed in the GTF file; used to name output files; can be set to “NCBI” or “GENCODE”.</td><td>String</td></tr><tr><td>gtf_annotation_version</td><td>Version or release of the reference genome listed in the GTF file; used to name STAR output files; ex.”M32”, “103”.</td><td>String</td></tr><tr><td>genome_build</td><td>Assembly accession (NCBI) or version (GENCODE) of the reference genome listed in the GTF file; used to name output files; ex. “GRCm39”, “GCF_003339765.1”.</td><td>String</td></tr><tr><td>organism</td><td>Organism of the reference genome; used to name the output files; can be set to “Macaque”, “Mouse”, “Human”, or any other organism matching the reference genome.</td><td>String</td></tr><tr><td>annotations_gtf</td><td>GTF file containing gene annotations; used to build the STAR reference files.</td><td>File</td></tr><tr><td>genome_fa</td><td>Genome FASTA file used for building indices.</td><td>File</td></tr><tr><td>biotypes</td><td>TSV file containing gene biotypes attributes to include in the modified GTF file; the first column contains the biotype and the second column contains “Y” to include or “N” to exclude the biotype; <a href="https://www.gencodegenes.org/pages/biotypes.html" target="_blank" rel="noopener noreferrer">GENCODE biotypes</a> are used for GENCODE references and RefSeq biotypes are used for NCBI references.</td><td>File</td></tr><tr><td>run_add_introns</td><td>Boolean to indicate whether to create reference files that include intronic regions; default is false.</td><td>Boolean</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="buildindices-tasks-and-tools">BuildIndices tasks and tools<a href="#buildindices-tasks-and-tools" class="hash-link" aria-label="Direct link to BuildIndices tasks and tools" title="Direct link to BuildIndices tasks and tools">​</a></h2><p>Overall, the BuildIndices workflow:</p><ol><li>Checks inputs, modifies reference files, and creates STAR index.</li><li>Calculates chromosome sizes.</li><li>Builds reference bundle for bwa-mem2.</li></ol><p>The tasks and tools used in the BuildIndices workflow are detailed in the table below. </p><p>To see specific tool parameters, select the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/build_indices/BuildIndices.wdl" target="_blank" rel="noopener noreferrer">workflow WDL link</a>; then find the task and view the <code>command {}</code> section of the task in the WDL script. To view or use the exact tool software, see the task&#x27;s Docker image which is specified in the task WDL <code># runtime values</code> section as <code>docker: </code>.</p><table><thead><tr><th>Task name</th><th>Tool</th><th>Software</th><th>Description</th></tr></thead><tbody><tr><td>BuildStarSingleNucleus</td><td><a href="https://github.com/broadinstitute/warp-tools/blob/develop/3rd-party-tools/build-indices/modify_gtf.py" target="_blank" rel="noopener noreferrer">modify_gtf.py</a>, STAR</td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop" target="_blank" rel="noopener noreferrer">warp-tools</a>, <a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">STAR</a></td><td>Checks that the input GTF file contains input genome source, genome build version, and annotation version with correct build source information, modifies files for the STAR aligner, and creates STAR index file. If &quot;Marmoset&quot; is selected as organism, a <a href="https://github.com/broadinstitute/warp-tools/blob/develop/3rd-party-tools/build-indices/modify_gtf_marmoset.py" target="_blank" rel="noopener noreferrer">Marmoset-specific custom script</a> is run to modify the GTF</td></tr><tr><td>CalculateChromosomeSizes</td><td>faidx</td><td><a href="http://www.htslib.org/" target="_blank" rel="noopener noreferrer">Samtools</a></td><td>Reads the genome FASTA file to create a FASTA index file that contains the genome chromosome sizes.</td></tr><tr><td>BuildBWAreference</td><td>index</td><td><a href="https://github.com/bwa-mem2/bwa-mem2" target="_blank" rel="noopener noreferrer">bwa-mem2</a></td><td>Builds the reference bundle for the bwa aligner.</td></tr><tr><td>SNSS2AddIntronsToGTF</td><td><a href="https://github.com/broadinstitute/warp-tools/blob/develop/3rd-party-tools/build-indices/add-introns-to-gtf.py" target="_blank" rel="noopener noreferrer">add-introns-to-gtf.py</a>, STAR</td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop" target="_blank" rel="noopener noreferrer">warp-tools</a>, <a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">STAR</a></td><td>Adds intron annotations to a GTF, regenerates a STAR genome index with the updated annotation and reference FASTA, and packages the index into a tar archive for downstream use.</td></tr></tbody></table><h4 class="anchor anchorWithStickyNavbar_LWe7" id="1-check-inputs-modify-reference-files-and-create-star-index-file">1. Check inputs, modify reference files, and create STAR index file<a href="#1-check-inputs-modify-reference-files-and-create-star-index-file" class="hash-link" aria-label="Direct link to 1. Check inputs, modify reference files, and create STAR index file" title="Direct link to 1. Check inputs, modify reference files, and create STAR index file">​</a></h4><p><strong>Check inputs</strong></p><p>The BuildStarSingleNucleus task reads the input GTF file and verifies that the <code>genome_source</code>, <code>genome_build</code>, and <code>gtf_annotation_version</code> listed in the file match the input values provided to the pipeline.</p><p><strong>Modify reference files and create STAR index</strong></p><p>The BuildStarSingleNucleus task uses a custom python script, <a href="https://github.com/broadinstitute/warp-tools/blob/develop/3rd-party-tools/build-indices/modify_gtf.py" target="_blank" rel="noopener noreferrer"><code>modify_gtf.py</code></a> or <a href="https://github.com/broadinstitute/warp-tools/blob/develop/3rd-party-tools/build-indices/modify_gtf_marmoset.py" target="_blank" rel="noopener noreferrer"><code>modify_get_marmoset</code></a>, and a list of biotypes (<a href="https://github.com/broadinstitute/warp-tools/blob/develop/3rd-party-tools/build-indices/Biotypes.tsv" target="_blank" rel="noopener noreferrer">example</a>) to filter the input GTF file for only the biotypes indicated in the list with the value “Y” in the second column. The defaults in the custom code produce reference outputs that are similar to those built with 10x Genomics reference scripts.</p><p>The task uses the filtered GTF file and STAR <code>--runMode genomeGenerate</code> to generate the index file for the STAR aligner. Outputs of the task include the modified GTF and compressed STAR index files.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="2-calculates-chromosome-sizes">2. Calculates chromosome sizes<a href="#2-calculates-chromosome-sizes" class="hash-link" aria-label="Direct link to 2. Calculates chromosome sizes" title="Direct link to 2. Calculates chromosome sizes">​</a></h4><p>The CalculateChromosomeSizes task uses Samtools to create and output a FASTA index file that contains the genome chromosome sizes, which can be used in downstream tools like SnapATAC2. </p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="3-builds-reference-bundle-for-bwa-mem2">3. Builds reference bundle for bwa-mem2<a href="#3-builds-reference-bundle-for-bwa-mem2" class="hash-link" aria-label="Direct link to 3. Builds reference bundle for bwa-mem2" title="Direct link to 3. Builds reference bundle for bwa-mem2">​</a></h4><p>The BuildBWAreference task uses the chromosome sizes file and bwa-mem2 to prepare the genome FASTA file for alignment and builds, compresses, and outputs the reference bundle for the bwa-mem2 aligner.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="outputs">Outputs<a href="#outputs" class="hash-link" aria-label="Direct link to Outputs" title="Direct link to Outputs">​</a></h2><p>The following table lists the output variables and files produced by the pipeline.</p><table><thead><tr><th>Output name</th><th>Filename, if applicable</th><th>Output format and description</th></tr></thead><tbody><tr><td>snSS2_star_index</td><td><code>modified_star2.7.10a-&lt;organism&gt;-&lt;genome_source&gt;-build-&lt;genome_build&gt;-&lt;gtf_annotation_version&gt;.tar</code></td><td>TAR file containing a species-specific reference genome and GTF file for <a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">STAR</a> alignment.</td></tr><tr><td>pipeline_version_out</td><td><code>BuildIndices_v&lt;pipeline_version&gt;</code></td><td>String describing the version of the BuildIndices pipeline used.</td></tr><tr><td>snSS2_annotation_gtf_modified</td><td><code>modified_v&lt;gtf_annotation_version&gt;.annotation.gtf</code></td><td>GTF file containing gene annotations filtered for selected biotypes.</td></tr><tr><td>reference_bundle</td><td><code>bwa-mem2-2.2.1-&lt;organism&gt;-&lt;genome_source&gt;-build-&lt;genome_build&gt;.tar</code></td><td>TAR file containing the reference index files for <a href="https://github.com/lh3/bwa" target="_blank" rel="noopener noreferrer">BWA-mem</a> alignment.</td></tr><tr><td>chromosome_sizes</td><td><code>chrom.sizes</code></td><td>Text file containing chromosome sizes for the genome build.</td></tr><tr><td>snSS2_annotation_gtf_with_introns</td><td><code>&lt;modified_annotation_gtf&gt;_with_introns.gtf</code></td><td>Optional GTF file containing gene annotations filtered for selected biotypes, with intronic regions included.</td></tr><tr><td>star_index_with_introns</td><td><code>&lt;modified_annotation_gtf&gt;_intron.tar</code></td><td>Optional TAR file containing a species-specific reference genome and GTF file containing intronic regions for <a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">STAR</a> alignment.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="versioning-and-testing">Versioning and testing<a href="#versioning-and-testing" class="hash-link" aria-label="Direct link to Versioning and testing" title="Direct link to Versioning and testing">​</a></h2><p>All BuildIndices pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/build_indices/BuildIndices.changelog.md" target="_blank" rel="noopener noreferrer">BuildIndices changelog</a> and tested manually using <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/skylab/build_indices" target="_blank" rel="noopener noreferrer">reference JSON files</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="citing-the-buildindices-pipeline">Citing the BuildIndices Pipeline<a href="#citing-the-buildindices-pipeline" class="hash-link" aria-label="Direct link to Citing the BuildIndices Pipeline" title="Direct link to Citing the BuildIndices Pipeline">​</a></h2><p>If you use the BuildIndices Pipeline in your research, please consider citing our preprint:</p><p>Degatano, K.; Awdeh, A.; Dingman, W.; Grant, G.; Khajouei, F.; Kiernan, E.; Konwar, K.; Mathews, K.; Palis, K.; Petrillo, N.; Van der Auwera, G.; Wang, C.; Way, J.; Pipelines, W. WDL Analysis Research Pipelines: Cloud-Optimized Workflows for Biological Data Processing and Reproducible Analysis. Preprints 2024, 2024012131. <a href="https://doi.org/10.20944/preprints202401.2131.v1" target="_blank" rel="noopener noreferrer">https://doi.org/10.20944/preprints202401.2131.v1</a></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="consortia-support">Consortia support<a href="#consortia-support" class="hash-link" aria-label="Direct link to Consortia support" title="Direct link to Consortia support">​</a></h2><p>This pipeline is supported by the <a href="https://braininitiative.nih.gov/" target="_blank" rel="noopener noreferrer">BRAIN Initiative</a> (BICCN and BICAN) and SCORCH. </p><p>If your organization also uses this pipeline, we would like to list you! Please reach out to us by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="example-references">Example references<a href="#example-references" class="hash-link" aria-label="Direct link to Example references" title="Direct link to Example references">​</a></h2><p>Example references are available in the Broad Public Reference bucket, a Google bucket that hosts reference files at no charge to the end-user.</p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="human">Human<a href="#human" class="hash-link" aria-label="Direct link to Human" title="Direct link to Human">​</a></h3><table><thead><tr><th>File Type</th><th>File Location</th></tr></thead><tbody><tr><td>Genomics Reference</td><td>GRCh38, primary assembly (PRI)</td></tr><tr><td>Gene annotation (PRI)</td><td>GENCODE Release 43 GRCh38.p13</td></tr><tr><td>Reference README</td><td><code>gs://gcp-public-data--broad-references/hg38/v0/star/v2_7_10a/v43_README.txt</code></td></tr><tr><td>STAR Index TAR</td><td><code>gs://gcp-public-data--broad-references/hg38/v0/star/v2_7_10a/modified_star2.7.10a-Human-GENCODE-build-GRCh38-43.tar</code></td></tr><tr><td>STAR Annotation GTF</td><td><code>gs://gcp-public-data--broad-references/hg38/v0/star/v2_7_10a/modified_v43.annotation.gtf</code></td></tr><tr><td>BWA-MEM2 Index TAR</td><td><code>gs://gcp-public-data--broad-references/hg38/v0/bwa/v2_2_1/bwa-mem2-2.2.1-Human-GENCODE-build-GRCh38.tar</code></td></tr><tr><td>Chromosome Sizes</td><td><code>gs://gcp-public-data--broad-references/hg38/v0/bwa/v2_2_1/chrom.sizes</code></td></tr></tbody></table><h3 class="anchor anchorWithStickyNavbar_LWe7" id="mouse">Mouse<a href="#mouse" class="hash-link" aria-label="Direct link to Mouse" title="Direct link to Mouse">​</a></h3><table><thead><tr><th>File Type</th><th>File Location</th></tr></thead><tbody><tr><td>Genomics Reference</td><td>GRCm39, primary assembly (PRI)</td></tr><tr><td>Gene annotation (PRI)</td><td>GENCODE Release 32</td></tr><tr><td>Reference README</td><td><code>gs://gcp-public-data--broad-references/GRCm39/star/v2_7_10a/M32_README.txt</code></td></tr><tr><td>STAR Index TAR</td><td><code>gs://gcp-public-data--broad-references/GRCm39/star/v2_7_10a/modified_star2.7.10a-Mouse-GENCODE-build-GRCm39-M32.tar</code></td></tr><tr><td>STAR Annotation GTF</td><td><code>gs://gcp-public-data--broad-references/GRCm39/star/v2_7_10a/modified_vM32.annotation.gtf</code></td></tr><tr><td>BWA-MEM2 Index TAR</td><td><code>gs://gcp-public-data--broad-references/GRCm39/bwa/v2_2_1/bwa-mem2-2.2.1-Mouse-GENCODE-build-GRCm39.tar</code></td></tr><tr><td>Chromosome Sizes</td><td><code>gs://gcp-public-data--broad-references/GRCm39/bwa/v2_2_1/chrom.sizes</code></td></tr></tbody></table><h3 class="anchor anchorWithStickyNavbar_LWe7" id="macaque">Macaque<a href="#macaque" class="hash-link" aria-label="Direct link to Macaque" title="Direct link to Macaque">​</a></h3><p>Inputs for the Macaque reference below were modified using a custom tool to handle nuclear mitochondrial inserts, <a href="https://github.com/nkschaefer/numty-dumpty" target="_blank" rel="noopener noreferrer">numty-dumpty</a>. See the README for the <a href="https://storage.cloud.google.com/gcp-public-data--broad-references/M.mulatta/Mmul_10/star/v2_7_10a/numty_dumpty/README_STAR.txt" target="_blank" rel="noopener noreferrer">STAR index</a> and the <!-- -->[bwa-mem2 index]<!-- -->
(<a href="https://storage.cloud.google.com/gcp-public-data--broad-references/M.mulatta/Mmul_10/bwa/v2_2_1/numty_dumpty/README_BWA.txt" target="_blank" rel="noopener noreferrer">https://storage.cloud.google.com/gcp-public-data--broad-references/M.mulatta/Mmul_10/bwa/v2_2_1/numty_dumpty/README_BWA.txt</a>).</p><table><thead><tr><th>File Type</th><th>File Location</th></tr></thead><tbody><tr><td>Genomics Reference</td><td>mmul10</td></tr><tr><td>Gene annotation</td><td>RefSeq annotation version 103</td></tr><tr><td>STAR Index TAR</td><td><code>gs://gcp-public-data--broad-references/M.mulatta/Mmul_10/star/v2_7_10a/10x_filtered/modified_star2.7.10a-Macaque-NCBI-build-GCF_003339765.1-103.tar</code></td></tr><tr><td>BWA Index TAR</td><td><code>gs://gcp-public-data--broad-references/M.mulatta/Mmul_10/bwa/v2_2_1/bwa-mem2-2.2.1-Macaque-NCBI-build-GCF_003339765.1.tar</code></td></tr><tr><td>GTF Annotation</td><td><code>gs://gcp-public-data--broad-references/M.mulatta/Mmul_10/star/v2_7_10a/10x_filtered/mt_v103_modified.annotation.gtf</code></td></tr><tr><td>Chromosome Sizes</td><td><code>gs://gcp-public-data--broad-references/M.mulatta/Mmul_10/bwa/v2_2_1/chrom.sizes</code></td></tr></tbody></table><p>This macaque reference works with the Optimus, Multiome, and Paired-tag workflows. However, mitochondrial genes are not demarcated with an &quot;mt-&quot; tag. A separate text file with MT genes is required. An example is the list below:</p><div class="codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#bfc7d5;--prism-background-color:#292d3e"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-text codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#bfc7d5"><span class="token plain">ND1</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">ND2</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">COX1</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">COX2</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">ATP8</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">ATP6</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">COX3</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">ND3</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">ND4L</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">ND4</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">ND5</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">ND6</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">CYTB</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>An example file with this list is located in a public Google bucket here: gs://warp-testing-public/references/BuildIndices_outs/Macaque_MT_genes.txt</p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="marmoset">Marmoset<a href="#marmoset" class="hash-link" aria-label="Direct link to Marmoset" title="Direct link to Marmoset">​</a></h3><p>Marmoset scripts expect a custom-modified input Marmoset GTF file. These inputs and accompanying README are located in a <a href="https://drive.google.com/drive/folders/15JcUhwOqkJwTVS8BOlA0yIdjh4RwJOdz" target="_blank" rel="noopener noreferrer">public Google Drive</a> maintained by Mike Debardine from the BICAN consortium.</p><table><thead><tr><th>File Type</th><th>File Location</th></tr></thead><tbody><tr><td>Genomics Reference</td><td>mCalJa1.2.pat.X (GenBank Accession GCA_011100555.2 and RefSeq Accession GCF_011100555.1)</td></tr><tr><td>Gene annotation</td><td>Custom (see note above table)</td></tr><tr><td>Chromosome Sizes</td><td><code>gs://gcp-public-data--broad-references/mCalJa1/mCalJa1.2.pat.X/chrom.sizes</code></td></tr><tr><td>GTF Annotation</td><td><code>gs://gcp-public-data--broad-references/mCalJa1/mCalJa1.2.pat.X/modified_vGCF_011100555.1-RS_2023_03.annotation.gtf</code></td></tr><tr><td>BWA-MEM2 Index TAR</td><td><code>gs://gcp-public-data--broad-references/mCalJa1/mCalJa1.2.pat.X/bwa-mem2-2.2.1-Marmoset-RefSeq-build-mCalJa1.2.pat.X.tar</code></td></tr><tr><td>STAR Index TAR</td><td><code>gs://gcp-public-data--broad-references/mCalJa1/mCalJa1.2.pat.X/modified_star2.7.10a-Marmoset-RefSeq-build-mCalJa1.2.pat.X-GCF_011100555.1-RS_2023_03.tar</code></td></tr></tbody></table><h3 class="anchor anchorWithStickyNavbar_LWe7" id="armadillo">Armadillo<a href="#armadillo" class="hash-link" aria-label="Direct link to Armadillo" title="Direct link to Armadillo">​</a></h3><table><thead><tr><th>File Type</th><th>File Location</th></tr></thead><tbody><tr><td>Genomic Reference</td><td>mDasNov1.hap2 (NCBI)</td></tr><tr><td>Gene Annotation</td><td>RefSeq GCF_030445035.1-RS_2023_07</td></tr><tr><td>BWA-MEM2 Index TAR</td><td><code>gs://gcp-public-data--broad-references/D.novemcinctus/mDasNov1.hap2/cleanome/bwa/v2_2_1/bwa-mem2-2.2.1-Armadillo-NCBI-build-mDasNov1.hap2.tar</code></td></tr><tr><td>Chromosome Sizes</td><td><code>gs://gcp-public-data--broad-references/D.novemcinctus/mDasNov1.hap2/cleanome/bwa/v2_2_1/chrom.sizes</code></td></tr><tr><td>STAR Index TAR</td><td><code>gs://gcp-public-data--broad-references/D.novemcinctus/mDasNov1.hap2/cleanome/star/v2_7_10a/modified_star2.7.10a-Armadillo-NCBI-build-mDasNov1.hap2-2.2.tar</code></td></tr><tr><td>GTF Annotation</td><td><code>gs://gcp-public-data--broad-references/D.novemcinctus/mDasNov1.hap2/cleanome/star/v2_7_10a/modified_v2.2.annotation.gtf</code></td></tr></tbody></table><h3 class="anchor anchorWithStickyNavbar_LWe7" id="opposum">Opposum<a href="#opposum" class="hash-link" aria-label="Direct link to Opposum" title="Direct link to Opposum">​</a></h3><table><thead><tr><th>File Type</th><th>File Location</th></tr></thead><tbody><tr><td>Genomic Reference</td><td>mMonDom1.pri (NCBI)</td></tr><tr><td>Gene Annotation</td><td>RefSeq GCF_027887165.1-RS_2023_05 (RefSeq link)</td></tr><tr><td>BWA-MEM2 Index TAR</td><td><code>gs://gcp-public-data--broad-references/M.domestica/mMonDom1.pri/cleanome/bwa/v2_2_1/bwa-mem2-2.2.1-Opossum-NCBI-build-mMonDom1.pri.tar</code></td></tr><tr><td>Chromosome Sizes</td><td><code>gs://gcp-public-data--broad-references/M.domestica/mMonDom1.pri/cleanome/bwa/v2_2_1/chrom.sizes</code></td></tr><tr><td>STAR Index TAR</td><td><code>gs://gcp-public-data--broad-references/M.domestica/mMonDom1.pri/cleanome/star/v2_7_10a/modified_star2.7.10a-Opossum-NCBI-build-mMonDom1.pri-2.2.tar</code></td></tr><tr><td>GTF Annotation</td><td><code>gs://gcp-public-data--broad-references/M.domestica/mMonDom1.pri/cleanome/star/v2_7_10a/modified_v2.2.annotation.gtf</code></td></tr></tbody></table><h3 class="anchor anchorWithStickyNavbar_LWe7" id="rat">Rat<a href="#rat" class="hash-link" aria-label="Direct link to Rat" title="Direct link to Rat">​</a></h3><table><thead><tr><th>File Type</th><th>File Location</th></tr></thead><tbody><tr><td>Genomic Reference</td><td>mRatBN7.2 (NCBI)</td></tr><tr><td>Gene Annotation</td><td>RefSeq GCF_015227675.2-RS_2023_06</td></tr><tr><td>BWA-MEM2 Index TAR</td><td><code>gs://gcp-public-data--broad-references/R.norvegicus/mRatBN7.2/cleanome/bwa/v2_2_1/bwa-mem2-2.2.1-Rat-NCBI-build-mRatBN7.2.tar</code></td></tr><tr><td>Chromosome Sizes</td><td><code>gs://gcp-public-data--broad-references/R.norvegicus/mRatBN7.2/cleanome/bwa/v2_2_1/chrom.sizes</code></td></tr><tr><td>STAR Index TAR</td><td><code>gs://gcp-public-data--broad-references/R.norvegicus/mRatBN7.2/cleanome/star/v2_7_10a/modified_star2.7.10a-Rat-NCBI-build-mRatBN7.2-2.2.tar</code></td></tr><tr><td>GTF Annotation</td><td><code>gs://gcp-public-data--broad-references/R.norvegicus/mRatBN7.2/cleanome/star/v2_7_10a/modified_v2.2.annotation.gtf</code></td></tr></tbody></table><h3 class="anchor anchorWithStickyNavbar_LWe7" id="pig">Pig<a href="#pig" class="hash-link" aria-label="Direct link to Pig" title="Direct link to Pig">​</a></h3><table><thead><tr><th>File Type</th><th>File Location</th></tr></thead><tbody><tr><td>Genomic Reference</td><td>Sscrofa11.1 (NCBI)</td></tr><tr><td>Gene Annotation</td><td>NCBI Annotation Release 106 (RefSeq GCF_000003025.6_Sscrofa11.1)</td></tr><tr><td>BWA-MEM2 Index TAR</td><td><code>gs://gcp-public-data--broad-references/S.scrofa/Sscrofa11.1/cleanome/bwa/v2_2_1/bwa-mem2-2.2.1-Pig-NCBI-build-Sscrofa11.1.tar</code></td></tr><tr><td>Chromosome Sizes</td><td><code>gs://gcp-public-data--broad-references/S.scrofa/Sscrofa11.1/cleanome/bwa/v2_2_1/chrom.sizes</code></td></tr><tr><td>STAR Index TAR</td><td><code>gs://gcp-public-data--broad-references/S.scrofa/Sscrofa11.1/cleanome/star/v2_7_10a/modified_star2.7.10a-Pig-NCBI-build-Sscrofa11.1-2.2.tar</code></td></tr><tr><td>GTF Annotation</td><td><code>gs://gcp-public-data--broad-references/S.scrofa/Sscrofa11.1/cleanome/star/v2_7_10a/modified_v2.2.annotation.gtf</code></td></tr></tbody></table><h3 class="anchor anchorWithStickyNavbar_LWe7" id="viral-references">Viral references<a href="#viral-references" class="hash-link" aria-label="Direct link to Viral references" title="Direct link to Viral references">​</a></h3><table><thead><tr><th>Species</th><th>Species reference</th><th>Species annotation</th><th>Viral reference</th><th>Viral annotation</th></tr></thead><tbody><tr><td>Human</td><td>Human-GENCODE-build-GRCh38-43</td><td>Gencode release 43 GRCh38</td><td>Modified HXB2 K03455.1 <br> <code>gs://fc-36d6b972-99f2-4ca2-a076-9c4c0e7567e1/viral-references/hiv.fa</code></td><td><code>gs://fc-36d6b972-99f2-4ca2-a076-9c4c0e7567e1/viral-references/HIV.gtf</code></td></tr><tr><td>Mouse</td><td>mm39/genecode.vM32</td><td>Gencode release M32 GRCm39</td><td>Modified <a href="https://www.ncbi.nlm.nih.gov/nuccore/MG470653.1" target="_blank" rel="noopener noreferrer">https://www.ncbi.nlm.nih.gov/nuccore/MG470653.1</a> (excluding 3’LTR) <br> <code>gs://fc-36d6b972-99f2-4ca2-a076-9c4c0e7567e1/viral-references/EcoHIV.fa</code></td><td><code>gs://fc-36d6b972-99f2-4ca2-a076-9c4c0e7567e1/viral-references/EcoHIV.gtf</code></td></tr><tr><td>Macaque</td><td>mmul10</td><td>Refseq annotation v 103</td><td>NCBI GenBank, accession number PP236443 <br> <code>gs://fc-36d6b972-99f2-4ca2-a076-9c4c0e7567e1/viral-references/siv.fa</code></td><td><code> gs://fc-36d6b972-99f2-4ca2-a076-9c4c0e7567e1/viral-references/SIV.gtf</code></td></tr><tr><td>Rat</td><td>mRatBN7.2</td><td>RefSeq GCF_015227675.2-RS_2023_06</td><td>AF324493.2 <br> <code>gs://fc-36d6b972-99f2-4ca2-a076-9c4c0e7567e1/viral-references/RAT-HIVTransgene.fa</code></td><td><code>gs://fc-36d6b972-99f2-4ca2-a076-9c4c0e7567e1/viral-references/RAT-HIVFullTranscript.gtf</code></td></tr></tbody></table><h3 class="anchor anchorWithStickyNavbar_LWe7" id="human--hiv">Human + HIV<a href="#human--hiv" class="hash-link" aria-label="Direct link to Human + HIV" title="Direct link to Human + HIV">​</a></h3><table><thead><tr><th>File Type</th><th>File Location</th></tr></thead><tbody><tr><td>Genomics Reference</td><td>GRCh38 + Modified HXB2 K03455.1</td></tr><tr><td>Gene annotation</td><td>GENCODE Release 43 GRCh38.p13 + HIV annotation</td></tr><tr><td>Human Reference</td><td><code>gs://gcp-public-data--broad-references/scorch/hg38/modified_star2.7.10a-Human-GENCODE-build-GRCh38-43.tar</code></td></tr><tr><td>Human Annotation GTF</td><td><code>gs://gcp-public-data--broad-references/scorch/hg38/modified_v43.annotation.gtf</code></td></tr><tr><td>BWA-MEM2 Index TAR</td><td><code>gs://gcp-public-data--broad-references/scorch/hg38/bwa-mem2-2.2.1-Human-GENCODE-build-GRCh38.tar</code></td></tr><tr><td>Chromosome Sizes</td><td><code>gs://gcp-public-data--broad-references/scorch/hg38/chrom.sizes</code></td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="feedback">Feedback<a href="#feedback" class="hash-link" aria-label="Direct link to Feedback" title="Direct link to Feedback">​</a></h2><p>Please help us make our tools better by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a> for pipeline-related suggestions or questions.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/BuildIndices_Pipeline/README.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2025-10-01T18:42:30.000Z">Oct 1, 2025</time></b> by <b>jsotobroad</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/Pipelines/ATAC/atac.methods"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">ATAC v2.3.1 Methods</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Exome Germline Single Sample Overview</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-buildindices-workflow" class="table-of-contents__link toc-highlight">Introduction to the BuildIndices workflow</a></li><li><a href="#quickstart-table" class="table-of-contents__link toc-highlight">Quickstart table</a></li><li><a href="#set-up" class="table-of-contents__link toc-highlight">Set-up</a><ul><li><a href="#buildindices-installation" class="table-of-contents__link toc-highlight">BuildIndices installation</a></li><li><a href="#inputs" class="table-of-contents__link toc-highlight">Inputs</a></li></ul></li><li><a href="#buildindices-tasks-and-tools" class="table-of-contents__link toc-highlight">BuildIndices tasks and tools</a></li><li><a href="#outputs" class="table-of-contents__link toc-highlight">Outputs</a></li><li><a href="#versioning-and-testing" class="table-of-contents__link toc-highlight">Versioning and testing</a></li><li><a href="#citing-the-buildindices-pipeline" class="table-of-contents__link toc-highlight">Citing the BuildIndices Pipeline</a></li><li><a href="#consortia-support" class="table-of-contents__link toc-highlight">Consortia support</a></li><li><a href="#example-references" class="table-of-contents__link toc-highlight">Example references</a><ul><li><a href="#human" class="table-of-contents__link toc-highlight">Human</a></li><li><a href="#mouse" class="table-of-contents__link toc-highlight">Mouse</a></li><li><a href="#macaque" class="table-of-contents__link toc-highlight">Macaque</a></li><li><a href="#marmoset" class="table-of-contents__link toc-highlight">Marmoset</a></li><li><a href="#armadillo" class="table-of-contents__link toc-highlight">Armadillo</a></li><li><a href="#opposum" class="table-of-contents__link toc-highlight">Opposum</a></li><li><a href="#rat" class="table-of-contents__link toc-highlight">Rat</a></li><li><a href="#pig" class="table-of-contents__link toc-highlight">Pig</a></li><li><a href="#viral-references" class="table-of-contents__link toc-highlight">Viral references</a></li><li><a href="#human--hiv" class="table-of-contents__link toc-highlight">Human + HIV</a></li></ul></li><li><a href="#feedback" class="table-of-contents__link toc-highlight">Feedback</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.f756d810.js"></script>
<script src="/warp/assets/js/main.affc3f00.js"></script>
</body>
</html>