<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Genomic Data Commons (GDC) Whole Genome Somatic Single Sample Overview | WARP</title>
    <meta name="generator" content="VuePress 1.7.1">
    <link rel="icon" href="/warp/logo.png">
    <meta name="description" content="WDL Analysis Research Pipelines">
    <meta name="theme-color" content="#3eaf7c">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <meta name="og:locale" content="en_US">
    
    <link rel="preload" href="/warp/assets/css/0.styles.647b5de0.css" as="style"><link rel="preload" href="/warp/assets/js/app.aea6aa50.js" as="script"><link rel="preload" href="/warp/assets/js/2.2043750e.js" as="script"><link rel="preload" href="/warp/assets/js/10.a9482500.js" as="script"><link rel="prefetch" href="/warp/assets/js/11.4a22fb00.js"><link rel="prefetch" href="/warp/assets/js/12.307a777e.js"><link rel="prefetch" href="/warp/assets/js/13.51fc7374.js"><link rel="prefetch" href="/warp/assets/js/14.d802e1b1.js"><link rel="prefetch" href="/warp/assets/js/15.8ef1d142.js"><link rel="prefetch" href="/warp/assets/js/16.bae06c91.js"><link rel="prefetch" href="/warp/assets/js/17.ed42aaa3.js"><link rel="prefetch" href="/warp/assets/js/18.3ed01c57.js"><link rel="prefetch" href="/warp/assets/js/19.a80d5c3b.js"><link rel="prefetch" href="/warp/assets/js/20.3ea0888d.js"><link rel="prefetch" href="/warp/assets/js/21.f2d33845.js"><link rel="prefetch" href="/warp/assets/js/22.e28d3bc6.js"><link rel="prefetch" href="/warp/assets/js/23.58fc4838.js"><link rel="prefetch" href="/warp/assets/js/24.4c0acda9.js"><link rel="prefetch" href="/warp/assets/js/25.1f9c686d.js"><link rel="prefetch" href="/warp/assets/js/26.4a0144c9.js"><link rel="prefetch" href="/warp/assets/js/27.f8195a98.js"><link rel="prefetch" href="/warp/assets/js/28.1728e2bc.js"><link rel="prefetch" href="/warp/assets/js/29.02800e8d.js"><link rel="prefetch" href="/warp/assets/js/3.dee46bae.js"><link rel="prefetch" href="/warp/assets/js/30.c81bb045.js"><link rel="prefetch" href="/warp/assets/js/31.a08f001b.js"><link rel="prefetch" href="/warp/assets/js/32.ef80a9d6.js"><link rel="prefetch" href="/warp/assets/js/33.a0e54ebc.js"><link rel="prefetch" href="/warp/assets/js/34.bfefd0e5.js"><link rel="prefetch" href="/warp/assets/js/35.49858f18.js"><link rel="prefetch" href="/warp/assets/js/36.800155ad.js"><link rel="prefetch" href="/warp/assets/js/37.768c7029.js"><link rel="prefetch" href="/warp/assets/js/38.9365f85d.js"><link rel="prefetch" href="/warp/assets/js/4.e5554209.js"><link rel="prefetch" href="/warp/assets/js/5.cb25b623.js"><link rel="prefetch" href="/warp/assets/js/6.57da0acf.js"><link rel="prefetch" href="/warp/assets/js/7.e72f1202.js"><link rel="prefetch" href="/warp/assets/js/8.1821baf8.js"><link rel="prefetch" href="/warp/assets/js/9.961e8595.js">
    <link rel="stylesheet" href="/warp/assets/css/0.styles.647b5de0.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/warp/" class="home-link router-link-active"><!----> <span class="site-name">WARP</span></a> <div class="links"><div class="search-box"><input aria-label="Search" autocomplete="off" spellcheck="false" value=""> <!----></div> <nav class="nav-links can-hide"><div class="nav-item"><a href="/warp/documentation/Intro.html" class="nav-link">
  Documentation
</a></div><div class="nav-item"><a href="/warp/blog/" class="nav-link">
  Blog
</a></div><div class="nav-item"><div class="dropdown-wrapper"><button type="button" aria-label="Learn More" class="dropdown-title"><span class="title">Learn More</span> <span class="arrow down"></span></button> <button type="button" aria-label="Learn More" class="mobile-dropdown-title"><span class="title">Learn More</span> <span class="arrow right"></span></button> <ul class="nav-dropdown" style="display:none;"><li class="dropdown-item"><h4>
          Explore
        </h4> <ul class="dropdown-subitem-wrapper"><li class="dropdown-subitem"><a href="/warp/research/" class="nav-link">
  Research
</a></li></ul></li><li class="dropdown-item"><h4>
          Resources
        </h4> <ul class="dropdown-subitem-wrapper"><li class="dropdown-subitem"><a href="/warp/contribution/" class="nav-link">
  Contribution Guide
</a></li><li class="dropdown-subitem"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="nav-link external">
  Changelog
  <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></li><li class="dropdown-subitem"><a href="/warp/privacy/" class="nav-link">
  Privacy
</a></li></ul></li></ul></div></div> <a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></nav></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><nav class="nav-links"><div class="nav-item"><a href="/warp/documentation/Intro.html" class="nav-link">
  Documentation
</a></div><div class="nav-item"><a href="/warp/blog/" class="nav-link">
  Blog
</a></div><div class="nav-item"><div class="dropdown-wrapper"><button type="button" aria-label="Learn More" class="dropdown-title"><span class="title">Learn More</span> <span class="arrow down"></span></button> <button type="button" aria-label="Learn More" class="mobile-dropdown-title"><span class="title">Learn More</span> <span class="arrow right"></span></button> <ul class="nav-dropdown" style="display:none;"><li class="dropdown-item"><h4>
          Explore
        </h4> <ul class="dropdown-subitem-wrapper"><li class="dropdown-subitem"><a href="/warp/research/" class="nav-link">
  Research
</a></li></ul></li><li class="dropdown-item"><h4>
          Resources
        </h4> <ul class="dropdown-subitem-wrapper"><li class="dropdown-subitem"><a href="/warp/contribution/" class="nav-link">
  Contribution Guide
</a></li><li class="dropdown-subitem"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="nav-link external">
  Changelog
  <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></li><li class="dropdown-subitem"><a href="/warp/privacy/" class="nav-link">
  Privacy
</a></li></ul></li></ul></div></div> <a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></nav>  <ul class="sidebar-links"><li><section class="sidebar-group collapsable depth-0"><p class="sidebar-heading"><span>Welcome to WARP</span> <span class="arrow right"></span></p> <!----></section></li><li><section class="sidebar-group collapsable depth-0"><p class="sidebar-heading"><span>About WARP</span> <span class="arrow right"></span></p> <!----></section></li><li><section class="sidebar-group depth-0"><p class="sidebar-heading open"><span>Pipelines</span> <!----></p> <ul class="sidebar-links sidebar-group-items"><li><section class="sidebar-group collapsable is-sub-group depth-1"><p class="sidebar-heading"><span>CEMBA MethylC Seq Pipeline</span> <span class="arrow right"></span></p> <!----></section></li><li><section class="sidebar-group collapsable is-sub-group depth-1"><p class="sidebar-heading"><span>Exome Germline Single Sample Pipeline</span> <span class="arrow right"></span></p> <!----></section></li><li><section class="sidebar-group collapsable is-sub-group depth-1"><p class="sidebar-heading open"><span>Genomic Data Commons Whole Genome Somatic Pipeline</span> <span class="arrow down"></span></p> <ul class="sidebar-links sidebar-group-items"><li><a href="/warp/documentation/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/" aria-current="page" class="active sidebar-link">Genomic Data Commons (GDC) Whole Genome Somatic Single Sample Overview</a><ul class="sidebar-sub-headers"><li class="sidebar-sub-header"><a href="/warp/documentation/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/#introduction-to-the-gdc-whole-genome-somatic-single-sample-pipeline" class="sidebar-link">Introduction to the GDC Whole Genome Somatic Single Sample Pipeline</a></li><li class="sidebar-sub-header"><a href="/warp/documentation/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/#set-up" class="sidebar-link">Set-up</a><ul class="sidebar-sub-headers"><li class="sidebar-sub-header"><a href="/warp/documentation/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/#workflow-installation-and-requirements" class="sidebar-link">Workflow Installation and Requirements</a></li><li class="sidebar-sub-header"><a href="/warp/documentation/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/#software-version-requirements" class="sidebar-link">Software Version Requirements</a></li><li class="sidebar-sub-header"><a href="/warp/documentation/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/#input-descriptions" class="sidebar-link">Input Descriptions</a></li><li class="sidebar-sub-header"><a href="/warp/documentation/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/#gdc-reference-genome" class="sidebar-link">GDC Reference Genome</a></li></ul></li><li class="sidebar-sub-header"><a href="/warp/documentation/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/#workflow-tasks-and-tools" class="sidebar-link">Workflow Tasks and Tools</a></li><li class="sidebar-sub-header"><a href="/warp/documentation/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/#workflow-outputs" class="sidebar-link">Workflow Outputs</a></li><li class="sidebar-sub-header"><a href="/warp/documentation/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/#important-notes" class="sidebar-link">Important Notes</a></li><li class="sidebar-sub-header"><a href="/warp/documentation/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/#contact-us" class="sidebar-link">Contact Us</a></li><li class="sidebar-sub-header"><a href="/warp/documentation/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/#licensing" class="sidebar-link">Licensing</a></li></ul></li></ul></section></li><li><section class="sidebar-group collapsable is-sub-group depth-1"><p class="sidebar-heading"><span>Illumina Genotyping Arrays Pipeline</span> <span class="arrow right"></span></p> <!----></section></li><li><section class="sidebar-group collapsable is-sub-group depth-1"><p class="sidebar-heading"><span>Optimus Pipeline</span> <span class="arrow right"></span></p> <!----></section></li><li><section class="sidebar-group collapsable is-sub-group depth-1"><p class="sidebar-heading"><span>Single Cell ATAC Seq Pipeline</span> <span class="arrow right"></span></p> <!----></section></li><li><section class="sidebar-group collapsable is-sub-group depth-1"><p class="sidebar-heading"><span>Smart-seq2 Multi Sample Pipeline</span> <span class="arrow right"></span></p> <!----></section></li><li><section class="sidebar-group collapsable is-sub-group depth-1"><p class="sidebar-heading"><span>Smart-seq2 Single Sample Pipeline</span> <span class="arrow right"></span></p> <!----></section></li><li><section class="sidebar-group collapsable is-sub-group depth-1"><p class="sidebar-heading"><span>Whole Genome Germline Single Sample Pipeline</span> <span class="arrow right"></span></p> <!----></section></li></ul></section></li></ul> </aside> <main class="page"> <div class="theme-default-content content__default"><h1 id="genomic-data-commons-gdc-whole-genome-somatic-single-sample-overview"><a href="#genomic-data-commons-gdc-whole-genome-somatic-single-sample-overview" class="header-anchor">#</a> Genomic Data Commons (GDC) Whole Genome Somatic Single Sample Overview</h1> <table><thead><tr><th style="text-align:center;">Pipeline Version</th> <th style="text-align:center;">Date Updated</th> <th style="text-align:center;">Documentation Author</th> <th style="text-align:center;">Questions or Feedback</th></tr></thead> <tbody><tr><td style="text-align:center;"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">GDCWholeGenomeSomaticSingleSample_v1.0.1<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td style="text-align:center;">January, 2021</td> <td style="text-align:center;"><a href="mailto:ekiernan@broadinstitute.org">Elizabeth Kiernan</a></td> <td style="text-align:center;">Please file GitHub issues in warp or contact <a href="mailto:kdegatano@broadinstitute.org">Kylee Degatano</a></td></tr></tbody></table> <h2 id="introduction-to-the-gdc-whole-genome-somatic-single-sample-pipeline"><a href="#introduction-to-the-gdc-whole-genome-somatic-single-sample-pipeline" class="header-anchor">#</a> Introduction to the GDC Whole Genome Somatic Single Sample Pipeline</h2> <p>The GDC Whole Genome Somatic Single Sample (abbreviated GDC here) pipeline is the alignment and preprocessing workflow for genomic data designed for the National Cancer Institute's <a href="https://gdc.cancer.gov/about-gdc" target="_blank" rel="noopener noreferrer">Genomic Data Commons<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>.</p> <p>A high-level overview of the pipeline in addition to tool parameters are available on the <a href="https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/DNA_Seq_Variant_Calling_Pipeline/" target="_blank" rel="noopener noreferrer">GDC Documentation site<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>.</p> <p>Overall, the pipeline converts reads (CRAM or BAM) to FASTQ and (re)aligns them to the latest human reference genome (see the <a href="#gdc-reference-genome">GDC Reference Genome</a> section below). Each read group is aligned separately. Read group alignments that belong to a single sample are then merged and duplicate reads are flagged for downstream variant calling.</p> <p><img src="/warp/assets/img/GDC.713fe745.png" alt=""></p> <h2 id="set-up"><a href="#set-up" class="header-anchor">#</a> Set-up</h2> <h3 id="workflow-installation-and-requirements"><a href="#workflow-installation-and-requirements" class="header-anchor">#</a> Workflow Installation and Requirements</h3> <p>The workflow is written in the Workflow Description Language WDL and can be downloaded by cloning the <a href="https://github.com/broadinstitute/warp/tree/master" target="_blank" rel="noopener noreferrer">warp repository<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> in GitHub. The workflow can be deployed using <a href="https://github.com/broadinstitute/cromwell" target="_blank" rel="noopener noreferrer">Cromwell<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>, a GA4GH compliant, flexible workflow management system that supports multiple computing platforms.</p> <p>For the latest workflow version and release notes, please see the <a href="https://github.com/broadinstitute/warp/blob/master/beta-pipelines/broad/somatic/single_sample/wgs/gdc_genome/GDCWholeGenomeSomaticSingleSample.changelog.md" target="_blank" rel="noopener noreferrer">changelog<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>.</p> <h3 id="software-version-requirements"><a href="#software-version-requirements" class="header-anchor">#</a> Software Version Requirements</h3> <ul><li>GATK 4.0.7</li> <li>Picard 2.18.11 (Custom Docker is used to run software on Cromwell 52)</li> <li>Samtools 1.11</li> <li>Python 3.0</li> <li>Cromwell version support
<ul><li>Successfully tested on v52</li> <li>Does not work on versions &lt; v23 due to output syntax</li></ul></li> <li>Papi version support
<ul><li>Successfully tested on Papi v2</li></ul></li></ul> <h3 id="input-descriptions"><a href="#input-descriptions" class="header-anchor">#</a> Input Descriptions</h3> <p>The table below describes each of the GDC pipeline inputs. The workflow requires a <strong>single</strong> aligned CRAM or BAM file, or a <strong>single</strong> unmapped BAM (uBAM) as input, and is set up to run on samples with reads greater than 75 bp.</p> <p>For examples of how to specify each input as well as cloud locations for different example input files, see the <a href="https://github.com/broadinstitute/warp/blob/master/beta-pipelines/broad/somatic/single_sample/wgs/gdc_genome/input_files/GDCWholeGenomeSomaticSingleSample.inputs.json" target="_blank" rel="noopener noreferrer">input configuration file (JSON)<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>.</p> <table><thead><tr><th>Input name</th> <th>Description</th> <th>Type</th></tr></thead> <tbody><tr><td>input_cram (optional)</td> <td>A single mapped CRAM file; alternatively input can be a mapped BAM (<code>input_bam</code>) or unmapped BAM (uBAM; <code>ubam</code>). If using a CRAM file aligned to a reference different than the reference chosen for the GDC workflow (the <code>ref_fasta</code> input), than you must specify both the <code>cram_ref_fasta</code> and the <code>cram_ref_fasta_index</code> inputs</td> <td>File</td></tr> <tr><td>input_bam (optional)</td> <td>A single mapped BAM file; alternatively input can be a CRAM (<code>input_cram</code>) or uBAM (<code>ubam</code>)</td> <td>File</td></tr> <tr><td>cram_ref_fasta (optional)</td> <td>The reference file that was used to align the CRAM input (if used); if unspecified, the workflow will use the <code>ref_fasta</code> input file by default</td> <td>File</td></tr> <tr><td>cram_ref_fasta_index (optional)</td> <td>CRAM reference FASTA index for the <code>cram_ref_fata</code> if CRAM is used as workflow input</td> <td>File</td></tr> <tr><td>output_map (optional)</td> <td>Tab-separated file containing two columns: a list of all the read group IDs found in the input_cram (or input_bam) and a list of the desired name of the uBAMs generated for read group</td> <td>File</td></tr> <tr><td>unmapped_bam_suffix (optional)</td> <td>Optional string used to name the output uBAM file</td> <td>String</td></tr> <tr><td>ubam (optional)</td> <td>A single uBAM file; alternatively, input can be a mapped BAM (<code>input_bam</code>) or CRAM (<code>input_cram</code>)</td> <td>File</td></tr> <tr><td>contamination_vcf</td> <td>VCF file of common variant sites that is used for the check_contamination task</td> <td>File</td></tr> <tr><td>contamination_vcf_index</td> <td>Index file for the <code>contamination_vcf</code> input</td> <td>File</td></tr> <tr><td>dbsnp_vcf</td> <td>VCF file of known variation sites that can be used to exclude these sites from the analysis; used for the gatk_baserecalibrator task</td> <td>File</td></tr> <tr><td>dbsnp_vcf_index</td> <td>Index file for the <code>dbsnp_vcf</code> input</td> <td>File</td></tr> <tr><td>ref_fasta</td> <td>Reference FASTA used to convert an input CRAM to BAM; if CRAM is not used, this file does not need to be specified</td> <td>File</td></tr> <tr><td>ref_fai</td> <td>Reference FASTA index for the <code>ref_fasta</code> input</td> <td>File</td></tr> <tr><td>ref_dict</td> <td>BWA reference dictionary used for alignment</td> <td>File</td></tr> <tr><td>ref_amb</td> <td>BWA reference file used for alignment</td> <td>File</td></tr> <tr><td>ref_ann</td> <td>BWA reference file used for alignment</td> <td>File</td></tr> <tr><td>ref_bwt</td> <td>BWA reference file used for alignment</td> <td>File</td></tr> <tr><td>ref_pac</td> <td>BWA reference file used for alignment</td> <td>File</td></tr> <tr><td>ref_sa</td> <td>BWA reference file used for alignment</td> <td>File</td></tr></tbody></table> <h3 id="gdc-reference-genome"><a href="#gdc-reference-genome" class="header-anchor">#</a> GDC Reference Genome</h3> <p>The GDC uses the human reference genome GRCh38.d1.vd1 for all data processing. Unlike the GRCh38 reference used by WARP pipelines for production, the GDC reference includes decoy viral sequences for ten types of human viral genomes. You can learn more about the reference from the <a href="https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/DNA_Seq_Variant_Calling_Pipeline/#reference-genome" target="_blank" rel="noopener noreferrer">GDC documentation<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>.</p> <p>The reference files required for the GDC workflow are hosted in a <a href="https://console.cloud.google.com/storage/browser/gcp-public-data--broad-references/hg38/v0/gdc" target="_blank" rel="noopener noreferrer">public Google Bucket<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>.</p> <h2 id="workflow-tasks-and-tools"><a href="#workflow-tasks-and-tools" class="header-anchor">#</a> Workflow Tasks and Tools</h2> <p>The <a href="https://github.com/broadinstitute/warp/blob/develop/beta-pipelines/broad/somatic/single_sample/wgs/gdc_genome/GDCWholeGenomeSomaticSingleSample.wdl" target="_blank" rel="noopener noreferrer">workflow<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> imports a series of tasks either from the Workflow script or the Broad <a href="https://github.com/broadinstitute/warp/tree/master/tasks/broad" target="_blank" rel="noopener noreferrer">tasks library<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>.</p> <table><thead><tr><th>Task Name in WDL</th> <th>Tool</th> <th>Software</th> <th>Description</th></tr></thead> <tbody><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/cram_to_unmapped_bams/CramToUnmappedBams.wdl" target="_blank" rel="noopener noreferrer">CramToUnmappedBams<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td>view, index, RevertSam, ValidateSamFile, SortSam</td> <td><a href="http://www.htslib.org/terms/" target="_blank" rel="noopener noreferrer">SamTools<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> and <a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td>If a CRAM file is used as input, the task converts to uBAM and generates an output map that is then used to split the uBAM by readgroup. The resulting BAM is sorted by query name using Picard</td></tr> <tr><td>bam_readgroup_to_contents</td> <td>view</td> <td><a href="http://www.htslib.org" target="_blank" rel="noopener noreferrer">Samtools<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td>Extracts all the readgroups from the BAM header and returns a WDL array where each row is a readgroup.</td></tr> <tr><td>biobambam_bamtofastq</td> <td>bamtofastq</td> <td><a href="https://gitlab.com/german.tischler/biobambam2" target="_blank" rel="noopener noreferrer">biobambam2<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td>Converts the uBAMs to FASTQ</td></tr> <tr><td>emit_pe_records/emit_se_records</td> <td>---</td> <td>---</td> <td>Associates the fasta file(s) generated by biobambam_bamtofastq with their respective readgroup; creates an array of structs to be used as input for alignment</td></tr> <tr><td>bwa_pe/ bwa_se</td> <td>BWA mem, view</td> <td><a href="http://bio-bwa.sourceforge.net/" target="_blank" rel="noopener noreferrer">BWA<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> and <a href="http://www.htslib.org" target="_blank" rel="noopener noreferrer">Samtools<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td>Aligns FASTQ reads to the reference genome, generating an aligned BAM file</td></tr> <tr><td>picard_markduplicates</td> <td><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-" target="_blank" rel="noopener noreferrer">MarkDuplicates<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td>Runs MarkDuplicates in silent mode to locate and tag duplicate reads. Outputs a tagged BAM and metrics file.</td></tr> <tr><td>sort_and_index_markdup_bam</td> <td>sort</td> <td><a href="http://www.htslib.org" target="_blank" rel="noopener noreferrer">Samtools<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td>Sorts and indexes the BAM file; outputs the sorted BAM and index file</td></tr> <tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/CheckContaminationSomatic.wdl" target="_blank" rel="noopener noreferrer">check_contamination<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037593451" target="_blank" rel="noopener noreferrer">GetPileUpSummaries<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> and <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360036888972" target="_blank" rel="noopener noreferrer">CalculateContamination<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td>Summarizes counts of reads that support reference, alternate and other alleles for given sites and then calculates the fraction of reads coming from cross-sample contamination</td></tr> <tr><td>gatk_baserecalibrator</td> <td><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360036898312" target="_blank" rel="noopener noreferrer">BaseRecalibrator<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td>Generates recalibration table for Base Quality Score Recalibration (BQSR)</td></tr> <tr><td>gatk_applybqsr</td> <td><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037055712" target="_blank" rel="noopener noreferrer">ApplyBQSR<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td>Recalibrates the base qualities of the input reads based on the recalibration table produced by the BaseRecalibrator tool and outputs a recalibrated BAM file; uses the emit_original_quals parameter to write the original base qualities under the BAM OQ tag</td></tr> <tr><td>collect_insert_size_metrics</td> <td><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037055772" target="_blank" rel="noopener noreferrer">CollectInsertSizeMetrics<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></td> <td>Generates metrics about insert size distribution in the form of both a histogram and table (txt file)</td></tr></tbody></table> <h2 id="workflow-outputs"><a href="#workflow-outputs" class="header-anchor">#</a> Workflow Outputs</h2> <table><thead><tr><th>Output Name</th> <th>Description</th> <th>Type</th></tr></thead> <tbody><tr><td>validation_report (optional)</td> <td>Samtools validation report(s); only returned if an aligned CRAM or BAM file is used as workflow input</td> <td>TXT</td></tr> <tr><td>unmapped_bams (optional)</td> <td>uBAM file only returned if an aligned CRAM or BAM is used as workflow input</td> <td>BAM</td></tr> <tr><td>bam</td> <td>Base recalibrated BAM file</td> <td>BAM</td></tr> <tr><td>bai</td> <td>Index file for the BAM</td> <td>BAI</td></tr> <tr><td>md_metrics</td> <td>Picard MarkDuplicates metrics</td> <td>TXT</td></tr> <tr><td>insert_size_metrics</td> <td>Picard insert size metrics</td> <td>TXT</td></tr> <tr><td>insert_size_histogram_pdf</td> <td>Histogram representation of insert size metrics</td> <td>PDF</td></tr> <tr><td>contamination</td> <td>File containing a value indicating the fraction contamination</td> <td>TXT</td></tr></tbody></table> <h2 id="important-notes"><a href="#important-notes" class="header-anchor">#</a> Important Notes</h2> <ul><li>Runtime parameters are optimized for Broad's Google Cloud Platform implementation.</li> <li>Please visit the <a href="https://gatk.broadinstitute.org/hc/en-us/categories/360002310591" target="_blank" rel="noopener noreferrer">GATK Technical Documentation<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> site for further documentation on GATK-related workflows and tools.</li></ul> <h2 id="contact-us"><a href="#contact-us" class="header-anchor">#</a> Contact Us</h2> <p>Please help us make our tools better by contacting <a href="mailto:kdegatano@broadinstitute.org">Kylee Degatano</a> for pipeline-related suggestions or questions.</p> <h2 id="licensing"><a href="#licensing" class="header-anchor">#</a> Licensing</h2> <p>Copyright Broad Institute, 2020 | BSD-3</p> <p>The workflow script is released under the <strong>WDL open source code license (BSD-3)</strong> (full license text at https://github.com/broadinstitute/warp/blob/master/LICENSE). However, please note that the programs it calls may be subject to different licenses. Users are responsible for checking that they are authorized to run all programs before running this script.</p> <ul><li><a href="https://software.broadinstitute.org/gatk/download/licensing.php" target="_blank" rel="noopener noreferrer">GATK<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></li> <li><a href="http://bio-bwa.sourceforge.net/bwa.shtml#13" target="_blank" rel="noopener noreferrer">BWA<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></li> <li><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></li> <li><a href="http://www.htslib.org/terms/" target="_blank" rel="noopener noreferrer">Samtools<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></li></ul></div> <footer class="page-edit"><div class="edit-link"><a href="https://github.com/broadinstitute/warp/edit/master/website/docs/documentation/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/README.md" target="_blank" rel="noopener noreferrer">Edit this page on GitHub</a> <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></div> <div class="last-updated"><span class="prefix">Last Updated:</span> <span class="time">4/20/2021, 3:55:18 PM</span></div></footer> <div class="page-nav"><p class="inner"><span class="prev">
      
      <a href="/warp/documentation/Pipelines/Exome_Germline_Single_Sample_Pipeline/" class="prev">
        Exome Germline Single Sample Overview
      </a></span> <span class="next"><a href="/warp/documentation/Pipelines/Illumina_Genotyping_Arrays_Pipeline/IlluminaGenotypingArray.documentation.html">
        Illumina Genotyping Array Overview
      </a>
      
    </span></p></div> </main></div><div class="global-ui"><!----></div></div>
    <script src="/warp/assets/js/app.aea6aa50.js" defer></script><script src="/warp/assets/js/2.2043750e.js" defer></script><script src="/warp/assets/js/10.a9482500.js" defer></script>
  </body>
</html>
