<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Pipelines/ImputationBeagle_Pipeline/README">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">Imputation Overview | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/ImputationBeagle_Pipeline/README"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Imputation Overview | WARP"><meta data-rh="true" name="description" content="|                                                  Pipeline Version                                                  | Date Updated |        Documentation Author        |                             Questions or Feedback                              |"><meta data-rh="true" property="og:description" content="|                                                  Pipeline Version                                                  | Date Updated |        Documentation Author        |                             Questions or Feedback                              |"><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/ImputationBeagle_Pipeline/README"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/ImputationBeagle_Pipeline/README" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/ImputationBeagle_Pipeline/README" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.e21075f0.js" as="script">
<link rel="preload" href="/warp/assets/js/main.845a3bc9.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/ATAC/README">ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/BuildIndices_Pipeline/README">BuildIndices</a><button aria-label="Toggle the collapsible sidebar category &#x27;BuildIndices&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README">Exome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Exome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/README">Illumina Genotyping Array</a><button aria-label="Toggle the collapsible sidebar category &#x27;Illumina Genotyping Array&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/ImputationBeagle_Pipeline/README">ImputationBeagle</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Imputation_Pipeline/README">Imputation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Imputation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/JointGenotyping_Pipeline/README">JointGenotyping</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Multiome_Pipeline/README">Multiome scATAC and GEX</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus</a><button aria-label="Toggle the collapsible sidebar category &#x27;Optimus&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/PairedTag_Pipeline/README">Paired-Tag</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README">RNA with UMIs</a><button aria-label="Toggle the collapsible sidebar category &#x27;RNA with UMIs&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/SlideSeq_Pipeline/README">Slide-seq</a><button aria-label="Toggle the collapsible sidebar category &#x27;Slide-seq&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/SlideTags_Pipeline/README">Slide-tags</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README">Smart-seq2 Single Nucleus Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Nucleus Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/snm3C/README">Single Nucleus Methyl-Seq and Chromatin Capture</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Nucleus Methyl-Seq and Chromatin Capture&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Ultima_Genomics_Whole_Genome_Germline_Pipeline/README">Ultima Genomics Whole Genome Germline</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/README">Whole Genome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Whole Genome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">WARP Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">ImputationBeagle</span><meta itemprop="position" content="2"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Imputation Overview</h1><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases?q=ImputationBeagle_v2.0.0&amp;expanded=true" target="_blank" rel="noopener noreferrer">ImputationBeagle_v2.0.0</a></td><td align="center">August, 2025</td><td align="center">Terra Scientific Pipeline Services</td><td align="center">Please <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">file an issue in WARP</a>.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="introduction-to-the-array-imputation-pipeline">Introduction to the Array Imputation pipeline<a href="#introduction-to-the-array-imputation-pipeline" class="hash-link" aria-label="Direct link to Introduction to the Array Imputation pipeline" title="Direct link to Introduction to the Array Imputation pipeline">​</a></h2><p>The Array Imputation pipeline imputes missing genotypes from either a multi-sample VCF or an array of single-sample VCFs using a large genomic reference panel. It uses <a href="https://faculty.washington.edu/browning/beagle/beagle.html" target="_blank" rel="noopener noreferrer">Beagle</a> as the imputation tool. Overall, the pipeline filters, phases, and performs imputation on a multi-sample VCF. It outputs the imputed VCF along with key imputation metrics.</p><p><img loading="lazy" src="/warp/assets/images/pipeline-513207ca464abdc89b555e79da1c58e7.png" width="1862" height="1422" class="img_ev3q"></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="set-up">Set-up<a href="#set-up" class="hash-link" aria-label="Direct link to Set-up" title="Direct link to Set-up">​</a></h2><h3 class="anchor anchorWithStickyNavbar_LWe7" id="workflow-installation-and-requirements">Workflow installation and requirements<a href="#workflow-installation-and-requirements" class="hash-link" aria-label="Direct link to Workflow installation and requirements" title="Direct link to Workflow installation and requirements">​</a></h3><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/arrays/imputation_beagle/ImputationBeagle.wdl" target="_blank" rel="noopener noreferrer">Array Imputation workflow</a> is written in the Workflow Description Language (WDL) and can be deployed using a WDL-compatible execution engine like <a href="https://github.com/broadinstitute/cromwell" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH compliant, flexible workflow management system that supports multiple computing platforms.
To identify the latest workflow version and release notes, please see the Imputation workflow <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/arrays/imputation_beagle/ImputationBeagle.changelog.md" target="_blank" rel="noopener noreferrer">changelog</a>.
The latest release of the workflow, example data, and dependencies are available from the WARP releases page. To discover and search releases, use the WARP command-line tool <a href="https://github.com/broadinstitute/warp/tree/develop/wreleaser" target="_blank" rel="noopener noreferrer">Wreleaser</a>.</p><p>Using the Array Imputation pipeline
This pipeline is used by the <a href="http://allofus-anvil-imputation.terra.bio" target="_blank" rel="noopener noreferrer">All of Us + AnVIL Imputation Service</a>. If you choose to use this service, you can impute your samples against the 515,000+ genomes in the All of Us + AnVIL reference panel, which can provide greater accuracy at more sites.</p><div class="theme-admonition theme-admonition-tip alert alert--success admonition_LlT9"><div class="admonitionHeading_tbUL"><span class="admonitionIcon_kALy"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Try the Imputation pipeline in Terra</div><div class="admonitionContent_S0QG"><p>You can alternatively run the pipeline with your own panel, using this <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/arrays/imputation_beagle/ImputationBeagle.wdl" target="_blank" rel="noopener noreferrer">WDL</a>.</p></div></div><h3 class="anchor anchorWithStickyNavbar_LWe7" id="input-descriptions">Input descriptions<a href="#input-descriptions" class="hash-link" aria-label="Direct link to Input descriptions" title="Direct link to Input descriptions">​</a></h3><p>The table below describes each of the Array Imputation pipeline inputs. The workflow requires a  multi-sample VCF. These samples must be from the same species and genotyping chip.</p><p>For examples of how to specify each input in a configuration file, as well as cloud locations for different example input files, see the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/arrays/imputation_beagle/test_inputs/Plumbing/NA12878_x10_hg38_arrays.json" target="_blank" rel="noopener noreferrer">example input configuration file (JSON)</a>.</p><table><thead><tr><th>Input name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>ChunkLength</td><td>Size of chunks; default set to 25 MB.</td><td>Int</td></tr><tr><td>chunkOverlaps</td><td>Padding adding to the beginning and end of each chunk to reduce edge effects; default set 5 MB.</td><td>Int</td></tr><tr><td>sample_chunk_size</td><td>Number of samples to chunk by when processing (default: 1,000)</td><td>Int</td></tr><tr><td>multi_sample_vcf</td><td>Merged VCF containing multiple samples; can also use an array of individual VCFs.</td><td>File</td></tr><tr><td>ref_dict</td><td>Reference dictionary.</td><td>File</td></tr><tr><td>contigs</td><td>Array of allowed contigs/chromosomes to process</td><td>Array of strings</td></tr><tr><td>reference_panel_path_prefix</td><td>Path to the cloud storage containing the reference panel files for all contigs.</td><td>String</td></tr><tr><td>genetics_maps_path</td><td>Path to the cloud storage containing the genetic map files for all contigs.</td><td>File</td></tr><tr><td>output_callset_name</td><td>Output callset name.</td><td>String</td></tr><tr><td>bcf_suffix</td><td>File extension used for the BED in the reference panel.</td><td>String</td></tr><tr><td>bref3_suffix</td><td>File extension used for the BREF3 in the reference panel.</td><td>String</td></tr><tr><td>error_count_override</td><td>Override for error check on chunk qc (set to 0 for workflow to continue no matter how many errors exist)</td><td>Int</td></tr><tr><td>min_dr2_for_inclusion</td><td>Min value of DR2 to include in final output (default: 0.0)</td><td></td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="workflow-tasks-and-tools">Workflow tasks and tools<a href="#workflow-tasks-and-tools" class="hash-link" aria-label="Direct link to Workflow tasks and tools" title="Direct link to Workflow tasks and tools">​</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/arrays/imputation_beagle/ImputationBeagle.wdl" target="_blank" rel="noopener noreferrer">Array Imputation workflow</a> imports a series of tasks from the ImputationTasks WDL and ImputationBeagleTasks WDL, which are hosted in the Broad <a href="https://github.com/broadinstitute/warp/tree/develop/tasks/wdl" target="_blank" rel="noopener noreferrer">tasks library</a>. The table below describes each workflow task, including the task name, tools, relevant software and non-default parameters. </p><table><thead><tr><th>Task name (alias) in WDL</th><th>Tool</th><th>Software</th><th>Description</th></tr></thead><tbody><tr><td>CountSamples</td><td>query</td><td>bcftools</td><td>Uses the merged input VCF file to count the number of samples and output a TXT file containing the count.</td></tr><tr><td>CreateVcfIndex</td><td>index</td><td>bcftools</td><td>Creates index of input multisample vcf</td></tr><tr><td>CalculateContigsToProcess</td><td>Determine which contigs will be processed by the workflow</td><td>multi_sample_vcf, contigs</td><td>Extracts contigs from input VCF and filters by allowed contigs</td></tr><tr><td>CalculateChromsomeLength</td><td>grep</td><td>bash</td><td>Reads chromosome lengths from the reference dictionary and uses these to generate chunk intervals for the GenerateChunk task.</td></tr><tr><td>GenerateChunk</td><td>SelectVariants</td><td>GATK</td><td>Performs site filtering by selecting SNPs only and excluding InDels, removing duplicate sites from the VCF, selecting biallelic variants, excluding symbolic/mixed variants, and removing sites with a maximum fraction of samples with no-call genotypes greater than 0.1. Also subsets to only a specified chunk of the genome.</td></tr><tr><td>CountVariantsInChunks</td><td>CountVariants, intersect</td><td>GATK, bedtools</td><td>Counts variants in the filtered VCF file; Returns the number of chunks in the array and in the reference file.</td></tr><tr><td>CheckChunks</td><td>convert</td><td>bcftools</td><td>Confirms that there are no chunks where less than 3 sites or less than 50% of the sites in the array are also in the reference panel; if valid, creates a new VCF output.</td></tr><tr><td>StoreChunksInfo</td><td></td><td>R</td><td>Gathers all results from CheckChunks</td></tr><tr><td>ErrorWithMessageIfErrorCountNotZero</td><td></td><td>bash</td><td>Fails workflow if any chunks fail qc check.  Can be overridden with error_count_override input</td></tr><tr><td>SelectSamplesWithCut</td><td>cut</td><td>bash</td><td>Chunks vcf by sample_chunk_size if more than sample_chunk_size samples exist in input vcf</td></tr><tr><td>Phase</td><td></td><td>Beagle</td><td>Performs phasing on the filtered, validated VCF</td></tr><tr><td>Impute</td><td></td><td>Beagle</td><td>Performs imputation on the prephased VCF;</td></tr><tr><td>LocalizeAndSubsetVcfToRegion</td><td>SelectVariants</td><td>GATK</td><td>Remove padding from imputed vcf</td></tr><tr><td>QuerySampleChunkedVcfForReannotation</td><td>query</td><td>bcftools</td><td>Query DS, AP1, AP2 from sample chunked VCFs to be used when merging samples together</td></tr><tr><td>RemoveAPAnnotations</td><td>annotate</td><td>bcftools</td><td>Remove AP1, AP2 annotations to reduce file size now that they’re no longer needed</td></tr><tr><td>RecalculateDR2AndAFChunked</td><td></td><td>python</td><td>Used query output to summarize DS, AP1, AP2 values</td></tr><tr><td>MergeSampleChunksVcfsWithPaste</td><td>Paste, view</td><td>Bash, bcftools</td><td>Merge sample chunked VCFs</td></tr><tr><td>IndexMergedSampleChunksVcfs</td><td>index</td><td>bcftools</td><td>Creates index for sample chunk merged VCF</td></tr><tr><td>AggregateChunkedDR2AndAF</td><td></td><td>python</td><td>Take summarized DS, AP1, AP2 data and calculate AF and DR2</td></tr><tr><td>ReannotateDR2AndAF</td><td>annotate</td><td>bcftools</td><td>Reannotate DR2 and AF for sample chunk merged vcf</td></tr><tr><td>FilterVcfByDR2</td><td>Filter variants by DR2 threshold</td><td>imputed VCF, min_dr2_for_inclusion</td><td>Removes variants below provided min_dr2_for_inclusion value</td></tr><tr><td>UpdateHeader</td><td>UpdateVCFSequenceDictionary</td><td>GATK</td><td>Updates the header of the imputed VCF; adds contig lengths</td></tr><tr><td>GatherVcfsNoIndex</td><td>GatherVCFs</td><td>GATK</td><td>Gathers the array of imputed VCFs and merges them into one VCF output.</td></tr><tr><td>CreateIndexForGatheredVcf</td><td>index</td><td>bcftools</td><td>Creates index for final output vcf</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="workflow-outputs">Workflow outputs<a href="#workflow-outputs" class="hash-link" aria-label="Direct link to Workflow outputs" title="Direct link to Workflow outputs">​</a></h2><p>The table below summarizes the workflow outputs. If running the workflow on Cromwell, these outputs are found in the task execution directory.</p><table><thead><tr><th>Output name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>imputed_multisample_vcf</td><td>VCF from the CreateIndexForGatheredVcf task; contains imputed variants as well as missing variants from the input VCF.</td><td>VCF</td></tr><tr><td>imputed_multisample_vcf_index</td><td>Index file for VCF from the CreateIndexForGatheredVcf task.</td><td>Index</td></tr><tr><td>chunks_info</td><td>TSV from StoreChunksInfo task; contains the chunk intervals as well as the number of variants in the array.</td><td>TSV</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="important-notes">Important notes<a href="#important-notes" class="hash-link" aria-label="Direct link to Important notes" title="Direct link to Important notes">​</a></h2><ul><li>Runtime parameters are optimized for Broad&#x27;s Google Cloud Platform implementation.</li></ul><h2 class="anchor anchorWithStickyNavbar_LWe7" id="citing-the-imputation-pipeline">Citing the Imputation Pipeline<a href="#citing-the-imputation-pipeline" class="hash-link" aria-label="Direct link to Citing the Imputation Pipeline" title="Direct link to Citing the Imputation Pipeline">​</a></h2><p>If you use the Imputation Pipeline in your research, please consider citing our preprint:</p><p>Degatano, K., Awdeh, A., Cox III, R.S., Dingman, W., Grant, G., Khajouei, F., Kiernan, E., Konwar, K., Mathews, K.L., Palis, K., et al. Warp Analysis Research Pipelines: Cloud-optimized workflows for biological data processing and reproducible analysis. Bioinformatics 2025; btaf494. <a href="https://doi.org/10.1093/bioinformatics/btaf494" target="_blank" rel="noopener noreferrer">https://doi.org/10.1093/bioinformatics/btaf494</a></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="contact-us">Contact us<a href="#contact-us" class="hash-link" aria-label="Direct link to Contact us" title="Direct link to Contact us">​</a></h2><p>Help us make our tools better by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>; we welcome pipeline-related suggestions or questions.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="licensing">Licensing<a href="#licensing" class="hash-link" aria-label="Direct link to Licensing" title="Direct link to Licensing">​</a></h2><p>Copyright Broad Institute, 2020 | BSD-3</p><p>The workflow script is released under the <strong>WDL open source code license (BSD-3)</strong> (full license text at <a href="https://github.com/broadinstitute/warp/blob/master/LICENSE" target="_blank" rel="noopener noreferrer">https://github.com/broadinstitute/warp/blob/master/LICENSE</a>). However, please note that the programs it calls may be subject to different licenses. Users are responsible for checking that they are authorized to run all programs before running this script.</p><ul><li><a href="https://github.com/broadinstitute/gatk/blob/master/LICENSE.TXT" target="_blank" rel="noopener noreferrer">GATK</a></li><li><a href="https://github.com/broadinstitute/picard/blob/master/LICENSE.txt" target="_blank" rel="noopener noreferrer">Picard</a></li><li><a href="https://faculty.washington.edu/browning/beagle/beagle.html" target="_blank" rel="noopener noreferrer">Beagle</a></li><li><a href="https://github.com/samtools/bcftools/blob/develop/LICENSE" target="_blank" rel="noopener noreferrer">bcftools</a></li><li><a href="http://vcftools.sourceforge.net/license.html" target="_blank" rel="noopener noreferrer">vcftools</a></li></ul></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/ImputationBeagle_Pipeline/README.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2025-11-06T20:46:34.000Z">Nov 6, 2025</time></b> by <b>Robert Sidney Cox</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/Illumina_genotyping_array_spec"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">VCF Overview: Illumina Genotyping Array</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/Pipelines/Imputation_Pipeline/README"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Imputation Overview</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-array-imputation-pipeline" class="table-of-contents__link toc-highlight">Introduction to the Array Imputation pipeline</a></li><li><a href="#set-up" class="table-of-contents__link toc-highlight">Set-up</a><ul><li><a href="#workflow-installation-and-requirements" class="table-of-contents__link toc-highlight">Workflow installation and requirements</a></li><li><a href="#input-descriptions" class="table-of-contents__link toc-highlight">Input descriptions</a></li></ul></li><li><a href="#workflow-tasks-and-tools" class="table-of-contents__link toc-highlight">Workflow tasks and tools</a></li><li><a href="#workflow-outputs" class="table-of-contents__link toc-highlight">Workflow outputs</a></li><li><a href="#important-notes" class="table-of-contents__link toc-highlight">Important notes</a></li><li><a href="#citing-the-imputation-pipeline" class="table-of-contents__link toc-highlight">Citing the Imputation Pipeline</a></li><li><a href="#contact-us" class="table-of-contents__link toc-highlight">Contact us</a></li><li><a href="#licensing" class="table-of-contents__link toc-highlight">Licensing</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.e21075f0.js"></script>
<script src="/warp/assets/js/main.845a3bc9.js"></script>
</body>
</html>