<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.5">
<link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Blog Atom Feed">
<link rel="preconnect" href="https://www.google-analytics.com">
<script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-181048107-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview")</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><title data-react-helmet="true">Exome Germline Single Sample Overview | WARP</title><meta data-react-helmet="true" name="twitter:card" content="summary_large_image"><meta data-react-helmet="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README"><meta data-react-helmet="true" name="docsearch:language" content="en"><meta data-react-helmet="true" name="docsearch:version" content="current"><meta data-react-helmet="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="Exome Germline Single Sample Overview | WARP"><meta data-react-helmet="true" name="description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><meta data-react-helmet="true" property="og:description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><link data-react-helmet="true" rel="shortcut icon" href="/warp/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README" hreflang="x-default"><link data-react-helmet="true" rel="preconnect" href="https://BH4D9OD16A-dsn.algolia.net" crossorigin="anonymous"><link rel="stylesheet" href="/warp/assets/css/styles.57c5b218.css">
<link rel="preload" href="/warp/assets/js/runtime~main.ccf0ddc7.js" as="script">
<link rel="preload" href="/warp/assets/js/main.1d35face.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#" class="skipToContent_1oUP">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle toggle_3Zt9 react-toggle--disabled"><div class="react-toggle-track" role="button" tabindex="-1"><div class="react-toggle-track-check"><span class="toggle_71bT">ðŸŒœ</span></div><div class="react-toggle-track-x"><span class="toggle_71bT">ðŸŒž</span></div><div class="react-toggle-thumb"></div></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div><div class="searchBox_1Doo"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docs-wrapper docs-doc-page"><div class="docPage_31aa"><button class="clean-btn backToTopButton_35hR" type="button"><svg viewBox="0 0 24 24" width="28"><path d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z" fill="currentColor"></path></svg></button><aside class="docSidebarContainer_3Kbt"><div class="sidebar_15mo"><nav class="menu thin-scrollbar menu_Bmed menuWithAnnouncementBar_2WvA"><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">About WARP</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#">WARP Pipelines</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">CEMBA</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#" tabindex="0">Exome Germline Single Sample</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README">Exome Germline Single Sample Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/exome.methods">Exome Germline Single Sample v3.0.0 Methods</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">GDC Whole Genome Somatic Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Illumina Genotyping Array</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Imputation</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Optimus</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">RNA with UMIs</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Single Cell ATAC</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Multi-Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Single Nucleus Multi-Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Ultima Genomics Whole Genome Germline</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Whole Genome Germline Single Sample</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">WDL Best Practices</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">Contribution Guide</a></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_1CGd"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_3E-R"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></aside><main class="docMainContainer_3ufF"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_3FnS"><div class="docItemContainer_33ec"><article><div class="tocCollapsible_1PrD tocMobile_3Hoh"><button type="button" class="clean-btn tocCollapsibleButton_2O1e">On this page</button></div><div class="markdown"><header><h1>Exome Germline Single Sample Overview</h1></header><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases?q=ExomeGermlineSingleSample_v3.0.0&amp;expanded=true" target="_blank" rel="noopener noreferrer">ExomeGermlineSingleSample_v3.0.0</a></td><td align="center">November, 2021</td><td align="center"><a href="mailto:ekiernan@broadinstitute.org" target="_blank" rel="noopener noreferrer">Elizabeth Kiernan</a></td><td align="center">Please file GitHub issues in WARP or contact <a href="mailto:warp-pipelines-help@broadinstitute.org" target="_blank" rel="noopener noreferrer">the WARP team</a></td></tr></tbody></table><p>The Exome Germline Single Sample pipeline implements data pre-processing and initial variant calling according to the GATK Best Practices for germline SNP and Indel discovery in human exome sequencing data.</p><p>For a broad overview of the pipeline processes, read the GATK Best Practices documentation for <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035535912" target="_blank" rel="noopener noreferrer">data pre-processing</a> and for <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035535932" target="_blank" rel="noopener noreferrer">germline short variant discovery</a>.</p><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Want to try the Exome Germline Single Sample pipeline?</h5></div><div class="admonition-content"><p>You can test the pipeline in Terra! Go the <a href="https://app.terra.bio/#workspaces/warp-pipelines/Exome-Analysis-Pipeline" target="_blank" rel="noopener noreferrer">Exome-Analysis-Pipeline workspace</a> which includes sample data and workflows for preprocessing and initial variant calling, sample map generation, and joint genotyping.</p></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="set-up"></a>Set-up<a class="hash-link" href="#set-up" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="workflow-installation-and-requirements"></a>Workflow Installation and Requirements<a class="hash-link" href="#workflow-installation-and-requirements" title="Direct link to heading">#</a></h3><p>The Exome Germline Single Sample workflow is written in the Workflow Description Language WDL and can be downloaded by cloning the <a href="https://github.com/broadinstitute/warp/tree/master" target="_blank" rel="noopener noreferrer">warp repository</a> in GitHub. The workflow can be deployed using <a href="https://github.com/broadinstitute/cromwell" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH compliant, flexible workflow management system that supports multiple computing platforms. For the latest workflow version and release notes, please see the Exome Germline Single Sample <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/broad/dna_seq/germline/single_sample/exome/ExomeGermlineSingleSample.changelog.md" target="_blank" rel="noopener noreferrer">changelog</a>.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="software-version-requirements"></a>Software Version Requirements<a class="hash-link" href="#software-version-requirements" title="Direct link to heading">#</a></h3><ul><li><a href="https://github.com/broadinstitute/gatk/releases/tag/4.1.8.0" target="_blank" rel="noopener noreferrer">GATK 4.1.8.0</a></li><li>Picard 2.23.8</li><li>Samtools 1.11</li><li>Python 3.0</li><li>Cromwell version support<ul><li>Successfully tested on v52</li><li>Does not work on versions &lt; v23 due to output syntax</li></ul></li><li>Papi version support<ul><li>Successfully tested on Papi v2</li></ul></li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="input-requirements-and-expectations"></a>Input Requirements and Expectations<a class="hash-link" href="#input-requirements-and-expectations" title="Direct link to heading">#</a></h3><ul><li>Human exome sequencing data in unmapped BAM (uBAM) format</li><li>One or more read groups, one per uBAM file, all belonging to a single sample (SM)</li><li>Input uBAM files must additionally comply with the following requirements:<ul><li>Filenames all have the same suffix (we use &quot;.unmapped.bam&quot;)</li><li>Files must pass validation by ValidateSamFile</li><li>Reads are provided in query-sorted order</li><li>All reads must have an RG tag</li></ul></li><li>Reblocked GVCF output names must end in &quot;.rb.g.vcf.gz&quot;</li><li>Reference genome must be Hg38 with ALT contigs</li><li>Unique exome calling, target, and bait <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531852" target="_blank" rel="noopener noreferrer">.interval_list</a> obtained from the sequencing provider. Generally the calling, target, and bait files will not be the same.</li></ul><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="workflow-tasks-and-tools"></a>Workflow Tasks and Tools<a class="hash-link" href="#workflow-tasks-and-tools" title="Direct link to heading">#</a></h2><p>The Exome Germline Single Sample <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/broad/dna_seq/germline/single_sample/exome/ExomeGermlineSingleSample.wdl" target="_blank" rel="noopener noreferrer">workflow</a> imports a series of tasks from the WARP <a href="https://github.com/broadinstitute/warp/tree/master/tasks/broad" target="_blank" rel="noopener noreferrer">tasks library</a> and a DNASeq struct (<a href="https://github.com/broadinstitute/warp/blob/master/structs/dna_seq/DNASeqStructs.wdl" target="_blank" rel="noopener noreferrer">DNASeqStructs.wdl</a>) containing reference files from the <a href="https://github.com/broadinstitute/warp/tree/master/structs/dna_seq" target="_blank" rel="noopener noreferrer">structs library</a>.</p><p>You can read more about the software tools implemented in these tasks by reading the GATK <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035535912" target="_blank" rel="noopener noreferrer">data pre-processing</a> and <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035535932" target="_blank" rel="noopener noreferrer">germline short variant discovery</a> documentation.</p><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Want to use the Exome Germline Single Sample workflow in your publication?</h5></div><div class="admonition-content"><p>Check out the workflow <a href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/exome.methods">Methods</a> to get started!</p></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="workflow-outputs"></a>Workflow Outputs<a class="hash-link" href="#workflow-outputs" title="Direct link to heading">#</a></h2><ul><li>CRAM, CRAM index, and CRAM MD5</li><li><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037593171" target="_blank" rel="noopener noreferrer">Reblocked</a> GVCF and its GVCF index (read more in the <a href="#reblocking">Reblocking</a> section below)</li><li>BQSR report</li><li>Summary metrics; to read more about any particular metric, you can search the metric using the <a href="https://gatk.broadinstitute.org/hc/en-us/categories/360002302312" target="_blank" rel="noopener noreferrer">GATK documentation search</a></li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="reblocking"></a>Reblocking<a class="hash-link" href="#reblocking" title="Direct link to heading">#</a></h3><p>Reblocking is a process that compresses a HaplotypeCaller GVCF by merging homRef blocks according to new genotype quality (GQ) bands and facilitates joint genotyping by removing alt alleles that do not appear in the called genotype. </p><p>As of November 2021, reblocking is a default task in the Exome pipeline. To skip reblocking, add the following to the workflow&#x27;s input configuration file (JSON):</p><div class="codeBlockContainer_K1bP"><div class="codeBlockContent_hGly WDL"><pre tabindex="0" class="prism-code language-WDL codeBlock_23N8 thin-scrollbar" style="color:#bfc7d5;background-color:#292d3e"><code class="codeBlockLines_39YC"><span class="token-line" style="color:#bfc7d5"><span class="token plain">&quot;ExomeGermlineSingleSample.BamToGvcf.skip_reblocking&quot;: true</span></span></code></pre><button type="button" aria-label="Copy code to clipboard" class="copyButton_Ue-o clean-btn">Copy</button></div></div><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/GermlineVariantDiscovery.wdl" target="_blank" rel="noopener noreferrer">Reblocking task</a> uses the GATK ReblockGVCF tool with the arguments:</p><div class="codeBlockContainer_K1bP"><div class="codeBlockContent_hGly WDL"><pre tabindex="0" class="prism-code language-WDL codeBlock_23N8 thin-scrollbar" style="color:#bfc7d5;background-color:#292d3e"><code class="codeBlockLines_39YC"><span class="token-line" style="color:#bfc7d5"><span class="token plain">-do-qual-approx -floor-blocks -GQB 20 -GQB 30 -GQB 40 </span></span></code></pre><button type="button" aria-label="Copy code to clipboard" class="copyButton_Ue-o clean-btn">Copy</button></div></div><p>The following summarizes how reblocking affects the Exome GVCF and downstream tools compared to the GVCF produced with the default HaplotypeCaller GQ bands:</p><ol><li><p>PLs are omitted for homozygous reference sites to save spaceâ€“ GQs are output for genotypes, PLs can be approximated as [0, GQ, 2*GQ].</p></li><li><p>GQ resolution for homozygous reference genotypes is reduced (i.e. homRef GQs will be underconfident) which may affect analyses like de novo calling where well-calibrated reference genotype qualities are important.</p></li><li><p>Alleles that arenâ€™t called in the sample genotype are dropped. Each variant should have no more than two non-symbolic alt alleles, with the majority having just one plus &lt;NON_REF&gt;.</p></li><li><p>New annotations enable merging data for filtering without using genotypes. For example:</p><ul><li>RAW_GT_COUNT(S) for doing ExcessHet calculation from a sites-only file.</li><li>QUALapprox and/or AS_QUALapprox for doing QUAL approximation/filling. </li><li>QUAL VCF field from a combined sites-only field.</li><li>VarDP and/or AS_VarDP used to calculate QualByDepth/QD annotation for VQSR.</li></ul></li><li><p>The MIN_DP has been removed.</p></li><li><p>Reblocked GVCFs have the following cost/scale improvements:</p><ul><li>A reduced storage footprint compared with HaplotypeCaller GVCF output.</li><li>Fewer VariantContexts (i.e. lines) per VCF which speeds up GenomicsDB/Hail import.</li><li>Fewer alternate alleles which reduce memory requirements for merging.</li></ul></li></ol><p>Additionally, the 4 GQ band schema has specific improvements compared with the 7-band schema:</p><ol><li>It does not drop GQ0s; reblocked GVCFs should cover all the positions that the input GVCF covers.</li><li>It has no overlaps; the only overlapping positions should be two variants (i.e. deletions) on separate haplotypes.</li><li>No more no-calls; all genotypes should be called. Positions with no data will be homRef with GQ0.</li></ol><p>Read more about the reblocked GVCFs in the <a href="https://broadinstitute.github.io/warp/blog/Nov21_ReblockedGVCF" target="_blank" rel="noopener noreferrer">WARP Blog</a>.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="base-quality-scores"></a>Base quality scores<a class="hash-link" href="#base-quality-scores" title="Direct link to heading">#</a></h3><p>The final CRAM files have base quality scores binned according to the <a href="https://github.com/CCDG/Pipeline-Standardization/blob/master/PipelineStandard.md#base-quality-score-binning-scheme" target="_blank" rel="noopener noreferrer">Functional Equivalence specification</a> (<a href="https://www.nature.com/articles/s41467-018-06159-4" target="_blank" rel="noopener noreferrer">Regier et al., 2018</a>).</p><table><thead><tr><th>Original Score</th><th>Score after BQSR recalibration</th></tr></thead><tbody><tr><td>1-6</td><td>unchanged</td></tr><tr><td>7-12</td><td>10</td></tr><tr><td>13-22</td><td>20</td></tr><tr><td>22-infinity</td><td>30</td></tr></tbody></table><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="important-notes"></a>Important Notes<a class="hash-link" href="#important-notes" title="Direct link to heading">#</a></h2><ul><li>Runtime parameters are optimized for Broad&#x27;s Google Cloud Platform implementation.</li><li>For help running workflows on the Google Cloud Platform or locally please
view the following tutorial <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035530952" target="_blank" rel="noopener noreferrer">(How to) Execute Workflows from the gatk-workflows Git Organization</a>.</li><li>Please visit the <a href="https://gatk.broadinstitute.org/hc/en-us/categories/360002310591" target="_blank" rel="noopener noreferrer">GATK Technical Documentation</a> site for further documentation on our workflows and tools.</li><li>You can access relevant reference and resource bundles in the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890811" target="_blank" rel="noopener noreferrer">GATK Resource Bundle</a>.</li></ul><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="contact-us"></a>Contact Us<a class="hash-link" href="#contact-us" title="Direct link to heading">#</a></h2><p>This material is provided by the Data Science Platform group at the Broad Institute. Please direct any questions or concerns to one of our forum sites : <a href="https://gatk.broadinstitute.org/hc/en-us/community/topics" target="_blank" rel="noopener noreferrer">GATK</a> or <a href="https://support.terra.bio/hc/en-us/community/topics/360000500432" target="_blank" rel="noopener noreferrer">Terra</a>.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="licensing"></a>Licensing<a class="hash-link" href="#licensing" title="Direct link to heading">#</a></h2><p>Copyright Broad Institute, 2020 | BSD-3</p><p>The workflow script is released under the <strong>WDL open source code license (BSD-3)</strong> (full license text at <a href="https://github.com/broadinstitute/warp/blob/master/LICENSE" target="_blank" rel="noopener noreferrer">https://github.com/broadinstitute/warp/blob/master/LICENSE</a>). However, please note that the programs it calls may be subject to different licenses. Users are responsible for checking that they are authorized to run all programs before running this script.</p><ul><li><a href="https://software.broadinstitute.org/gatk/download/licensing.php" target="_blank" rel="noopener noreferrer">GATK</a></li><li><a href="http://bio-bwa.sourceforge.net/bwa.shtml#13" target="_blank" rel="noopener noreferrer">BWA</a></li><li><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a></li><li><a href="http://www.htslib.org/terms/" target="_blank" rel="noopener noreferrer">Samtools</a></li></ul></div><footer class="docusaurus-mt-lg"><div class="row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_2_ui" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_13-_">Last updated on <b><time datetime="2023-01-26T18:35:44.000Z">1/26/2023</time></b> by <b>ekiernan</b></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/CEMBA.methods"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Â« CEMBA_v1.1.0 Publication Methods</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/exome.methods"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Exome Germline Single Sample v3.0.0 Methods Â»</div></a></div></nav></div></div><div class="col col--3"><div class="tableOfContents_35-E thin-scrollbar"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#set-up" class="table-of-contents__link">Set-up</a><ul><li><a href="#workflow-installation-and-requirements" class="table-of-contents__link">Workflow Installation and Requirements</a></li><li><a href="#software-version-requirements" class="table-of-contents__link">Software Version Requirements</a></li><li><a href="#input-requirements-and-expectations" class="table-of-contents__link">Input Requirements and Expectations</a></li></ul></li><li><a href="#workflow-tasks-and-tools" class="table-of-contents__link">Workflow Tasks and Tools</a></li><li><a href="#workflow-outputs" class="table-of-contents__link">Workflow Outputs</a><ul><li><a href="#reblocking" class="table-of-contents__link">Reblocking</a></li><li><a href="#base-quality-scores" class="table-of-contents__link">Base quality scores</a></li></ul></li><li><a href="#important-notes" class="table-of-contents__link">Important Notes</a></li><li><a href="#contact-us" class="table-of-contents__link">Contact Us</a></li><li><a href="#licensing" class="table-of-contents__link">Licensing</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item"><span>GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_3J9K"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></span></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright Â© 2023 Copyright Â© Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.ccf0ddc7.js"></script>
<script src="/warp/assets/js/main.1d35face.js"></script>
</body>
</html>