<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Deprecated_Pipelines/Smart-seq2_Single_Sample_Pipeline/Loom_schema">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">What&#x27;s in the Smart-seq2 Single Sample Pipeline Loom File? | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Deprecated_Pipelines/Smart-seq2_Single_Sample_Pipeline/Loom_schema"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="What&#x27;s in the Smart-seq2 Single Sample Pipeline Loom File? | WARP"><meta data-rh="true" name="description" content="The Loom file is an HDF5 file generated using Loompy v.3.0.6. It contains global attributes describing the Loom file and workflow that generated it (Table 1), column attributes detailing metrics for the individual cells (Table 2) and row attributes detailing metrics for individual genes (Table 3). The main matrix contains the RSEM TPMs and there is an additional layer that contains RSEM expectedcounts (the named &quot;estimatedcounts&quot;)."><meta data-rh="true" property="og:description" content="The Loom file is an HDF5 file generated using Loompy v.3.0.6. It contains global attributes describing the Loom file and workflow that generated it (Table 1), column attributes detailing metrics for the individual cells (Table 2) and row attributes detailing metrics for individual genes (Table 3). The main matrix contains the RSEM TPMs and there is an additional layer that contains RSEM expectedcounts (the named &quot;estimatedcounts&quot;)."><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Deprecated_Pipelines/Smart-seq2_Single_Sample_Pipeline/Loom_schema"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Deprecated_Pipelines/Smart-seq2_Single_Sample_Pipeline/Loom_schema" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Deprecated_Pipelines/Smart-seq2_Single_Sample_Pipeline/Loom_schema" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.4d6b11e4.js" as="script">
<link rel="preload" href="/warp/assets/js/main.845a3bc9.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">CEMBA</a><button aria-label="Toggle the collapsible sidebar category &#x27;CEMBA&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/README">GDC Whole Genome Somatic Single Sample</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Single_Cell_ATAC_Seq_Pipeline/README">Single Cell ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Cell ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Multi_Sample_Pipeline/README">Smart-seq2 Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Sample_Pipeline/README">Smart-seq2 Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Deprecated_Pipelines/Smart-seq2_Single_Sample_Pipeline/Loom_schema">What&#x27;s in the Smart-seq2 Single Sample Pipeline Loom File?</a></li></ul></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">Deprecated Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/warp/docs/Pipelines/Smart-seq2_Single_Sample_Pipeline/README"><span itemprop="name">Smart-seq2 Single Sample</span></a><meta itemprop="position" content="2"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">What&#x27;s in the Smart-seq2 Single Sample Pipeline Loom File?</span><meta itemprop="position" content="3"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>What&#x27;s in the Smart-seq2 Single Sample Pipeline Loom File?</h1><p>The Loom file is an HDF5 file generated using <a href="http://loompy.org/" target="_blank" rel="noopener noreferrer">Loompy v.3.0.6</a>. It contains global attributes describing the Loom file and workflow that generated it (<a href="#table-1-global-attributes">Table 1</a>), column attributes detailing metrics for the individual cells (<a href="#table-2-column-attributes-cell-metrics">Table 2</a>) and row attributes detailing metrics for individual genes (<a href="#table-3-row-attributes-gene-metrics">Table 3</a>). The main matrix contains the RSEM TPMs and there is an additional layer that contains RSEM expected_counts (the named &quot;estimated_counts&quot;).</p><p>The tables below document these metrics, list which tools generate them, and define them. This Loom file is the default matrix output of the Smart-seq2 Single Sample pipeline.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="table-1-global-attributes">Table 1. Global Attributes<a href="#table-1-global-attributes" class="hash-link" aria-label="Direct link to Table 1. Global Attributes" title="Direct link to Table 1. Global Attributes">​</a></h2><p>The global attributes in the Loom apply to the whole file, not any specific part.</p><table><thead><tr><th align="left">Attribute</th><th align="left">Details</th></tr></thead><tbody><tr><td align="left">pipeline_version</td><td align="left">String with the pipeline version</td></tr><tr><td align="left">LOOM_SPEC_VERSION</td><td align="left">String with the loom file spec version</td></tr><tr><td align="left">CreationDate</td><td align="left">Date Loom file was generated</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="table-2-column-attributes-cell-metrics">Table 2. Column Attributes (Cell Metrics)<a href="#table-2-column-attributes-cell-metrics" class="hash-link" aria-label="Direct link to Table 2. Column Attributes (Cell Metrics)" title="Direct link to Table 2. Column Attributes (Cell Metrics)">​</a></h2><table><thead><tr><th align="left">Cell Metrics</th><th align="center">Program</th><th align="left">Details</th></tr></thead><tbody><tr><td align="left"><code>CellID</code></td><td align="center">---</td><td align="left">The unique identifier for each cell; read from the input_id</td></tr><tr><td align="left"><code>cell_names</code></td><td align="center">---</td><td align="left">The unique identifier for each cell; read from the input_id and identical to <code>CellID</code></td></tr><tr><td align="left"><code>input_id</code></td><td align="center">---</td><td align="left">The input_id listed in the pipeline configuration file</td></tr><tr><td align="left"><code>input_id_metadata_field</code></td><td align="center">---</td><td align="left">Optional identifier for the metadata field used for the <code>input_id</code></td></tr><tr><td align="left"><code>input_name</code></td><td align="center">---</td><td align="left">The optional string provided in the pipeline configuration to further identify samples</td></tr><tr><td align="left"><code>input_name_metadata_field</code></td><td align="center">---</td><td align="left">Optional identifier for the metadata field used for the <code>input_name</code></td></tr><tr><td align="left"><code>ACCUMULATION_LEVEL</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>ALIGNED_READS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>AT_DROPOUT</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>Aligned 0 time</code></td><td align="center"><a href="http://daehwankimlab.github.io/hisat2/manual/" target="_blank" rel="noopener noreferrer">HISAT2</a></td><td align="left">Number and percent reads aligned 0 times</td></tr><tr><td align="left"><code>Aligned 1 time</code></td><td align="center"><a href="http://daehwankimlab.github.io/hisat2/manual/" target="_blank" rel="noopener noreferrer">HISAT2</a></td><td align="left">Number and percent reads aligned 1 time</td></tr><tr><td align="left"><code>Aligned &gt;1 times</code></td><td align="center"><a href="http://daehwankimlab.github.io/hisat2/manual/" target="_blank" rel="noopener noreferrer">HISAT2</a></td><td align="left">Number and percent reads aligned more than 1 time</td></tr><tr><td align="left"><code>BAD_CYCLES.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>CODING_BASES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>CORRECT_STRAND_READS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>ESTIMATED_LIBRARY_SIZE</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>GC_DROPOUT</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>GC_NC_0_19</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>GC_NC_20_39</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>GC_NC_40_59</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>GC_NC_60_79</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>GC_NC_80_100</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>IGNORED_READS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>INCORRECT_STRAND_READS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>INTERGENIC_BASES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>INTRONIC_BASES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>MEAN_READ_LENGTH.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>MEDIAN_3PRIME_BIAS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>MEDIAN_5PRIME_BIAS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>MEDIAN_5PRIME_TO_3PRIME_BIAS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>MEDIAN_CV_COVERAGE</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>NUM_R1_TRANSCRIPT_STRAND_READS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>NUM_R2_TRANSCRIPT_STRAND_READS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>NUM_UNEXPLAINED_READS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>Overall alignment rate</code></td><td align="center"><a href="http://daehwankimlab.github.io/hisat2/manual/" target="_blank" rel="noopener noreferrer">HISAT2</a></td><td align="left">Overall percent of reads that aligned</td></tr><tr><td align="left"><code>PCT_ADAPTER.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_CHIMERAS.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_CODING_BASES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_CORRECT_STRAND_READS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_INTERGENIC_BASES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_INTRONIC_BASES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_MRNA_BASES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_PF_READS.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_PF_READS_ALIGNED.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_PF_READS_IMPROPER_PAIRS.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_R1_TRANSCRIPT_STRAND_READS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_R2_TRANSCRIPT_STRAND_READS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_READS_ALIGNED_IN_PAIRS.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_RIBOSOMAL_BASES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_USABLE_BASES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PCT_UTR_BASES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PERCENT_DUPLICATION</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PF_ALIGNED_BASES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PF_ALIGNED_BASES.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PF_BASES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PF_HQ_ALIGNED_BASES.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PF_HQ_ALIGNED_Q20_BASES.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PF_HQ_ALIGNED_READS.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PF_HQ_ERROR_RATE.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PF_HQ_MEDIAN_MISMATCHES.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PF_INDEL_RATE.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PF_MISMATCH_RATE.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PF_NOISE_READS.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PF_READS.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PF_READS_ALIGNED.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>PF_READS_IMPROPER_PAIRS.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>READS_ALIGNED_IN_PAIRS.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>READS_USED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>READ_PAIRS_EXAMINED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>READ_PAIR_DUPLICATES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>READ_PAIR_OPTICAL_DUPLICATES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>RIBOSOMAL_BASES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>SECONDARY_OR_SUPPLEMENTARY_RDS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>STRAND_BALANCE.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>TOTAL_CLUSTERS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>TOTAL_READS.UNPAIRED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>Total reads</code></td><td align="center"><a href="http://daehwankimlab.github.io/hisat2/manual/" target="_blank" rel="noopener noreferrer">HISAT2</a></td><td align="left">Total number of aligned reads</td></tr><tr><td align="left"><code>UNMAPPED_READS</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>UNPAIRED_READS_EXAMINED</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>UNPAIRED_READ_DUPLICATES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>UTR_BASES</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>WINDOW_SIZE</code></td><td align="center"><a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a></td><td align="left">See <a href="https://broadinstitute.github.io/picard/picard-metric-definitions.html" target="_blank" rel="noopener noreferrer">Picard</a> documentation</td></tr><tr><td align="left"><code>alignable reads</code></td><td align="center"><a href="https://github.com/deweylab/RSEM/blob/master/cnt_file_description.txt" target="_blank" rel="noopener noreferrer">RSEM cnt file</a></td><td align="left">The number of alignable reads</td></tr><tr><td align="left"><code>filtered reads</code></td><td align="center"><a href="https://github.com/deweylab/RSEM/blob/master/cnt_file_description.txt" target="_blank" rel="noopener noreferrer">RSEM cnt file</a></td><td align="left">The number of filtered reads due to too many alignments</td></tr><tr><td align="left"><code>multiple mapped</code></td><td align="center"><a href="https://github.com/deweylab/RSEM/blob/master/cnt_file_description.txt" target="_blank" rel="noopener noreferrer">RSEM cnt file</a></td><td align="left">The number of reads aligned to multiple genes</td></tr><tr><td align="left"><code>strand</code></td><td align="center"><a href="https://github.com/deweylab/RSEM/blob/master/cnt_file_description.txt" target="_blank" rel="noopener noreferrer">RSEM cnt file</a></td><td align="left">The RSEM read_type; describes if data is single- or paired-end</td></tr><tr><td align="left"><code>total alignments</code></td><td align="center"><a href="https://github.com/deweylab/RSEM/blob/master/cnt_file_description.txt" target="_blank" rel="noopener noreferrer">RSEM cnt file</a></td><td align="left">The RSEM nHits; the number of total alignments</td></tr><tr><td align="left"><code>total reads</code></td><td align="center"><a href="https://github.com/deweylab/RSEM/blob/master/cnt_file_description.txt" target="_blank" rel="noopener noreferrer">RSEM cnt file</a></td><td align="left">The number of total alignable reads</td></tr><tr><td align="left"><code>unalignable reads</code></td><td align="center"><a href="https://github.com/deweylab/RSEM/blob/master/cnt_file_description.txt" target="_blank" rel="noopener noreferrer">RSEM cnt file</a></td><td align="left">The number of  reads unalignable</td></tr><tr><td align="left"><code>uncertain reads</code></td><td align="center"><a href="https://github.com/deweylab/RSEM/blob/master/cnt_file_description.txt" target="_blank" rel="noopener noreferrer">RSEM cnt file</a></td><td align="left">The number of  reads aligned to multiple locations</td></tr><tr><td align="left"><code>unique aligned</code></td><td align="center"><a href="https://github.com/deweylab/RSEM/blob/master/cnt_file_description.txt" target="_blank" rel="noopener noreferrer">RSEM cnt file</a></td><td align="left">The number of reads uniquely alignable to one gene</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="table-3-row-attributes-gene-metrics">Table 3. Row Attributes (Gene Metrics)<a href="#table-3-row-attributes-gene-metrics" class="hash-link" aria-label="Direct link to Table 3. Row Attributes (Gene Metrics)" title="Direct link to Table 3. Row Attributes (Gene Metrics)">​</a></h2><table><thead><tr><th>Gene Metrics</th><th>Program</th><th>Details</th></tr></thead><tbody><tr><td><code>ensembl_ids</code></td><td><a href="https://www.gencodegenes.org/" target="_blank" rel="noopener noreferrer">GENCODE GTF</a></td><td>The gene_id listed in the GENCODE GTF</td></tr><tr><td><code>gene_names</code></td><td><a href="https://www.gencodegenes.org/" target="_blank" rel="noopener noreferrer">GENCODE GTF</a></td><td>The unique gene_name provided in the GENCODE GTF</td></tr><tr><td><code>Gene</code></td><td><a href="https://www.gencodegenes.org/" target="_blank" rel="noopener noreferrer">GENCODE GTF</a></td><td>The unique gene_name provided in the GENCODE GTF; identical to attribute in <code>gene_names</code></td></tr></tbody></table></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Deprecated_Pipelines/Smart-seq2_Single_Sample_Pipeline/Loom_schema.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2025-11-12T16:18:32.000Z">Nov 12, 2025</time></b> by <b>M. Morgan Aster</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/Pipelines/Smart-seq2_Single_Sample_Pipeline/README"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Smart-seq2 Single Sample Overview</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/contribution/README"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Welcome</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#table-1-global-attributes" class="table-of-contents__link toc-highlight">Table 1. Global Attributes</a></li><li><a href="#table-2-column-attributes-cell-metrics" class="table-of-contents__link toc-highlight">Table 2. Column Attributes (Cell Metrics)</a></li><li><a href="#table-3-row-attributes-gene-metrics" class="table-of-contents__link toc-highlight">Table 3. Row Attributes (Gene Metrics)</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.4d6b11e4.js"></script>
<script src="/warp/assets/js/main.845a3bc9.js"></script>
</body>
</html>