<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.5">
<link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Blog Atom Feed">
<link rel="preconnect" href="https://www.google-analytics.com">
<script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-181048107-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview")</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><title data-react-helmet="true">Slide-seq Overview | WARP</title><meta data-react-helmet="true" name="twitter:card" content="summary_large_image"><meta data-react-helmet="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/SlideSeq_Pipeline/README"><meta data-react-helmet="true" name="docsearch:language" content="en"><meta data-react-helmet="true" name="docsearch:version" content="current"><meta data-react-helmet="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="Slide-seq Overview | WARP"><meta data-react-helmet="true" name="description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><meta data-react-helmet="true" property="og:description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><link data-react-helmet="true" rel="shortcut icon" href="/warp/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/SlideSeq_Pipeline/README"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/SlideSeq_Pipeline/README" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/SlideSeq_Pipeline/README" hreflang="x-default"><link data-react-helmet="true" rel="preconnect" href="https://BH4D9OD16A-dsn.algolia.net" crossorigin="anonymous"><link rel="stylesheet" href="/warp/assets/css/styles.57c5b218.css">
<link rel="preload" href="/warp/assets/js/runtime~main.702083b4.js" as="script">
<link rel="preload" href="/warp/assets/js/main.2b70a255.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#" class="skipToContent_1oUP">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle toggle_3Zt9 react-toggle--disabled"><div class="react-toggle-track" role="button" tabindex="-1"><div class="react-toggle-track-check"><span class="toggle_71bT">üåú</span></div><div class="react-toggle-track-x"><span class="toggle_71bT">üåû</span></div><div class="react-toggle-thumb"></div></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div><div class="searchBox_1Doo"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docs-wrapper docs-doc-page"><div class="docPage_31aa"><button class="clean-btn backToTopButton_35hR" type="button"><svg viewBox="0 0 24 24" width="28"><path d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z" fill="currentColor"></path></svg></button><aside class="docSidebarContainer_3Kbt"><div class="sidebar_15mo"><nav class="menu thin-scrollbar menu_Bmed menuWithAnnouncementBar_2WvA"><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">About WARP</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#">WARP Pipelines</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">CEMBA</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Exome Germline Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">GDC Whole Genome Somatic Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Illumina Genotyping Array</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Imputation</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Optimus</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">RNA with UMIs</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Single Cell ATAC</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#" tabindex="0">Slide-seq</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/SlideSeq_Pipeline/README">Slide-seq Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/SlideSeq_Pipeline/count-matrix-overview">Slide-seq Count Matrix Overview</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Multi-Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Single Nucleus Multi-Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Ultima Genomics Whole Genome Germline</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Whole Genome Germline Single Sample</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">WDL Best Practices</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">Contribution Guide</a></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_1CGd"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_3E-R"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></aside><main class="docMainContainer_3ufF"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_3FnS"><div class="docItemContainer_33ec"><article><div class="tocCollapsible_1PrD tocMobile_3Hoh"><button type="button" class="clean-btn tocCollapsibleButton_2O1e">On this page</button></div><div class="markdown"><header><h1>Slide-seq Overview</h1></header><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">SlideSeq v1.0.1</a></td><td align="center">March, 2023</td><td align="center">Elizabeth Kiernan &amp; Kaylee Mathews</td><td align="center">Please file GitHub issues in warp or contact <a href="mailto:warp-pipelines-help@broadinstitute.org" target="_blank" rel="noopener noreferrer">documentation authors</a></td></tr></tbody></table><p><img alt="SlideSeq_diagram" src="/warp/assets/images/slide-seq_diagram-0aab9bd56ca06869890d086d59b19c1d.png"></p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="introduction-to-the-slide-seq-workflow"></a>Introduction to the Slide-seq workflow<a class="hash-link" href="#introduction-to-the-slide-seq-workflow" title="Direct link to heading">#</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/slideseq/SlideSeq.wdl" target="_blank" rel="noopener noreferrer">Slide-seq workflow</a> is an open-source, cloud-optimized pipeline developed in collaboration with the <a href="https://biccn.org/" target="_blank" rel="noopener noreferrer">BRAIN Initiative Cell Census Network</a> (BICCN) and the BRAIN Initiative Cell Atlas Network (BICAN). It supports the processing of spatial transcriptomic data generated with the <a href="https://www.science.org/doi/10.1126/science.aaw1219" target="_blank" rel="noopener noreferrer">Slide-seq</a> (commercialized as <a href="https://curiobioscience.com/product/" target="_blank" rel="noopener noreferrer">Curio Seeker</a>) assay.</p><p>Overall, the workflow corrects bead barcodes, aligns reads to the genome, generates a count matrix, calculates summary metrics for genes, barcodes, and UMIs, and returns read outputs in BAM format.</p><p>Slide-seq has been validated for analyzing mouse datasets generated with the Slide-seq assay. Learn more in the <a href="#validation-against-on-prem-pipeline">validation section</a>.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="where-did-the-pipeline-come-from"></a>Where did the pipeline come from?<a class="hash-link" href="#where-did-the-pipeline-come-from" title="Direct link to heading">#</a></h4><p>The Broad Pipelines Development team worked closely with members of the Evan Macosko Laboratory to adapt an on-premise pipeline into the cloud-based Slide-seq pipeline described here. Thank you to James Webber, Jonah Langlieb, Nina Sachdev, and the rest of the Evan Macosko Lab for their commitment to <a href="https://www.nature.com/articles/sdata201618" target="_blank" rel="noopener noreferrer">FAIR Principles</a> and for helping create the cloud-based Slide-seq pipeline.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="quickstart-table"></a>Quickstart table<a class="hash-link" href="#quickstart-table" title="Direct link to heading">#</a></h2><p>The following table provides a quick glance at the Slide-seq pipeline features:</p><table><thead><tr><th>Pipeline features</th><th>Description</th><th>Source</th></tr></thead><tbody><tr><td>Assay type</td><td>Slide-seq</td><td><a href="https://www.science.org/doi/10.1126/science.aaw1219" target="_blank" rel="noopener noreferrer">Rodriques et al. 2019</a></td></tr><tr><td>Overall workflow</td><td>Quality control and transcriptome quantification module</td><td>Code available from <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/slideseq/SlideSeq.wdl" target="_blank" rel="noopener noreferrer">GitHub</a></td></tr><tr><td>Workflow language</td><td>WDL 1.0</td><td><a href="https://github.com/openwdl/wdl" target="_blank" rel="noopener noreferrer">openWDL</a></td></tr><tr><td>Genomic Reference Sequence</td><td>M23 (GRCm38.p6) mouse genome primary sequence</td><td>GENCODE <a href="https://www.gencodegenes.org/mouse/release_M23.html" target="_blank" rel="noopener noreferrer">mouse reference files</a></td></tr><tr><td>Transcriptomic reference annotation</td><td>M23 mouse transcriptome built with the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/build_indices/BuildIndices.wdl" target="_blank" rel="noopener noreferrer">BuildIndices workflow</a></td><td>GENCODE <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/gencode.vM23.annotation.gff3.gz" target="_blank" rel="noopener noreferrer">mouse GTF</a>; <a href="https://console.cloud.google.com/storage/browser/_details/gcp-public-data--broad-references/mm10/v0/single_nucleus/star/modified_star_2.7.9a_primary_gencode_mouse_vM23.tar;tab=live_object" target="_blank" rel="noopener noreferrer">modified version</a> available in Broad‚Äôs public reference bucket</td></tr><tr><td>Aligner and transcript quantification</td><td>STARsolo</td><td><a href="https://www.biorxiv.org/content/10.1101/2021.05.05.442755v1" target="_blank" rel="noopener noreferrer">Kaminow et al. 2021</a></td></tr><tr><td>Data input file format</td><td>File format in which sequencing data is provided</td><td><a href="https://academic.oup.com/nar/article/38/6/1767/3112533" target="_blank" rel="noopener noreferrer">FASTQ</a></td></tr><tr><td>Data output file format</td><td>File formats in which Slide-seq output is provided</td><td><a href="http://samtools.github.io/hts-specs/" target="_blank" rel="noopener noreferrer">BAM</a>, Python NumPy arrays, and Loom (generated with <a href="http://loompy.org/" target="_blank" rel="noopener noreferrer">Loompy)</a></td></tr></tbody></table><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="set-up"></a>Set-up<a class="hash-link" href="#set-up" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="slide-seq-installation"></a>Slide-seq installation<a class="hash-link" href="#slide-seq-installation" title="Direct link to heading">#</a></h3><p>To download the latest Slide-seq release, see the release tags prefixed with &quot;SlideSeq&quot; on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a>. All Slide-seq pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/slideseq/SlideSeq.changelog.md" target="_blank" rel="noopener noreferrer">Slide-seq changelog</a>. </p><p>To search releases of this and other pipelines, use the WARP command-line tool <a href="https://github.com/broadinstitute/warp/tree/master/wreleaser" target="_blank" rel="noopener noreferrer">Wreleaser</a>.</p><p>If you‚Äôre running a Slide-seq workflow version prior to the latest release, the accompanying documentation for that release may be downloaded with the source code on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a> (see the source code folder <code>website/pipelines/skylab/slideseq</code>.</p><p>The Slide-seq pipeline can be deployed using <a href="https://cromwell.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH compliant, flexible workflow management system that supports multiple computing platforms. The workflow can also be run in <a href="https://app.terra.bio" target="_blank" rel="noopener noreferrer">Terra</a>, a cloud-based analysis platform. The <a href="https://app.terra.bio/#workspaces/warp-pipelines/Slide-seq" target="_blank" rel="noopener noreferrer">Slide-seq public workspace</a> on Terra contains the Slide-seq workflow, workflow configuration, required reference data and other inputs, and example testing data.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="inputs"></a>Inputs<a class="hash-link" href="#inputs" title="Direct link to heading">#</a></h3><p>The Slide-seq workflow inputs are specified in JSON configuration files. Example configuration files can be found in the <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/skylab/slideseq/test_inputs" target="_blank" rel="noopener noreferrer">test_inputs</a> folder in the WARP repository.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="input-descriptions"></a>Input descriptions<a class="hash-link" href="#input-descriptions" title="Direct link to heading">#</a></h4><table><thead><tr><th>Parameter name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>r1_fastq</td><td>Array of Read 1 FASTQ files split across multiple sequencing lanes; forward reads containing bead barcodes and Unique Molecular Identifiers (UMIs); order of files must match the <code>r2_fastq</code> array.</td><td>Array[File]</td></tr><tr><td>r2_fastq</td><td>Array of Read 2 FASTQ files split across multiple sequencing lanes; reverse reads containing the cDNA fragment generated from captured mRNA; order of files must match the <code>r1_fastq</code> array.</td><td>Array[File]</td></tr><tr><td>i1_fastq</td><td>Optional array of i1 (index) FASTQ files; index reads used for demultiplexing of multiple samples on one flow cell.</td><td>Array[File]</td></tr><tr><td>input_id</td><td>Unique identifier describing the biological sample or replicate that corresponds with the FASTQ files; inserted into read group header and used to name output files.</td><td>String</td></tr><tr><td>read_structure</td><td>Description of the UMI (M) and Barcode (C) positions in the Read 1 FASTQ; used to trim spacer sequences (X) for use by STARsolo; ex. &quot;8C18X6C9M1X&quot;.</td><td>String</td></tr><tr><td>tar_star_reference</td><td>TAR file containing a species-specific reference genome and GTF; generated using the <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/skylab/build_indices/BuildIndices.wdl" target="_blank" rel="noopener noreferrer">BuildIndices workflow</a>.</td><td>File</td></tr><tr><td>annotations_gtf</td><td>GTF containing gene annotations used for gene tagging (must match GTF in STAR reference).</td><td>File</td></tr><tr><td>output_bam_basename</td><td>Optional string used for the output BAM file basename.</td><td>String</td></tr><tr><td>count_exons</td><td>Optional boolean indicating if the workflow should calculate exon counts; default is set to ‚Äútrue‚Äù and produces a Loom file containing both whole-gene counts and exon counts in an additional layer; when set to ‚Äúfalse‚Äù, a Loom file containing only whole-gene counts is produced.</td><td>Boolean</td></tr><tr><td>bead_locations</td><td>Whitelist TSV file containing bead barcodes and XY coordinates on a single line for each bead; determined by sequencing prior to mRNA transfer and library preparation.</td><td>File</td></tr></tbody></table><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="pseudogene-handling"></a>Pseudogene handling<a class="hash-link" href="#pseudogene-handling" title="Direct link to heading">#</a></h4><p>The example Slide-seq reference files are created using the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/build_indices/BuildIndices.wdl" target="_blank" rel="noopener noreferrer">BuildIndices pipeline</a>. The BuildIndices pipeline modifies a given GTF downloaded from GENCODE to only include biotypes that are listed in a tab separated file (<a href="https://github.com/broadinstitute/warp/blob/master/dockers/skylab/snss2-build-indices/Biotypes.tsv" target="_blank" rel="noopener noreferrer">biotypes.tsv</a>). The example references do not include the pseudogene biotype. Learn more about Ensembl biotypes in the <a href="https://m.ensembl.org/info/genome/genebuild/biotypes.html" target="_blank" rel="noopener noreferrer">Ensembl overview</a>.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="sample-inputs-for-analyses-in-a-terra-workspace"></a>Sample inputs for analyses in a Terra Workspace<a class="hash-link" href="#sample-inputs-for-analyses-in-a-terra-workspace" title="Direct link to heading">#</a></h4><p>The Slide-seq pipeline is currently available on the cloud-based platform Terra. After registering, you can access the <a href="https://app.terra.bio/#workspaces/warp-pipelines/Slide-seq" target="_blank" rel="noopener noreferrer">Slide-seq public workspace</a> which is preloaded with instructions and sample data. Please view the <a href="https://support.terra.bio/hc/en-us" target="_blank" rel="noopener noreferrer">Support Center</a> for more information on using the Terra platform.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="slide-seq-tasks-and-tools"></a>Slide-seq tasks and tools<a class="hash-link" href="#slide-seq-tasks-and-tools" title="Direct link to heading">#</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/slideseq/SlideSeq.wdl" target="_blank" rel="noopener noreferrer">Slide-seq workflow</a> imports individual &quot;tasks,&quot; also written in WDL script, from the WARP <a href="https://github.com/broadinstitute/warp/tree/develop/tasks/skylab" target="_blank" rel="noopener noreferrer">tasks folder</a>. </p><p>Overall, the Slide-seq workflow:</p><ol><li>Calculates prealignment metrics.</li><li>Uses sctools to filter, trim, and split reads into &lt; 30 GB FASTQs.</li><li>Uses STARsolo to correct bead barcodes, align reads, and count genes.</li><li>Calculates metrics.</li><li>Merges the STAR outputs into NPY and NPZ arrays.</li><li>Merges gene counts and metrics into a Loom-formatted matrix.</li></ol><p>The tools each Slide-seq task employs are detailed in the table below. </p><p>To see specific tool parameters, select the task WDL link in the table; then find the task and view the <code>command {}</code> section of the task in the WDL script. To view or use the exact tool software, see the task&#x27;s Docker image which is specified in the task WDL <code># runtime values</code> section as <code>String docker =</code>.</p><table><thead><tr><th>Task name and WDL link</th><th>Tool</th><th>Software</th><th>Description</th></tr></thead><tbody><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/Metrics.wdl" target="_blank" rel="noopener noreferrer">Metrics.FastqMetricsSlideSeq (alias = FastqMetrics)</a></td><td>fastq_metrics</td><td><a href="https://github.com/broadinstitute/warp-tools" target="_blank" rel="noopener noreferrer">warp-tools</a></td><td>Calculates prealignment metrics from the input FASTQ files. Outputs metrics files including the bead barcode distribution, UMI distribution, number of reads per cell and number of UMIs per cell.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/FastqProcessing.wdl" target="_blank" rel="noopener noreferrer">FastqProcessing.FastqProcessingSlidSeq (alias = SplitFastq)</a></td><td>fastq_slideseq</td><td><a href="https://github.com/broadinstitute/warp-tools" target="_blank" rel="noopener noreferrer">warp-tools</a></td><td>Filters reads, trims bead barcodes in read 1 to remove spacer sequences, and partitions the input FASTQ files by bead barcode to create an array of FASTQ files that are each ~ 30 GB. The function keeps all reads belonging to the same bead barcode in the same partitioned FASTQ file.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/StarAlign.wdl" target="_blank" rel="noopener noreferrer">StarAlign.STARsoloFastqSlideSeq (alias = STARsoloFastqSlideSeq)</a></td><td>STARsolo</td><td><a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">STAR</a></td><td>For each of the partitioned FASTQ files, performs bead barcode correction, poly(A) tail trimming from adapters, alignment, gene annotation, and gene counting. Produces a count matrix and BAM file for each partition.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/MergeSortBam.wdl" target="_blank" rel="noopener noreferrer">Merge.MergeSortBamFiles (alias= MergeBam)</a></td><td>MergeSamFiles</td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a></td><td>Merges the array of BAM files into a single BAM and sorts in coordinate order.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/Metrics.wdl" target="_blank" rel="noopener noreferrer">Metrics.CalculateGeneMetrics (alias = GeneMetrics)</a></td><td>TagSort</td><td><a href="https://github.com/broadinstitute/warp-tools" target="_blank" rel="noopener noreferrer">warp-tools</a></td><td>Sorts the BAM file by gene using the bead barcode (CB), molecule barcode (UB), and gene ID (GX) tags and computes gene metrics.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/Metrics.wdl" target="_blank" rel="noopener noreferrer">Metrics.CalculateUMIsMetrics (alias = UMIsMetrics)</a></td><td>TagSort</td><td><a href="https://github.com/broadinstitute/warp-tools" target="_blank" rel="noopener noreferrer">warp-tools</a></td><td>Sorts the BAM file by gene using the bead barcode (CB), molecule barcode (UB), and gene ID (GX) tags and computes gene metrics.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/Metrics.wdl" target="_blank" rel="noopener noreferrer">Metrics.CalculateCellMetrics (alias = CellMetrics)</a></td><td>TagSort</td><td><a href="https://github.com/broadinstitute/warp-tools" target="_blank" rel="noopener noreferrer">warp-tools</a></td><td>Sorts the BAM file by bead barcode (CB), molecule barcode (UB), and gene ID (GX) tags and computes bead barcode metrics.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/StarAlign.wdl" target="_blank" rel="noopener noreferrer">StarAlign.MergeStarOutput (alias = MergeStarOutputsExons)</a></td><td>create-npz-output.py</td><td><a href="https://www.python.org/" target="_blank" rel="noopener noreferrer">Python 3</a></td><td>Creates a compressed raw NPY or NPZ file containing the STARsolo output features (NPY), barcodes (NPZ) and counts (NPZ). By default, <code>count_exons</code> is true and exon counts are included in output files. When <code>count_exons</code> is false, exon counts are excluded.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/LoomUtils.wdl" target="_blank" rel="noopener noreferrer">LoomUtils.SingleNucleusOptimusLoomOutput (alias = SlideseqLoomGenerationWithExons)</a></td><td>create_loom_slide_seq.py</td><td><a href="https://www.python.org/" target="_blank" rel="noopener noreferrer">Python 3</a></td><td>Merges the gene counts, bead barcode metrics, and gene metrics data into a Loom formatted bead-by-gene matrix. By default, the Loom file contains whole-gene counts with exon counts in an additional layer. When <code>count_exons</code> is false, the task is run as <code>SlideseqLoomGeneration</code> and exon counts are excluded.</td></tr></tbody></table><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="1-calculating-prealignment-metrics"></a>1. Calculating prealignment metrics<a class="hash-link" href="#1-calculating-prealignment-metrics" title="Direct link to heading">#</a></h4><p>The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/FastqProcessing.wdl" target="_blank" rel="noopener noreferrer">FastqMetricsSlideSeq</a> task calculates prealignment metrics used for assessing data quality from the input FASTQ files. These metrics include the bead barcode distribution,  UMI distribution, number of reads per cell and number of UMIs per cell. These metrics are included in the final outputs of the workflow. </p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="2-filtering-reads-trimming-barcodes-and-splitting-fastqs-with-sctools"></a>2. Filtering reads, trimming barcodes, and splitting FASTQs with sctools<a class="hash-link" href="#2-filtering-reads-trimming-barcodes-and-splitting-fastqs-with-sctools" title="Direct link to heading">#</a></h4><p><strong>Read filtering</strong></p><p>Although the function of the bead barcodes is to identify the spatial location of gene expression in the tissue, barcode errors can arise during sequencing that make it difficult to assign reads to the proper location. </p><p>The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/FastqProcessing.wdl" target="_blank" rel="noopener noreferrer">FastqProcessingSlidSeq</a> task uses <a href="https://github.com/broadinstitute/warp-tools" target="_blank" rel="noopener noreferrer">warp-tools</a> to evaluate barcode errors by comparing the Read 1 FASTQ sequences against a whitelist of barcodes created by sequencing prior to the mRNA transfer and library preparation steps of the <a href="https://www.science.org/doi/10.1126/science.aaw1219" target="_blank" rel="noopener noreferrer">Slide-Seq</a> protocol.</p><p>Barcodes that are more than one edit distance (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5410656/" target="_blank" rel="noopener noreferrer">Hamming distance</a>) from any of the barcode sequences in the whitelist are filtered at this step and not included in the output. The remaining barcodes are corrected by the <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/StarAlign.wdl" target="_blank" rel="noopener noreferrer">STARsoloFastqSlideSeq</a> task.</p><p><strong>Barcode trimming</strong></p><p>The task uses sctools to trim spacer sequences from bead barcodes and UMIs for use by STARsolo, which requires continuous sample barcodes without spacer sequences between them. The input <code>read_structure</code> is used to parse the barcodes and remove any bases with tags other than C or M, which represent the bead barcode and UMI, respectively. For example, with a <code>read_structure</code> of 8C18X6C9M1X, bases represented by 18X and 1X are removed from the reads and the string of bases is rewritten with the structure 14C9M. Bases represented by tags other than X will also be removed during this step, so long as they are not C or M.</p><p><strong>FASTQ splitting</strong></p><p>For faster processing of large datasets, the task divides the input array of FASTQ files into multiple FASTQs, each containing reads from the same bead barcodes such that the FASTQ files are approximately 30 GB in size. </p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="3-correcting-bead-barcodes-trimming-and-aligning-reads-and-counting-genes-with-starsolo"></a>3. Correcting bead barcodes, trimming and aligning reads, and counting genes with STARsolo<a class="hash-link" href="#3-correcting-bead-barcodes-trimming-and-aligning-reads-and-counting-genes-with-starsolo" title="Direct link to heading">#</a></h4><p><strong>Bead barcode correction</strong></p><p>The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/StarAlign.wdl" target="_blank" rel="noopener noreferrer">STARsoloFastqSlideSeq</a> task uses the STAR aligner to evaluate barcode errors by comparing the Read 1 FASTQ sequences against a whitelist of barcodes created by sequencing prior to the mRNA transfer and library preparation steps of the <a href="https://www.science.org/doi/10.1126/science.aaw1219" target="_blank" rel="noopener noreferrer">Slide-seq</a> protocol.</p><p>Corrected barcodes are those that fall within one edit distance (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5410656/" target="_blank" rel="noopener noreferrer">Hamming distance</a>) of one of the barcode sequences in the whitelist. Note that prior to running STARsolo, uncorrectable barcodes (with more than one mismatch) are filtered out by the <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/FastqProcessing.wdl" target="_blank" rel="noopener noreferrer">FastqProcessingSlidSeq</a> task.</p><p><strong>Read trimming</strong></p><p>Poly(A) tails are trimmed from reads using the STARsolo parameter <code>--clip3pAdapterSeq AAAAAA</code>.</p><p><strong>Alignment</strong></p><p>STAR maps barcoded reads to the genome primary assembly reference (see the <a href="https://broadinstitute.github.io/warp/docs/Pipelines/Slide-seq_Pipeline/README#quickstart-table" target="_blank" rel="noopener noreferrer">Quickstart table</a> above for version information). The example references for the Slide-seq workflow were generated using the <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/skylab/build_indices/BuildIndices.wdl" target="_blank" rel="noopener noreferrer">BuildIndices pipeline</a>.</p><p><strong>Gene annotation and counting</strong></p><p>Prior to gene counting, STARsolo adds gene annotations which will vary depending on the value of <code>count_exons</code>. By default, <code>count_exons</code> is true and STARsolo will run with the parameter <code>--soloFeatures Gene GeneFull</code> to produce both whole-gene and exon counts. If <code>count_exons</code> is false, STARsolo will run with the parameter <code>--soloFeatures GeneFull</code> to produce whole-gene counts.</p><p>The resulting BAM files are merged together into a single BAM using the <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/MergeSortBam.wdl" target="_blank" rel="noopener noreferrer">MergeSortBamFiles</a> task.</p><p><strong>STARsolo outputs</strong></p><p>The task‚Äôs output includes a coordinate-sorted BAM file containing the bead barcode-corrected reads and SAM attributes UB UR UY CR CB CY NH GX GN. Additionally, after counting, the task outputs three intermediate TSV files (features, barcodes, and matrix) used for downstream Loom matrix generation.   </p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="4-calculating-metrics"></a>4. Calculating metrics<a class="hash-link" href="#4-calculating-metrics" title="Direct link to heading">#</a></h4><p>The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/Metrics.wdl" target="_blank" rel="noopener noreferrer">CalculateGeneMetrics</a>, <a href="https://github.com/broadinstitute/warp/blob/develop/master/skylab/Metrics.wdl" target="_blank" rel="noopener noreferrer">CalculateUMIsMetrics</a>, and <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/Metrics.wdl" target="_blank" rel="noopener noreferrer">CalculateCellMetrics</a> tasks use <a href="https://github.com/broadinstitute/warp-tools" target="_blank" rel="noopener noreferrer">warp-tools</a> to calculate summary metrics that help assess the per-bead and per-UMI quality of the data output each time this pipeline is run. </p><p>These metrics output from both tasks are included in the output Loom matrix. A detailed list of these metrics is found in the <a href="/warp/docs/Pipelines/SlideSeq_Pipeline/count-matrix-overview">Slide-seq Count Matrix Overview</a>.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="5-merging-the-star-outputs-into-npy-and-npz-arrays"></a>5. Merging the STAR outputs into NPY and NPZ arrays<a class="hash-link" href="#5-merging-the-star-outputs-into-npy-and-npz-arrays" title="Direct link to heading">#</a></h4><p>The STARsolo output includes a features, barcodes, and matrix TSV for each of the partitioned FASTQ input files. The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/StarAlign.wdl" target="_blank" rel="noopener noreferrer">MergeStarOutput task</a> merges each respective TSV. It uses a custom python script to convert the merged matrix, features, and barcodes output from STARsolo into an NPY (features and barcodes)- and NPZ (the matrix)-formatted file.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="6-merging-counts-and-metrics-data-into-loom-formatted-matrix"></a>6. Merging counts and metrics data into Loom-formatted matrix<a class="hash-link" href="#6-merging-counts-and-metrics-data-into-loom-formatted-matrix" title="Direct link to heading">#</a></h4><p>The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/LoomUtils.wdl" target="_blank" rel="noopener noreferrer">SingleNucleusOptimusLoomOutput</a> task uses a custom python script to merge the converted STARsolo count matrix and the cell (bead) and gene metrics into a Loom-formatted bead-by-gene matrix. <strong>These counts are raw and unfiltered.</strong></p><p>Read full details for all the metrics in the <a href="/warp/docs/Pipelines/SlideSeq_Pipeline/count-matrix-overview">Slide-seq Count Matrix Overview</a>.</p><p><strong>Gene counts</strong></p><p>The type of gene counts in the Loom will vary depending on the value of the Slide-seq workflow input, <code>count_exons</code>. By default, <code>count_exons</code> is set to true and the output Loom will contain whole-gene counts with exon counts in an additional layer.</p><p>If the workflow is run with <code>count_exons</code> set to false, the output Loom file will contain whole-gene counts. Running the workflow in this configuration will cause the Loom matrix to have fewer columns (bead barcodes) due to the difference in STARsolo counting mode.</p><p>You can determine which type of counts are in the Loom by looking at the global attribute <code>expression_data_type</code>.</p><p>After running the pipeline with <code>count_exons</code> set to true, you can access whole-gene and exonic counts using Loompy&#x27;s <code>layers()</code> method. For example, <code>loompy.connect.layers[‚Äú‚Äù]</code> will return the whole-gene counts from the output Loom file. Similarly, <code>loompy.connect.layers[‚Äúexon_counts‚Äù]</code> will return the exonic counts from the output Loom. </p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="6-outputs"></a>6. Outputs<a class="hash-link" href="#6-outputs" title="Direct link to heading">#</a></h4><p>Output files of the pipeline include:</p><ol><li>Bead x Gene unnormalized count matrices in Loom format.</li><li>Unfiltered, sorted BAM file with barcode and downstream analysis tags.</li><li>Bead metadata, including bead metrics.</li><li>Gene metadata, including gene metrics.</li></ol><p>The following table lists the output files produced from the pipeline. For samples that have sequenced over multiple lanes, the pipeline will output one merged version of each listed file.</p><table><thead><tr><th>Output Name</th><th>Filename, if applicable</th><th>Output Type</th><th>Output Format</th></tr></thead><tbody><tr><td>pipeline_version_out</td><td>N/A</td><td>Version of the processing pipeline run on this data.</td><td>String</td></tr><tr><td>genomic_reference_version</td><td><code>reference_version.txt</code></td><td>File containing genomic reference source, build, and annotation versions.</td><td>TXT</td></tr><tr><td>bam</td><td><code>&lt;output_bam_basename&gt;.bam</code></td><td>Aligned BAM</td><td>BAM</td></tr><tr><td>matrix</td><td><code>&lt;input_id&gt;_sparse_counts.npz</code></td><td>Converted sparse matrix file from the MergeStarOutputs task.</td><td>NPZ</td></tr><tr><td>matrix_row_index</td><td><code>&lt;input_id&gt;_sparse_counts_row_index.npy</code></td><td>Index of beads in count matrix.</td><td>NPY</td></tr><tr><td>matrix_col_index</td><td><code>&lt;input_id&gt;_sparse_counts_col_index.npy</code></td><td>Index of genes in count matrix.</td><td>NPY</td></tr><tr><td>cell_metrics</td><td><code>cell-metrics.csv.gz</code></td><td>Cell metrics based on bead barcodes</td><td>Compressed CSV</td></tr><tr><td>gene_metrics</td><td><code>gene-metrics.csv.gz</code></td><td>Gene metrics</td><td>Compressed CSV</td></tr><tr><td>umi_metrics</td><td><code>umi-metrics.csv.gz</code></td><td>UMI metrics</td><td>Compressed CSV</td></tr><tr><td>fastq_barcode_distribution</td><td><code>&lt;input_id&gt;.barcode_distribution_XC.txt</code></td><td>Metric file containing the distribution of reads per bead barcode that were calculated prior to alignment.</td><td>TXT</td></tr><tr><td>fastq_umi_distribution</td><td><code>&lt;input_id&gt;.barcode_distribution_XM.txt</code></td><td>Metric file containing the distribution of reads per UMI that were calculated prior to alignment.</td><td>TXT</td></tr><tr><td>fastq_reads_per_cell</td><td><code>&lt;input_id&gt;.numReads_perCell_XC.txt</code></td><td>Metric file containing the number of reads per barcode that were calculated prior to alignment.</td><td>TXT</td></tr><tr><td>fastq_reads_per_umi</td><td><code>&lt;input_id&gt;.numReads_perCell_XM.txt</code></td><td>Metric file containing the number of reads per UMI that were calculated prior to alignment.</td><td>TXT</td></tr><tr><td>loom_output_file</td><td><code>&lt;input_id&gt;.loom</code></td><td>Loom file containing count data and metadata.</td><td>Loom</td></tr></tbody></table><p>The Loom matrix is the default output. See the <a href="https://github.com/broadinstitute/warp-tools/blob/develop/scripts/create_loom_optimus.py" target="_blank" rel="noopener noreferrer">create_loom_slide_seq.py</a> script for the detailed code. This matrix contains the unnormalized (unfiltered) count matrices, as well as the gene and bead barcode metrics detailed in the <a href="/warp/docs/Pipelines/SlideSeq_Pipeline/count-matrix-overview">Slide-seq Count Matrix Overview</a>.</p><p>The output Loom matrix can be converted to an H5AD file for downstream processing using a <a href="https://github.com/broadinstitute/warp-tools/blob/develop/scripts/loom_to_h5ad.py" target="_blank" rel="noopener noreferrer">custom script</a> available in the <a href="https://github.com/broadinstitute/warp-tools" target="_blank" rel="noopener noreferrer">warp-tools GitHub repository</a>.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="validation-against-on-prem-pipeline"></a>Validation against on-prem pipeline<a class="hash-link" href="#validation-against-on-prem-pipeline" title="Direct link to heading">#</a></h2><p>Slide-seq has been validated for processing mouse spatial transcriptomic data generated with the Slide-seq assay. Slide-seq results are compared to those of an on-premise pipeline using BCL input files and python scripts. See the <a href="https://docs.google.com/document/d/1cmfG3Y4uGvfJUJZV9MZvgN5Yw6-bFajpS7lonMKxdso/edit?usp=sharing" target="_blank" rel="noopener noreferrer">validation report</a> for more information.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="versioning-and-testing"></a>Versioning and testing<a class="hash-link" href="#versioning-and-testing" title="Direct link to heading">#</a></h2><p>All Slide-seq pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/slideseq/SlideSeq.changelog.md" target="_blank" rel="noopener noreferrer">Slide-seq changelog</a> and tested using <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/slideseq/test_inputs/test_data_overview.md" target="_blank" rel="noopener noreferrer">plumbing and scientific test data</a>. To learn more about WARP pipeline testing, see <a href="https://broadinstitute.github.io/warp/docs/About_WARP/TestingPipelines" target="_blank" rel="noopener noreferrer">Testing Pipelines</a>.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="citing-the-slide-seq-pipeline"></a>Citing the Slide-seq Pipeline<a class="hash-link" href="#citing-the-slide-seq-pipeline" title="Direct link to heading">#</a></h2><p>Please identify the pipeline in your methods section using the Slide-seq Pipeline&#x27;s <a href="https://scicrunch.org/resources/data/record/nlx_144509-1/SCR_023379/resolver?q=%22Slide-seq%22&amp;l=%22Slide-seq%22&amp;i=rrid:scr_023379" target="_blank" rel="noopener noreferrer">SciCrunch resource identifier</a>.</p><ul><li>Ex: <em>Slide-seq Pipeline (RRID:SCR_023379)</em></li></ul><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="consortia-support"></a>Consortia support<a class="hash-link" href="#consortia-support" title="Direct link to heading">#</a></h2><p>This pipeline is supported by the <a href="https://biccn.org/" target="_blank" rel="noopener noreferrer">BRAIN Initiative Cell Census Network</a> (BICCN) and BRAIN Initiative Cell Atlas Network (BICAN). </p><p>If your organization also uses this pipeline, we would like to list you! Please reach out to us by contacting the <a href="mailto:warp-pipelines-help@broadinstitute.org" target="_blank" rel="noopener noreferrer">WARP Pipelines team</a>.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="feedback"></a>Feedback<a class="hash-link" href="#feedback" title="Direct link to heading">#</a></h2><p>Please help us make our tools better by contacting the <a href="mailto:warp-pipelines-help@broadinstitute.org" target="_blank" rel="noopener noreferrer">WARP Pipelines Team</a> for pipeline-related suggestions or questions.</p></div><footer class="docusaurus-mt-lg"><div class="row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/SlideSeq_Pipeline/README.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_2_ui" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_13-_">Last updated on <b><time datetime="2023-05-25T20:13:47.000Z">5/25/2023</time></b> by <b>Kevin Palis</b></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/warp/docs/Pipelines/Single_Cell_ATAC_Seq_Pipeline/scatac.methods"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">¬´ scATAC v1.2.0 Methods</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/warp/docs/Pipelines/SlideSeq_Pipeline/count-matrix-overview"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Slide-seq Count Matrix Overview ¬ª</div></a></div></nav></div></div><div class="col col--3"><div class="tableOfContents_35-E thin-scrollbar"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-slide-seq-workflow" class="table-of-contents__link">Introduction to the Slide-seq workflow</a></li><li><a href="#quickstart-table" class="table-of-contents__link">Quickstart table</a></li><li><a href="#set-up" class="table-of-contents__link">Set-up</a><ul><li><a href="#slide-seq-installation" class="table-of-contents__link">Slide-seq installation</a></li><li><a href="#inputs" class="table-of-contents__link">Inputs</a></li></ul></li><li><a href="#slide-seq-tasks-and-tools" class="table-of-contents__link">Slide-seq tasks and tools</a></li><li><a href="#validation-against-on-prem-pipeline" class="table-of-contents__link">Validation against on-prem pipeline</a></li><li><a href="#versioning-and-testing" class="table-of-contents__link">Versioning and testing</a></li><li><a href="#citing-the-slide-seq-pipeline" class="table-of-contents__link">Citing the Slide-seq Pipeline</a></li><li><a href="#consortia-support" class="table-of-contents__link">Consortia support</a></li><li><a href="#feedback" class="table-of-contents__link">Feedback</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item"><span>GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_3J9K"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></span></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright ¬© 2023 Copyright ¬© Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.702083b4.js"></script>
<script src="/warp/assets/js/main.2b70a255.js"></script>
</body>
</html>