<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.5">
<link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Blog Atom Feed">
<link rel="preconnect" href="https://www.google-analytics.com">
<script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-181048107-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview")</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><title data-react-helmet="true">Optimus Overview | WARP</title><meta data-react-helmet="true" name="twitter:card" content="summary_large_image"><meta data-react-helmet="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/README"><meta data-react-helmet="true" name="docsearch:language" content="en"><meta data-react-helmet="true" name="docsearch:version" content="current"><meta data-react-helmet="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="Optimus Overview | WARP"><meta data-react-helmet="true" name="description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><meta data-react-helmet="true" property="og:description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><link data-react-helmet="true" rel="shortcut icon" href="/warp/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/README"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/README" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/README" hreflang="x-default"><link data-react-helmet="true" rel="preconnect" href="https://BH4D9OD16A-dsn.algolia.net" crossorigin="anonymous"><link rel="stylesheet" href="/warp/assets/css/styles.57c5b218.css">
<link rel="preload" href="/warp/assets/js/runtime~main.71584178.js" as="script">
<link rel="preload" href="/warp/assets/js/main.1d35face.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#" class="skipToContent_1oUP">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle toggle_3Zt9 react-toggle--disabled"><div class="react-toggle-track" role="button" tabindex="-1"><div class="react-toggle-track-check"><span class="toggle_71bT">üåú</span></div><div class="react-toggle-track-x"><span class="toggle_71bT">üåû</span></div><div class="react-toggle-thumb"></div></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div><div class="searchBox_1Doo"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docs-wrapper docs-doc-page"><div class="docPage_31aa"><button class="clean-btn backToTopButton_35hR" type="button"><svg viewBox="0 0 24 24" width="28"><path d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z" fill="currentColor"></path></svg></button><aside class="docSidebarContainer_3Kbt"><div class="sidebar_15mo"><nav class="menu thin-scrollbar menu_Bmed menuWithAnnouncementBar_2WvA"><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">About WARP</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#">WARP Pipelines</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">CEMBA</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Exome Germline Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">GDC Whole Genome Somatic Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Illumina Genotyping Array</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Imputation</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#" tabindex="0">Optimus</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/optimus.methods">Optimus v5.3.0 Methods</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/Bam_tags">What tags are included in an Optimus BAM file?</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing">Consortia Data Processing</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">RNA with UMIs</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Single Cell ATAC</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Multi-Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Single Nucleus Multi-Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Ultima Genomics Whole Genome Germline</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Whole Genome Germline Single Sample</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">WDL Best Practices</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">Contribution Guide</a></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_1CGd"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_3E-R"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></aside><main class="docMainContainer_3ufF"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_3FnS"><div class="docItemContainer_33ec"><article><div class="tocCollapsible_1PrD tocMobile_3Hoh"><button type="button" class="clean-btn tocCollapsibleButton_2O1e">On this page</button></div><div class="markdown"><header><h1>Optimus Overview</h1></header><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases?q=optimus&amp;expanded=true" target="_blank" rel="noopener noreferrer">optimus_v5.6.0</a></td><td align="center">November, 2022</td><td align="center"><a href="mailto:ekiernan@broadinstitute.org" target="_blank" rel="noopener noreferrer">Elizabeth Kiernan</a></td><td align="center">Please file GitHub issues in warp or contact <a href="mailto:warp-pipelines-help@broadinstitute.org" target="_blank" rel="noopener noreferrer">the WARP team</a></td></tr></tbody></table><p><img alt="Optimus_diagram" src="/warp/assets/images/Optimus_diagram-8b47ebeeb96d96cfbf40b37cc193a9e5.png"></p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="introduction-to-the-optimus-workflow"></a>Introduction to the Optimus workflow<a class="hash-link" href="#introduction-to-the-optimus-workflow" title="Direct link to heading">#</a></h2><p>Optimus is an open-source, cloud-optimized pipeline developed by the Data Coordination Platform (DCP) of the <a href="https://data.humancellatlas.org/" target="_blank" rel="noopener noreferrer">Human Cell Atlas (HCA) Project</a> as well as the <a href="https://biccn.org/" target="_blank" rel="noopener noreferrer">BRAIN Initiative Cell Census Network</a> (BICCN). It supports the processing of any 3&#x27; single-cell and single-nucleus count data generated with the <a href="https://www.10xgenomics.com/solutions/single-cell/" target="_blank" rel="noopener noreferrer">10x Genomics v2 or v3 assay</a>.</p><p>It is an alignment and transcriptome quantification pipeline that corrects cell barcodes (CBs), aligns reads to the genome, corrects Unique Molecular Identifiers (UMIs), generates a count matrix in a UMI-aware manner, calculates summary metrics for genes and cells, detects empty droplets, returns read outputs in BAM format, and returns cell gene counts in numpy matrix and Loom file formats.</p><p>In addition to providing commonly used metrics such as empty drop detection and mitochondrial reads, Optimus takes special care to <strong>keep all reads in the output BAM that may be useful to the downstream user</strong>, such as unaligned reads or reads with uncorrectable barcodes. This design provides flexibility to the downstream user and allows for alternative filtering or leveraging the data for novel methodological development. </p><p>Optimus has been validated for analyzing both human and mouse single-cell or single-nucleus datasets. Learn more in the <a href="#validation-against-cell-ranger">validation section</a>.</p><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Want to use the Optimus pipeline for your publication?</h5></div><div class="admonition-content"><p>Check out the <a href="/warp/docs/Pipelines/Optimus_Pipeline/optimus.methods">Optimus Publication Methods</a> to get started!</p></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="quickstart-table"></a>Quickstart table<a class="hash-link" href="#quickstart-table" title="Direct link to heading">#</a></h2><p>The following table provides a quick glance at the Optimus pipeline features:</p><table><thead><tr><th>Pipeline features</th><th>Description</th><th>Source</th></tr></thead><tbody><tr><td>Assay type</td><td>10x single cell or single nucleus expression (v2 and v3)</td><td><a href="https://www.10xgenomics.com" target="_blank" rel="noopener noreferrer">10x Genomics</a></td></tr><tr><td>Overall workflow</td><td>Quality control module and transcriptome quantification module</td><td>Code available from <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/Optimus.wdl" target="_blank" rel="noopener noreferrer">GitHub</a></td></tr><tr><td>Workflow language</td><td>WDL 1.0</td><td><a href="https://github.com/openwdl/wdl" target="_blank" rel="noopener noreferrer">openWDL</a></td></tr><tr><td>Genomic Reference Sequence</td><td>GRCh38 human genome primary sequence and M21 (GRCm38.p6) mouse genome primary sequence</td><td>GENCODE <a href="https://www.gencodegenes.org/human/release_27.html" target="_blank" rel="noopener noreferrer">human reference files</a> and <a href="https://www.gencodegenes.org/mouse/release_M21.html" target="_blank" rel="noopener noreferrer">mouse reference files</a></td></tr><tr><td>Transcriptomic reference annotation</td><td>V27 GENCODE human transcriptome and M21 mouse transcriptome</td><td>GENCODE <a href="ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_27/gencode.v27.annotation.gtf.gz" target="_blank" rel="noopener noreferrer">human GTF</a> and <a href="ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M21/gencode.vM21.annotation.gff3.gz" target="_blank" rel="noopener noreferrer">mouse GTF</a></td></tr><tr><td>Aligner and transcript quantification</td><td>STARsolo</td><td><a href="https://www.biorxiv.org/content/10.1101/2021.05.05.442755v1" target="_blank" rel="noopener noreferrer">Dobin, et al.,2021</a></td></tr><tr><td>Data input file format</td><td>File format in which sequencing data is provided</td><td><a href="https://academic.oup.com/nar/article/38/6/1767/3112533" target="_blank" rel="noopener noreferrer">FASTQ</a></td></tr><tr><td>Data output file format</td><td>File formats in which Optimus output is provided</td><td><a href="http://samtools.github.io/hts-specs/" target="_blank" rel="noopener noreferrer">BAM</a>, Python numpy arrays (internal), Loom (generated with <a href="http://loompy.org/" target="_blank" rel="noopener noreferrer">Loompy v.3.0.6)</a></td></tr></tbody></table><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="set-up"></a>Set-up<a class="hash-link" href="#set-up" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="optimus-installation"></a>Optimus installation<a class="hash-link" href="#optimus-installation" title="Direct link to heading">#</a></h3><p>To download the latest Optimus release, see the release tags prefixed with &quot;Optimus&quot; on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a>. All Optimus pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/skylab/optimus/Optimus.changelog.md" target="_blank" rel="noopener noreferrer">Optimus changelog</a>. </p><p>To discover and search releases, use the WARP command-line tool <a href="https://github.com/broadinstitute/warp/tree/master/wreleaser" target="_blank" rel="noopener noreferrer">Wreleaser</a>.</p><p>If you‚Äôre running an Optimus workflow version prior to the latest release, the accompanying documentation for that release may be downloaded with the source code on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a> (see the source code folder ‚Äúwebsite/pipelines/skylab/optimus).</p><p>Optimus can be deployed using <a href="https://cromwell.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH compliant, flexible workflow management system that supports multiple computing platforms. The workflow can also be run in <a href="https://app.terra.bio" target="_blank" rel="noopener noreferrer">Terra</a>, a cloud-based analysis platform. The Terra <a href="https://app.terra.bio/#workspaces/featured-workspaces-hca/HCA_Optimus_Pipeline" target="_blank" rel="noopener noreferrer">Optimus Featured Workspace</a> contains the Optimus workflow, workflow configurations, required reference data and other inputs, and example testing data.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="inputs"></a>Inputs<a class="hash-link" href="#inputs" title="Direct link to heading">#</a></h3><p>Optimus pipeline inputs are detailed in JSON format configuration files. There are five downsampled example configuration files available for running the pipeline:</p><ul><li><a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/example_inputs/human_v2_example.json" target="_blank" rel="noopener noreferrer">human_v2_example</a>: An example human 10x v2 single-cell dataset</li><li><a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/example_inputs/human_v3_example.json" target="_blank" rel="noopener noreferrer">human_v3_example</a>: An example human 10x v3 single-cell dataset</li><li><a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/example_inputs/mouse_v2_example.json" target="_blank" rel="noopener noreferrer">mouse_v2_example</a>: An example mouse 10x v2 single-cell dataset</li><li><a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/example_inputs/mouse_v2_snRNA_example.json" target="_blank" rel="noopener noreferrer">mouse_v2_snRNA_example</a>: An example mouse v2 single-nucleus dataset</li></ul><p>Additionally, there are multiple full-size example datasets available in the <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/skylab/optimus/example_inputs/test_optimus_full_datasets" target="_blank" rel="noopener noreferrer">test_optimus_full_datasets</a> folder. Unlike the example configuration files above, the full-size configuration files may not reflect updated Optimus parameters. However, you can still access the FASTQ files for each dataset at the Google bucket locations listed in the dataset configuration files.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="sample-data-input"></a>Sample data input<a class="hash-link" href="#sample-data-input" title="Direct link to heading">#</a></h4><p>Each 10x v2 and v3 3‚Äô sequencing experiment generates triplets of FASTQ files for any given sample. Optimus takes the FASTQs listed below as input.</p><ol><li>Forward reads (<code>r1_fastq</code>) containing the unique molecular identifier (UMI) and CB sequences</li><li>Reverse reads (<code>r2_fastq</code>) containing the alignable genomic information from the mRNA transcript</li><li>Optional index FASTQ (<code>i1_fastq</code>) containing the sample barcodes, when provided by the sequencing facility</li></ol><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Optimus is currently a single sample pipeline</h5></div><div class="admonition-content"><p>However, it can take in multiple sets of FASTQs for a sample that has been split over multiple lanes of sequencing. For an example configuration file with multiple lanes, please see the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/example_inputs/mouse_v2_example.json" target="_blank" rel="noopener noreferrer">mouse_v2_example.json</a>. Additionally, Optimus does not support demultiplexing even though it accepts index FASTQ files.</p></div></div><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="additional-reference-inputs"></a>Additional reference inputs<a class="hash-link" href="#additional-reference-inputs" title="Direct link to heading">#</a></h4><p>The example configuration files also contain metadata for the reference files, described in the table below.</p><table><thead><tr><th>Parameter name</th><th>Description</th><th>Optional strings (when applicable)</th></tr></thead><tbody><tr><td>whitelist</td><td>List of known CBs from <a href="https://www.10xgenomics.com/" target="_blank" rel="noopener noreferrer">10x Genomics</a> that corresponds to the v2 or v3 chemistry.</td><td>N/A</td></tr><tr><td>tar_star_reference</td><td>TAR file containing a species-specific reference genome and GTF; it is generated using the <a href="https://github.com/broadinstitute/warp/tree/develop/pipelines/skylab/build_indices/BuildIndices.wdl" target="_blank" rel="noopener noreferrer">BuildIndices workflow</a>.</td><td>N/A</td></tr><tr><td>input_id</td><td>Unique identifier describing the biological sample or replicate that corresponds with the FASTQ files; can be a human-readable name or UUID.</td><td>N/A</td></tr><tr><td>input_name</td><td>Optional string that can be used to further identify the original biological sample.</td><td>N/A</td></tr><tr><td>input_id_metadata_field</td><td>Optional string describing, when applicable, the metadata field containing the input_id.</td><td>N/A</td></tr><tr><td>input_name_metadata_field</td><td>Optional string describing, when applicable, the metadata field containing the input_name.</td><td>N/A</td></tr><tr><td>annotations_gtf</td><td>GTF containing gene annotations used for gene tagging (must match GTF in STAR reference).</td><td>N/A</td></tr><tr><td>chemistry</td><td>Optional string describing whether data was generated with 10x v2 or v3 chemistry. Optimus validates this string. If the string does not match one of the optional strings, the pipeline will fail. You can remove the checks by setting &quot;force_no_check = true&quot; in the input JSON.</td><td>&quot;tenX_v2&quot; (default) or &quot;tenX_v3&quot;</td></tr><tr><td>mt_genes</td><td>Optional file containing mitochondrial gene names for a specific species. This is used for calculating gene metrics.</td><td>N/A</td></tr><tr><td>counting_mode</td><td>String describing whether data is single-cell or single-nucleus. Single-cell mode counts reads aligned to the gene transcript, whereas single-nucleus counts whole transcript to account for nuclear pre-mRNA.</td><td>&quot;sc_rna&quot; or &quot;sn_rna&quot;</td></tr><tr><td>output_bam_basename</td><td>String used as a basename for output BAM file; the default is set to the string used for the <code>input_id</code> parameter.</td><td>N/A</td></tr><tr><td>use_strand_info</td><td>Optional string for reading stranded data. Default is &quot;false&quot;; set to &quot;true&quot; to count reads in stranded mode.</td><td>&quot;true&quot; or &quot;false&quot; (default)</td></tr><tr><td>emptydrops_lower</td><td>UMI threshold for emptyDrops detection; default is 100.</td><td>N/A</td></tr><tr><td>count_exons</td><td>Boolean indicating if the workflow should calculate exon counts <strong>when in single-nucleus (sn_rna) mode</strong>. If true, this option will output an additional layer for the Loom file. By default, it is set to &quot;false&quot;. If the parameter is true and used with sc_rnamode, the workflow will return an error.</td><td>&quot;true&quot; or &quot;false&quot; (default)</td></tr></tbody></table><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="pseudogene-handling"></a>Pseudogene handling<a class="hash-link" href="#pseudogene-handling" title="Direct link to heading">#</a></h4><p>The example Optimus reference files are downloaded directly from GENCODE (see Quickstart table) and are not modified to remove pseudogenes. This is in contrast to the <a href="https://support.10xgenomics.com/single-cell-multiome-atac-gex/software/release-notes/references#header" target="_blank" rel="noopener noreferrer">references created for Cell Ranger</a> which remove pseudogenes and small RNAs.</p><p>In the case of multi-mapped pseudogenes, Optimus and Cell Ranger will produce different results. Optimus does not count multi-mapped reads in the final count matrix, whereas Cell Ranger will keep potential multi-mapped reads because it does not identify the pseudogene reads.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="sample-inputs-for-analyses-in-a-terra-workspace"></a>Sample inputs for analyses in a Terra Workspace<a class="hash-link" href="#sample-inputs-for-analyses-in-a-terra-workspace" title="Direct link to heading">#</a></h4><p>The Optimus pipeline is currently available on the cloud-based platform Terra. After registering, you can access the <a href="https://app.terra.bio/#workspaces/featured-workspaces-hca/HCA_Optimus_Pipeline" target="_blank" rel="noopener noreferrer">Optimus featured workspace</a>. The workspace is preloaded with instructions and sample data. Please view the <a href="https://support.terra.bio/hc/en-us" target="_blank" rel="noopener noreferrer">Support Center</a> for more information on using the Terra platform.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="optimus-tasks-and-tools"></a>Optimus tasks and tools<a class="hash-link" href="#optimus-tasks-and-tools" title="Direct link to heading">#</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/Optimus.wdl" target="_blank" rel="noopener noreferrer">Optimus workflow</a> imports individual &quot;tasks,&quot; also written in  WDL script, from the WARP <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab" target="_blank" rel="noopener noreferrer">tasks folder</a>. </p><p>Overall, the Optimus workflow:</p><ol><li>Partitions FASTQs by CB.</li><li>Corrects CBs, aligns reads, corrects UMIs, and counts genes with STAR.</li><li>Merges the Star outputs into NPY and NPZ arrays.</li><li>Calculates gene metrics.</li><li>Calculates cell metrics.</li><li>Runs emptyDrops.</li><li>Merges gene counts, metrics, and emptyDrops data into a Loom-formatted matrix.</li></ol><p>The tools each Optimus task employs are detailed in the table below. </p><p>To see specific tool parameters, select the task WDL link in the table; then view the <code>command {}</code> section of the task in the WDL script. To view or use the exact tool software, see the task&#x27;s Docker image which is specified in the task WDL <code># runtime values</code> section as <code>String docker =</code>.</p><table><thead><tr><th>Task name and WDL link</th><th>Tool</th><th>Software</th><th>Description</th></tr></thead><tbody><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/FastqProcessing.wdl" target="_blank" rel="noopener noreferrer">FastqProcessing.FastqProcessing (alias = SplitFastq)</a></td><td>fastqprocess</td><td>sctools</td><td>Partitions the input FASTQ files by CB to create an array of FASTQ files that are each ~ 30 GB. This task is skipped if the input files are smaller than 30 GB.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/StarAlign.wdl" target="_blank" rel="noopener noreferrer">StarAlign.STARsoloFastq (alias = STARsoloFastq)</a></td><td>STAR</td><td><a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">Star</a></td><td>Uses the partitioned FASTQ files to perform CB correction, adaptor trimming, alignment, gene annotation, UMI correction, and gene counting.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/MergeSortBam.wdl" target="_blank" rel="noopener noreferrer">Merge.MergeSortBamFiles (alias= MergeBam)</a></td><td>MergeSamFiles</td><td>Picard</td><td>Merges the array of BAM files into a single BAM and sorts in coordinate order.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/StarAlign.wdl" target="_blank" rel="noopener noreferrer">StarAlign.MergeStarOutput (alias = MergeStarOutputs)</a></td><td>create-npz-output.py</td><td>Python3</td><td>Creates a compressed raw NPY or NPZ file containing the STARsolo output features (NPY), barcodes (NPZ) and counts (NPZ).</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/Metrics.wdl" target="_blank" rel="noopener noreferrer">Metrics.CalculateGeneMetrics (alias = GeneMetrics)</a></td><td>TagSort</td><td>sctools</td><td>Sorts the BAM file by gene using the cell barcode (CB), molecule barcode (UB) and gene ID (GX) tags and computes gene metrics.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/Metrics.wdl" target="_blank" rel="noopener noreferrer">Metrics.CalculateCellMetrics (alias = CellMetrics)</a></td><td>TagSort</td><td>sctools</td><td>Sorts the BAM file by cell using the cell barcode (CB), molecule barcode (UB) and gene ID (GX) tags and computes cell metrics.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/RunEmptyDrops.wdl" target="_blank" rel="noopener noreferrer">RunEmptyDrops.RunEmptyDrops</a></td><td>npz2rds.sh, emptyDropsWrapper.R, emptyDrops</td><td><a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" target="_blank" rel="noopener noreferrer">DropletUtils</a></td><td>Runs custom scripts to convert the NPY and NPZ files to RDS and then uses emptyDrops to identify empty lipid droplets. This step only runs when <code>counting_mode</code> = &quot;sc_rna&quot;.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/LoomUtils.wdl" target="_blank" rel="noopener noreferrer">LoomUtils.OptimusLoomGeneration</a></td><td>create_loom_optimus.py</td><td>Python3</td><td>Merges the gene counts, cell metrics, gene metrics, and emptyDrops data into a Loom formatted cell-by-gene matrix. The Loom contains exon counts when using sc_rna mode, and whole-gene counts when running in sn_rna mode. It optionally contains an additional layer for exon counts when running sn_rna mode with <code>exon_counts</code> set to true.</td></tr></tbody></table><p>More information about the different tags used to flag the data can be found in the <a href="/warp/docs/Pipelines/Optimus_Pipeline/Bam_tags">Bam_tags documentation</a>.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="1-partition-cbs"></a>1. Partition CBs<a class="hash-link" href="#1-partition-cbs" title="Direct link to heading">#</a></h4><p>To optimally scale for large files, the workflow first uses the <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/FastqProcessing.wdl" target="_blank" rel="noopener noreferrer">FastqProcessing task</a> to partition the input array of forward and reverse read FASTQ files by uncorrected CBs. The resulting array contains FASTQ files that are ~30 GB in size. If the input files are smaller than 30 GB, this task is omitted.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="2-correct-cbs-trims-reads-align-annotate-genes-correct-umis-and-count-genes"></a>2. Correct CBs, trims reads, align, annotate genes, correct UMIs, and count genes<a class="hash-link" href="#2-correct-cbs-trims-reads-align-annotate-genes-correct-umis-and-count-genes" title="Direct link to heading">#</a></h4><p>The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/StarAlign.wdl" target="_blank" rel="noopener noreferrer">STARsoloFastq task</a> uses the partitioned FASTQ files to perform CB correction, alignment, gene annotation, and counting. </p><p><strong>CB correction</strong></p><p>Although the function of the CBs is to identify unique cells, barcode errors can arise during sequencing, such as the incorporation of the barcode into contaminating DNA or sequencing and PCR errors. This makes it difficult to distinguish unique cells from barcode artifacts. The STARsoloFastq task uses the STAR aligner to evaluate barcode errors by comparing the R1 FASTQ sequences against a 10x chemistry-specific whitelist of known barcode sequences. </p><p>Corrected barcodes are those that come within one edit distance (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5410656/" target="_blank" rel="noopener noreferrer">Hamming  distance</a>) of matching the whitelist of barcode sequences. This is specified in the STAR parameter  <code>--soloCBmatchWLtype 1MM_multi_Nbase_pseudocounts</code>, which allows multiple matches in the whitelist with 1 mismatched base and uses posterior probability calculation to choose one of the matches. Additionally, it adds pseudocounts of 1 to all whitelist barcodes and allows multi-matching of CBs with N-bases to the whitelist.</p><p>Correct barcodes are assigned a ‚ÄúCB‚Äù tag in the BAM. Uncorrectable barcodes (with more than one error) are preserved and given a ‚ÄúCR‚Äù (Cell barcode Raw) tag. CB quality scores are also preserved in the file under the ‚ÄúCY‚Äù tag. </p><p><strong>Read trimming</strong></p><p>Read trimming removes Illumina adapter sequences. This is set to match the read trimming performed by CellRanger4 and is specified using the parameter <code>--clipAdapterType CellRanger4</code> and  <code>--outFilterScoreMin 30</code>.</p><p><strong>Alignment</strong></p><p>STAR maps barcoded reads to the genome primary assembly reference (see the Quickstart table above for version information). The example references for Optimus were generated using the <a href="https://github.com/broadinstitute/warp/tree/develop/pipelines/skylab/build_indices/BuildIndices.wdl" target="_blank" rel="noopener noreferrer">BuildIndices workflow</a>. The strandedness for alignment is specified in STAR with the <code>--soloStrand</code> parameter, which is set to unstranded by default. </p><p><strong>Gene annotation</strong></p><p>Prior to gene counting, STARsolo adds gene annotations which will vary depending on the counting_mode (&quot;sc_rna&quot; or &quot;sn_rna&quot;) specified in the Optimus workflow. With <code>sc_rna</code>, STARsolo runs with the ‚ÄúGene‚Äù COUNTING_MODE, which is specific to exons.  With the <code>sn_rna</code> mode, STARsolo runs the ‚ÄúGeneFull‚Äù COUNTING_MODE and has the additional option to run the &quot;Gene&quot; mode when the input parameter <code>count_exons</code> is set to &quot;true&quot;. </p><p>Genes that overlap an alignment are stored with the GX BAM tag; for sc_rna mode, this will include the gene that corresponds to an exon or UTR, whereas for sn_rna mode, this will include the gene corresponding to an exon, UTR, and intron.</p><p>All tags are detailed in the pipeline&#x27;s <a href="/warp/docs/Pipelines/Optimus_Pipeline/Bam_tags">BAM_tag documentation</a>.</p><p>The resulting BAM files are merged together into a single BAM using the <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/skylab/MergeSortBam.wdl" target="_blank" rel="noopener noreferrer">MergeSortBamFiles (alias= MergeBam)</a>.</p><p><strong>UMI correction and gene counting</strong></p><p>UMIs are designed to distinguish unique transcripts present in the cell at lysis from those arising from PCR amplification of these same transcripts. But, like CBs, UMIs can also be incorrectly sequenced or amplified. </p><p>By specifying the <code>--soloUMIdedup 1MM_Directional_UMItools</code>, STARsolo applies a network-based, &quot;directional&quot; correction method (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5340976/" target="_blank" rel="noopener noreferrer">Smith, et al., 2017</a>) to account for such errors. Additionally, the UMI correction task requires the length of the UMI, which differs between v2 and v3 chemistry; v2 is 10 bp whereas v3 is 12 bp. The task will add a &#x27;UB&#x27; tag for UMI-corrected barcodes. </p><p>Deduplicated UMIs are counted towards their assigned gene/cells, producing a raw count matrix. </p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="3-merge-starsolo-count-matrices"></a>3. Merge STARsolo count matrices<a class="hash-link" href="#3-merge-starsolo-count-matrices" title="Direct link to heading">#</a></h4><p>The STARsolo output will include a features, barcodes, and matrix TSV for each of the partitioned FASTQ input. The <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/skylab/StarAlign.wdl" target="_blank" rel="noopener noreferrer">MergeStarOutput task</a> merges each respective TSV. It uses a custom python script to convert the merged matrix, features, and barcodes output from STARsolo into an NPY (features and barcodes)- and NPZ (the matrix)-formatted file for downstream empty drops detection and Loom matrix generation. </p><p><strong>STARsolo outputs</strong></p><p>The task‚Äôs output includes a merged, coordinate-sorted BAM file containing the CB-corrected reads and SAM attributes UB UR UY CR CB CY NH GX GN. Additionally, after counting, the task outputs three intermediate TSV files (features, barcodes, and matrix) used for downstream empty droplet detection and Loom matrix generation. </p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="4-calculate-gene-metrics"></a>4. Calculate gene metrics<a class="hash-link" href="#4-calculate-gene-metrics" title="Direct link to heading">#</a></h4><p>The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/Metrics.wdl" target="_blank" rel="noopener noreferrer">CalculateGeneMetrics</a> task uses <a href="https://github.com/HumanCellAtlas/sctools" target="_blank" rel="noopener noreferrer">sctools</a> to calculate summary metrics that help assess the quality of the data output each time this pipeline is run. These metrics are included in the output Loom matrix. A detailed list of these metrics is found in the <a href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a>.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="5-calculate-cell-metrics"></a>5. Calculate cell metrics<a class="hash-link" href="#5-calculate-cell-metrics" title="Direct link to heading">#</a></h4><p>The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/Metrics.wdl" target="_blank" rel="noopener noreferrer">CalculateCellMetrics</a> task uses <a href="https://github.com/HumanCellAtlas/sctools" target="_blank" rel="noopener noreferrer">sctools</a> to calculate summary metrics that help assess the per-cell quality of the data output each time this pipeline is run. These metrics are included in the output Loom matrix. A detailed list of these metrics is found in the <a href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a>.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="6-run-emptydrops"></a>6. Run emptyDrops<a class="hash-link" href="#6-run-emptydrops" title="Direct link to heading">#</a></h4><p>Empty droplets are lipid droplets that did not encapsulate a cell during 10x sequencing, but instead acquired cell-free RNA (secreted RNA or RNA released during cell lysis) from the solution in which the cells resided (<a href="https://www.ncbi.nlm.nih.gov/pubmed/?term=30902100" target="_blank" rel="noopener noreferrer">Lun, et al., 2018</a>. This ambient RNA can serve as a substrate for reverse transcription, leading to a small number of background reads. The Optimus pipeline calls the <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/RunEmptyDrops.wdl" target="_blank" rel="noopener noreferrer">RunEmptyDrops</a> task which uses the <a href="http://bioconductor.org/packages/release/bioc/html/DropletUtils.html" target="_blank" rel="noopener noreferrer">DropletUtils</a> R package to flag CBs that represent empty droplets rather than cells. A cell will be flagged if it contains fewer than 100 molecules.
emptyDrops metrics for single-cell data (not single-nucleus; see note below) are stored in the columns of the output Loom matrix.  Read full details for all the metrics in the <a href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Matrix Overview</a>.</p><p>EmptyDrops relies on a visual knee point inflection (described in <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1662-y" target="_blank" rel="noopener noreferrer">Lun et al. (2019)</a>) to differentiate ambient-like cells from empty droplets. <strong>If the single-cell data (counting mode set to <code>sc_rna</code>) does not produce a knee point inflection when running emptyDrops, the Loom columns for emptyDrops data will contain &quot;NA&quot;s.</strong></p><div class="admonition admonition-warning alert alert--danger"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M5.05.31c.81 2.17.41 3.38-.52 4.31C3.55 5.67 1.98 6.45.9 7.98c-1.45 2.05-1.7 6.53 3.53 7.7-2.2-1.16-2.67-4.52-.3-6.61-.61 2.03.53 3.33 1.94 2.86 1.39-.47 2.3.53 2.27 1.67-.02.78-.31 1.44-1.13 1.81 3.42-.59 4.78-3.42 4.78-5.56 0-2.84-2.53-3.22-1.25-5.61-1.52.13-2.03 1.13-1.89 2.75.09 1.08-1.02 1.8-1.86 1.33-.67-.41-.66-1.19-.06-1.78C8.18 5.31 8.68 2.45 5.05.32L5.03.3l.02.01z"></path></svg></span>RunEmptyDrops output not included for single-nucleus data</h5></div><div class="admonition-content"><p>Often snRNAseq data does not produce a visual knee point inflection when running emptyDrops and the tool can not accurately distinguish ambient-like cells from empty droplets. For this reason, emptyDrops is not used if Optimus counting_mode is set to <code>sn_rna</code>, and the output Loom matrix will not contain emptyDrops metrics.</p></div></div><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="7--matrix-construction"></a>7.  Matrix construction<a class="hash-link" href="#7--matrix-construction" title="Direct link to heading">#</a></h4><p>The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/LoomUtils.wdl" target="_blank" rel="noopener noreferrer">OptimusLoomGeneration</a> task uses a custom python script to merge the converted STARsolo count matrix, the emptyDrops results, and the cell and gene metrics into a Loom-formatted cell-by-gene matrix. <strong>These counts are raw and unfiltered.</strong></p><p>Read full details for all the metrics in the <a href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a>.</p><p>The type of gene counts in the Loom will vary depending on the Optimus workflow counting_mode. If running single-cell data (sc_rna mode), the counts will include only exonic gene counts. </p><p>If running single-nucleus data (sn_rna mode), the counts in the main matrix will be whole transcript and, when <code>count_exons</code> is set to true, the counts in an additional layer will be exonic. Using the <code>count_exons</code> parameter will cause the Loom matrix to have additional columns (cell barcodes) due to the difference in STARsolo counting mode.</p><p>You can determine which type of counts are in the loom by looking at the global attribute <code>expression_data_type</code>.</p><p>For sn_rna mode, you can also access whole transcript and exonic counts using Loompy&#x27;s <code>layers()</code> method. For example, <code>loompy.connect.layers[‚Äú‚Äù]</code> will return the whole transcript counts from the output Loom file. Similarly, <code>loompy.connect.layers[‚Äúexon_counts‚Äù]</code> will return the exonic counts from the output Loom. </p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="8-outputs"></a>8. Outputs<a class="hash-link" href="#8-outputs" title="Direct link to heading">#</a></h4><p>Output files of the pipeline include:</p><ol><li>Cell x Gene unnormalized, but UMI-corrected, count matrices in Loom format.</li><li>Unfiltered, sorted BAM file with barcode and downstream analysis <a href="/warp/docs/Pipelines/Optimus_Pipeline/Bam_tags">tags</a>.</li><li>Cell metadata, including cell metrics.</li><li>Gene metadata, including gene metrics.</li></ol><p>The following table lists the output files produced from the pipeline. For samples that have sequenced over multiple lanes, the pipeline will output one merged version of each listed file.</p><table><thead><tr><th>Output Variable Name</th><th>Filename, if applicable</th><th>Output Type</th><th>Output Format</th></tr></thead><tbody><tr><td>pipeline_version_out</td><td>N/A</td><td>Version of the processing pipeline run on this data.</td><td>String</td></tr><tr><td>bam</td><td><code>&lt;input_id&gt;.bam</code></td><td>Aligned BAM</td><td>BAM</td></tr><tr><td>matrix</td><td><code>&lt;input_id&gt;_sparse_counts.npz</code></td><td>Converted sparse matrix file from the MergeStarOutputs task.</td><td>NPZ</td></tr><tr><td>matrix_row_index</td><td><code>&lt;input_id&gt;_sparse_counts_row_index.npy</code></td><td>Index of cells in count matrix.</td><td>NPY</td></tr><tr><td>matrix_col_index</td><td><code>&lt;input_id&gt;_sparse_counts_col_index.npy</code></td><td>Index of genes in count matrix.</td><td>NPY</td></tr><tr><td>cell_metrics</td><td>cell-metrics.csv.gz</td><td>Cell metrics</td><td>Compressed CSV</td></tr><tr><td>gene_metrics</td><td>gene-metrics.csv.gz</td><td>Gene metrics</td><td>Compressed CSV</td></tr><tr><td>cell_calls</td><td>empty_drops_result.csv</td><td>emptyDrops results from the RunEmptyDrops task.</td><td>CSV</td></tr><tr><td>loom_output_file</td><td><code>&lt;input_id&gt;.loom</code></td><td>Loom</td><td>Loom</td></tr></tbody></table><p>The Loom matrix is the default output. See the <a href="https://github.com/broadinstitute/warp/blob/master/dockers/skylab/loom-output/create_loom_optimus.py" target="_blank" rel="noopener noreferrer">create_loom_optimus.py</a> for the detailed code. This matrix contains the unnormalized (unfiltered), UMI-corrected count matrices, as well as the gene and cell metrics detailed in the <a href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a>.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="try-the-optimus-matrix-with-community-tools"></a>Try the Optimus matrix with community tools<a class="hash-link" href="#try-the-optimus-matrix-with-community-tools" title="Direct link to heading">#</a></h4><p>The matrix is compatible with multiple downstream community analysis tools, including <a href="https://satijalab.org/seurat/index.html" target="_blank" rel="noopener noreferrer">Seurat</a>, <a href="https://scanpy.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Scanpy</a>, <a href="https://cumulus.readthedocs.io/en/latest/index.html" target="_blank" rel="noopener noreferrer">Cumulus</a>, and <a href="https://pegasus.readthedocs.io/en/stable/#" target="_blank" rel="noopener noreferrer">Pegasus</a>. To try a tutorial using the Optimus matrix with these tools, register for the open-source platform <a href="https://app.terra.bio" target="_blank" rel="noopener noreferrer">Terra</a> and then navigate to the public <a href="https://app.terra.bio/#workspaces/featured-workspaces-hca/Intro-to-HCA-data-on-Terra" target="_blank" rel="noopener noreferrer">Intro-to-HCA-data-on-Terra workspace</a>. You can also view the accompanying <a href="https://support.terra.bio/hc/en-us/articles/360060041772" target="_blank" rel="noopener noreferrer">step-by-step guide</a> without registration.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="validation-against-cell-ranger"></a>Validation against Cell Ranger<a class="hash-link" href="#validation-against-cell-ranger" title="Direct link to heading">#</a></h2><p>Optimus has been validated for processing both human and mouse single-cell and single-nucleus data (see links to validation reports in the table below). For each validation, Optimus results are compared to those of Cell Ranger (see the <a href="#faqs">FAQ</a> for more on Cell Ranger comparisons).</p><table><thead><tr><th>Workflow configuration</th><th>Link to Report</th></tr></thead><tbody><tr><td>Human 10x v2 single-cell</td><td><a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/benchmarking/v1_Apr2019/optimus_report.rst" target="_blank" rel="noopener noreferrer">Report</a></td></tr><tr><td>Mouse 10x v2 single-cell</td><td><a href="https://docs.google.com/document/d/1_3oO0ZQSrwEoe6D3GgKdSmAQ9qkzH_7wrE7x6_deL10/edit" target="_blank" rel="noopener noreferrer">Report</a></td></tr><tr><td>Human and mouse 10x v3 single-cell</td><td><a href="https://docs.google.com/document/d/1-hwfXkqtL8MblgDWFzk-HsVRYiy4PS8ZhJqAGlHBWYE/edit#heading=h.4uokn64v1s5m" target="_blank" rel="noopener noreferrer">Report</a></td></tr><tr><td>Human and Mouse 10x v2/v3 single-nucleus</td><td><a href="https://docs.google.com/document/d/1rv2M7vfpOzIOsMnMfNyKB4HV18lQ9dnOGHK2tPikiH0/edit" target="_blank" rel="noopener noreferrer">Report</a></td></tr><tr><td>Optimus STARsolo (v5.0.0 and later)</td><td><a href="https://docs.google.com/document/d/1B6Ux6HICD4ZL4Z0TG9LO-X43gOdF3sbq5qw_L2GA6fg/edit" target="_blank" rel="noopener noreferrer">Report</a></td></tr></tbody></table><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="versioning"></a>Versioning<a class="hash-link" href="#versioning" title="Direct link to heading">#</a></h2><p>All Optimus pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/skylab/optimus/Optimus.changelog.md" target="_blank" rel="noopener noreferrer">Optimus changelog</a>.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="citing-the-optimus-pipeline"></a>Citing the Optimus pipeline<a class="hash-link" href="#citing-the-optimus-pipeline" title="Direct link to heading">#</a></h2><p>Please identify the pipeline in your methods section using the Optimus Pipeline&#x27;s <a href="https://scicrunch.org/scicrunch/Resources/record/nlx_144509-1/SCR_018908/resolver?q=SCR_018908&amp;l=SCR_018908" target="_blank" rel="noopener noreferrer">SciCrunch resource identifier</a>.</p><ul><li>Ex: <em>Optimus Pipeline (RRID:SCR_018908)</em></li></ul><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="consortia-support"></a>Consortia support<a class="hash-link" href="#consortia-support" title="Direct link to heading">#</a></h2><p>This pipeline is supported and used by the <a href="https://www.humancellatlas.org/" target="_blank" rel="noopener noreferrer">Human Cell Atlas</a> (HCA) project and the <a href="https://biccn.org/" target="_blank" rel="noopener noreferrer">BRAIN Initiative Cell Census Network</a> (BICCN). </p><p>Each consortia may use slightly different reference files for data analysis or have different post-processing steps. Learn more by reading the <a href="/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing">Consortia Processing</a> overview.</p><p>If your organization also uses this pipeline, we would like to list you! Please reach out to us by contacting <a href="mailto:warp-pipelines-help@broadinstitute.org" target="_blank" rel="noopener noreferrer">the WARP team</a>.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="feedback"></a>Feedback<a class="hash-link" href="#feedback" title="Direct link to heading">#</a></h2><p>Please help us make our tools better by contacting <a href="mailto:warp-pipelines-help@broadinstitute.org" target="_blank" rel="noopener noreferrer">the WARP team</a> for pipeline-related suggestions or questions.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="faqs"></a>FAQs<a class="hash-link" href="#faqs" title="Direct link to heading">#</a></h2><div class="admonition admonition-note alert alert--secondary"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="14" height="16" viewBox="0 0 14 16"><path fill-rule="evenodd" d="M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"></path></svg></span>Question Can I run Optimus in Terra?</h5></div><div class="admonition-content"><p>Yes! We have a Terra workspace that is preconfigured with the latest Optimus workflow and is preloaded with human and mouse sample data. You can access the <a href="https://app.terra.bio/#workspaces/featured-workspaces-hca/HCA_Optimus_Pipeline" target="_blank" rel="noopener noreferrer">workspace</a>. You will need a Google account to set up Terra. Please see <a href="https://support.terra.bio/hc/en-us" target="_blank" rel="noopener noreferrer">Terra Support</a> for documents on getting started.</p></div></div><div class="admonition admonition-note alert alert--secondary"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="14" height="16" viewBox="0 0 14 16"><path fill-rule="evenodd" d="M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"></path></svg></span>Question Is the output count matrix filtered or normalized?</h5></div><div class="admonition-content"><p>No, we do not filter. We keep as much data as possible so that the researcher can make their own filtering and normalization choices. We do, however, output some information that may be helpful for filtering, like UMI counts per cell and calls on whether or not a cell is empty from emptyDrops software. For the emptyDrops call, a cell will be flagged as possibly empty if it contains fewer than 100 molecules.</p></div></div><div class="admonition admonition-note alert alert--secondary"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="14" height="16" viewBox="0 0 14 16"><path fill-rule="evenodd" d="M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"></path></svg></span>Question How does the workflow change when using the single-cell RNA-seq (counting_mode = &#x27;sc_rna&#x27;) vs. the single-nucleus (counting_mode = &#x27;sn_rna&#x27;) parameters?</h5></div><div class="admonition-content"><p>The counting_mode parameter is used to specify the STARsolo COUNTING_MODE; when sn_rna is specified, STARsolo will tag gene exons, UTRs, AND introns with the GX tag.  Additionally, the Optimus uses the counting_mode to determine whether to run emptyDrops; no emptyDrops data is calculated for the sn_rna mode.</p></div></div><div class="admonition admonition-note alert alert--secondary"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="14" height="16" viewBox="0 0 14 16"><path fill-rule="evenodd" d="M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"></path></svg></span>Question Where can I find example Optimus datasets and parameters to test the pipeline?</h5></div><div class="admonition-content"><p>There are four example configuration JSON files available for you to test the pipeline- the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/example_inputs/human_v2_example.json" target="_blank" rel="noopener noreferrer">human_v2_example.json</a>, the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/example_inputs/human_v3_example.json" target="_blank" rel="noopener noreferrer">human_v3_example.json</a>, the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/example_inputs/mouse_v2_snRNA_example.json" target="_blank" rel="noopener noreferrer">mouse_v2_snRNA_example.json</a>, and the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/example_inputs/mouse_v2_snRNA_example.json" target="_blank" rel="noopener noreferrer">mouse_v2_snRNA_example.json</a>(see the Inputs section). Each of these configuration files can be run in the Optimus Featured Workspace in Terra at <a href="https://app.terra.bio/#workspaces/featured-workspaces-hca/HCA_Optimus_Pipeline" target="_blank" rel="noopener noreferrer">https://app.terra.bio/#workspaces/featured-workspaces-hca/HCA_Optimus_Pipeline</a>, but you should note that the workspace comes preloaded with the same data and configurations. We also have multiple example datasets available in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/example_inputs/test_optimus_full_datasets/" target="_blank" rel="noopener noreferrer">test_optimus_full_datasets</a> folder. These datasets contain configuration files listing the cloud location of the dataset FASTQ files; however, the configuration files may not be updated with all the workflow parameters. For the most up-to-date configuration examples, see the four example files listed above.</p></div></div><div class="admonition admonition-note alert alert--secondary"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="14" height="16" viewBox="0 0 14 16"><path fill-rule="evenodd" d="M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"></path></svg></span>Question What outputs are expected if my sample has been sequenced over multiple lanes?</h5></div><div class="admonition-content"><p>The Optimus pipeline is a single sample pipeline, but it can accept multiple FASTQ files if a sample is sequenced across lanes. In this case, the pipeline will merge the results from each lane into single output files. There will only be one merged file for each output type (i.e one Loom, etc.). If you would like to view an example configuration file for a multi-lane dataset, please see the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/example_inputs/mouse_v2_example.json" target="_blank" rel="noopener noreferrer">mouse_v2_example.json</a>.  Additionally, you can view sample outputs in the Optimus featured workspace on <a href="https://app.terra.bio/#workspaces/featured-workspaces-hca/HCA_Optimus_Pipeline" target="_blank" rel="noopener noreferrer">Terra</a>.</p></div></div><div class="admonition admonition-note alert alert--secondary"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="14" height="16" viewBox="0 0 14 16"><path fill-rule="evenodd" d="M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"></path></svg></span>Question How do I find which parameters and Docker images were used for the different tasks (i.e. STAR alignment, emptyDrops, etc.)</h5></div><div class="admonition-content"><p>Parameters are listed in each task WDL. For a list of the tasks, see the table in the <a href="#optimus-task-summary">Task Summary Section</a>. Select the link for the task of interest and then view the parameters in the task WDL &quot;command {}&quot; section. For the task Docker image, see task WDL &quot;# runtime values&quot; section; the Docker is listed as &quot;String docker =  &quot;. If you want to learn more about all the different parameters available for a software tool, please select the relevant link in the table&#x27;s &quot;Tool&quot; column.</p></div></div><div class="admonition admonition-note alert alert--secondary"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="14" height="16" viewBox="0 0 14 16"><path fill-rule="evenodd" d="M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"></path></svg></span>Question Does Optimus have any read length requirements?</h5></div><div class="admonition-content"><p>For Read 1 sequences, the only minimum requirement is that reads are the combined lengths of the CB and UMIs (which will vary between 10x V1, V2, and V3 chemistry).</p><p>For Read 2 sequences, there is no read length requirement and read lengths will vary.</p></div></div><div class="admonition admonition-note alert alert--secondary"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="14" height="16" viewBox="0 0 14 16"><path fill-rule="evenodd" d="M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"></path></svg></span>Question How does Optimus compare to Cell Ranger?</h5></div><div class="admonition-content"><p>Cell Ranger is a commonly used set of analysis pipelines developed by <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" target="_blank" rel="noopener noreferrer">10x Genomics</a>. Optimus and Cell Ranger share many features and additionally, Optimus results are validated against Cell Ranger results (see our <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/benchmarking/v1_Apr2019/optimus_report.rst" target="_blank" rel="noopener noreferrer">human validation report</a>). </p><p><em>So why develop an independent pipeline for 10x data analyses?</em></p><p>For three reasons:
1) Need for an open-source, cloud-optimized pipeline. When Optimus was developed, Cell Ranger software was not yet open source, nor was it optimized for the cloud. To date, the Cell Ranger open-source code is still not regularly updated with Cell Ranger releases. In consequence, using the latest Cell Ranger (which is not open source yet) limits our ability to harness the breadth of tools available in the scientific community.</p><p>2) Flexibility to process data similar, but not identical, to 10x. We wanted the ability to evolve our pipeline to process non-10x data types that might use similar features such as combinatorial indexing.</p><p>3) Addition of metrics. We wanted the pipeline to calculate key metrics that would be useful to the scientific community, such as emptyDrops calculations, mitochondrial read metrics, etc.</p><p><em>Reference differences between Optimus and Cell Ranger</em></p><p>Unlike Cell Ranger references, Optimus references are downloaded directly from GENCODE and not modified to remove pseudogenes and small RNAs. Learn more about Cell Ranger references on the <a href="https://support.10xgenomics.com/single-cell-multiome-atac-gex/software/release-notes/references#header" target="_blank" rel="noopener noreferrer">10x website</a>. </p><p>In the case of multi-mapped pseudogenes, Optimus and Cell Ranger will produce different results. Optimus does not count multi-mapped reads in the final count matrix, whereas Cell Ranger will keep potential multi-mapped reads because it does not identify the pseudogene reads.</p></div></div></div><footer class="docusaurus-mt-lg"><div class="row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/Optimus_Pipeline/README.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_2_ui" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_13-_">Last updated on <b><time datetime="2023-01-12T16:10:05.000Z">1/12/2023</time></b> by <b>dependabot[bot]</b></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/warp/docs/Pipelines/Imputation_Pipeline/references_overview"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">¬´ Creating the Imputation Pipeline&#x27;s Modified 1000 Genomes Reference</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/warp/docs/Pipelines/Optimus_Pipeline/optimus.methods"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Optimus v5.3.0 Methods ¬ª</div></a></div></nav></div></div><div class="col col--3"><div class="tableOfContents_35-E thin-scrollbar"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-optimus-workflow" class="table-of-contents__link">Introduction to the Optimus workflow</a></li><li><a href="#quickstart-table" class="table-of-contents__link">Quickstart table</a></li><li><a href="#set-up" class="table-of-contents__link">Set-up</a><ul><li><a href="#optimus-installation" class="table-of-contents__link">Optimus installation</a></li><li><a href="#inputs" class="table-of-contents__link">Inputs</a></li></ul></li><li><a href="#optimus-tasks-and-tools" class="table-of-contents__link">Optimus tasks and tools</a></li><li><a href="#validation-against-cell-ranger" class="table-of-contents__link">Validation against Cell Ranger</a></li><li><a href="#versioning" class="table-of-contents__link">Versioning</a></li><li><a href="#citing-the-optimus-pipeline" class="table-of-contents__link">Citing the Optimus pipeline</a></li><li><a href="#consortia-support" class="table-of-contents__link">Consortia support</a></li><li><a href="#feedback" class="table-of-contents__link">Feedback</a></li><li><a href="#faqs" class="table-of-contents__link">FAQs</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item"><span>GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_3J9K"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></span></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright ¬© 2023 Copyright ¬© Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.71584178.js"></script>
<script src="/warp/assets/js/main.1d35face.js"></script>
</body>
</html>