<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Pipelines/Optimus_Pipeline/consortia-processing">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">Consortia Data Processing | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Consortia Data Processing | WARP"><meta data-rh="true" name="description" content="Human Cell Atlas Data Coordination Platform Matrix Processing"><meta data-rh="true" property="og:description" content="Human Cell Atlas Data Coordination Platform Matrix Processing"><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.728576f2.js" as="script">
<link rel="preload" href="/warp/assets/js/main.868095ac.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/ATAC/README">ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/BuildIndices_Pipeline/README">BuildIndices</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README">Exome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Exome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/README">Illumina Genotyping Array</a><button aria-label="Toggle the collapsible sidebar category &#x27;Illumina Genotyping Array&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Imputation_Pipeline/README">Imputation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Imputation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/JointGenotyping_Pipeline/README">JointGenotyping</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Multiome_Pipeline/README">Multiome scATAC and GEX</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus</a><button aria-label="Toggle the collapsible sidebar category &#x27;Optimus&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/optimus.methods">Optimus v7.8.1 Methods</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/Bam_tags">What tags are included in an Optimus BAM file?</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/Library-metrics">Optimus Library-level metrics</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing">Consortia Data Processing</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/starsolo-metrics">STAR Aligner Metrics</a></li></ul></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/PairedTag_Pipeline/README">Paired-Tag</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README">RNA with UMIs</a><button aria-label="Toggle the collapsible sidebar category &#x27;RNA with UMIs&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/SlideSeq_Pipeline/README">Slide-seq</a><button aria-label="Toggle the collapsible sidebar category &#x27;Slide-seq&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README">Smart-seq2 Single Nucleus Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Nucleus Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/snm3C/README">Single Nucleus Methyl-Seq and Chromatin Capture</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Nucleus Methyl-Seq and Chromatin Capture&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Ultima_Genomics_Whole_Genome_Germline_Pipeline/README">Ultima Genomics Whole Genome Germline</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/README">Whole Genome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Whole Genome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">WARP Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/warp/docs/Pipelines/Optimus_Pipeline/README"><span itemprop="name">Optimus</span></a><meta itemprop="position" content="2"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Consortia Data Processing</span><meta itemprop="position" content="3"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Consortia Data Processing</h1><h2 class="anchor anchorWithStickyNavbar_LWe7" id="human-cell-atlas-data-coordination-platform-matrix-processing">Human Cell Atlas Data Coordination Platform Matrix Processing<a href="#human-cell-atlas-data-coordination-platform-matrix-processing" class="hash-link" aria-label="Direct link to Human Cell Atlas Data Coordination Platform Matrix Processing" title="Direct link to Human Cell Atlas Data Coordination Platform Matrix Processing">​</a></h2><p>Optimus supports data processing for the Human Cell Atlas (HCA) Data Coordination Platform (DCP). Learn more about the DCP at the <a href="https://data.humancellatlas.org/" target="_blank" rel="noopener noreferrer">HCA Data Portal</a>).</p><p>All DCP Projects processed with Optimus have matrices containing the standard metrics and counts detailed in the <a href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a>, but also have additional post-processing to combine project data and incorporate DCP-curated metadata.</p><div class="theme-admonition theme-admonition-warning alert alert--danger admonition_LlT9"><div class="admonitionHeading_tbUL"><span class="admonitionIcon_kALy"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M5.05.31c.81 2.17.41 3.38-.52 4.31C3.55 5.67 1.98 6.45.9 7.98c-1.45 2.05-1.7 6.53 3.53 7.7-2.2-1.16-2.67-4.52-.3-6.61-.61 2.03.53 3.33 1.94 2.86 1.39-.47 2.3.53 2.27 1.67-.02.78-.31 1.44-1.13 1.81 3.42-.59 4.78-3.42 4.78-5.56 0-2.84-2.53-3.22-1.25-5.61-1.52.13-2.03 1.13-1.89 2.75.09 1.08-1.02 1.8-1.86 1.33-.67-.41-.66-1.19-.06-1.78C8.18 5.31 8.68 2.45 5.05.32L5.03.3l.02.01z"></path></svg></span>danger</div><div class="admonitionContent_S0QG"><p>This section details matrices produced for the Human Cell Atlas (HCA) <a href="https://data.humancellatlas.org/" target="_blank" rel="noopener noreferrer">Data Coordination Platform (DCP)2.0</a>, which includes matrices processed with Optimus v4.1.7 and later. The DCP is currently reprocessing data generated with earlier Optimus versions and will deprecate previous matrices once reprocessing is complete.</p></div></div><h3 class="anchor anchorWithStickyNavbar_LWe7" id="dcp-project-matrix-post-processing">DCP project matrix post-processing<a href="#dcp-project-matrix-post-processing" class="hash-link" aria-label="Direct link to DCP project matrix post-processing" title="Direct link to DCP project matrix post-processing">​</a></h3><p>DCP project matrices combine project data from individual library preparations that contain the same species, developmental age, and sequencing technology. </p><p>Key differences between the standard Optimus Loom matrix described in the <a href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a> and the DCP Project matrix include: </p><ul><li>Combined project data in the DCP matrix</li><li>Filtering: to reduce file size, DCP project matrices are in sparse format and minimally filtered so that <strong>only cells with 100 molecules or more</strong> are retained.</li><li>DCP-curated metadata in the Loom global attributes: see table below.</li><li>input_ids in the global attributes: all input_ids representing each library preparation in the matrix are added as a comma-separated string.</li></ul><p>DCP project matrices contain DCP-curated metadata in the Loom global attributes which may be useful when exploring the data and linking it back to the Project metadata. </p><p>Read more about each metadata field in the DCP <a href="https://data.humancellatlas.org/metadata" target="_blank" rel="noopener noreferrer">Metadata Dictionary</a>.</p><table><thead><tr><th>Metadata Attribute Name in Count Matrix</th><th>Metadata Description</th></tr></thead><tbody><tr><td><code>donor_organism.genus_species</code></td><td>species information; human or mouse</td></tr><tr><td><code>library_preparation_protocol.library_construction_approach</code></td><td>technology used for library preparation, i.e 10x or SS2</td></tr><tr><td><code>specimen_from_organism.organ</code></td><td>organ</td></tr><tr><td><code>project.project_core.project_name</code></td><td>project name</td></tr><tr><td><code>project.provenance.document_id</code></td><td>project id</td></tr><tr><td><code>input_id</code></td><td>metadata values for  <code>sequencing_process.provenance.document_id</code>; unique ID to demarcate the library prep</td></tr><tr><td><code>input_name</code></td><td>metadata values for <code>sequencing_input.biomaterial_core.biomaterial_id</code>; unique ID for the biomaterial</td></tr><tr><td><code>input_id_metadata_field</code></td><td>string describing the DCP-curated metadata field used for input_id: <code>sequencing_process.provenance.document_id</code></td></tr><tr><td><code>input_name_metadata_field</code></td><td>string describing the DCP-curated metadata field used for input_name: <code>sequencing_input.biomaterial_core.biomaterial_id</code></td></tr></tbody></table><h3 class="anchor anchorWithStickyNavbar_LWe7" id="post-processing-steps">Post-processing steps<a href="#post-processing-steps" class="hash-link" aria-label="Direct link to Post-processing steps" title="Direct link to Post-processing steps">​</a></h3><p>To create the DCP project matrices, Loom outputs from individual 10x library preparations, each with their own <code>input_id</code>, are combined into a single Loom file.</p><p><img loading="lazy" src="/warp/assets/images/post_processing-256dd7dbd7f417796924b928983dadc3.png" width="1400" height="420" class="img_ev3q"></p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="making-project-matrix-cell-barcodes-unique">Making project matrix cell barcodes unique<a href="#making-project-matrix-cell-barcodes-unique" class="hash-link" aria-label="Direct link to Making project matrix cell barcodes unique" title="Direct link to Making project matrix cell barcodes unique">​</a></h3><p>Since DCP project matrices often contain combined data from multiple library preparations, the project matrix cell barcodes are modified so that they are unique for each library preparation, allowing the barcodes to be used by downstream community tools like Cumulus and Seurat.</p><p>In the standard Optimus matrix, cell barcodes are listed in both the Loom columns <code>CellID</code> and the <code>cell_names</code>.</p><p>For DCP projects, however, <strong>the cell barcodes in the  <code>cell_names</code> column are modified</strong> so that each cell barcode belonging to an individual library preparation is unique. This is done by adding a numerical suffix to the barcodes that corresponds to the <code>input_id</code> for the library preparation from which the cell barcodes came. </p><p>The <code>input_ids</code> are listed in the matrix global attributes. The order of the input_ids serves as an index for the cell barcode suffix.</p><p><img loading="lazy" src="/warp/assets/images/unique_barcodes-9a129aad53dcd952b00b44ad268b51f3.png" width="800" height="938" class="img_ev3q"></p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="example-cell-barcodes-from-a-dcp-project-matrix">Example cell barcodes from a DCP project matrix<a href="#example-cell-barcodes-from-a-dcp-project-matrix" class="hash-link" aria-label="Direct link to Example cell barcodes from a DCP project matrix" title="Direct link to Example cell barcodes from a DCP project matrix">​</a></h4><p>Let&#x27;s use loompy to look at the global attribute <code>input_id</code> for a DCP project matrix (loom format):</p><div class="language-python codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#bfc7d5;--prism-background-color:#292d3e"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-python codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#bfc7d5"><span class="token operator" style="color:rgb(137, 221, 255)">&gt;&gt;</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> ds</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token plain">attrs</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token plain">input_id</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;166c1b1a-ad9c-4476-a4ec-8b52eb5032c7, 22b7da3d-a301-433e-99e1-e67266c1ee8b, 337a48c5-e363-45aa-886f-ccd4425edc2b, 40630e8b-c3a3-4813-b1e4-b156637c5cc3, 58d703d1-d366-42d0-af44-a3bb836838a5, 70c8d647-7984-4d03-912a-f2437aa1ba4f, 7c86cf30-4284-4a0d-817f-6047560c05c3, 8ef7aca4-be00-4c03-8576-1b2eff4ce7af, ae0cfa6e-e7cb-4a88-9f89-1c44abaa2291, cbd23025-b1bf-4e9e-a297-ddab4a217b76, df049da4-3d20-4da7-a1d7-7d6e8f7740ff, e17bf5ea-788b-4756-a008-a07aec091e10&#x27;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;&gt;</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> </span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Notice the attribute&#x27;s value is a string of comma- <strong>and space-</strong> separated UUIDs.</p><p>Each of these UUIDs represents one library preparation. This matrix contains data from 12 library preparations total. </p><p>Now let&#x27;s look at the <code>cell_names</code> column attribute which contains the unique cell barcodes:</p><div class="language-python codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#bfc7d5;--prism-background-color:#292d3e"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-python codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#bfc7d5"><span class="token operator" style="color:rgb(137, 221, 255)">&gt;&gt;</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> ds</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token plain">ca</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token plain">cell_names</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">array</span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token punctuation" style="color:rgb(199, 146, 234)">[</span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;GGACAAGAGTGCGTGA-0&#x27;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"> </span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;GATCGATCACCAGGTC-0&#x27;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"> </span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;AGCGGTCAGGGCTTGA-0&#x27;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"></span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">       </span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"> </span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;GTACGTAAGCTATGCT-11&#x27;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"> </span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;CAGAATCTCTGAGTGT-11&#x27;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"></span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">       </span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;AACACGTAGTGTTTGC-11&#x27;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">]</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"> dtype</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token builtin" style="color:rgb(130, 170, 255)">object</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token plain"></span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;&gt;</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> </span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The suffix appended to the barcodes in the <code>cell_names</code> column is the index for the <code>input_id</code> UUID to which the cell barcodes belong. </p><p>For example, cell barcodes with a &quot;-0&quot; suffix belong to the library preparation represented by the first UUID, <code>166c1b1a-ad9c-4476-a4ec-8b52eb5032c7</code>, whereas cell barcodes with a &quot;-11&quot; suffix represent the 10th UUID, <code>cbd23025-b1bf-4e9e-a297-ddab4a217b76</code>.</p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="mapping-dcp-project-matrix-data-to-the-metadata-manifest">Mapping DCP project matrix data to the metadata manifest<a href="#mapping-dcp-project-matrix-data-to-the-metadata-manifest" class="hash-link" aria-label="Direct link to Mapping DCP project matrix data to the metadata manifest" title="Direct link to Mapping DCP project matrix data to the metadata manifest">​</a></h3><p>While the project matrices contain some project metadata (listed in the table above), there is additionally useful metadata in the project&#x27;s metadata manifest, a TSV file containing all of a project&#x27;s metadata, including donor and disease state information.</p><p>In addition to the global attribute <code>input_id</code>, each project matrix has an <code>input_id</code> column that can be useful for mapping matrix data back to the DCP metadata manifest. </p><p>The values listed in the <code>input_id</code> column match the library preparation UUID in the metadata manifest column <code>sequencing_process.provenance.document_id</code>. </p><p>Read more about the metadata manifest in the DCP <a href="https://data.humancellatlas.org/guides" target="_blank" rel="noopener noreferrer">Exploring Projects guide</a>.</p><div class="theme-admonition theme-admonition-tip alert alert--success admonition_LlT9"><div class="admonitionHeading_tbUL"><span class="admonitionIcon_kALy"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Explore HCA Project matrices in Terra</div><div class="admonitionContent_S0QG"><p>HCA matrices produced with Optimus are compatible with multiple downstream community analysis tools. For a tutorial on using the Optimus matrix with <a href="https://satijalab.org/seurat/index.html" target="_blank" rel="noopener noreferrer">Seurat</a>, <a href="https://scanpy.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Scanpy</a>, <a href="https://cumulus.readthedocs.io/en/latest/index.html" target="_blank" rel="noopener noreferrer">Cumulus</a>, or <a href="https://pegasus.readthedocs.io/en/stable/#" target="_blank" rel="noopener noreferrer">Pegasus</a>, see the public <a href="https://app.terra.bio/#workspaces/featured-workspaces-hca/Intro-to-HCA-data-on-Terra" target="_blank" rel="noopener noreferrer">Intro-to-HCA-data-on-Terra workspace</a> (login required) and its accompanying <a href="https://support.terra.bio/hc/en-us/articles/360060041772" target="_blank" rel="noopener noreferrer">step-by-step guide</a>.</p></div></div><h3 class="anchor anchorWithStickyNavbar_LWe7" id="mapping-dcp-project-matrix-data-to-the-contributor-matrices">Mapping DCP project matrix data to the contributor matrices<a href="#mapping-dcp-project-matrix-data-to-the-contributor-matrices" class="hash-link" aria-label="Direct link to Mapping DCP project matrix data to the contributor matrices" title="Direct link to Mapping DCP project matrix data to the contributor matrices">​</a></h3><p>Contributor matrices contain data analyzed and provided by the original project contributors. While they vary in format and content from project to project, they often include cell type annotations and additional metadata such as donor information and cell barcodes. </p><p><em>If the contributor matrix contains donor metadata that matches a field in the project metadata manifest</em>, the matrix can be linked to the DCP-generated project matrix in a two-step process.</p><ul><li><p>First, map the DCP matrix to the metadata manifest using the Loom&#x27;s <code>input_id</code> <strong>column</strong>; this column contains the same library preparation/donor IDs as the project metadata manifest&#x27;s <code>sequencing_process.provenance.document_id</code> column. </p></li><li><p>Second, map the contributor matrix to the metadata manifest using the contributor matrix column that matches the metadata manifest. </p></li></ul><p>Contributor matrices might contain a column for cell barcodes for each library/preparation donor. These barcodes should match the non-unique barcodes listed in the DCP project matrix, with the exception of cells that might have been filtered out of the Loom matrix due to low UMIs. The Loom&#x27;s non-unique barcodes are listed in the <code>CellID</code> column. </p><p><img loading="lazy" src="/warp/assets/images/mapping_matrices-431db5e1dfb615be0c1741fbb863465e.png" width="1720" height="763" class="img_ev3q"></p><p>For example code showing how to link a contributor matrix to a DCP project matrix, see the <a target="_blank" href="/warp/assets/files/Matrix_matching-2c4edc3d0b1881f352a39bd44e22513d.ipynb">Matrix_matching Jupyter Notebook</a>.</p><p>If you have any questions related to the contributor matrix and content, reach out to the individual project contributors listed on the Project page.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="brain-initiative-cell-census-network-processing">Brain Initiative Cell Census Network Processing<a href="#brain-initiative-cell-census-network-processing" class="hash-link" aria-label="Direct link to Brain Initiative Cell Census Network Processing" title="Direct link to Brain Initiative Cell Census Network Processing">​</a></h2><p>The Optimus pipeline supports data processing for the <a href="https://biccn.org/" target="_blank" rel="noopener noreferrer">BRAIN Initiative Cell Census Network (BICCN)</a>. An overview of the BICCN pipeline resources is available on the BICCN&#x27;s <a href="https://biccn.org/tools/biccn-pipelines" target="_blank" rel="noopener noreferrer">Pipelines page</a>.</p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="optimus-reference-files-for-biccn-data-processing">Optimus reference files for BICCN data processing<a href="#optimus-reference-files-for-biccn-data-processing" class="hash-link" aria-label="Direct link to Optimus reference files for BICCN data processing" title="Direct link to Optimus reference files for BICCN data processing">​</a></h3><p>The BICCN 2.0 Whole Mouse Brain Working Group uses the Ensembl GRCm38 reference for alignment and a modified GTF for gene annotation (see table below). All Optimus pipeline reference inputs were created with the <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/skylab/build_indices" target="_blank" rel="noopener noreferrer">BuildIndices workflow</a>.</p><p>BICCN processes single-nucleus data, which is enriched in pre-mRNAs containing introns. To account for this, the BuildIndices workflow uses the <code>BuildStarSingleNucleus</code> task to add intron annotations to the GTF with a custom <a href="https://github.com/broadinstitute/warp-tools/blob/develop/3rd-party-tools/build-indices/add-introns-to-gtf.py" target="_blank" rel="noopener noreferrer">python script</a>. The GTF contains all annotations for any <code>gene_id</code> that has at least one transcript. This reduces the number of genes in the GTF to <!-- -->~<!-- -->32,000. </p><p>All reference files are available in a public Google bucket (see table below) and are accompanied by a README that details reference provenance (gs://gcp-public-data--broad-references/mm10/v0/README_mm10_singlecell_gencode.txt). </p><table><thead><tr><th>Optimus reference input name</th><th>Google bucket URI</th><th>Reference source</th><th>Description</th></tr></thead><tbody><tr><td><code>annotations_gtf</code></td><td>gs://gcp-public-data--broad-references/mm10/v0/single_nucleus/modified_gencode.vM23.primary_assembly.annotation.gtf</td><td><a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/gencode.vM23.annotation.gtf.gzf" target="_blank" rel="noopener noreferrer">https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/gencode.vM23.annotation.gtf.gzf</a></td><td>Modified GENCODE GTF including intron annotations that can be used for intron counting with featureCounts.</td></tr><tr><td><code>ref_genome_fasta</code></td><td>gs://gcp-public-data--broad-references/mm10/v0/single_nucleus/modified_mm10.primary_assembly.genome.fa</td><td><a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/GRCm38.p6.genome.fa.gz" target="_blank" rel="noopener noreferrer">https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/GRCm38.p6.genome.fa.gz</a></td><td>FASTA filed used to create the STAR reference files.</td></tr><tr><td><code>tar_star_reference</code></td><td>gs://gcp-public-data--broad-references/mm10/v0/single_nucleus/star/modified_star_2.7.9a_primary_gencode_mouse_vM23.tar</td><td>NA — built with the BuildIndices workflow.</td><td>Reference files used for alignment with STAR.</td></tr><tr><td><code>whitelist</code></td><td>gs://broad-gotc-test-storage/Optimus/truth/plumbing/master/inputs/737K-august-2016.txt (for v2 chemistry) and gs://broad-gotc-test-storage/Optimus/truth/plumbing/master/inputs/3M-february-2018.txt (for v3 chemisty)</td><td>See <a href="https://kb.10xgenomics.com/hc/en-us/articles/115004506263-What-is-a-barcode-whitelist" target="_blank" rel="noopener noreferrer">10x barcode descriptions</a>.</td><td>List of barcode sequences included in 10x library preparations. Each barcode demarcates an individual cell.</td></tr></tbody></table></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/Optimus_Pipeline/consortia-processing.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2024-12-05T15:09:24.000Z">Dec 5, 2024</time></b> by <b>Robert Sidney Cox</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/Pipelines/Optimus_Pipeline/Library-metrics"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Optimus Library-level metrics</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/Pipelines/Optimus_Pipeline/starsolo-metrics"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">STAR Aligner Metrics</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#human-cell-atlas-data-coordination-platform-matrix-processing" class="table-of-contents__link toc-highlight">Human Cell Atlas Data Coordination Platform Matrix Processing</a><ul><li><a href="#dcp-project-matrix-post-processing" class="table-of-contents__link toc-highlight">DCP project matrix post-processing</a></li><li><a href="#post-processing-steps" class="table-of-contents__link toc-highlight">Post-processing steps</a></li><li><a href="#making-project-matrix-cell-barcodes-unique" class="table-of-contents__link toc-highlight">Making project matrix cell barcodes unique</a></li><li><a href="#mapping-dcp-project-matrix-data-to-the-metadata-manifest" class="table-of-contents__link toc-highlight">Mapping DCP project matrix data to the metadata manifest</a></li><li><a href="#mapping-dcp-project-matrix-data-to-the-contributor-matrices" class="table-of-contents__link toc-highlight">Mapping DCP project matrix data to the contributor matrices</a></li></ul></li><li><a href="#brain-initiative-cell-census-network-processing" class="table-of-contents__link toc-highlight">Brain Initiative Cell Census Network Processing</a><ul><li><a href="#optimus-reference-files-for-biccn-data-processing" class="table-of-contents__link toc-highlight">Optimus reference files for BICCN data processing</a></li></ul></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2024 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.728576f2.js"></script>
<script src="/warp/assets/js/main.868095ac.js"></script>
</body>
</html>