<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Pipelines/Optimus_Pipeline/Loom_schema">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">Optimus Count Matrix Overview | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Optimus Count Matrix Overview | WARP"><meta data-rh="true" name="description" content="The Loom matrix is deprecated and the default matrix is now h5ad."><meta data-rh="true" property="og:description" content="The Loom matrix is deprecated and the default matrix is now h5ad."><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.a2ee2c75.js" as="script">
<link rel="preload" href="/warp/assets/js/main.435ed8cb.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/ATAC/README">ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/BuildIndices_Pipeline/README">BuildIndices</a><button aria-label="Toggle the collapsible sidebar category &#x27;BuildIndices&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README">Exome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Exome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/README">Illumina Genotyping Array</a><button aria-label="Toggle the collapsible sidebar category &#x27;Illumina Genotyping Array&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/ImputationBeagle_Pipeline/README">ImputationBeagle</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Imputation_Pipeline/README">Imputation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Imputation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/JointGenotyping_Pipeline/README">JointGenotyping</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Multiome_Pipeline/README">Multiome scATAC and GEX</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus</a><button aria-label="Toggle the collapsible sidebar category &#x27;Optimus&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/optimus.methods">Optimus v7.8.1 Methods</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/Bam_tags">What tags are included in an Optimus BAM file?</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/Library-metrics">Optimus Library-level metrics</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing">Consortia Data Processing</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/starsolo-metrics">STAR Aligner Metrics</a></li></ul></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/PairedTag_Pipeline/README">Paired-Tag</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README">RNA with UMIs</a><button aria-label="Toggle the collapsible sidebar category &#x27;RNA with UMIs&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/SlideSeq_Pipeline/README">Slide-seq</a><button aria-label="Toggle the collapsible sidebar category &#x27;Slide-seq&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/SlideTags_Pipeline/README">Slide-tags</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README">Smart-seq2 Single Nucleus Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Nucleus Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/snm3C/README">Single Nucleus Methyl-Seq and Chromatin Capture</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Nucleus Methyl-Seq and Chromatin Capture&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Ultima_Genomics_Whole_Genome_Germline_Pipeline/README">Ultima Genomics Whole Genome Germline</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/README">Whole Genome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Whole Genome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">WARP Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/warp/docs/Pipelines/Optimus_Pipeline/README"><span itemprop="name">Optimus</span></a><meta itemprop="position" content="2"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Optimus Count Matrix Overview</span><meta itemprop="position" content="3"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Optimus Count Matrix Overview</h1><div class="theme-admonition theme-admonition-warning alert alert--danger admonition_LlT9"><div class="admonitionHeading_tbUL"><span class="admonitionIcon_kALy"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M5.05.31c.81 2.17.41 3.38-.52 4.31C3.55 5.67 1.98 6.45.9 7.98c-1.45 2.05-1.7 6.53 3.53 7.7-2.2-1.16-2.67-4.52-.3-6.61-.61 2.03.53 3.33 1.94 2.86 1.39-.47 2.3.53 2.27 1.67-.02.78-.31 1.44-1.13 1.81 3.42-.59 4.78-3.42 4.78-5.56 0-2.84-2.53-3.22-1.25-5.61-1.52.13-2.03 1.13-1.89 2.75.09 1.08-1.02 1.8-1.86 1.33-.67-.41-.66-1.19-.06-1.78C8.18 5.31 8.68 2.45 5.05.32L5.03.3l.02.01z"></path></svg></span>danger</div><div class="admonitionContent_S0QG"><p>The Loom matrix is deprecated and the default matrix is now h5ad.</p></div></div><p>The Optimus pipeline&#x27;s default count matrix output is a h5ad file generated using <a href="https://anndata.readthedocs.io/en/latest/index.html" target="_blank" rel="noopener noreferrer">AnnData</a>. </p><p>It contains the raw, but UMI-corrected cell by gene counts, which vary depending on the workflow&#x27;s <code>counting_mode</code> and <code>count_exons</code> parameters. If running single-cell data (<code>counting_mode</code> is <code>sc_rna</code>), the counts will include only exonic gene counts. If running single-nucleus data (<code>counting_mode</code> is <code>sn_rna</code>), the counts will be whole transcript counts. Additionally, if <code>count_exons</code> is set to <code>true</code> in <code>sn_rna</code> mode, the h5ad will contain the whole transcript counts as well as an additional layer with exonic counts.</p><p>You can determine which type of counts are in the h5ad file by looking at the unstructured metadata (the <code>anndata.uns</code> property of the matrix) <code>expression_data_type</code> key (see <a href="#table-1-global-attributes">Table 1</a> below).</p><p>The matrix also contains multiple metrics for both individual cells (the <code>anndata.obs</code> property of the matrix; <a href="#table-2-cell-metrics">Table 2</a>) and individual genes (the <code>anndata.var</code> property of the matrix; <a href="#table-3-gene-metrics">Table 3</a>). </p><div class="theme-admonition theme-admonition-tip alert alert--success admonition_LlT9"><div class="admonitionHeading_tbUL"><span class="admonitionIcon_kALy"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Additional Matrix Processing for Consortia</div><div class="admonitionContent_S0QG"><p>Previous Loom files generated by Optimus for consortia, such as the Human Cell Atlas (HCA) or the BRAIN Initiative Cell Census Network (BICCN), may have additional processing steps. Read the <a href="/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing#hca-data-coordination-platform-matrix-processing">Consortia Processing Overview</a> for details on consortia-specific matrix changes.</p></div></div><h2 class="anchor anchorWithStickyNavbar_LWe7" id="table-1-global-attributes">Table 1. Global attributes<a href="#table-1-global-attributes" class="hash-link" aria-label="Direct link to Table 1. Global attributes" title="Direct link to Table 1. Global attributes">​</a></h2><p>The global attributes (unstuctured metadata) in the h5ad apply to the whole file, not any specific part. </p><table><thead><tr><th align="left">Attribute</th><th align="left">Details</th></tr></thead><tbody><tr><td align="left"><code>expression_data_type</code></td><td align="left">String describing if the pipeline counts exonic or whole transcript (exonic and intronic) reads. For the single-cell mode (<code>counting_mode = sc_rna</code>), the value will be <code>exonic</code>; for the single-nucleus mode (<code>counting_mode = sn_rna</code>), the value will be <code>whole_transcript</code>.</td></tr><tr><td align="left"><code>input_id</code></td><td align="left">The sample or cell ID listed in the pipeline configuration file. This can be any string, but we recommend it be consistent with any sample metadata.</td></tr><tr><td align="left"><code>input_name</code></td><td align="left">Optional string that can be used to further describe the input.</td></tr><tr><td align="left"><code>input_id_metadata_field</code></td><td align="left">Optional string that describes, when applicable, the metadata field containing the <code>input_id</code>.</td></tr><tr><td align="left"><code>input_name_metadata_field</code></td><td align="left">Optional string that describes, when applicable, the metadata field containing the <code>input_name</code>.</td></tr><tr><td align="left"><code>pipeline_version</code></td><td align="left">String describing the version of the Optimus pipeline run on the data.</td></tr><tr><td align="left"><code>NHashID</code></td><td align="left">String that represents NHashID (an optional library aliquot identifier) if specified during the worfklow run.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="table-2-cell-metrics">Table 2. Cell metrics<a href="#table-2-cell-metrics" class="hash-link" aria-label="Direct link to Table 2. Cell metrics" title="Direct link to Table 2. Cell metrics">​</a></h2><table><thead><tr><th align="left">Cell Metrics</th><th align="center">Program</th><th align="left">Details</th></tr></thead><tbody><tr><td align="left"><code>CellID</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The unique identifier for each cell based on cell barcodes (sequences used to identify unique cells); identical to <code>cell_names</code>. Learn more about cell barcodes in the <a href="#definitions">Definitions</a> section below.</td></tr><tr><td align="left"><code>cell_names</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The unique identifier for each cell based on cell barcodes; identical to <code>CellID</code>.</td></tr><tr><td align="left"><code>input_id</code></td><td align="center">Provided as pipeline input</td><td align="left">The sample or cell ID listed in the pipeline configuration file. This can be any string, but we recommend it be consistent with any sample metadata.</td></tr><tr><td align="left"><code>star_IsCell</code></td><td align="center">STARsolo</td><td align="left">A true/false flag demarcating if the STARsolo aligner called a cell barcode as a cell.</td></tr><tr><td align="left"><code>n_reads</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of reads associated with the cell. Like all metrics, <code>n_reads</code> is calculated from the Optimus output BAM file. Prior to alignment, reads are checked against the whitelist and any within one edit distance (Hamming distance) are corrected. These CB-corrected reads are aligned using STARsolo, where they get further CB correction. For this reason, most reads in the aligned BAM file have both <code>CB</code> and <code>UB</code> tags. Therefore, <code>n_reads</code> represents CB-corrected reads, rather than all reads in the input FASTQ files.</td></tr><tr><td align="left"><code>tso_reads</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of reads that have 20 or more bp of TSO sequence clipped from the 5&#x27; end. Calculated using the first number of cN tag in the BAM, which is specific to the number of TSO nucleotides clipped.</td></tr><tr><td align="left"><code>noise_reads</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">Number of reads that are categorized by 10x Genomics Cell Ranger as &quot;noise&quot;. Refers to long polymers, or reads with high numbers of N (ambiguous) nucleotides.</td></tr><tr><td align="left"><code>perfect_molecule_barcodes</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of reads with molecule barcodes (sequences used to identify unique transcripts) that have no errors. Learn more about UMIs in the <a href="#definitions">Definitions</a> section below.</td></tr><tr><td align="left"><code>reads_mapped_exonic</code></td><td align="center">STARsolo and <a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of unique reads counted as exon; counted when BAM file&#x27;s <code>sF</code> tag is assigned to <code>1</code> or <code>3</code> and the <code>NH:i</code> tag is <code>1</code>; mitochondrial reads are excluded.</td></tr><tr><td align="left"><code>reads_mapped_exonic_as</code></td><td align="center">STARsolo and <a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of reads counted as exon in the antisense direction; counted when the BAM&#x27;s <code>sF</code> is assigned to a <code>2</code> or <code>4</code> and the <code>NH:i</code> tag is <code>1</code>; mitochondrial reads are excluded.</td></tr><tr><td align="left"><code>reads_mapped_intronic</code></td><td align="center">STARsolo and <a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of unique reads counted as intron; counted when the BAM files&#x27;s <code>sF</code> tag is assigned to a <code>5</code> and the <code>NH:i</code> tag is <code>1</code>; mitochondrial reads are excluded.</td></tr><tr><td align="left"><code>reads_mapped_intronic_as</code></td><td align="center">STARsolo and <a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of unique reads counted as intron in the antisense direction; counted when the BAM file&#x27;s <code>sF</code> tag is assigned to a <code>6</code> and the <code>NH:i</code> tas is <code>1</code>; mitochondrial reads are excluded.</td></tr><tr><td align="left"><code>duplicate_reads</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">Not currently calculated for Optimus output; number of duplicate reads.</td></tr><tr><td align="left"><code>n_mitochondrial_genes</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of mitochondrial genes detected by this cell.</td></tr><tr><td align="left"><code>n_mitochondrial_molecules</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of molecules from mitochondrial genes detected for this cell.</td></tr><tr><td align="left"><code>pct_mitochondrial_molecules</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The percentage of molecules from mitochondrial genes detected for this cell.</td></tr><tr><td align="left"><code>reads_mapped_uniquely</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of reads mapped to a single unambiguous location in the genome; mitochondrial reads are excluded.</td></tr><tr><td align="left"><code>reads_mapped_multiple</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of reads mapped to multiple genomic positions with equal confidence; mitochondrial reads are excluded.</td></tr><tr><td align="left"><code>spliced_reads</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of reads that overlap splicing junctions.</td></tr><tr><td align="left"><code>antisense_reads</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">Not calculated for Optimus outputs; see <code>reads_mapped_exonic_as</code> or <code>reads_mapped_intronic_as</code> for antisense counts.</td></tr><tr><td align="left"><code>molecule_barcode_fraction_bases_above_30_mean</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The average fraction of bases in molecule barcodes that receive quality scores greater than 30 across the reads of the cell.</td></tr><tr><td align="left"><code>molecule_barcode_fraction_bases_above_30_variance</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The variance in the fraction of bases in molecule barcodes that receive quality scores greater than 30 across the reads of the cell.</td></tr><tr><td align="left"><code>genomic_reads_fraction_bases_quality_above_30_mean</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The average fraction of bases in the genomic read that receive quality scores greater than 30 across the reads of the cell (included for 10x Cell Ranger count comparison).</td></tr><tr><td align="left"><code>genomic_reads_fraction_bases_quality_above_30_variance</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The variance in the fraction of bases in the genomic read that receive quality scores greater than 30 across the reads of the cell (included for 10x Cell Ranger count comparison).</td></tr><tr><td align="left"><code>genomic_read_quality_mean</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">Average quality of Illumina base calls in the genomic reads corresponding to the cell.</td></tr><tr><td align="left"><code>genomic_read_quality_variance</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">Variance in quality of Illumina base calls in the genomic reads corresponding to the cell.</td></tr><tr><td align="left"><code>n_molecules</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">Number of molecules corresponding to the cell (only reflects reads with CB and UB tags).</td></tr><tr><td align="left"><code>n_fragments</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">Number of fragments (distinct segments of reads that align to a specific location on the reference genome), corresponding to the cell barcode. Learn more in the <a href="#definitions">Definitions</a> section below.</td></tr><tr><td align="left"><code>reads_per_fragment</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The average number of reads associated with each fragment in the cell.</td></tr><tr><td align="left"><code>fragments_per_molecule</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The average number of fragments associated with each molecule in the cell.</td></tr><tr><td align="left"><code>fragments_with_single_read_evidence</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of fragments associated with the cell that are observed by only one read.</td></tr><tr><td align="left"><code>molecules_with_single_read_evidence</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of molecules associated with the cell that are observed by only one read.</td></tr><tr><td align="left"><code>reads_mapped_mitochondrial</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number unique reads (NH:i:1 BAM tag) that come from mitochondrial genes.</td></tr><tr><td align="left"><code>perfect_cell_barcodes</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of reads whose cell barcodes contain no error.</td></tr><tr><td align="left"><code>reads_mapped_too_many_loci</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of reads that were mapped to too many loci across the genome and as a consequence, are reported unmapped by the aligner.</td></tr><tr><td align="left"><code>cell_barcode_fraction_bases_above_30_variance</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The variance of the fraction of Illumina base calls for the cell barcode sequence that are greater than 30, across molecules.</td></tr><tr><td align="left"><code>cell_barcode_fraction_bases_above_30_mean</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The average fraction of Illumina base calls for the cell barcode sequences that are greater than 30, across molecules.</td></tr><tr><td align="left"><code>n_genes</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of genes detected by this cell.</td></tr><tr><td align="left"><code>genes_detected_multiple_observations</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of genes that are observed by more than one read in this cell.</td></tr><tr><td align="left"><code>emptydrops_FDR</code></td><td align="center"><a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" target="_blank" rel="noopener noreferrer">dropletUtils</a></td><td align="left">False Discovery Rate (FDR) for being a non-empty droplet; single-cell data will read <code>NA</code> if task is unable to detect knee point inflection. Column is not included for data run in the <code>sn_rna</code> mode.</td></tr><tr><td align="left"><code>emptydrops_IsCell</code></td><td align="center"><a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" target="_blank" rel="noopener noreferrer">dropletUtils</a></td><td align="left">Binarized call of cell/background based on predefined FDR cutoff; single-cell data will read <code>NA</code> if task is unable to detect knee point inflection. Column is not included for data run in the <code>sn_rna</code> mode</td></tr><tr><td align="left"><code>emptydrops_Limited</code></td><td align="center"><a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" target="_blank" rel="noopener noreferrer">dropletUtils</a></td><td align="left">Indicates whether a lower p-value could be obtained by increasing the number of iterations; single-cell data will read <code>NA</code> if task is unable to detect knee point inflection. Column is not included for data run in the <code>sn_rna</code> mode.</td></tr><tr><td align="left"><code>emptydrops_LogProb</code></td><td align="center"><a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" target="_blank" rel="noopener noreferrer">dropletUtils</a></td><td align="left">The log-probability of observing the barcode’s count vector under the null model; single-cell data will read <code>NA</code> if the task is unable to detect knee point inflection. Column is not included for data run in the <code>sn_rna</code> mode.</td></tr><tr><td align="left"><code>emptydrops_PValue</code></td><td align="center"><a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" target="_blank" rel="noopener noreferrer">dropletUtils</a></td><td align="left">The Monte Carlo p-value against the null model; single-cell data will read <code>NA</code> if task is unable to detect knee point inflection. Column is not included for data run in the <code>sn_rna</code> mode</td></tr><tr><td align="left"><code>emptydrops_Total</code></td><td align="center"><a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" target="_blank" rel="noopener noreferrer">dropletUtils</a></td><td align="left">The total read counts for each barcode; single-cell data will read <code>NA</code> if task is unable to detect knee point inflection. Column is not included for data run in the <code>sn_rna</code> mode.</td></tr><tr><td align="left"><code>reads_mapped_intergenic</code></td><td align="center">STARsolo and <a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The number of reads counted as intergenic; counted when the BAM file&#x27;s <code>sF</code> tag is assigned to a <code>7</code> and the <code>NH:i</code> tag is <code>1</code>.</td></tr><tr><td align="left"><code>reads_unmapped</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The total number of reads that are unmapped; counted when the BAM file&#x27;s <code>sF</code> tag is <code>0</code>.</td></tr><tr><td align="left"><code>reads_per_molecule</code></td><td align="center"><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td align="left">The average number of reads associated with each molecule in the cell.</td></tr><tr><td align="left"><code>doublet_score</code></td><td align="center">Modified version of <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6853612/" target="_blank" rel="noopener noreferrer">DoubletFinder</a></td><td align="left">A score produced by a modified version of the DoubletFinder software that normalizes data using scanpy and then uses the k-nearest neighbors algorithm to determine cells. This program is non-deterministic, so results will vary across runs of the workflow. The metrics are used to determine overall library quality.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="table-3-gene-metrics">Table 3. Gene metrics<a href="#table-3-gene-metrics" class="hash-link" aria-label="Direct link to Table 3. Gene metrics" title="Direct link to Table 3. Gene metrics">​</a></h2><table><thead><tr><th>Gene Metrics</th><th>Program</th><th>Details</th></tr></thead><tbody><tr><td><code>ensembl_ids</code></td><td><a href="https://www.gencodegenes.org/" target="_blank" rel="noopener noreferrer">GENCODE GTF</a></td><td>The <code>gene_id</code> listed in the GENCODE GTF file.</td></tr><tr><td><code>Gene</code></td><td><a href="https://www.gencodegenes.org/" target="_blank" rel="noopener noreferrer">GENCODE GTF</a></td><td>The unique <code>gene_name</code> provided in the GENCODE GTF file; identical to the <code>gene_names</code> attribute.</td></tr><tr><td><code>gene_names</code></td><td><a href="https://www.gencodegenes.org/" target="_blank" rel="noopener noreferrer">GENCODE GTF</a></td><td>The unique <code>gene_name</code> provided in the GENCODE GTF file; identical to the <code>Gene</code> attribute.</td></tr><tr><td><code>n_reads</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of reads associated with this gene.</td></tr><tr><td><code>tso_reads</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of reads that have 20 or more bp of TSO sequence clipped from the 5&#x27; end. Calculated using the first number of cN tag in the BAM, which is specific to the number of TSO nucleotides clipped.</td></tr><tr><td><code>noise_reads</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>Not currently calculated for Optimus output; number of reads that are categorized by 10x Genomics Cell Ranger as &quot;noise&quot;; refers to long polymers, or reads with high numbers of N (ambiguous) nucleotides.</td></tr><tr><td><code>perfect_molecule_barcodes</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of reads with molecule barcodes (sequences used to identify unique transcripts) that have no errors. Learn more about UMIs in the <a href="#definitions">Definitions</a> section below.</td></tr><tr><td><code>reads_mapped_exonic</code></td><td>STARsolo and <a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of unique reads counted as exon; counted when BAM file&#x27;s <code>sF</code> tag is assigned to <code>1</code> or <code>3</code> and the <code>NH:i</code> tag is <code>1</code>; mitochondrial reads are excluded.</td></tr><tr><td><code>reads_mapped_exonic_as</code></td><td>STARsolo and <a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of reads counted as exon in the antisense direction; counted when the BAM file&#x27;s <code>sF</code> tag is assigned to a <code>2</code> or <code>4</code> and the <code>NH:i</code> tag is <code>1</code>; mitochondrial reads are excluded.</td></tr><tr><td><code>reads_mapped_intronic</code></td><td>STARsolo and <a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of reads counted as intron; counted when the BAM file&#x27;s <code>sF</code> tag is assigned to a <code>5</code> and the <code>NH:i</code> tag is <code>1</code>; mitochondrial reads are excluded.</td></tr><tr><td><code>reads_mapped_intronic_as</code></td><td>STARsolo and <a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of reads counted as intron in the antisense direction; counted when the BAM file&#x27;s <code>sF</code> tag is assigned to a <code>6</code> and the <code>NH:i</code> tag is <code>1</code>; mitochondrial reads are excluded.</td></tr><tr><td><code>reads_mapped_uniquely</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of reads mapped to a single unambiguous location in the genome; mitochondrial reads are excluded.</td></tr><tr><td><code>reads_mapped_multiple</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of reads mapped to multiple genomic positions with equal confidence; mitochondrial reads are excluded.</td></tr><tr><td><code>spliced_reads</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of reads that overlap splicing junctions.</td></tr><tr><td><code>antisense_reads</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of reads that are mapped to the antisense strand instead of the transcribed strand.</td></tr><tr><td><code>duplicate_reads</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of duplicate reads.</td></tr><tr><td><code>molecule_barcode_fraction_bases_above_30_mean</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The average fraction of bases in molecule barcodes that receive quality scores greater than 30 across the reads of the gene.</td></tr><tr><td><code>molecule_barcode_fraction_bases_above_30_variance</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The variance in the fraction of bases in molecule barcodes that receive quality scores greater than 30 across the reads of the gene.</td></tr><tr><td><code>genomic_reads_fraction_bases_quality_above_30_mean</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The average fraction of bases in the genomic read that receive quality scores greater than 30 across the reads of the gene (included for 10x Cell Ranger count comparison).</td></tr><tr><td><code>genomic_reads_fraction_bases_quality_above_30_variance</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The variance in the fraction of bases in the genomic read that receive quality scores greater than 30 across the reads of the gene (included for 10x Cell Ranger count comparison).</td></tr><tr><td><code>genomic_read_quality_mean</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>Average quality of Illumina base calls in the genomic reads corresponding to the gene.</td></tr><tr><td><code>genomic_read_quality_variance</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>Variance in quality of Illumina base calls in the genomic reads corresponding to the gene.</td></tr><tr><td><code>n_molecules</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>Number of molecules corresponding to the gene (only reflects reads with <code>CB</code> and <code>UB</code> tags).</td></tr><tr><td><code>n_fragments</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>Number of fragments (distinct segments of reads that align to a specific location on the reference genome), corresponding to the gene. Learn more in the <a href="#definitions">Definitions</a> section below.</td></tr><tr><td><code>reads_per_molecule</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The average number of reads associated with each molecule in the gene.</td></tr><tr><td><code>reads_per_fragment</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The average number of reads associated with each fragment in the gene.</td></tr><tr><td><code>fragments_per_molecule</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The average number of fragments associated with each molecule in the gene.</td></tr><tr><td><code>fragments_with_single_read_evidence</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of fragments associated with the gene that are observed by only one read.</td></tr><tr><td><code>molecules_with_single_read_evidence</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of molecules associated with the gene that are observed by only one read.</td></tr><tr><td><code>reads_mapped_mitochondrial</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number unique reads (NH:i:1 BAM tag) that come from mitochondrial genes.</td></tr><tr><td><code>number_cells_detected_multiple</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of cells which observe more than one read of the gene.</td></tr><tr><td><code>number_cells_expressing</code></td><td><a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a></td><td>The number of cells that detect the gene.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="definitions">Definitions<a href="#definitions" class="hash-link" aria-label="Direct link to Definitions" title="Direct link to Definitions">​</a></h2><ul><li>Cell Barcode: Short nucleotide sequence used to label and distinguish which reads come from each unique cell, allowing for tracking of many cells simultaneously.</li><li>Fragment: A distinct segment of a read that aligns to a specific location on the reference genome. The <a href="https://github.com/broadinstitute/warp-tools/tree/develop/tools/TagSort" target="_blank" rel="noopener noreferrer">TagSort</a> function defines fragments based on:  1) the presence of a combined <code>UMI/GX/CB</code> tag, 2) the reference (Chr1, Chr2, etc.), 3) the nuleotide position, and 4) the strand. While some cells may have more <code>n_fragments</code> than <code>n_reads</code> (for example, when an RNA read overlaps an exon-exon junction), some barcodes may have fewer fragments than reads (for example, when a cell has multiple reads that overlap).</li><li>Unique Molecular Identifier (UMI): Short nucleotide sequence used to label and distinguish which reads come from each unique transcript present in the cell at the time of lysis, allowing for tracking of many transcripts simultaneously.</li></ul></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/Optimus_Pipeline/Loom_schema.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2026-02-11T20:27:07.000Z">Feb 11, 2026</time></b> by <b>jsotobroad</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/Pipelines/Optimus_Pipeline/Bam_tags"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">What tags are included in an Optimus BAM file?</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/Pipelines/Optimus_Pipeline/Library-metrics"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Optimus Library-level metrics</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#table-1-global-attributes" class="table-of-contents__link toc-highlight">Table 1. Global attributes</a></li><li><a href="#table-2-cell-metrics" class="table-of-contents__link toc-highlight">Table 2. Cell metrics</a></li><li><a href="#table-3-gene-metrics" class="table-of-contents__link toc-highlight">Table 3. Gene metrics</a></li><li><a href="#definitions" class="table-of-contents__link toc-highlight">Definitions</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2026 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.a2ee2c75.js"></script>
<script src="/warp/assets/js/main.435ed8cb.js"></script>
</body>
</html>