<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Pipelines/ATAC/README">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">ATAC Overview | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/ATAC/README"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="ATAC Overview | WARP"><meta data-rh="true" name="description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><meta data-rh="true" property="og:description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/ATAC/README"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/ATAC/README" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/ATAC/README" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.f756d810.js" as="script">
<link rel="preload" href="/warp/assets/js/main.affc3f00.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible menu__list-item-collapsible--active"><a class="menu__link menu__link--sublist menu__link--active" aria-current="page" aria-expanded="true" tabindex="0" href="/warp/docs/Pipelines/ATAC/README">ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/ATAC/count-matrix-overview">ATAC Count Matrix Overview</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/ATAC/library-metrics">ATAC Library Metrics Overview</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/ATAC/atac.methods">ATAC v2.3.1 Methods</a></li></ul></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/BuildIndices_Pipeline/README">BuildIndices</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README">Exome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Exome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/README">Illumina Genotyping Array</a><button aria-label="Toggle the collapsible sidebar category &#x27;Illumina Genotyping Array&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/ImputationBeagle_Pipeline/README">ImputationBeagle</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Imputation_Pipeline/README">Imputation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Imputation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/JointGenotyping_Pipeline/README">JointGenotyping</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Multiome_Pipeline/README">Multiome scATAC and GEX</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus</a><button aria-label="Toggle the collapsible sidebar category &#x27;Optimus&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/PairedTag_Pipeline/README">Paired-Tag</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README">RNA with UMIs</a><button aria-label="Toggle the collapsible sidebar category &#x27;RNA with UMIs&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/SlideSeq_Pipeline/README">Slide-seq</a><button aria-label="Toggle the collapsible sidebar category &#x27;Slide-seq&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/SlideTags_Pipeline/README">Slide-tags</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README">Smart-seq2 Single Nucleus Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Nucleus Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/snm3C/README">Single Nucleus Methyl-Seq and Chromatin Capture</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Nucleus Methyl-Seq and Chromatin Capture&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Ultima_Genomics_Whole_Genome_Germline_Pipeline/README">Ultima Genomics Whole Genome Germline</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/README">Whole Genome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Whole Genome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">WARP Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">ATAC</span><meta itemprop="position" content="2"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>ATAC Overview</h1><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">2.5.1</a></td><td align="center">November, 2024</td><td align="center">WARP Pipelines</td><td align="center">Please <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">file an issue in WARP</a>.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="introduction-to-the-atac-workflow">Introduction to the ATAC workflow<a href="#introduction-to-the-atac-workflow" class="hash-link" aria-label="Direct link to Introduction to the ATAC workflow" title="Direct link to Introduction to the ATAC workflow">​</a></h2><p>ATAC is an open-source, cloud-optimized pipeline developed in collaboration with members of the <a href="https://braininitiative.nih.gov/" target="_blank" rel="noopener noreferrer">BRAIN Initiative</a> (BICCN and <a href="https://brainblog.nih.gov/brain-blog/brain-issues-suite-funding-opportunities-advance-brain-cell-atlases-through-centers" target="_blank" rel="noopener noreferrer">BICAN</a> Sequencing Working Group) and <a href="https://nida.nih.gov/about-nida/organization/divisions/division-neuroscience-behavior-dnb/basic-research-hiv-substance-use-disorder/scorch-program" target="_blank" rel="noopener noreferrer">SCORCH</a> (see <a href="#acknowledgements">Acknowledgements</a> below). It supports the processing of 10x single-nucleus data generated with 10x Multiome <a href="https://www.10xgenomics.com/products/single-cell-multiome-atac-plus-gene-expression" target="_blank" rel="noopener noreferrer">ATAC-seq (Assay for Transposase-Accessible Chromatin)</a>, a technique used in molecular biology to assess genome-wide chromatin accessibility. </p><p>This workflow is the ATAC component of the <a href="/warp/docs/Pipelines/Multiome_Pipeline/README">Mutiome wrapper workflow</a>. It corrects cell barcodes (CBs), aligns reads to the genome, and produces a fragment file as well as <a href="/warp/docs/Pipelines/ATAC/count-matrix-overview">per barcode metrics</a> and <a href="/warp/docs/Pipelines/ATAC/library-metrics">library-level metrics</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="quickstart-table">Quickstart table<a href="#quickstart-table" class="hash-link" aria-label="Direct link to Quickstart table" title="Direct link to Quickstart table">​</a></h2><p>The following table provides a quick glance at the ATAC pipeline features:</p><table><thead><tr><th>Pipeline features</th><th>Description</th><th>Source</th></tr></thead><tbody><tr><td>Assay type</td><td>10x single cell or single nucleus ATAC</td><td><a href="https://www.10xgenomics.com" target="_blank" rel="noopener noreferrer">10x Genomics</a></td></tr><tr><td>Overall workflow</td><td>Barcode correction, read alignment, and fragment quantification</td><td>Code available from <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/atac/atac.wdl" target="_blank" rel="noopener noreferrer">GitHub</a></td></tr><tr><td>Workflow language</td><td>WDL 1.0</td><td><a href="https://github.com/openwdl/wdl" target="_blank" rel="noopener noreferrer">openWDL</a></td></tr><tr><td>Genomic Reference Sequence</td><td>GRCh38 human genome primary sequence</td><td>GENCODE</td></tr><tr><td>Aligner</td><td>bwa-mem2</td><td><a href="http://www.ncbi.nlm.nih.gov/pubmed/19451168" target="_blank" rel="noopener noreferrer">Li H. and Durbin R., 2009</a></td></tr><tr><td>Fragment quantification</td><td>SnapATAC2</td><td><a href="https://pubmed.ncbi.nlm.nih.gov/34774128/" target="_blank" rel="noopener noreferrer">Zhang, K. et al., 2021</a></td></tr><tr><td>Data input file format</td><td>File format in which sequencing data is provided</td><td><a href="https://academic.oup.com/nar/article/38/6/1767/3112533" target="_blank" rel="noopener noreferrer">FASTQ</a></td></tr><tr><td>Data output file format</td><td>File formats in which ATAC output is provided</td><td>TSV, h5ad, BAM</td></tr><tr><td>Library-level metrics</td><td>The <a href="/warp/docs/Pipelines/ATAC/README">ATAC</a> pipeline uses <a href="https://github.com/kaizhang/SnapATAC2" target="_blank" rel="noopener noreferrer">SnapATAC2</a> to generate library-level metrics in CSV format.</td><td><a href="/warp/docs/Pipelines/ATAC/library-metrics">Library-level metrics</a></td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="set-up">Set-up<a href="#set-up" class="hash-link" aria-label="Direct link to Set-up" title="Direct link to Set-up">​</a></h2><h3 class="anchor anchorWithStickyNavbar_LWe7" id="atac-installation">ATAC installation<a href="#atac-installation" class="hash-link" aria-label="Direct link to ATAC installation" title="Direct link to ATAC installation">​</a></h3><p>To download the latest ATAC release, see the release tags prefixed with &quot;Multiome&quot; on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a>. All ATAC pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/multiome/atac.changelog.md" target="_blank" rel="noopener noreferrer">ATAC changelog</a>. </p><p>To discover and search releases, use the WARP command-line tool <a href="https://github.com/broadinstitute/warp/tree/master/wreleaser" target="_blank" rel="noopener noreferrer">Wreleaser</a>.</p><p>ATAC can be deployed using <a href="https://cromwell.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH-compliant, flexible workflow management system that supports multiple computing platforms. The workflow can also be run in <a href="https://app.terra.bio" target="_blank" rel="noopener noreferrer">Terra</a>, a cloud-based analysis platform. </p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="input-variables">Input Variables<a href="#input-variables" class="hash-link" aria-label="Direct link to Input Variables" title="Direct link to Input Variables">​</a></h2><p>The following describes the inputs of the ATAC workflow. For more details on how default inputs are set for the Multiome workflow, see the <a href="/warp/docs/Pipelines/Multiome_Pipeline/README">Multiome overview</a>.</p><table><thead><tr><th>Variable name</th><th>Description</th></tr></thead><tbody><tr><td>read1_fastq_gzipped</td><td>Fastq inputs (array of compressed read 1 FASTQ files).</td></tr><tr><td>read2_fastq_gzipped</td><td>Fastq inputs (array of compressed read 2 FASTQ files containing cellular barcodes).</td></tr><tr><td>read3_fastq_gzipped</td><td>Fastq inputs (array of compressed read 3 FASTQ files).</td></tr><tr><td>input_id</td><td>Output prefix/base name for all intermediate files and pipeline outputs.</td></tr><tr><td>cloud_provider</td><td>String describing the cloud provider that should be used to run the workflow; value should be &quot;gcp&quot; or &quot;azure&quot;.</td></tr><tr><td>preindex</td><td>Boolean used for paired-tag data and not applicable to ATAC data types; default is set to false.</td></tr><tr><td>atac_expected_cells</td><td>Number of cells loaded to create the ATAC library; default is set to 3000.</td></tr><tr><td>tar_bwa_reference</td><td>BWA reference (tar file containing reference fasta and corresponding files).</td></tr><tr><td>num_threads_bwa</td><td>Optional integer defining the number of CPUs per node for the BWA-mem alignment task (default: 128).</td></tr><tr><td>mem_size_bwa</td><td>Optional integer defining the memory size for the BWA-mem alignment task in GB (default: 512).</td></tr><tr><td>cpu_platform_bwa</td><td>Optional string defining the CPU platform for the BWA-mem alignment task (default: &quot;Intel Ice Lake&quot;).</td></tr><tr><td>annotations_gtf</td><td>CreateFragmentFile input variable: GTF file for SnapATAC2 to calculate TSS sites of fragment file.</td></tr><tr><td>chrom_sizes</td><td>CreateFragmentFile input variable: Text file containing chrom_sizes for genome build (i.e., hg38)</td></tr><tr><td>whitelist</td><td>Whitelist file for ATAC cellular barcodes.</td></tr><tr><td>adapter_seq_read1</td><td>TrimAdapters input: Sequence adapter for read 1 fastq.</td></tr><tr><td>adapter_seq_read3</td><td>TrimAdapters input: Sequence adapter for read 3 fastq.</td></tr><tr><td>vm_size</td><td>String defining the Azure virtual machine family for the workflow (default: &quot;Standard_M128s&quot;).</td></tr><tr><td>atac_nhash_id</td><td>String that represents an optional library aliquot identifier. When used, it is echoed in the h5ad unstructured data.</td></tr><tr><td>peak_calling</td><td>Optional boolean used to determine if the ATAC pipeline should run Peak Calling; default is &quot;false&quot;.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="atac-tasks-and-tools">ATAC tasks and tools<a href="#atac-tasks-and-tools" class="hash-link" aria-label="Direct link to ATAC tasks and tools" title="Direct link to ATAC tasks and tools">​</a></h2><p>Overall, the ATAC workflow:</p><ol><li>Identifies optimal parameters for performing CB correction and alignment. </li><li>Corrects CBs and partitions FASTQs by CB.</li><li>Aligns reads.</li><li>Generates a fragment file.</li><li>Calculates per cell barcode fragment metrics.</li></ol><p>The tools each ATAC task employs are detailed in the table below. </p><p>To see specific tool parameters, select the task WDL link in the table; then view the <code>command {}</code> section of the task in the WDL script. To view or use the exact tool software, see the task&#x27;s Docker image which is specified in the task WDL <code># runtime values</code> section as <code>String docker =</code>.</p><table><thead><tr><th>Task name and WDL link</th><th>Tool</th><th>Software</th><th>Description</th></tr></thead><tbody><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/atac/atac.wdl" target="_blank" rel="noopener noreferrer">GetNumSplits</a></td><td>Bash</td><td>Bash</td><td>Uses the virtual machine type to determine the optimal number of FASTQ files for performing the BWA-mem alignment step. This allows BWA-mem to run in parallel on multiple FASTQ files in the subsequent workflow steps.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/skylab/FastqProcessing.wdl" target="_blank" rel="noopener noreferrer">FastqProcessing as SplitFastq</a></td><td>fastqprocess</td><td>custom</td><td>Dynamically selects the correct barcode orientation, corrects cell barcodes, and splits FASTQ files by the optimal number determined in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/atac/atac.wdl" target="_blank" rel="noopener noreferrer">GetNumSplits</a> task. The smaller FASTQ files are grouped by cell barcode with each read having the corrected (CB) and raw barcode (CR) in the read name.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/atac/atac.wdl" target="_blank" rel="noopener noreferrer">TrimAdapters</a></td><td>Cutadapt v4.4</td><td>cutadapt</td><td>Trims adaptor sequences.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/atac/atac.wdl" target="_blank" rel="noopener noreferrer">BWAPairedEndAlignment</a></td><td>bwa-mem2</td><td>mem</td><td>Aligns reads from each set of partitioned FASTQ files to the genome and outputs a BAM with ATAC barcodes in the CB:Z tag.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/atac/atac.wdl" target="_blank" rel="noopener noreferrer">CreateFragmentFile</a></td><td>make_fragment_file, import_data</td><td>SnapATAC2</td><td>Generates a fragment file from the final aligned BAM and outputs per barcode quality metrics in h5ad. A detailed list of these metrics is found in the <a href="/warp/docs/Pipelines/ATAC/count-matrix-overview">ATAC Count Matrix Overview</a>. This task is nondeterministic.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/atac/atac.wdl" target="_blank" rel="noopener noreferrer">PeakCalling</a></td><td>macs3</td><td>SnapATAC2</td><td>Generates two h5ad files (<code>cellbybin.h5ad</code> and <code>cellbypeak.h5ad</code>) from the CreateFragmentFile h5ad output file (<code>metrics.h5ad</code>). The <code>cellbybin.h5ad</code> contains the peak called per cluster in the macs3 unstructured metadata and <code>cellbypeak.h5ad</code> contains the merged peaks and the count matrix of peaks per fragment. A detailed list of these metrics is found in the <a href="/warp/docs/Pipelines/ATAC/count-matrix-overview">ATAC Count Matrix Overview</a>. This task is nondeterministic.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="output-variables">Output variables<a href="#output-variables" class="hash-link" aria-label="Direct link to Output variables" title="Direct link to Output variables">​</a></h2><table><thead><tr><th>Output variable name</th><th>Filename, if applicable</th><th>Output format and description</th></tr></thead><tbody><tr><td>bam_aligned_output</td><td><code>&lt;input_id&gt;</code>.bam</td><td>BAM containing aligned reads from ATAC workflow.</td></tr><tr><td>fragment_file</td><td><code>&lt;input_id&gt;</code>.fragments.sorted.tsv.gz</td><td>Bgzipped TSV containing fragment start and stop coordinates per barcode. In order, the columns are &quot;Chromosome&quot;, &quot;Start&quot;, &quot;Stop&quot;, &quot;ATAC Barcode&quot;, and &quot;Number Reads&quot;.</td></tr><tr><td>snap_metrics</td><td><code>&lt;input_id</code>.metrics.h5ad</td><td>h5ad (Anndata) containing per barcode metrics from <a href="https://github.com/kaizhang/SnapATAC2" target="_blank" rel="noopener noreferrer">SnapATAC2</a>. A detailed list of these metrics is found in the <a href="/warp/docs/Pipelines/ATAC/count-matrix-overview">ATAC Count Matrix Overview</a>.</td></tr><tr><td>library_metrics</td><td><code>&lt;input_id&gt;</code>_<code>&lt;atac_nhash_id&gt;</code>_library_metrics.csv</td><td>CSV file containing library-level metrics. Read more in the <a href="/warp/docs/Pipelines/ATAC/library-metrics">Library Metrics Overview</a></td></tr><tr><td>snap_cellbybin</td><td><code>&lt;input_id&gt;</code>.cellbybin.h5ad</td><td>h5ad (Anndata) containing peaks (called by MACS3) per cluster. <a href="https://github.com/kaizhang/SnapATAC2" target="_blank" rel="noopener noreferrer">SnapATAC2</a>.</td></tr><tr><td>snap_cellbypeak</td><td><code>&lt;input_id&gt;</code>.cellbypeak.h5ad</td><td>h5ad (Anndata) containing merged peaks (called by MACS3) per cluster and count matrix of insertion sites per peak and cell. <a href="https://github.com/kaizhang/SnapATAC2" target="_blank" rel="noopener noreferrer">SnapATAC2</a>.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="versioning-and-testing">Versioning and testing<a href="#versioning-and-testing" class="hash-link" aria-label="Direct link to Versioning and testing" title="Direct link to Versioning and testing">​</a></h2><p>All ATAC pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/multiome/atac.changelog.md" target="_blank" rel="noopener noreferrer">ATAC changelog</a> and tested using <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/skylab/multiome/test_inputs" target="_blank" rel="noopener noreferrer">plumbing and scientific test data</a>. To learn more about WARP pipeline testing, see <a href="https://broadinstitute.github.io/warp/docs/About_WARP/TestingPipelines" target="_blank" rel="noopener noreferrer">Testing Pipelines</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="citing-the-atac-pipeline">Citing the ATAC Pipeline<a href="#citing-the-atac-pipeline" class="hash-link" aria-label="Direct link to Citing the ATAC Pipeline" title="Direct link to Citing the ATAC Pipeline">​</a></h2><p>If you use the ATAC Pipeline in your research, please identify the pipeline in your methods section using the <a href="https://scicrunch.org/resources/data/record/nlx_144509-1/SCR_024656/resolver?q=SCR_024656%2A&amp;l=SCR_024656%2A&amp;i=rrid:scr_024656" target="_blank" rel="noopener noreferrer">ATAC SciCrunch resource identifier</a>.</p><ul><li>Ex: <em>ATAC Pipeline (RRID:SCR_024656)</em></li></ul><p>When citing WARP, please use the following:</p><p>Kylee Degatano, Aseel Awdeh, Robert Sidney Cox III, Wes Dingman, George Grant, Farzaneh Khajouei, Elizabeth Kiernan, Kishori Konwar, Kaylee L Mathews, Kevin Palis, Nikelle Petrillo, Geraldine Van der Auwera, Chengchen (Rex) Wang, Jessica Way. &quot;Warp Analysis Research Pipelines: Cloud-optimized workflows for biological data processing and reproducible analysis.&quot; <em>Bioinformatics</em>, 2025; <a href="https://doi.org/10.1093/bioinformatics/btaf494" target="_blank" rel="noopener noreferrer">https://doi.org/10.1093/bioinformatics/btaf494</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="acknowledgements">Acknowledgements<a href="#acknowledgements" class="hash-link" aria-label="Direct link to Acknowledgements" title="Direct link to Acknowledgements">​</a></h2><p>We are immensely grateful to the members of the BRAIN Initiative (BICAN Sequencing Working Group) and SCORCH for their invaluable and exceptional contributions to this pipeline. Our heartfelt appreciation goes to Alex Dobin, Aparna Bhaduri, Alec Wysoker, Anish Chakka, Brian Herb, Daofeng Li, Fenna Krienen, Guo-Long Zuo, Jeff Goldy, Kai Zhang, Khalid Shakir, Bo Li, Mariano Gabitto, Michael DeBerardine, Mengyi Song, Melissa Goldman, Nelson Johansen, James Nemesh, and Theresa Hodges for their unwavering dedication and remarkable efforts.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/ATAC/README.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2025-10-01T18:42:30.000Z">Oct 1, 2025</time></b> by <b>jsotobroad</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/About_WARP/ForkingWARP"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Forking WARP</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/Pipelines/ATAC/count-matrix-overview"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">ATAC Count Matrix Overview</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-atac-workflow" class="table-of-contents__link toc-highlight">Introduction to the ATAC workflow</a></li><li><a href="#quickstart-table" class="table-of-contents__link toc-highlight">Quickstart table</a></li><li><a href="#set-up" class="table-of-contents__link toc-highlight">Set-up</a><ul><li><a href="#atac-installation" class="table-of-contents__link toc-highlight">ATAC installation</a></li></ul></li><li><a href="#input-variables" class="table-of-contents__link toc-highlight">Input Variables</a></li><li><a href="#atac-tasks-and-tools" class="table-of-contents__link toc-highlight">ATAC tasks and tools</a></li><li><a href="#output-variables" class="table-of-contents__link toc-highlight">Output variables</a></li><li><a href="#versioning-and-testing" class="table-of-contents__link toc-highlight">Versioning and testing</a></li><li><a href="#citing-the-atac-pipeline" class="table-of-contents__link toc-highlight">Citing the ATAC Pipeline</a></li><li><a href="#acknowledgements" class="table-of-contents__link toc-highlight">Acknowledgements</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.f756d810.js"></script>
<script src="/warp/assets/js/main.affc3f00.js"></script>
</body>
</html>