<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.1">
<link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Blog Atom Feed">
<link rel="preconnect" href="https://www.google-analytics.com">
<script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-181048107-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview")</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><title data-react-helmet="true">VCF Overview: Illumina Genotyping Array | WARP</title><meta data-react-helmet="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/Illumina_genotyping_array_spec"><meta data-react-helmet="true" name="docsearch:language" content="en"><meta data-react-helmet="true" name="docsearch:version" content="current"><meta data-react-helmet="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="VCF Overview: Illumina Genotyping Array | WARP"><meta data-react-helmet="true" name="description" content="The Illumina Genotyping Array Pipeline v1.11.0 pipeline produces a VCF (Variant Call Format) output with data processing and sample-specific genotype information. The VCF follows the format listed in the VCF 4.2 specification, but additionally contains fields and attributes that are unique to the Arrays pipeline."><meta data-react-helmet="true" property="og:description" content="The Illumina Genotyping Array Pipeline v1.11.0 pipeline produces a VCF (Variant Call Format) output with data processing and sample-specific genotype information. The VCF follows the format listed in the VCF 4.2 specification, but additionally contains fields and attributes that are unique to the Arrays pipeline."><link data-react-helmet="true" rel="shortcut icon" href="/warp/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/Illumina_genotyping_array_spec"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/Illumina_genotyping_array_spec" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/Illumina_genotyping_array_spec" hreflang="x-default"><link data-react-helmet="true" rel="preconnect" href="https://BH4D9OD16A-dsn.algolia.net" crossorigin="anonymous"><link rel="stylesheet" href="/warp/assets/css/styles.e4a54f34.css">
<link rel="preload" href="/warp/assets/js/runtime~main.8c93c99c.js" as="script">
<link rel="preload" href="/warp/assets/js/main.b32885c2.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#main" class="skipToContent_1oUP shadow--md">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle displayOnlyInLargeViewport_GrZ2 react-toggle--disabled"><div class="react-toggle-track" role="button" tabindex="-1"><div class="react-toggle-track-check"><span class="toggle_71bT">ðŸŒœ</span></div><div class="react-toggle-track-x"><span class="toggle_71bT">ðŸŒž</span></div><div class="react-toggle-thumb"></div></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div><div class="searchBox_1Doo"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div><div class="navbar-sidebar"><div class="navbar-sidebar__brand"><a class="navbar__brand" href="/warp/"><b class="navbar__title">WARP Pipelines</b></a></div><div class="navbar-sidebar__items"><div class="menu"><ul class="menu__list"><li class="menu__list-item"><a class="menu__link navbar__link--active" href="/warp/docs/get-started">Documentation</a></li><li class="menu__list-item"><a class="menu__link" href="/warp/blog">Blog</a></li><li class="menu__list-item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="menu__link header-github-link" aria-label="GitHub repository"></a></li></ul></div></div></div></nav><div class="main-wrapper docs-wrapper doc-page"><div class="docPage_31aa"><aside class="docSidebarContainer_3Kbt"><div class="sidebar_15mo"><nav class="menu menu--responsive thin-scrollbar menu_Bmed" aria-label="Sidebar navigation"><button aria-label="Open menu" aria-haspopup="true" class="button button--secondary button--sm menu__button" type="button"><svg class="sidebarMenuIcon_fgN0" width="24" height="24" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!">About WARP</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/About_WARP/BestPractices">Best Practices for Building Data Processing Pipelines</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/About_WARP/PipelineRequirements">Pipeline Requirements</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/About_WARP/TestingPipelines">Testing Pipelines</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/About_WARP/VersionAndReleasePipelines">Version and Release Pipelines</a></li></ul></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!">WARP Pipelines</a><ul class="menu__list"><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">CEMBA</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">CEMBA Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/CEMBA.methods">CEMBA_v1.1.0 Publication Methods</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Exome Germline Single Sample</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README">Exome Germline Single Sample Overview</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">GDC Whole Genome Somatic Single Sample</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/README">Genomic Data Commons (GDC) Whole Genome Somatic Single Sample Overview</a></li></ul></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!" tabindex="0">Illumina Genotyping Array</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/IlluminaGenotypingArray.documentation">Illumina Genotyping Array Overview</a></li><li class="menu__list-item"><a aria-current="page" class="menu__link menu__link--active active" tabindex="0" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/Illumina_genotyping_array_spec">VCF Overview: Illumina Genotyping Array</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Optimus</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Optimus_Pipeline/optimus.methods">Optimus v4.2.4 Methods</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Optimus_Pipeline/Bam_tags">What tags are included in an Optimus BAM file?</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Single Cell ATAC</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Single_Cell_ATAC_Seq_Pipeline/README">scATAC Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Single_Cell_ATAC_Seq_Pipeline/scatac.methods">scATAC v1.2.0 Methods</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Smart-seq2 Multi-Sample</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Smart-seq2_Multi_Sample_Pipeline/README">Smart-seq2 Multi-Sample Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Smart-seq2_Multi_Sample_Pipeline/smart-seq2.methods">Smart-seq2 Multi-Sample v2.2.1 Publication Methods</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Smart-seq2_Multi_Sample_Pipeline/Loom_schema">What&#x27;s in the Smart-seq2 Multi Sample Pipeline Loom File?</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Smart-seq2 Single Nucleus Multi-Sample</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README">Smart-seq2 Single Nucleus Multi-Sample Overview</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Smart-seq2 Single Nucleus</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Pipeline/README">Smart-seq2 Single Nucleus Overview</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Smart-seq2 Single Sample</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Smart-seq2_Single_Sample_Pipeline/README">Smart-seq2 Single Sample Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Smart-seq2_Single_Sample_Pipeline/Loom_schema">What&#x27;s in the Smart-seq2 Single Sample Pipeline Loom File?</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Whole Genome Germline Single Sample</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/README">Whole Genome Germline Single Sample Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/wgs.methods">Whole Genome Germline Single Sample v2.2.0 Methods</a></li></ul></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!">Contribution Guide</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/contribution/README">Welcome</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="-1">Contribute to WARP</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/contribution/contribute_to_warp/changelog_style">Changelog Style Guide Overview</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="-1">Contribute to WARP Documentation</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/contribution/contribute_to_warp_docs/docsite_maintenance">WARP Documentation Maintenance Guide</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/warp/docs/contribution/contribute_to_warp_docs/doc_style">Documentation Style Guide</a></li></ul></li></ul></li></ul></nav></div></aside><main class="docMainContainer_3ufF"><div class="container padding-top--md padding-bottom--lg docItemWrapper_3FMP"><div class="row"><div class="col docItemCol_3FnS"><div class="docItemContainer_33ec"><article><div class="markdown"><header><h1 class="h1Heading_27L5">VCF Overview: Illumina Genotyping Array</h1></header><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/broad/genotyping/illumina/IlluminaGenotypingArray.wdl" target="_blank" rel="noopener noreferrer">Illumina Genotyping Array Pipeline</a> v1.11.0 pipeline produces a VCF (Variant Call Format) output with data processing and sample-specific genotype information. The VCF follows the format listed in the <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf" target="_blank" rel="noopener noreferrer">VCF 4.2 specification</a>, but additionally contains fields and attributes that are unique to the Arrays pipeline.</p><p>This document describes the Array pipelineâ€™s unique VCF fields and attributes that are not listed in the standard VCF specification. To learn more about the pipeline, see the <a href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/IlluminaGenotypingArray.documentation">Illumina Genotyping Array Pipeline Overview</a>.</p><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>How do I view a VCF file?</h5></div><div class="admonition-content"><p>A VCF is a tab delimited text file that you can open with either a text editor or with a spreadsheet editor (like Microsoft Excel).</p></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2LWZ" id="meta-information-fields"></a>Meta information fields<a class="hash-link" href="#meta-information-fields" title="Direct link to heading">#</a></h2><p>Each VCF has meta information fields with attributes that generally describe the processing of the sample in the array.</p><ul><li>arrayType - Name of the genotyping array that was processed.</li><li>autocallDate - Date that the genotyping array was processed by â€˜autocallâ€™ (aka gencall), the Illumina genotype calling software</li><li>autocallGender - Gender (sex) that autocall determined for the sample processed</li><li>autocallVersion - Version of the autocall/gencall software used</li><li>chipWellBarcode - Chip well barcode (a unique identifier for sample as processed on a specific location on the Illumina genotyping array)</li><li>clusterFile - Cluster file used</li><li>extendedIlluminaManifestVersion - Version of the â€˜extended Illumina manifestâ€™ used by the VCF - generation software.</li><li>extendedManifestFile - File name of the â€˜extended Illumina manifestâ€™ used by the VCF generation software</li><li>fingerprintGender - Gender (sex) determined using an orthogonal fingerprinting technology, populated by an optional parameter used by the VCF generation software</li><li>gtcCallRate - GTC call rate of the sample processed that is generated by the autocall/gencall software and represents the fraction of callable loci that had valid calls</li><li>imagingDate - Creation date for the chip well barcode IDATs (raw image scans)</li><li>manifestFile - Name of the Illumina manifest (.bpm) file used by the VCF generation software</li><li>sampleAlias - Sample name</li></ul><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Illumina-specific meta information</h5></div><div class="admonition-content"><p>In addition to the above attributes, there are several attributes specific to Illumina control values that are not described here: Biotin, DNP, Extension, Hyb, NP, NSB, Restore, String, TargetRemoval.</p></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2LWZ" id="header-line-columns"></a>Header line columns<a class="hash-link" href="#header-line-columns" title="Direct link to heading">#</a></h2><p>Following the meta-information, there are 8 standard header line columns:</p><ol><li>#CHROM</li><li>POS</li><li>ID</li><li>REF</li><li>ALT</li><li>QUAL</li><li>FILTER</li><li>INFO</li></ol><p>These columns are standard to each VCF, but the Illumina Genotyping Array Pipeline has an additional FORMAT column. Pipeline-specific column attributes are described below:</p><h3><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2LWZ" id="filter"></a>FILTER<a class="hash-link" href="#filter" title="Direct link to heading">#</a></h3><ul><li>DUPE - Filter applied if there are multiple rows in the VCF for the same loci and alleles. For example, if there are two or more rows that share the same chromosome, position, ref allele and alternate alleles, all but one of them will have the â€˜DUPEâ€™ filter set</li><li>TRIALLELIC - Filter applied if there is a site at which there are two alternate alleles and neither of them is the same as the reference allele</li><li>ZEROED_OUT_ASSAY - Filter applied if the variant at the site was â€˜zeroed outâ€™ in the Illumina cluster file; typically applied when the calls at the site are intentionally marked as unusual. Genotypes called for sites that are â€˜zeroed outâ€™ will always be no-calls</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2LWZ" id="format-genotype"></a>FORMAT (genotype)<a class="hash-link" href="#format-genotype" title="Direct link to heading">#</a></h3><ul><li>GT - GENOTYPE; standard field described in the <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf" target="_blank" rel="noopener noreferrer">VCF specification</a></li><li>IGC - Illumina GenCall Confidence Score. Measure of the call confidence</li><li>X - Raw X intensity as scanned from the original genotyping array</li><li>Y - Raw Y intensity as scanned from the original genotyping array</li><li>NORMX - Normalized X intensity</li><li>NORMY - Normalized Y intensity</li><li>R - Normalized R Value (one of the polar coordinates after the transformation of NORMX and NORMY)</li><li>THETA - Normalized Theta value (one of the polar coordinates after the transformation of NORMX and NORMY)</li><li>LRR - Log R Ratio</li><li>BAF - B Allele Frequency</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2LWZ" id="info"></a>INFO<a class="hash-link" href="#info" title="Direct link to heading">#</a></h3><p>This column contains attributes specific to the array&#x27;s probe.</p><ul><li>AC - Allele Count in genotypes for each ALT allele; standard field in the VCF specification</li><li>AF - Allele Frequency; standard field in the VCF specification</li><li>AN - Allele Number; standard field in the VCF specification</li><li>ALLELE_A - Allele A as annotated in the Illumina manifest (a *suffix indicates this is the reference allele)</li><li>ALLELE_B - Allele B as annotated in the Illumina manifest (a *suffix indicates this is the reference allele)</li><li>BEADSET_ID - BeadSet ID; an Illumina identifier used for normalization</li><li>GC_SCORE - Illumina GenTrain Score; a quality score describing the probe design</li><li>ILLUMINA_BUILD - Genome Build for the design probe sequence, as annotated in the Illumina manifest</li><li>ILLUMINA_CHR - Chromosome of the design probe sequence as annotated in the Illumina manifest</li><li>ILLUMINA_POS - Position of the design probe sequence (on ILLUMINA_CHR) as annotated in the Illumina manifest</li><li>ILLUMINA_STRAND - Strand for the design probe sequence as annotated in the Illumina manifest</li><li>PROBE_A - Allele A probe sequence as annotated in the Illumina manifest</li><li>PROBE_B - Allele B probe sequence as annotated in the Illumina manifest; only present on strand ambiguous SNPs</li><li>SOURCE - Probe source as annotated in the Illumina manifest</li><li>refSNP - dbSNP rsId for this probe</li></ul><p>The remaining attributes describe the cluster definitions provided in the cluster file used for calling the genotype for a SNP.</p><ul><li>N_AA - Number of AA calls in training set</li><li>N_AB - Number of AB calls in training set</li><li>N_BB - Number of BB calls in training set</li><li>devR_AA - Standard deviation of normalized R for AA cluster</li><li>devR_AB - Standard deviation of normalized R for AB cluster</li><li>devR_BB - Standard deviation of normalized R for BB cluster</li><li>devTHETA_AA - Standard deviation of normalized THETA for AA cluster</li><li>devTHETA_AB - Standard deviation of normalized THETA for AB cluster</li><li>devTHETA_BB - Standard deviation of normalized THETA for BB cluster</li><li>devX_AA - Standard deviation of normalized X for AA cluster</li><li>devX_AB - Standard deviation of normalized X for AB cluster</li><li>devX_BB - Standard deviation of normalized X for BB cluster</li><li>devY_AA - Standard deviation of normalized Y for AA cluster</li><li>devY_AB - Standard deviation of normalized Y for AB cluster</li><li>devY_BB - Standard deviation of normalized Y for BB cluster</li><li>meanR_AA - Mean of normalized R for AA cluster</li><li>meanR_AB - Mean of normalized R for AB cluster</li><li>meanR_BB - Mean of normalized R for BB cluster</li><li>meanTHETA_AA - Mean of normalized THETA for AA cluster</li><li>meanTHETA_AB - Mean of normalized THETA for AB cluster</li><li>meanTHETA_BB - Mean of normalized THETA for BB cluster</li><li>meanX_AA - Mean of normalized X for AA cluster</li><li>meanX_AB - Mean of normalized X for AB cluster</li><li>meanX_BB - Mean of normalized X for BB cluster</li><li>meanY_AA - Mean of normalized Y for AA cluster</li><li>meanY_AB - Mean of normalized Y for AB cluster</li><li>meanY_BB - Mean of normalized Y for BB cluster</li></ul></div><footer class="row docusaurus-mt-lg"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/Illumina_genotyping_array_spec.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_2_ui" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_3DPF">Last updated on <b><time datetime="2021-06-24T20:15:53.000Z">6/24/2021</time></b> by <b>dependabot[bot]</b></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/IlluminaGenotypingArray.documentation"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Â« Illumina Genotyping Array Overview</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/warp/docs/Pipelines/Optimus_Pipeline/README"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Optimus Overview Â»</div></a></div></nav></div></div><div class="col col--3"><div class="tableOfContents_35-E thin-scrollbar"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#meta-information-fields" class="table-of-contents__link">Meta information fields</a></li><li><a href="#header-line-columns" class="table-of-contents__link">Header line columns</a><ul><li><a href="#filter" class="table-of-contents__link">FILTER</a></li><li><a href="#format-genotype" class="table-of-contents__link">FORMAT (genotype)</a></li><li><a href="#info" class="table-of-contents__link">INFO</a></li></ul></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub</a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright Â© 2021 Copyright Â© Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.8c93c99c.js"></script>
<script src="/warp/assets/js/main.b32885c2.js"></script>
</body>
</html>