<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.5">
<link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Blog Atom Feed">
<link rel="preconnect" href="https://www.google-analytics.com">
<script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-181048107-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview")</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><title data-react-helmet="true">One post tagged with &quot;genomes&quot; | WARP</title><meta data-react-helmet="true" property="og:title" content="One post tagged with &quot;genomes&quot; | WARP"><meta data-react-helmet="true" name="twitter:card" content="summary_large_image"><meta data-react-helmet="true" property="og:url" content="https://broadinstitute.github.io/warp/blog/tags/genomes"><meta data-react-helmet="true" name="docsearch:language" content="en"><meta data-react-helmet="true" name="docsearch:docusaurus_tag" content="blog_tags_posts"><link data-react-helmet="true" rel="shortcut icon" href="/warp/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://broadinstitute.github.io/warp/blog/tags/genomes"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/blog/tags/genomes" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/blog/tags/genomes" hreflang="x-default"><link data-react-helmet="true" rel="preconnect" href="https://BH4D9OD16A-dsn.algolia.net" crossorigin="anonymous"><link rel="stylesheet" href="/warp/assets/css/styles.57c5b218.css">
<link rel="preload" href="/warp/assets/js/runtime~main.e306bdb5.js" as="script">
<link rel="preload" href="/warp/assets/js/main.143bdc00.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#" class="skipToContent_1oUP">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a class="navbar__item navbar__link" href="/warp/docs/get-started">Documentation</a><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle toggle_3Zt9 react-toggle--disabled"><div class="react-toggle-track" role="button" tabindex="-1"><div class="react-toggle-track-check"><span class="toggle_71bT">üåú</span></div><div class="react-toggle-track-x"><span class="toggle_71bT">üåû</span></div><div class="react-toggle-thumb"></div></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div><div class="searchBox_1Doo"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper blog-wrapper blog-tags-post-list-page"><div class="container margin-vert--lg"><div class="row"><aside class="col col--3"><nav class="sidebar_2ahu thin-scrollbar" aria-label="Blog recent posts navigation"><div class="sidebarItemTitle_2hhb margin-bottom--md">Recent posts</div><ul class="sidebarItemList_2xAf"><li class="sidebarItem_2UVv"><a class="sidebarItemLink_1RT6" href="/warp/blog/Nov21_ReblockedGVCF">WARP Whole Genome and Exome Pipelines Produce Reblocked GVCFs</a></li></ul></nav></aside><main class="col col--7" itemscope="" itemtype="http://schema.org/Blog"><header class="margin-bottom--xl"><h1>One post tagged with &quot;genomes&quot;</h1><a href="/warp/blog/tags">View All Tags</a></header><article class="margin-bottom--xl" itemprop="blogPost" itemscope="" itemtype="http://schema.org/BlogPosting"><header><h2 class="blogPostTitle_GeHD" itemprop="headline"><a itemprop="url" href="/warp/blog/Nov21_ReblockedGVCF">WARP Whole Genome and Exome Pipelines Produce Reblocked GVCFs</a></h2><div class="blogPostData_291c margin-vert--md"><time datetime="2021-11-09T00:00:00.000Z" itemprop="datePublished">November 9, 2021</time> ¬∑ 3 min read</div><div class="row margin-top--md margin-bottom--sm"><div class="col col--6 authorCol_1R69"><div class="avatar margin-bottom--sm"><a class="avatar__photo-link avatar__photo"><img class="image_1yU8" src="https://avatars.githubusercontent.com/u/14366016?v=4" alt="Liz Kiernan"></a><div class="avatar__intro" itemprop="author" itemscope="" itemtype="https://schema.org/Person"><div class="avatar__name"><a itemprop="url"><span itemprop="name">Liz Kiernan</span></a></div><small class="avatar__subtitle" itemprop="description">Senior Science Writer, Data Sciences Platform</small></div></div></div></div></header><div class="markdown" itemprop="articleBody"><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="november-9-2021"></a>November 9, 2021<a class="hash-link" href="#november-9-2021" title="Direct link to heading">#</a></h4><p>We are announcing the latest release of the Whole Genome and Exome production pipelines, which now produce reblocked GVCFs, a compressed form of GVCF files, as default outputs. </p><p>This is a major release, which in this case means:</p><ol><li>The data output from the pipeline is scientifically different than in previous versions.</li><li>Data output from this pipeline should not be co-analyzed with data from previous versions (for example, in joint calling), without first reblocking the GVCFs. </li><li>Users of the pipeline will see a change in both input and output interfaces.</li></ol><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="reblocking--what-is-it-and-why-use-it"></a>Reblocking ‚Äì what is it and why use it?<a class="hash-link" href="#reblocking--what-is-it-and-why-use-it" title="Direct link to heading">#</a></h2><p>Reblocking is a process of compressing GVCFs that reduces file storage footprint and facilitates joint genotyping by removing alt alleles that do not appear in the called genotype. Normal GVCFs contain blocks of adjacent homozygous reference (homRef) genotype calls that are stored as a single entry when they have the exact same value up to a Phred-scaled confidence of 60. The vast majority of downstream analyses discard homRef genotypes with a confidence of less than 20, so we can compress that data without losing any fidelity in downstream analysis. </p><p>Additionally, because most whole-genome sequencing aims for 30X coverage, most confidences are around 90 or higher. With that, we can represent large intervals of high coverage with a single entry, indicating all high-quality homRef genotypes.</p><p>The reblocked GVCFs are sufficient for multiple downstream applications. While some applications, like population variation surveys, can establish variant frequency using binary (i.e. good/bad) homRef confidences, other applications, like de novo analysis, require more nuance. For the latter kind of applications where the reference status of the parents is critical, the reblocked GVCFs provide multiple intermediate categories for homRef confidences. </p><p>Overall, reblocked GVCFs reduce the storage and I/O required to operate on variant data. This leads to faster, more efficient, and less expensive call sets. It is standard practice at the Broad to reblock GVCFs before joint calling. Now, we are making that step a standard part of the GVCF generation in the single sample pipeline. </p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="how-is-reblocking-implemented-in-the-pipelines"></a>How is reblocking implemented in the pipelines?<a class="hash-link" href="#how-is-reblocking-implemented-in-the-pipelines" title="Direct link to heading">#</a></h2><p>We‚Äôve added reblocking as a default subtask, called ‚ÄúReblock‚Äù, in the WARP <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/broad/dna_seq/germline/variant_calling/VariantCalling.wdl" target="_blank" rel="noopener noreferrer">VariantCalling WDL script</a> which is imported by both the Exome and Whole Genome workflows. The Reblock task performs reblocking by using the GATK tool <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037593171" target="_blank" rel="noopener noreferrer">ReblockGVCF</a>. </p><p>While the reblocking option is turned on by default for any pipeline using the VariantCalling task, you can optionally turn it off by setting the <code>skip_reblocking</code> input in the workflows input configuration file (JSON). For example, if you‚Äôre working with the Exome pipeline, just add the parameter:</p><div class="codeBlockContainer_K1bP"><div class="codeBlockContent_hGly WDL"><pre tabindex="0" class="prism-code language-WDL codeBlock_23N8 thin-scrollbar" style="color:#bfc7d5;background-color:#292d3e"><code class="codeBlockLines_39YC"><span class="token-line" style="color:#bfc7d5"><span class="token plain">ExomeGermlineSingleSample.BamToGvcf.skip_reblocking&quot;: true</span></span></code></pre><button type="button" aria-label="Copy code to clipboard" class="copyButton_Ue-o clean-btn">Copy</button></div></div><p>We recognize that some researchers might want to perform reblocking independent of the Exome or Whole Genome pipelines, so we also created a utility workflow to reblock GVCFs. This pipeline takes in GVCFs produced by the single sample pipelines and outputs equivalently reblocked GVCFs. It is available in both the <a href="https://github.com/broadinstitute/warp/tree/develop/pipelines/broad/dna_seq/germline/joint_genotyping/reblocking" target="_blank" rel="noopener noreferrer">WARP repository</a> and in a Terra workspace, and costs less than $0.02 per genome to run.</p><p>If you‚Äôre curious as to whether your WARP pipeline output is a reblocked GVCF, just look at the filename suffix; a reblocked GVCF will have the <code>.rb.g.vcf</code> ending. </p><p>As always, if you have any questions about working with the updated workflows, <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">file an issue in WARP</a> or reach out to <a href="mailto:kdegatano@broadinstitute.org" target="_blank" rel="noopener noreferrer">Kylee Degatano</a>.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="additional-resources"></a>Additional resources<a class="hash-link" href="#additional-resources" title="Direct link to heading">#</a></h2><p>Learn more about the reblocking tool on the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/4405443600667" target="_blank" rel="noopener noreferrer">GATK Support website</a>.
Learn more about the reblocked GVCFs in the <a href="https://broadinstitute.github.io/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README" target="_blank" rel="noopener noreferrer">Exome</a> and <a href="https://broadinstitute.github.io/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/README" target="_blank" rel="noopener noreferrer">Whole-Genome</a> Overviews.</p></div><footer class="row docusaurus-mt-lg"><div class="col col--9"><b>Tags:</b><ul class="tags_2ga9 padding--none margin-left--sm"><li class="tag_11ep"><a class="tag_1Okp tagRegular_3MiF" href="/warp/blog/tags/warp">warp</a></li><li class="tag_11ep"><a class="tag_1Okp tagRegular_3MiF" href="/warp/blog/tags/wdl">WDL</a></li><li class="tag_11ep"><a class="tag_1Okp tagRegular_3MiF" href="/warp/blog/tags/genomes">genomes</a></li><li class="tag_11ep"><a class="tag_1Okp tagRegular_3MiF" href="/warp/blog/tags/exomes">exomes</a></li><li class="tag_11ep"><a class="tag_1Okp tagRegular_3MiF" href="/warp/blog/tags/germline">germline</a></li><li class="tag_11ep"><a class="tag_1Okp tagRegular_3MiF" href="/warp/blog/tags/joint-calling">joint calling</a></li><li class="tag_11ep"><a class="tag_1Okp tagRegular_3MiF" href="/warp/blog/tags/reblock">reblock</a></li></ul></div><div class="col col--3 text--right"><a aria-label="Read more about WARP Whole Genome and Exome Pipelines Produce Reblocked GVCFs" href="/warp/blog/Nov21_ReblockedGVCF"><b>Read More</b></a></div></footer></article></main></div></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item"><span>GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_3J9K"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></span></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright ¬© 2022 Copyright ¬© Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.e306bdb5.js"></script>
<script src="/warp/assets/js/main.143bdc00.js"></script>
</body>
</html>