<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Pipelines/JointGenotyping/README">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">JointGenotyping Overview | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/JointGenotyping_Pipeline/README"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="JointGenotyping Overview | WARP"><meta data-rh="true" name="description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><meta data-rh="true" property="og:description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/JointGenotyping_Pipeline/README"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/JointGenotyping_Pipeline/README" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/JointGenotyping_Pipeline/README" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.728576f2.js" as="script">
<link rel="preload" href="/warp/assets/js/main.868095ac.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/ATAC/README">ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/BuildIndices_Pipeline/README">BuildIndices</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README">Exome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Exome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/README">Illumina Genotyping Array</a><button aria-label="Toggle the collapsible sidebar category &#x27;Illumina Genotyping Array&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Imputation_Pipeline/README">Imputation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Imputation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/JointGenotyping_Pipeline/README">JointGenotyping</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Multiome_Pipeline/README">Multiome scATAC and GEX</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus</a><button aria-label="Toggle the collapsible sidebar category &#x27;Optimus&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/PairedTag_Pipeline/README">Paired-Tag</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README">RNA with UMIs</a><button aria-label="Toggle the collapsible sidebar category &#x27;RNA with UMIs&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/SlideSeq_Pipeline/README">Slide-seq</a><button aria-label="Toggle the collapsible sidebar category &#x27;Slide-seq&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README">Smart-seq2 Single Nucleus Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Nucleus Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/snm3C/README">Single Nucleus Methyl-Seq and Chromatin Capture</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Nucleus Methyl-Seq and Chromatin Capture&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Ultima_Genomics_Whole_Genome_Germline_Pipeline/README">Ultima Genomics Whole Genome Germline</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/README">Whole Genome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Whole Genome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">WARP Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">JointGenotyping</span><meta itemprop="position" content="2"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>JointGenotyping Overview</h1><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">JointGenotyping_v1.6.10</a></td><td align="center">February, 2024</td><td align="center">Elizabeth Kiernan &amp; Kaylee Mathews</td><td align="center">Please <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">file an issue in WARP</a>.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="introduction-to-the-jointgenotyping-workflow">Introduction to the JointGenotyping workflow<a href="#introduction-to-the-jointgenotyping-workflow" class="hash-link" aria-label="Direct link to Introduction to the JointGenotyping workflow" title="Direct link to Introduction to the JointGenotyping workflow">​</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/broad/dna_seq/germline/joint_genotyping/JointGenotyping.wdl" target="_blank" rel="noopener noreferrer">JointGenotyping workflow</a> is an open-source, cloud-optimized pipeline that implements joint variant calling, filtering, and (optional) fingerprinting.</p><p>The pipeline can be configured to run using one of the following GATK joint genotyping methods:</p><ul><li><strong><a href="https://gatk.broadinstitute.org/hc/en-us/articles/21905118377755" target="_blank" rel="noopener noreferrer">GenotypeGVCFs</a></strong> (default method) performs joint genotyping on GVCF files stored in GenomicsDB and pre-called with HaplotypeCaller.</li><li><strong><a href="https://gatk.broadinstitute.org/hc/en-us/articles/21904951112091" target="_blank" rel="noopener noreferrer">GnarlyGenotyper</a></strong> performs scalable, “quick and dirty” joint genotyping on a set of GVCF files stored in GenomicsDB and pre-called with HaplotypeCaller.</li></ul><p>The pipeline can be configured to run using one of the following GATK variant filtering techniques:</p><ul><li><strong><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531612" target="_blank" rel="noopener noreferrer">Variant Quality Score Recalibration (VQSR)</a></strong> (default method) uses the VariantRecalibrator and ApplyVQSR tools to filter variants according to <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035535932" target="_blank" rel="noopener noreferrer">GATK Best Practices</a>.</li><li><strong>Variant Extract-Train-Score (VETS)</strong> uses the ExtractVariantAnnotations, TrainVariantAnnotationsModel, and ScoreVariantAnnotations tools called in the <a href="https://github.com/broadinstitute/gatk/blob/master/scripts/vcf_site_level_filtering_wdl/JointVcfFiltering.wdl" target="_blank" rel="noopener noreferrer">VETS subworkflow</a> to score variant annotations.</li></ul><p>The pipeline takes in a sample map file listing GVCF files produced by HaplotypeCaller in GVCF mode and produces a filtered VCF file (with index) containing genotypes for all samples present in the input VCF files. All sites that are present in the input VCF file are retained. Filtered sites are annotated as such in the FILTER field. If you are new to VCF files, see the <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf" target="_blank" rel="noopener noreferrer">file type specification</a>.</p><p>The JointGenotyping pipeline can be adapted to run on Microsoft Azure instead of Google Cloud. For more information, see the <a href="https://github.com/broadinstitute/azure-warp-joint-calling" target="_blank" rel="noopener noreferrer">azure-warp-joint-calling GitHub repository</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="set-up">Set-up<a href="#set-up" class="hash-link" aria-label="Direct link to Set-up" title="Direct link to Set-up">​</a></h2><h3 class="anchor anchorWithStickyNavbar_LWe7" id="jointgenotyping-installation-and-requirements">JointGenotyping Installation and Requirements<a href="#jointgenotyping-installation-and-requirements" class="hash-link" aria-label="Direct link to JointGenotyping Installation and Requirements" title="Direct link to JointGenotyping Installation and Requirements">​</a></h3><p>To download the latest JointGenotyping release, see the release tags prefixed with &quot;JointGenotyping&quot; on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a>. All JointGenotyping pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/broad/dna_seq/germline/joint_genotyping/JointGenotyping.changelog.md" target="_blank" rel="noopener noreferrer">JointGenotyping changelog</a>. </p><p>To search releases of this and other pipelines, use the WARP command-line tool <a href="https://github.com/broadinstitute/warp/tree/master/wreleaser" target="_blank" rel="noopener noreferrer">Wreleaser</a>.</p><p>If you’re running a JointGenotyping workflow version prior to the latest release, the accompanying documentation for that release may be downloaded with the source code on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a> (see the folder <code>website/docs/Pipelines/JointGenotyping</code>).</p><p>The JointGenotyping pipeline can be deployed using <a href="https://cromwell.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH-compliant, flexible workflow management system that supports multiple computing platforms. The workflow can also be run in <a href="https://app.terra.bio" target="_blank" rel="noopener noreferrer">Terra</a>, a cloud-based analysis platform. The Terra <a href="https://app.terra.bio/#workspaces/warp-pipelines/Whole-Genome-Analysis-Pipeline" target="_blank" rel="noopener noreferrer">Whole-Genome-Analysis-Pipeline</a> and <a href="https://app.terra.bio/#workspaces/warp-pipelines/Exome-Analysis-Pipeline" target="_blank" rel="noopener noreferrer">Exome-Analysis-Pipeline</a> workspaces contain the JointGenotyping pipeline, as well as workflows for preprocessing, initial variant calling, and sample map generation, workflow configurations, required reference data and other inputs, and example testing data.</p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="inputs">Inputs<a href="#inputs" class="hash-link" aria-label="Direct link to Inputs" title="Direct link to Inputs">​</a></h3><p>The JointGenotyping workflow inputs are specified in JSON configuration files. Example configuration files can be found in the <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/broad/dna_seq/germline/joint_genotyping/test_inputs" target="_blank" rel="noopener noreferrer">test_inputs</a> folder in the WARP repository.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="default-joint-calling-input-descriptions">Default joint calling input descriptions<a href="#default-joint-calling-input-descriptions" class="hash-link" aria-label="Direct link to Default joint calling input descriptions" title="Direct link to Default joint calling input descriptions">​</a></h4><p>The table below describes the pipeline inputs that apply when the pipeline is run with default parameters and uses GenotypeGVCFs for joint calling and VQSR for variant filtering:</p><table><thead><tr><th>Parameter name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>unpadded_intervals_file</td><td>Describes the intervals for which VCF output will be written; exome data will have different captures/targets.</td><td>File</td></tr><tr><td>callset_name</td><td>Identifier for the group of VCF files used for joint calling.</td><td>String</td></tr><tr><td>sample_name_map</td><td>Path to file containing the sample names and the cloud location of the individual GVCF files.</td><td>String</td></tr><tr><td>ref_fasta</td><td>Reference FASTA file used for joint calling; must agree with reference for <code>unpadded_intervals_file</code>.</td><td>File</td></tr><tr><td>ref_fasta_index</td><td>Index for reference FASTA file used for joint calling; must agree with reference for <code>unpadded_intervals_file</code>.</td><td>File</td></tr><tr><td>ref_dict</td><td>Reference dictionary file used for joint calling; must agree with reference for <code>unpadded_intervals_file</code>.</td><td>File</td></tr><tr><td>dbsnp_vcf</td><td>Resource VCF file containing common SNPs and indels used for annotating the VCF file after joint calling.</td><td>File</td></tr><tr><td>dbsnp_vcf_index</td><td>Index for <code>dbsnp_vcf</code>.</td><td>File</td></tr><tr><td>snp_recalibration_tranche_values</td><td>Set of sensitivity levels used when running the pipeline using VQSR; value should match estimated sensitivity of truth resource passed as <code>hapmap_resource_vcf</code> to the <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">SNPsVariantRecalibratorCreateModel</a> and <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">SNPsVariantRecalibrator as SNPsVariantRecalibratorScattered</a> tasks; filter cutoff based on sensitivity to common variants (more sensitivity = more false positives); required when <code>run_vets</code> is “false”.</td><td>Array<!-- -->[String]</td></tr><tr><td>snp_recalibration_annotation_values</td><td>Features used for filtering model (annotations in VCF file); all allele-specific versions.</td><td>Array<!-- -->[String]</td></tr><tr><td>indel_recalibration_tranche_values</td><td>Set of sensitivity levels used when running the pipeline using VQSR; value should match estimated sensitivity of truth resource passed as <code>mills_resource_vcf</code> to the <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">IndelsVariantRecalibrator</a> task; filter cutoff based on sensitivity to common variants (more sensitivity = more false positives); required when <code>run_vets</code> is “false”.</td><td>Array<!-- -->[String]</td></tr><tr><td>indel_recalibration_annotation_values</td><td>Features used for filtering model when running the pipeline using VQSR; required when <code>run_vets</code> is “false”.</td><td>Array<!-- -->[String]</td></tr><tr><td>eval_interval_list</td><td>Subset of the unpadded intervals file used for metrics.</td><td>File</td></tr><tr><td>hapmap_resource_vcf</td><td>Used for SNP variant recalibration; see the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890811" target="_blank" rel="noopener noreferrer">GATK Resource Bundle</a> for more information.</td><td>File</td></tr><tr><td>hapmap_resource_vcf_index</td><td>Used for SNP variant recalibration; see the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890811" target="_blank" rel="noopener noreferrer">GATK Resource Bundle</a> for more information.</td><td>File</td></tr><tr><td>omni_resource_vcf</td><td>Used for SNP recalibration; see the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890811" target="_blank" rel="noopener noreferrer">GATK Resource Bundle</a> for more information.</td><td>File</td></tr><tr><td>omni_resource_vcf_index</td><td>Used for SNP recalibration; see the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890811" target="_blank" rel="noopener noreferrer">GATK Resource Bundle</a> for more information.</td><td>File</td></tr><tr><td>one_thousand_genomes_resource_vcf</td><td>Used for SNP recalibration; see the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890811" target="_blank" rel="noopener noreferrer">GATK Resource Bundle</a> for more information.</td><td>File</td></tr><tr><td>one_thousand_genomes_resource_vcf_index</td><td>Used for SNP recalibration; see the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890811" target="_blank" rel="noopener noreferrer">GATK Resource Bundle</a> for more information.</td><td>File</td></tr><tr><td>mills_resource_vcf</td><td>Used for indel variant recalibration; see the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890811" target="_blank" rel="noopener noreferrer">GATK Resource Bundle</a> for more information.</td><td>File</td></tr><tr><td>mills_resource_vcf_index</td><td>Used for indel variant recalibration; see the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890811" target="_blank" rel="noopener noreferrer">GATK Resource Bundle</a> for more information.</td><td>File</td></tr><tr><td>axiomPoly_resource_vcf</td><td>Used for indel variant recalibration; see the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890811" target="_blank" rel="noopener noreferrer">GATK Resource Bundle</a> for more information.</td><td>File</td></tr><tr><td>axiomPoly_resource_vcf_index</td><td>Used for indel variant recalibration; see the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890811" target="_blank" rel="noopener noreferrer">GATK Resource Bundle</a> for more information.</td><td>File</td></tr><tr><td>dbsnp_resource_vcf</td><td>Optional file used for SNP/indel variant recalibration; set to <code>dbsnp_vcf</code> by default; see the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890811" target="_blank" rel="noopener noreferrer">GATK Resource Bundle</a> for more information.</td><td>File</td></tr><tr><td>dbsnp_resource_vcf_index</td><td>Optional file used for SNP/indel variant recalibration; set to <code>dbsnp_vcf_index</code> by default; see the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890811" target="_blank" rel="noopener noreferrer">GATK Resource Bundle</a> for more information.</td><td>File</td></tr><tr><td>excess_het_threshold</td><td>Optional float used for hard filtering joint calls; phred-scaled p-value; set to <code>54.69</code> by default to cut off quality scores greater than a z-score of -4.5 (p-value of 3.4e-06).</td><td>Float</td></tr><tr><td>vqsr_snp_filter_level</td><td>Used for applying the recalibration model when running the pipeline using VQSR; required when <code>run_vets</code> is “false”.</td><td>Float</td></tr><tr><td>vqsr_indel_filter_level</td><td>Used for applying the recalibration model when running the pipeline using VQSR; required when <code>run_vets</code> is “false”.</td><td>Float</td></tr><tr><td>snp_vqsr_downsampleFactor</td><td>The downsample factor used for SNP variant recalibration if the number of GVCF files is greater than the <code> snps_variant_recalibration_threshold</code> when running the pipeline using VQSR; required when <code>run_vets</code> is “false”.</td><td>Int</td></tr><tr><td>top_level_scatter_count</td><td>Optional integer used to determine how many files the input interval list should be split into; default will split the interval list into 2 files.</td><td>Int</td></tr><tr><td>gather_vcfs</td><td>Optional boolean; “true” is used for small callsets containing less than 100,000 GVCF files.</td><td>Boolean</td></tr><tr><td>snps_variant_recalibration_threshold</td><td>Optional integer that sets the threshold for the number of callset VCF files used to perform recalibration on a single file; if the number of VCF files exceeds the threshold, variants will be downsampled to enable parallelization; default is “500000”.</td><td>Int</td></tr><tr><td>rename_gvcf_samples</td><td>Optional boolean describing whether GVCF samples should be renamed; default is “true”.</td><td>Boolean</td></tr><tr><td>unbounded_scatter_count_scale_factor</td><td>Optional float used to scale the scatter count when <code>top_level_scatter_count</code> is not provided as input; default is “0.15”.</td><td>Float</td></tr><tr><td>use_allele_specific_annotations</td><td>Optional boolean used for SNP and indel variant recalibration when running the pipeline using VQSR; set to “true” by default.</td><td>Boolean</td></tr></tbody></table><h4 class="anchor anchorWithStickyNavbar_LWe7" id="gnarlygenotyper-joint-calling-input-descriptions">GnarlyGenotyper joint calling input descriptions<a href="#gnarlygenotyper-joint-calling-input-descriptions" class="hash-link" aria-label="Direct link to GnarlyGenotyper joint calling input descriptions" title="Direct link to GnarlyGenotyper joint calling input descriptions">​</a></h4><p>The table below describes the additional pipeline inputs that apply when the pipeline is run with GnarlyGenotyper for joint calling:</p><table><thead><tr><th>Parameter name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>gnarly_scatter_count</td><td>Optional integer used to determine how many files to split the interval list into when using GnarlyGenotyper; default is “10”.</td><td>Int</td></tr><tr><td>use_gnarly_genotyper</td><td>Optional boolean describing whether GnarlyGenotyper should be used; default is “false”.</td><td>Boolean</td></tr></tbody></table><h4 class="anchor anchorWithStickyNavbar_LWe7" id="vets-variant-filtering-input-descriptions">VETS variant filtering input descriptions<a href="#vets-variant-filtering-input-descriptions" class="hash-link" aria-label="Direct link to VETS variant filtering input descriptions" title="Direct link to VETS variant filtering input descriptions">​</a></h4><p>The table below describes the additional pipeline inputs that apply when the pipeline is run with VETS for variant filtering:</p><table><thead><tr><th>Parameter name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>targets_interval_list</td><td>Describes the intervals for which the filtering model will be trained when running the pipeline using VETS; for more details, see the associated <a href="https://storage.googleapis.com/gcp-public-data--broad-references/hg38/v0/bge_exome_calling_regions.v1.1.interval_list.README.md" target="_blank" rel="noopener noreferrer">README</a>; required when <code>run_vets</code> is “true”.</td><td>File</td></tr><tr><td>run_vets</td><td>Optional boolean used to describe whether the pipeline will use VQSR (<code>run_vets = false</code>) or VETS (<code>run_vets = true</code>) to create the filtering model; default is “false”.</td><td>Boolean</td></tr></tbody></table><h4 class="anchor anchorWithStickyNavbar_LWe7" id="fingerprinting-input-descriptions">Fingerprinting input descriptions<a href="#fingerprinting-input-descriptions" class="hash-link" aria-label="Direct link to Fingerprinting input descriptions" title="Direct link to Fingerprinting input descriptions">​</a></h4><p>The table below describes the pipeline inputs that apply to fingerprinting:</p><table><thead><tr><th>Parameter name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>haplotype_database</td><td>Haplotype reference used for fingerprinting (see the CrosscheckFingerprints task).</td><td>File</td></tr><tr><td>cross_check_fingerprints</td><td>Optional boolean describing whether or not the pipeline should check fingerprints; default is “true”.</td><td>Boolean</td></tr><tr><td>scatter_cross_check_fingerprints</td><td>Optional boolean describing whether <code>CrossCheckFingerprintsScattered</code> or <code>CrossCheckFingerprintsSolo</code> should be run; default is “false” and <code>CrossCheckFingerprintsSolo</code> will be run.</td><td>Boolean</td></tr></tbody></table><h4 class="anchor anchorWithStickyNavbar_LWe7" id="runtime-parameter-input-descriptions">Runtime parameter input descriptions<a href="#runtime-parameter-input-descriptions" class="hash-link" aria-label="Direct link to Runtime parameter input descriptions" title="Direct link to Runtime parameter input descriptions">​</a></h4><p>The table below describes the pipeline inputs used for setting runtime parameters of tasks:</p><table><thead><tr><th>Parameter name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>small_disk</td><td>Disk size; dependent on cohort size; requires user input; see example JSON configuration files found in the WARP <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/broad/dna_seq/germline/joint_genotyping/test_inputs" target="_blank" rel="noopener noreferrer">test_inputs</a> folder for recommendations.</td><td>Int</td></tr><tr><td>medium_disk</td><td>Disk size; dependent on cohort size; requires user input; see example JSON configuration files found in the WARP <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/broad/dna_seq/germline/joint_genotyping/test_inputs" target="_blank" rel="noopener noreferrer">test_inputs</a> folder for recommendations.</td><td>Int</td></tr><tr><td>large_disk</td><td>Disk size; dependent on cohort size; requires user input; see example JSON configuration files found in the WARP <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/broad/dna_seq/germline/joint_genotyping/test_inputs" target="_blank" rel="noopener noreferrer">test_inputs</a> folder for recommendations.</td><td>Int</td></tr><tr><td>huge_disk</td><td>Disk size; dependent on cohort size; requires user input; see example JSON configuration files found in the WARP <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/broad/dna_seq/germline/joint_genotyping/test_inputs" target="_blank" rel="noopener noreferrer">test_inputs</a> folder for recommendations.</td><td>Int</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="jointgenotyping-tasks-and-tools">JointGenotyping tasks and tools<a href="#jointgenotyping-tasks-and-tools" class="hash-link" aria-label="Direct link to JointGenotyping tasks and tools" title="Direct link to JointGenotyping tasks and tools">​</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/broad/dna_seq/germline/joint_genotyping/JointGenotyping.wdl" target="_blank" rel="noopener noreferrer">JointGenotyping workflow</a> imports individual &quot;tasks,&quot; also written in WDL script, from the WARP <a href="https://github.com/broadinstitute/warp/tree/master/tasks/broad" target="_blank" rel="noopener noreferrer">tasks folder</a>. </p><p>Overall, the JointGenotyping workflow:</p><ol><li>Splits the input interval list and imports GVCF files.</li><li>Performs joint genotyping using GATK GenotypeGVCFs (default) or GnarlyGenotyper.</li><li>Creates single site-specific VCF and index files.</li><li>Creates and applies a variant filtering model using GATK VQSR (default) or VETS.</li><li>Collects variant calling metrics.</li><li>Checks fingerprints (optional).</li></ol><p>The tasks and tools used in the JointGenotyping workflow are detailed in the table below. </p><p>To see specific tool parameters, select the task WDL link in the table; then find the task and view the <code>command {}</code> section of the task in the WDL script. To view or use the exact tool software, see the task&#x27;s Docker image which is specified in the task WDL <code># runtime values</code> section as <code>String docker =</code>.</p><table><thead><tr><th>Task</th><th>Tool</th><th>Software</th><th>Description</th></tr></thead><tbody><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">CheckSamplesUnique</a></td><td>bash</td><td>bash</td><td>Checks that there are more than 50 unique samples in <code>sample_name_map</code>.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">SplitIntervalList</a></td><td>SplitIntervals</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Splits the unpadded interval list for scattering.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">ImportGVCFs</a></td><td>GenomicsDBImport</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Imports single-sample GVCF files into GenomicsDB before joint genotyping.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">SplitIntervalList as GnarlyIntervalScatterDude</a></td><td>SplitIntervals</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>use_gnarly_genotyper</code> is “true” (default is “false”), splits the unpadded interval list for scattering; otherwise, this task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GnarlyGenotyper</a></td><td>GnarlyGenotyper</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>use_gnarly_genotyper</code> is “true” (default is “false”), performs scalable, “quick and dirty” joint genotyping on a set of GVCF files stored in GenomicsDB; otherwise, this task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GatherVcfs as TotallyRadicalGatherVcfs</a></td><td>GatherVcfsCloud</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>use_gnarly_genotyper</code> is “true” (default is “false”), compiles the site-specific VCF files generated for each interval into one VCF output and index; otherwise, this task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GenotypeGVCFs</a></td><td>GenotypeGVCFs</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>use_gnarly_genotyper</code> is “false” (default is “false”), performs joint genotyping on GVCF files stored in GenomicsDB; otherwise this task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">HardFilterAndMakeSitesOnlyVcf</a></td><td>VariantFiltration, MakeSitesOnlyVcf</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Uses the VCF files to hard filter the variant calls; outputs a VCF file with the site-specific (but not genotype) information.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GatherVcfs as SitesOnlyGatherVcf</a></td><td>GatherVcfsCloud</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Compiles the site-specific VCF files generated for each interval into one VCF output file and index.</td></tr><tr><td><a href="https://github.com/broadinstitute/gatk/blob/master/scripts/vcf_site_level_filtering_wdl/JointVcfFiltering.wdl" target="_blank" rel="noopener noreferrer">JointVcfFiltering as TrainAndApplyVETS</a></td><td>ExtractVariantAnnotations, TrainVariantAnnotationsModel, ScoreVariantAnnotations</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>run_vets</code> is “true” (default is “false”), calls the <code>JointVcfFiltering.wdl</code> subworkflow to extract variant-level annotations, trains a model for variant scoring, and scores variants; otherwise, this task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">IndelsVariantRecalibrator</a></td><td>VariantRecalibrator</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>run_vets</code> is “false” (default is “false”), uses the compiled VCF file to build a recalibration model to score indel variant quality; produces a recalibration table.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">SNPsVariantRecalibratorCreateModel</a></td><td>VariantRecalibrator</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>run_vets</code> is “false” (default is “false”) and the number of input GVCF files is greater than <code>snps_variant_recalibration_threshold</code>, builds a recalibration model to score variant quality; otherwise this task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">SNPsVariantRecalibrator as SNPsVariantRecalibratorScattered</a></td><td>VariantRecalibrator</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>run_vets</code> is “false” (default is “false”) and the number of input GVCF files is greater than <code>snps_variant_recalibration_threshold</code>, builds a scattered recalibration model to score variant quality; otherwise this task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">Tasks.GatherTranches as SNPGatherTranches</a></td><td>GatherTranches</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>run_vets</code> is “false” (default is “false”) and the number of input GVCF files is greater than <code>snps_variant_recalibration_threshold</code>, gathers tranches into a single file; otherwise this task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">SNPsVariantRecalibrator as SNPsVariantRecalibratorClassic</a></td><td>VariantRecalibrator</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>run_vets</code> is “false” (default is “false”) and the number of input GVCF files is not greater than <code>snps_variant_recalibration_threshold</code>, builds a recalibration model to score variant quality; otherwise this task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">ApplyRecalibration</a></td><td>ApplyVQSR</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>run_vets</code> is “false” (default is “false”), scatters the site-specific VCF file and applies a filtering threshold.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">CollectVariantCallingMetrics as CollectMetricsSharded</a></td><td>CollectVariantCallingMetrics</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If the callset has at least 1000 GVCF files, returns detail and summary metrics for each of the scattered VCF files. If the number is small, will return metrics for a merged VCF file produced in the <code>GatherVcfs as FinalGatherVcf</code> task (listed below).</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GatherVcfs as FinalGatherVcf</a></td><td>GatherVcfsCloud</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If the callset has fewer than 1000 GVCF files, compiles the VCF files prior to collecting metrics in the <code>CollectVariantCallingMetrics as CollectMetricsOnFullVcf</code> task (listed below).</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">CollectVariantCallingMetrics as CollectMetricsOnFullVcf</a></td><td>CollectVariantCallingMetrics</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If the callset has fewer than 1000 GVCF files, returns metrics for the merged VCF file produced in the <code>GatherVcfs as FinalGatherVcf</code> task.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GatherVariantCallingMetrics</a></td><td>AccumulateVariantCallingMetrics</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If the callset has at least 1000 GVCF files, gathers metrics produced for each VCF file.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GetFingerprintingIntervalIndices</a></td><td>IntervalListTools</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>cross_check_fingerprints</code> is “true” (default is “true”) and <code>scatter_cross_check_fingerprints</code> is “true” (default is “false”), gets and sorts indices for fingerprint intervals; otherwise the task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GatherVcfs as GatherFingerprintingVcfs</a></td><td>GatherVcfsCloud</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>cross_check_fingerprints</code> is “true” (default is “true”) and <code>scatter_cross_check_fingerprints</code> is “true” (default is “false”), compiles the fingerprint VCF files; otherwise the task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">SelectFingerprintSiteVariants</a></td><td>SelectVariants</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>cross_check_fingerprints</code> is “true”  (default is “true”)and <code>scatter_cross_check_fingerprints</code> is “true” (default is “false”), selects variants from the fingerprint VCF file; otherwise the task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">PartitionSampleNameMap</a></td><td>bash</td><td>bash</td><td>If <code>cross_check_fingerprints</code> is “true” (default is “true”) and <code>scatter_cross_check_fingerprints</code> is “true” (default is “false”), partitions the sample name map and files are scattered by the partition; otherwise the task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">CrossCheckFingerprint as CrossCheckFingerprintsScattered</a></td><td>CrosscheckFingerprints</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>cross_check_fingerprints</code> is “true” (default is “true”) and <code>scatter_cross_check_fingerprints</code> is “true” (default is “false”), checks fingerprints for the VCFs in the scattered partitions and produces a metrics file; otherwise the task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GatherPicardMetrics as GatherFingerprintingMetrics</a></td><td>bash</td><td>bash</td><td>If <code>cross_check_fingerprints</code> is “true” (default is “true”) and <code>scatter_cross_check_fingerprints</code> is “true” (default is “false”), combines the fingerprint metrics files into a single metrics file; otherwise the task is skipped.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">CrossCheckFingerprint as CrossCheckFingerprintSolo</a></td><td>CrosscheckFingerprints</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>If <code>cross_check_fingerprints</code> is “true” (default is “true”) and <code>scatter_cross_check_fingerprints</code> is “false” (default is “false”), checks fingerprints for the single VCF file and produces a metrics file; otherwise the task is skipped.</td></tr></tbody></table><h4 class="anchor anchorWithStickyNavbar_LWe7" id="1-splits-the-input-interval-list-and-imports-gvcf-files">1. Splits the input interval list and imports GVCF files<a href="#1-splits-the-input-interval-list-and-imports-gvcf-files" class="hash-link" aria-label="Direct link to 1. Splits the input interval list and imports GVCF files" title="Direct link to 1. Splits the input interval list and imports GVCF files">​</a></h4><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">SplitIntervalList</a> task uses GATK’s SplitIntervals tool to split the input interval list into two or more interval files. The number of output interval files can be specified using the <code>top_level_scatter_count</code> input parameter or by specifying <code>unbounded_scatter_count_scale_factor</code>, which will scale the number of output files based on the number of input GVCF files.</p><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">ImportGVCFs</a> task uses GATK’s GenomicsDBImport tool and the input sample map file to import single-sample GVCF files into GenomicsDB before joint genotyping.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="2-performs-joint-genotyping-using-gatk-genotypegvcfs-default-or-gnarlygenotyper">2. Performs joint genotyping using GATK GenotypeGVCFs (default) or GnarlyGenotyper<a href="#2-performs-joint-genotyping-using-gatk-genotypegvcfs-default-or-gnarlygenotyper" class="hash-link" aria-label="Direct link to 2. Performs joint genotyping using GATK GenotypeGVCFs (default) or GnarlyGenotyper" title="Direct link to 2. Performs joint genotyping using GATK GenotypeGVCFs (default) or GnarlyGenotyper">​</a></h4><p><strong>GenotypeGVCFs (default)</strong></p><p>When <code>use_gnarly_genotyper</code> is “false”, the <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GenotypeGVCFs</a> task uses GATK’s GenotypeGVCFs tool to perform joint genotyping on GVCF files stored in GenomicsDB that have been pre-called with HaplotypeCaller.</p><p><strong>GnarlyGenotyper</strong></p><p>When <code>use_gnarly_genotyper</code> is “true”, the <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">SplitIntervalList as GnarlyIntervalScatterDude</a> task splits the unpadded interval list for scattering using GATK’s SplitIntervals tool. The output is used as input for the <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GnarlyGenotyper</a> task which performs joint genotyping on the set of GVCF files and outputs an array of VCF and index files using the GnarlyGenotyper tool. Those VCF and index files are gathered in the next task, <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GatherVcfs as TotallyRadicalGatherVcfs</a>, which uses the GatherVcfsCloud tool.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="3-creates-single-site-specific-vcf-and-index-files">3. Creates single site-specific VCF and index files<a href="#3-creates-single-site-specific-vcf-and-index-files" class="hash-link" aria-label="Direct link to 3. Creates single site-specific VCF and index files" title="Direct link to 3. Creates single site-specific VCF and index files">​</a></h4><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">HardFilterAndMakeSitesOnlyVcf</a> task takes in the output VCF and index files produced by either GnarlyGenotyper or GenotypeGVCFs. The task uses the <code>excess_het_threshold</code> input value to hard filter the variant calls using GATK’s VariantFiltration tool. After filtering, the site-specific VCF files are generated from the filtered VCF files by removing all sample-specific genotype information, leaving only the site-level summary information at each site.</p><p>Next, the site-specific VCF and index files for each interval are gathered into a single site-specific VCF and index file by the <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GatherVcfs as SitesOnlyGatherVcf</a> task, which uses the GatherVcfsCloud tool.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="4-creates-and-applies-a-variant-filtering-model-using-gatk-vqsr-default-or-vets">4. Creates and applies a variant filtering model using GATK VQSR (default) or VETS<a href="#4-creates-and-applies-a-variant-filtering-model-using-gatk-vqsr-default-or-vets" class="hash-link" aria-label="Direct link to 4. Creates and applies a variant filtering model using GATK VQSR (default) or VETS" title="Direct link to 4. Creates and applies a variant filtering model using GATK VQSR (default) or VETS">​</a></h4><p><strong>VQSR (default)</strong></p><p>If <code>run_vets</code> is “false”, the <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">IndelsVariantRecalibrator</a> task takes in the site-specific VCF and index files generated in <a href="#3-creates-single-site-specific-vcf-and-index-files">Step 3</a> and uses GATK’s VariantRecalibrator tool to perform the first step of the Variant Quality Score Recalibration (VQSR) technique of filtering variants. The tool builds a model to be used to score and filter indels and produces a recalibration table as output.</p><p>After building the indel filtering model, the workflow uses the VariantRecalibrator tool to build a model to be used to score and filter SNPs. If the number of input GVCF files is greater than <code>snps_variant_recalibration_threshold</code>, the <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">SNPsVariantRecalibratorCreateModel</a>, <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">SNPsVariantRecalibrator as SNPsVariantRecalibratorScattered</a>, and <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">Tasks.GatherTranches as SNPGatherTranches</a> tasks are called to scatter the site-specific VCF and index files, build the SNP model, and gather scattered tranches into a single file. If the number of input GVCF files is less than <code>snps_variant_recalibration_threshold</code>, the <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">SNPsVariantRecalibrator as SNPsVariantRecalibratorClassic</a> task is called to build the SNP model.</p><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">ApplyRecalibration</a> task uses GATK’s ApplyVQSR tool to scatter the site-specific VCF file, apply the indel and SNP filtering models, and output a recalibrated VCF and index file. </p><p><strong>VETS</strong></p><p>If <code>run_vets</code> is “true”, the <a href="https://github.com/broadinstitute/gatk/blob/master/scripts/vcf_site_level_filtering_wdl/JointVcfFiltering.wdl" target="_blank" rel="noopener noreferrer">JointVcfFiltering as TrainAndApplyVETS</a> task takes in the hard filtered and site-specific VCF and index files generated in <a href="#3-creates-single-site-specific-vcf-and-index-files">Step 3</a> and calls the <code>JointVcfFiltering.wdl</code> subworkflow. This workflow uses the Variant Extract-Train-Score (VETS) algorithm to extract variant-level annotations, train a filtering model, and score variants based on the model. The subworkflow uses the GATK ExtractVariantAnnotations, TrainVariantAnnotationsModel, and ScoreVariantAnnotations tools to create extracted and scored VCF and index files. The output VCF and index files are not filtered by the score assigned by the model. The score is included in the output VCF files in the INFO field as an annotation called “SCORE”.</p><p>The VETS algorithm trains the model only over target regions, rather than including exon tails which can lead to poor-quality data. However, the model is applied everywhere including the exon tails.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="5-collects-variant-calling-metrics">5. Collects variant calling metrics<a href="#5-collects-variant-calling-metrics" class="hash-link" aria-label="Direct link to 5. Collects variant calling metrics" title="Direct link to 5. Collects variant calling metrics">​</a></h4><p>Summary and per-sample metrics are collected using Picard’s CollectVariantCallingMetrics tool. For large callsets (at least 1000 GVCF files), the workflow calls the <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">CollectVariantCallingMetrics as CollectMetricsSharded</a> followed by the <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GatherVariantCallingMetrics</a> task to compute and gather the variant calling metrics into single output files. For small callsets (less than 1000 GVCF files), the workflow calls the <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">GatherVcfs as FinalGatherVcf</a> task followed by the <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">CollectVariantCallingMetrics as CollectMetricsOnFullVcf</a> task to first compile the VCF files and then compute the variant calling metrics. Detail and summary metrics files are produced as outputs of these tasks.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="6-checks-fingerprints-optional">6. Checks fingerprints (optional)<a href="#6-checks-fingerprints-optional" class="hash-link" aria-label="Direct link to 6. Checks fingerprints (optional)" title="Direct link to 6. Checks fingerprints (optional)">​</a></h4><p>If <code>cross_check_fingerprints</code> is “true”, the workflow will use Picard to determine the likelihood that the input and output data were generated from the same individual to verify that the pipeline didn’t swap any of the samples during processing. The <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/JointGenotypingTasks.wdl" target="_blank" rel="noopener noreferrer">SelectFingerprintSiteVariants</a> task uses GATK’s SelectVariants tool to select variants in the site-specific VCF file based on the variants present in the <code>haplotype_database</code> and outputs a fingerprint VCF and index file. Next, the workflow cross-checks the fingerprints and creates an output metrics file using the CrosscheckFingerprints tool.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="outputs">Outputs<a href="#outputs" class="hash-link" aria-label="Direct link to Outputs" title="Direct link to Outputs">​</a></h2><p>The following table lists the output variables and files produced by the pipeline.</p><table><thead><tr><th>Output name</th><th>Filename, if applicable</th><th>Output format and description</th></tr></thead><tbody><tr><td>detail_metrics_file</td><td><code>&lt;callset_name&gt;.variant_calling_detail_metrics</code></td><td>Detail metrics file produced using Picard.</td></tr><tr><td>summary_metrics_file</td><td><code>&lt;callset_name&gt;.variant_calling_summary_metrics</code></td><td>Summary metrics file produced using Picard.</td></tr><tr><td>output_vcfs</td><td><code>&lt;callset_name&gt;.vcf.gz</code> or <code>&lt;callset_name&gt;.filtered.&lt;idx&gt;.vcf.gz</code></td><td>Array of all site-specific output VCF files.</td></tr><tr><td>output_vcf_indices</td><td><code>&lt;callset_name&gt;.vcf.gz.tbi</code> or <code>&lt;callset_name&gt;.filtered.&lt;idx&gt;.vcf.gz.tbi</code></td><td>Array of all output VCF index files.</td></tr><tr><td>output_intervals</td><td><code>scatterDir/&lt;output_intervals_files&gt;</code></td><td>Interval list file produced by the workflow.</td></tr><tr><td>crosscheck_fingerprint_check</td><td><code>&lt;callset_name&gt;.fingerprintcheck</code></td><td>Fingerprint metrics</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="versioning-and-testing">Versioning and testing<a href="#versioning-and-testing" class="hash-link" aria-label="Direct link to Versioning and testing" title="Direct link to Versioning and testing">​</a></h2><p>All JointGenotyping pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/broad/dna_seq/germline/joint_genotyping/JointGenotyping.changelog.md" target="_blank" rel="noopener noreferrer">JointGenotyping changelog</a> and tested using <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/broad/dna_seq/germline/joint_genotyping/test_data_overview.md" target="_blank" rel="noopener noreferrer">plumbing and scientific test data</a>. To learn more about WARP pipeline testing, see <a href="https://broadinstitute.github.io/warp/docs/About_WARP/TestingPipelines" target="_blank" rel="noopener noreferrer">Testing Pipelines</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="citing-the-jointgenotyping-pipeline">Citing the JointGenotyping Pipeline<a href="#citing-the-jointgenotyping-pipeline" class="hash-link" aria-label="Direct link to Citing the JointGenotyping Pipeline" title="Direct link to Citing the JointGenotyping Pipeline">​</a></h2><p>If you use the JointGenotyping Pipeline in your research, please consider citing our preprint:</p><p>Degatano, K.; Awdeh, A.; Dingman, W.; Grant, G.; Khajouei, F.; Kiernan, E.; Konwar, K.; Mathews, K.; Palis, K.; Petrillo, N.; Van der Auwera, G.; Wang, C.; Way, J.; Pipelines, W. WDL Analysis Research Pipelines: Cloud-Optimized Workflows for Biological Data Processing and Reproducible Analysis. Preprints 2024, 2024012131. <a href="https://doi.org/10.20944/preprints202401.2131.v1" target="_blank" rel="noopener noreferrer">https://doi.org/10.20944/preprints202401.2131.v1</a></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="feedback">Feedback<a href="#feedback" class="hash-link" aria-label="Direct link to Feedback" title="Direct link to Feedback">​</a></h2><p>Please help us make our tools better by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>; we welcome pipeline-related suggestions or questions.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/JointGenotyping/README.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2024-12-05T15:09:24.000Z">Dec 5, 2024</time></b> by <b>Robert Sidney Cox</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/Pipelines/Imputation_Pipeline/references_overview"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Creating the Imputation Pipeline&#x27;s Modified 1000 Genomes Reference</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/Pipelines/Multiome_Pipeline/README"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Multiome Overview</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-jointgenotyping-workflow" class="table-of-contents__link toc-highlight">Introduction to the JointGenotyping workflow</a></li><li><a href="#set-up" class="table-of-contents__link toc-highlight">Set-up</a><ul><li><a href="#jointgenotyping-installation-and-requirements" class="table-of-contents__link toc-highlight">JointGenotyping Installation and Requirements</a></li><li><a href="#inputs" class="table-of-contents__link toc-highlight">Inputs</a></li></ul></li><li><a href="#jointgenotyping-tasks-and-tools" class="table-of-contents__link toc-highlight">JointGenotyping tasks and tools</a></li><li><a href="#outputs" class="table-of-contents__link toc-highlight">Outputs</a></li><li><a href="#versioning-and-testing" class="table-of-contents__link toc-highlight">Versioning and testing</a></li><li><a href="#citing-the-jointgenotyping-pipeline" class="table-of-contents__link toc-highlight">Citing the JointGenotyping Pipeline</a></li><li><a href="#feedback" class="table-of-contents__link toc-highlight">Feedback</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2024 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.728576f2.js"></script>
<script src="/warp/assets/js/main.868095ac.js"></script>
</body>
</html>