<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Pipelines/Multiome_Pipeline/README">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">Multiome Overview | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/Multiome_Pipeline/README"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Multiome Overview | WARP"><meta data-rh="true" name="description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><meta data-rh="true" property="og:description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/Multiome_Pipeline/README"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Multiome_Pipeline/README" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Multiome_Pipeline/README" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.efee88c0.js" as="script">
<link rel="preload" href="/warp/assets/js/main.affc3f00.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/ATAC/README">ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/BuildIndices_Pipeline/README">BuildIndices</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README">Exome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Exome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/README">Illumina Genotyping Array</a><button aria-label="Toggle the collapsible sidebar category &#x27;Illumina Genotyping Array&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/ImputationBeagle_Pipeline/README">ImputationBeagle</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Imputation_Pipeline/README">Imputation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Imputation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/JointGenotyping_Pipeline/README">JointGenotyping</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/Multiome_Pipeline/README">Multiome scATAC and GEX</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus</a><button aria-label="Toggle the collapsible sidebar category &#x27;Optimus&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/PairedTag_Pipeline/README">Paired-Tag</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README">RNA with UMIs</a><button aria-label="Toggle the collapsible sidebar category &#x27;RNA with UMIs&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/SlideSeq_Pipeline/README">Slide-seq</a><button aria-label="Toggle the collapsible sidebar category &#x27;Slide-seq&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/SlideTags_Pipeline/README">Slide-tags</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README">Smart-seq2 Single Nucleus Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Nucleus Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/snm3C/README">Single Nucleus Methyl-Seq and Chromatin Capture</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Nucleus Methyl-Seq and Chromatin Capture&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Ultima_Genomics_Whole_Genome_Germline_Pipeline/README">Ultima Genomics Whole Genome Germline</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/README">Whole Genome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Whole Genome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">WARP Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Multiome scATAC and GEX</span><meta itemprop="position" content="2"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Multiome Overview</h1><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">Multiome v5.11.0</a></td><td align="center">February, 2025</td><td align="center">WARP Pipelines</td><td align="center">Please <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">file an issue in WARP</a>.</td></tr></tbody></table><p><img loading="lazy" alt="Multiome_diagram" src="/warp/assets/images/multiome_diagram-9f63b411312ab30ae8d54dca539a30fa.png" width="1099" height="1984" class="img_ev3q"></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="introduction-to-the-multiome-workflow">Introduction to the Multiome workflow<a href="#introduction-to-the-multiome-workflow" class="hash-link" aria-label="Direct link to Introduction to the Multiome workflow" title="Direct link to Introduction to the Multiome workflow">​</a></h2><p>Multiome is an open-source, cloud-optimized pipeline developed in collaboration with members of the <a href="https://braininitiative.nih.gov/" target="_blank" rel="noopener noreferrer">BRAIN Initiative</a> (BICCN and <a href="https://brainblog.nih.gov/brain-blog/brain-issues-suite-funding-opportunities-advance-brain-cell-atlases-through-centers" target="_blank" rel="noopener noreferrer">BICAN</a> Sequencing Working Group) and <a href="https://nida.nih.gov/about-nida/organization/divisions/division-neuroscience-behavior-dnb/basic-research-hiv-substance-use-disorder/scorch-program" target="_blank" rel="noopener noreferrer">SCORCH</a> (see <a href="#acknowledgements">Acknowledgements</a> below). It supports the processing of 10x 3&#x27; single-cell and single-nucleus gene expression (GEX) and chromatin accessibility (ATAC) data generated with the <a href="https://www.10xgenomics.com/products/single-cell-multiome-atac-plus-gene-expression" target="_blank" rel="noopener noreferrer">10x Genomics Multiome assay</a>.</p><p>The workflow is a wrapper WDL script that calls two subworkflows: the <a href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus workflow</a> for single-cell GEX data and the <a href="/warp/docs/Pipelines/ATAC/README">ATAC workflow</a> for single-cell ATAC data. </p><p>The GEX component corrects cell barcodes (CBs) and Unique Molecular Identifiers (UMIs), aligns reads to the genome, calculates per-barcode and per-gene quality metrics, and produces a raw cell-by-gene count matrix. It also produces <a href="/warp/docs/Pipelines/Optimus_Pipeline/Library-metrics">library-level metrics</a> calculated from STARsolo aligner metrics. </p><p>The ATAC component corrects CBs, aligns reads to the genome, calculates <a href="/warp/docs/Pipelines/ATAC/count-matrix-overview">per-barcode quality metrics</a>, <a href="/warp/docs/Pipelines/ATAC/library-metrics">library-level metrics</a> and produces a fragment file.</p><p>The wrapper WDL is available in the WARP repository (see the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/skylab/multiome/Multiome.wdl" target="_blank" rel="noopener noreferrer">code here</a>).</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="quickstart-table">Quickstart table<a href="#quickstart-table" class="hash-link" aria-label="Direct link to Quickstart table" title="Direct link to Quickstart table">​</a></h2><p>The following table provides a quick glance at the Multiome pipeline features:</p><table><thead><tr><th>Pipeline features</th><th>Description</th><th>Source</th></tr></thead><tbody><tr><td>Assay type</td><td>10x single cell or single nucleus gene expression (GEX) and ATAC</td><td><a href="https://www.10xgenomics.com" target="_blank" rel="noopener noreferrer">10x Genomics</a></td></tr><tr><td>Overall workflow</td><td>Barcode correction, read alignment, gene and fragment quantification</td><td></td></tr><tr><td>Workflow language</td><td>WDL 1.0</td><td><a href="https://github.com/openwdl/wdl" target="_blank" rel="noopener noreferrer">openWDL</a></td></tr><tr><td>Genomic Reference Sequence</td><td>GRCh38 human genome primary sequence</td><td>GENCODE <a href="https://www.gencodegenes.org/human/release_43.html" target="_blank" rel="noopener noreferrer">human reference files</a></td></tr><tr><td>Gene annotation reference (GTF)</td><td>Reference containing gene annotations</td><td>GENCODE <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_43/gencode.v43.annotation.gtf.gz" target="_blank" rel="noopener noreferrer">human GTF</a></td></tr><tr><td>Aligners</td><td>STARsolo (GEX), BWA-mem2 (ATAC)</td><td><a href="https://www.biorxiv.org/content/10.1101/2021.05.05.442755v1" target="_blank" rel="noopener noreferrer">Kaminow et al. 2021</a>, <a href="https://ieeexplore.ieee.org/document/8820962" target="_blank" rel="noopener noreferrer">Vasimuddin et al. 2019</a></td></tr><tr><td>Transcript and fragment quantification</td><td>STARsolo (GEX), SnapATAC2 (ATAC)</td><td><a href="https://www.biorxiv.org/content/10.1101/2021.05.05.442755v1" target="_blank" rel="noopener noreferrer">Kaminow et al. 2021</a>, <a href="https://kzhang.org/SnapATAC2/" target="_blank" rel="noopener noreferrer">SnapATAC2</a></td></tr><tr><td>Data input file format</td><td>File format in which sequencing data is provided</td><td><a href="https://academic.oup.com/nar/article/38/6/1767/3112533" target="_blank" rel="noopener noreferrer">FASTQ</a></td></tr><tr><td>Data output file format</td><td>File formats in which Multiome output is provided</td><td><a href="http://samtools.github.io/hts-specs/" target="_blank" rel="noopener noreferrer">BAM</a> and <a href="https://anndata.readthedocs.io/en/latest/" target="_blank" rel="noopener noreferrer">h5ad</a></td></tr><tr><td>Library-level metrics</td><td>Library-level metrics produced by the Optimus and ATAC workflows</td><td><a href="/warp/docs/Pipelines/Optimus_Pipeline/Library-metrics">Optimus ibrary-level metrics</a> and <a href="/warp/docs/Pipelines/ATAC/library-metrics">ATAC library-level metrics</a></td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="set-up">Set-up<a href="#set-up" class="hash-link" aria-label="Direct link to Set-up" title="Direct link to Set-up">​</a></h2><h3 class="anchor anchorWithStickyNavbar_LWe7" id="multiome-installation">Multiome installation<a href="#multiome-installation" class="hash-link" aria-label="Direct link to Multiome installation" title="Direct link to Multiome installation">​</a></h3><p>To download the latest Multiome release, see the release tags prefixed with &quot;Multiome&quot; on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a>. All Multiome pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/skylab/multiome/Multiome.changelog.md" target="_blank" rel="noopener noreferrer">Multiome changelog</a>. </p><p>To discover and search releases, use the WARP command-line tool <a href="https://github.com/broadinstitute/warp/tree/master/wreleaser" target="_blank" rel="noopener noreferrer">Wreleaser</a>.</p><p>If you’re running a Multiome workflow version prior to the latest release, the accompanying documentation for that release may be downloaded with the source code on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a> (see the source code folder).</p><p>Multiome can be deployed using <a href="https://cromwell.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH-compliant, flexible workflow management system that supports multiple computing platforms. The workflow can also be run in <a href="https://app.terra.bio" target="_blank" rel="noopener noreferrer">Terra</a>, a cloud-based analysis platform. The <a href="https://app.terra.bio/#workspaces/warp-pipelines/Multiome" target="_blank" rel="noopener noreferrer">Multiome public workspace</a> on Terra contains the Multiome workflow, workflow configuration, required reference data and other inputs, and example testing data.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="inputs">Inputs<a href="#inputs" class="hash-link" aria-label="Direct link to Inputs" title="Direct link to Inputs">​</a></h2><table><thead><tr><th>Input name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>input_id</td><td>Unique identifier describing the biological sample or replicate that corresponds with the FASTQ files; can be a human-readable name or UUID.</td><td>String</td></tr><tr><td>cloud_provider</td><td>String describing the cloud provider that should be used to run the workflow; value should be &quot;gcp&quot; or &quot;azure&quot;.</td><td>String</td></tr><tr><td>gex_nhash_id</td><td>Optional identifier for the library aliquot; when specified, the gene expression workflow will echo the ID in the gene expression output h5ads (in the adata.uns section) and in the library-level metrics CSV.</td><td></td></tr><tr><td>atac_nhash_id</td><td>Optional identifier for the library aliquot; when specified, the workflow will echo the ID in the ATAC output h5ads (in the adata.uns section) and in the library-level metrics CSV.</td><td></td></tr><tr><td>expected_cells</td><td>Number of cells loaded for library preparation; default is set to 3000.</td><td>Integer</td></tr><tr><td>annotations_gtf</td><td>GTF file containing gene annotations used for GEX cell metric calculation and ATAC fragment metrics; must match the GTF used to build the STAR aligner.</td><td>File</td></tr><tr><td>gex_r1_fastq</td><td>Array of read 1 FASTQ files representing a single GEX 10x library.</td><td>Array<!-- -->[File]</td></tr><tr><td>gex_r2_fastq</td><td>Array of read 2 FASTQ files representing a single GEX 10x library.</td><td>Array<!-- -->[File]</td></tr><tr><td>gex_i1_fastq</td><td>Optional array of index FASTQ files representing a single GEX 10x library; multiplexed samples are not currently supported, but the file may be passed to the pipeline.</td><td>Array<!-- -->[File]</td></tr><tr><td>tar_star_reference</td><td>TAR file containing a species-specific reference genome and GTF for Optimus (GEX) pipeline.</td><td>File</td></tr><tr><td>mt_genes</td><td>Optional file for the Optimus (GEX) pipeline containing mitochondrial gene names used for metric calculation; default assumes &#x27;mt&#x27; prefix in GTF (case insensitive).</td><td>File</td></tr><tr><td>counting_mode</td><td>Optional string that determines whether the Optimus (GEX) pipeline should be run in single-cell mode (sc_rna) or single-nucleus mode (sn_rna); default is &quot;sn_rna&quot;.</td><td>String</td></tr><tr><td>tenx_chemistry_version</td><td>Optional integer for the Optimus (GEX) pipeline specifying the 10x version chemistry the data was generated with; validated by examination of the first read 1 FASTQ file read structure; default is &quot;3&quot;.</td><td>Integer</td></tr><tr><td>emptydrops_lower</td><td>Optional threshold for UMIs for the Optimus (GEX) pipeline that empty drops tool should consider for determining cell; data below threshold is not removed; default is &quot;100&quot;.</td><td>Integer</td></tr><tr><td>force_no_check</td><td>Optional boolean for the Optimus (GEX) pipeline indicating if the pipeline should perform checks; default is &quot;false&quot;.</td><td>Boolean</td></tr><tr><td>ignore_r1_read_length</td><td>Optional boolean for the Optimus (GEX) pipeline indicating if the pipeline should ignore barcode chemistry check; if &quot;true&quot;, the workflow will not ensure the <code>10x_chemistry_version</code> input matches the chemistry in the read 1 FASTQ; default is &quot;false&quot;.</td><td>Boolean</td></tr><tr><td>star_strand_mode</td><td>Optional string for the Optimus (GEX) pipeline for performing STARsolo alignment on forward stranded, reverse stranded, or unstranded data; default is &quot;Forward&quot;.</td><td>String</td></tr><tr><td>count_exons</td><td>Optional boolean for the Optimus (GEX) pipeline indicating if the workflow should calculate exon counts <strong>when in single-nucleus (sn_rna) mode</strong>; if &quot;true&quot; in sc_rna mode, the workflow will return an error; default is &quot;false&quot;.</td><td>Boolean</td></tr><tr><td>soloMultiMappers</td><td>Optional string describing whether or not the Optimus (GEX) pipeline should run STARsolo with the <code>--soloMultiMappers</code> flag.</td><td>String</td></tr><tr><td>atac_r1_fastq</td><td>Array of read 1 paired-end FASTQ files representing a single 10x multiome ATAC library.</td><td>Array<!-- -->[File]</td></tr><tr><td>atac_r2_fastq</td><td>Array of barcodes FASTQ files representing a single 10x multiome ATAC library.</td><td>Array<!-- -->[File]</td></tr><tr><td>atac_r3_fastq</td><td>Array of read 2 paired-end FASTQ files representing a single 10x multiome ATAC library.</td><td>Array<!-- -->[File]</td></tr><tr><td>tar_bwa_reference</td><td>TAR file containing the reference index files for BWA-mem alignment for the ATAC pipeline.</td><td>File</td></tr><tr><td>chrom_sizes</td><td>File containing the genome chromosome sizes; used to calculate ATAC fragment file metrics.</td><td>File</td></tr><tr><td>adapter_seq_read1</td><td>Optional string describing the adapter sequence for ATAC read 1 paired-end reads to be used during adapter trimming with Cutadapt; default is &quot;GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG&quot;.</td><td>String</td></tr><tr><td>adapter_seq_read3</td><td>Optional string describing the adapter sequence for ATAC read 2 paired-end reads to be used during adapter trimming with Cutadapt; default is &quot;TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG&quot;.</td><td>String</td></tr><tr><td>run_cellbender</td><td>Optional boolean used to determine if the Optimus (GEX) pipeline should run CellBender on the output gene expression h5ad file, <code>h5ad_output_file_gex</code>; default is &quot;false&quot;.</td><td>Boolean</td></tr><tr><td>run_peak_calling</td><td>Optional boolean used to determine if the ATAC pipeline should run Peak Calling; default is <code>false</code>. When set to true, the pipeline takes the ATAC h5ad produced by the JoinBarcodes task and performs peak calling to produce a cell by bin matrix and a cell by peak matrix.</td><td>Boolean</td></tr><tr><td>vm_size</td><td>String defining the Azure virtual machine family for the workflow (default: &quot;Standard_M128s&quot;).</td><td>String</td></tr></tbody></table><h4 class="anchor anchorWithStickyNavbar_LWe7" id="sample-inputs-for-analyses-in-a-terra-workspace">Sample inputs for analyses in a Terra Workspace<a href="#sample-inputs-for-analyses-in-a-terra-workspace" class="hash-link" aria-label="Direct link to Sample inputs for analyses in a Terra Workspace" title="Direct link to Sample inputs for analyses in a Terra Workspace">​</a></h4><p>The Multiome pipeline is currently available on the cloud-based platform Terra. After registering, you can access the <a href="https://app.terra.bio/#workspaces/warp-pipelines/Multiome" target="_blank" rel="noopener noreferrer">Multiome public workspace</a> which is preloaded with instructions and sample data. Please view the <a href="https://support.terra.bio/hc/en-us" target="_blank" rel="noopener noreferrer">Support Center</a> for more information on using the Terra platform.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="tasks">Tasks<a href="#tasks" class="hash-link" aria-label="Direct link to Tasks" title="Direct link to Tasks">​</a></h2><p>The Multiome workflow calls two WARP subworkflows, one external subworkflow (optional), and an additional task, which are described briefly in the table below. For more details on each subworkflow and task, see the documentation and WDL scripts linked in the table.</p><table><thead><tr><th>Subworkflow</th><th>Software</th><th>Description</th></tr></thead><tbody><tr><td>ATAC (<a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/skylab/atac/atac.wdl" target="_blank" rel="noopener noreferrer">WDL</a> and <a href="/warp/docs/Pipelines/ATAC/README">documentation</a>)</td><td>fastqprocess, bwa-mem, SnapATAC2</td><td>Workflow used to analyze 10x single-cell ATAC data.</td></tr><tr><td>Optimus (<a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/skylab/optimus/Optimus.wdl" target="_blank" rel="noopener noreferrer">WDL</a> and <a href="/warp/docs/Pipelines/Optimus_Pipeline/README">documentation</a>)</td><td>fastqprocess, STARsolo, Emptydrops</td><td>Workflow used to analyze 10x single-cell GEX data.</td></tr><tr><td>JoinMultiomeBarcodes as JoinBarcodes (<a href="https://github.com/broadinstitute/warp/blob/develop/tasks/skylab/H5adUtils.wdl" target="_blank" rel="noopener noreferrer">WDL</a>)</td><td>Python3</td><td>Task that adds an extra column to the Optimus metrics <code>h5ad.obs</code> property that lists the respective ATAC barcodes for each gene expression barcode. It also adds an extra column to the ATAC metrics <code>h5ad.obs</code> property to link ATAC barcodes to gene expression barcodes.</td></tr><tr><td>CellBender.run_cellbender_remove_background_gpu as CellBender (<a href="https://raw.githubusercontent.com/broadinstitute/CellBender/v0.3.0/wdl/cellbender_remove_background.wdl" target="_blank" rel="noopener noreferrer">WDL</a>)</td><td>CellBender</td><td>Optional task that runs the <code>cellbender_remove_background.wdl</code> WDL script directly from the <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">CellBender GitHub repository</a>, depending on whether the input <code>run_cellbender</code> is &quot;true&quot; or &quot;false&quot;.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="outputs">Outputs<a href="#outputs" class="hash-link" aria-label="Direct link to Outputs" title="Direct link to Outputs">​</a></h2><table><thead><tr><th>Output variable name</th><th>Filename, if applicable</th><th>Output format and description</th></tr></thead><tbody><tr><td>multiome_pipeline_version_out</td><td>N.A.</td><td>String describing the version of the Multiome pipeline used.</td></tr><tr><td>bam_aligned_output_atac</td><td><code>&lt;input_id&gt;_atac.bam</code></td><td>BAM file containing aligned reads from ATAC workflow.</td></tr><tr><td>fragment_file_atac</td><td><code>&lt;input_id&gt;_atac.fragments.sorted.tsv.gz</code></td><td>Sorted and bgzipped TSV file containing fragment start and stop coordinates per barcode. The columns are &quot;Chromosome&quot;, &quot;Start&quot;, &quot;Stop&quot;, &quot;ATAC Barcode&quot;, &quot;Number of reads&quot;, and &quot;GEX Barcode&quot;.</td></tr><tr><td>fragment_file_index</td><td><code>&lt;input_id&gt;_atac.fragments.sorted.tsv.gz.csi</code></td><td>Tabix CSI index file for the fragment file.</td></tr><tr><td>snap_metrics_atac</td><td><code>&lt;input_id&gt;_atac.metrics.h5ad</code></td><td>h5ad (Anndata) file containing per-barcode metrics from SnapATAC2. Also contains the equivalent gene expression barcode for each ATAC barcode in the <code>gex_barcodes</code> column of the <code>h5ad.obs</code> property. See the <a href="/warp/docs/Pipelines/ATAC/count-matrix-overview">ATAC Count Matrix Overview</a> for more details.</td></tr><tr><td>atac_library_metrics</td><td><code>&lt;input_id&gt;_atac_&lt;nhash_id&gt;_library_metrics.csv</code></td><td>CSV with library-level metrics produced by SnapATAC2. See the ATAC <a href="/warp/docs/Pipelines/ATAC/library-metrics">Library Level Metrics Overview</a> for more details.</td></tr><tr><td>genomic_reference_version_gex</td><td><code>&lt;reference_version&gt;.txt</code></td><td>File containing the Genome build, source and GTF annotation version.</td></tr><tr><td>bam_gex</td><td><code>&lt;input_id&gt;_gex.bam</code></td><td>BAM file containing aligned reads from Optimus workflow.</td></tr><tr><td>matrix_gex</td><td><code>&lt;input_id&gt;_gex_sparse_counts.npz</code></td><td>NPZ file containing raw gene by cell counts.</td></tr><tr><td>matrix_row_index_gex</td><td><code>&lt;input_id&gt;_gex_sparse_counts_row_index.npy</code></td><td>NPY file containing the row indices.</td></tr><tr><td>matrix_col_index_gex</td><td><code>&lt;input_id&gt;_gex_sparse_counts_col_index.npy</code></td><td>NPY file containing the column indices.</td></tr><tr><td>cell_metrics_gex</td><td><code>&lt;input_id&gt;_gex.cell_metrics.csv.gz</code></td><td>CSV file containing the per-cell (barcode) metrics.</td></tr><tr><td>gene_metrics_gex</td><td><code>&lt;input_id&gt;_gex.gene_metrics.csv.gz</code></td><td>CSV file containing the per-gene metrics.</td></tr><tr><td>cell_calls_gex</td><td><code>&lt;input_id&gt;_gex.emptyDrops</code></td><td>TSV file containing the EmptyDrops results when the Optimus workflow is run in sc_rna mode.</td></tr><tr><td>h5ad_output_file_gex</td><td><code>&lt;input_id&gt;_gex.h5ad</code></td><td>h5ad (Anndata) file containing the raw cell-by-gene count matrix, gene metrics, cell metrics, and global attributes. Also contains equivalent ATAC barcode for each gene expression barcode in the <code>atac_barcodes</code> column of the <code>h5ad.obs</code> property. See the <a href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a> for more details.</td></tr><tr><td>multimappers_EM_matrix</td><td><code>UniqueAndMult-EM.mtx</code></td><td>Optional output produced when <code>soloMultiMappers</code> is &quot;EM&quot;; see STARsolo <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md#multi-gene-reads" target="_blank" rel="noopener noreferrer">documentation</a> for more information.</td></tr><tr><td>multimappers_Uniform_matrix</td><td><code>UniqueAndMult-Uniform.mtx</code></td><td>Optional output produced when <code>soloMultiMappers</code> is &quot;Uniform&quot;; see STARsolo <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md#multi-gene-reads" target="_blank" rel="noopener noreferrer">documentation</a> for more information.</td></tr><tr><td>multimappers_Rescue_matrix</td><td><code>UniqueAndMult-Rescue.mtx</code></td><td>Optional output produced when <code>soloMultiMappers</code> is &quot;Rescue&quot;; see STARsolo <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md#multi-gene-reads" target="_blank" rel="noopener noreferrer">documentation</a> for more information.</td></tr><tr><td>multimappers_PropUnique_matrix</td><td><code>UniqueAndMult-PropUnique.mtx</code></td><td>Optional output produced when <code>soloMultiMappers</code> is &quot;PropUnique&quot;; see STARsolo <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md#multi-gene-reads" target="_blank" rel="noopener noreferrer">documentation</a> for more information.</td></tr><tr><td>gex_aligner_metrics</td><td><code>&lt;input_id&gt;_gex.star_metrics.tar</code></td><td>Text file containing per barcode metrics (<code>CellReads.stats</code>) produced by the GEX pipeline STARsolo aligner.</td></tr><tr><td>library_metrics</td><td><code>&lt;input_id&gt;_gex_&lt;gex_nhash_id&gt;_library_metrics.csv</code></td><td>Optional CSV file containing all library-level metrics calculated with STARsolo for gene expression data.</td></tr><tr><td>mtx_files</td><td><code>&lt;input_id&gt;_gex.mtx_files.tar</code></td><td>TAR file with STARsolo matrix market files (barcodes.tsv, features.tsv, and matrix.mtx)</td></tr><tr><td>cell_barcodes_csv</td><td><code>&lt;cell_csv&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td></tr><tr><td>checkpoint_file</td><td><code>&lt;ckpt_file&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td></tr><tr><td>h5_array</td><td><code>&lt;h5_array&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td></tr><tr><td>html_report_array</td><td><code>&lt;report_array&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td></tr><tr><td>log</td><td><code>&lt;log&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td></tr><tr><td>metrics_csv_array</td><td><code>&lt;metrics_array&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td></tr><tr><td>output_directory</td><td><code>&lt;output_dir&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td></tr><tr><td>summary_pdf</td><td><code>&lt;pdf&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td></tr><tr><td>cellbybin_h5ad_file</td><td>h5ad</td><td>Cell by bin matrix produced by SnapATAC2 peak calling. This matrix contains (unmerged) peaks in the MACS3 unstructured metadata (adata.uns<!-- -->[&#x27;MACS3&#x27;]<!-- -->). The matrix consists of insertion counts per 500 bp genomic bin and cell barcode.</td></tr><tr><td>cellbypeak_h5ad_file</td><td>h5ad</td><td>Cell by peak matrix produced by SnapATAC2 peak calling. This matrix contains insertion counts per (merged) peak coordinates and per cell barcode.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="versioning-and-testing">Versioning and testing<a href="#versioning-and-testing" class="hash-link" aria-label="Direct link to Versioning and testing" title="Direct link to Versioning and testing">​</a></h2><p>All Multiome pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/skylab/multiome/Multiome.changelog.md" target="_blank" rel="noopener noreferrer">Multiome changelog</a> and tested using <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/skylab/multiome/test_inputs" target="_blank" rel="noopener noreferrer">plumbing and scientific test data</a>. To learn more about WARP pipeline testing, see <a href="https://broadinstitute.github.io/warp/docs/About_WARP/TestingPipelines" target="_blank" rel="noopener noreferrer">Testing Pipelines</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="citing-the-multiome-pipeline">Citing the Multiome Pipeline<a href="#citing-the-multiome-pipeline" class="hash-link" aria-label="Direct link to Citing the Multiome Pipeline" title="Direct link to Citing the Multiome Pipeline">​</a></h2><p>If you use the Multiome Pipeline in your research, please identify the pipeline in your methods section using the <a href="https://scicrunch.org/resources/data/record/nlx_144509-1/SCR_024217/resolver?q=SCR_024217&amp;l=SCR_024217&amp;i=rrid:scr_024217" target="_blank" rel="noopener noreferrer">Multiome SciCrunch resource identifier</a>.</p><ul><li>Ex: <em>Multiome Pipeline (RRID:SCR_024217)</em></li></ul><p>When citing WARP, please use the following:</p><p>Kylee Degatano, Aseel Awdeh, Robert Sidney Cox III, Wes Dingman, George Grant, Farzaneh Khajouei, Elizabeth Kiernan, Kishori Konwar, Kaylee L Mathews, Kevin Palis, Nikelle Petrillo, Geraldine Van der Auwera, Chengchen (Rex) Wang, Jessica Way. &quot;Warp Analysis Research Pipelines: Cloud-optimized workflows for biological data processing and reproducible analysis.&quot; <em>Bioinformatics</em>, 2025; <a href="https://doi.org/10.1093/bioinformatics/btaf494" target="_blank" rel="noopener noreferrer">https://doi.org/10.1093/bioinformatics/btaf494</a> .</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="consortia-support">Consortia support<a href="#consortia-support" class="hash-link" aria-label="Direct link to Consortia support" title="Direct link to Consortia support">​</a></h2><p>This pipeline is supported by the <a href="https://braininitiative.nih.gov/" target="_blank" rel="noopener noreferrer">BRAIN Initiative</a> (BICCN and BICAN). </p><p>If your organization also uses this pipeline, we would like to list you! Please reach out to us by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="acknowledgements">Acknowledgements<a href="#acknowledgements" class="hash-link" aria-label="Direct link to Acknowledgements" title="Direct link to Acknowledgements">​</a></h2><p>We are immensely grateful to the members of the BRAIN Initiative (BICAN Sequencing Working Group) and SCORCH for their invaluable and exceptional contributions to this pipeline. Our heartfelt appreciation goes to Alex Dobin, Aparna Bhaduri, Alec Wysoker, Anish Chakka, Brian Herb, Daofeng Li, Fenna Krienen, Guo-Long Zuo, Jeff Goldy, Kai Zhang, Khalid Shakir, Bo Li, Mariano Gabitto, Michael DeBerardine, Mengyi Song, Melissa Goldman, Nelson Johansen, James Nemesh, and Theresa Hodges for their unwavering dedication and remarkable efforts. </p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="feedback">Feedback<a href="#feedback" class="hash-link" aria-label="Direct link to Feedback" title="Direct link to Feedback">​</a></h2><p>Please help us make our tools better by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>; we welcome pipeline-related suggestions or questions.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/Multiome_Pipeline/README.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2025-10-01T14:49:24.000Z">Oct 1, 2025</time></b> by <b>Nikelle Petrillo</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/Pipelines/JointGenotyping_Pipeline/README"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">JointGenotyping Overview</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/Pipelines/Optimus_Pipeline/README"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Optimus Overview</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-multiome-workflow" class="table-of-contents__link toc-highlight">Introduction to the Multiome workflow</a></li><li><a href="#quickstart-table" class="table-of-contents__link toc-highlight">Quickstart table</a></li><li><a href="#set-up" class="table-of-contents__link toc-highlight">Set-up</a><ul><li><a href="#multiome-installation" class="table-of-contents__link toc-highlight">Multiome installation</a></li></ul></li><li><a href="#inputs" class="table-of-contents__link toc-highlight">Inputs</a></li><li><a href="#tasks" class="table-of-contents__link toc-highlight">Tasks</a></li><li><a href="#outputs" class="table-of-contents__link toc-highlight">Outputs</a></li><li><a href="#versioning-and-testing" class="table-of-contents__link toc-highlight">Versioning and testing</a></li><li><a href="#citing-the-multiome-pipeline" class="table-of-contents__link toc-highlight">Citing the Multiome Pipeline</a></li><li><a href="#consortia-support" class="table-of-contents__link toc-highlight">Consortia support</a></li><li><a href="#acknowledgements" class="table-of-contents__link toc-highlight">Acknowledgements</a></li><li><a href="#feedback" class="table-of-contents__link toc-highlight">Feedback</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.efee88c0.js"></script>
<script src="/warp/assets/js/main.affc3f00.js"></script>
</body>
</html>