<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Pipelines/SlideTags_Pipeline/README">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">Slide-tags Overview | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/SlideTags_Pipeline/README"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Slide-tags Overview | WARP"><meta data-rh="true" name="description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><meta data-rh="true" property="og:description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/SlideTags_Pipeline/README"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/SlideTags_Pipeline/README" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/SlideTags_Pipeline/README" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.f756d810.js" as="script">
<link rel="preload" href="/warp/assets/js/main.affc3f00.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/ATAC/README">ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/BuildIndices_Pipeline/README">BuildIndices</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README">Exome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Exome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/README">Illumina Genotyping Array</a><button aria-label="Toggle the collapsible sidebar category &#x27;Illumina Genotyping Array&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/ImputationBeagle_Pipeline/README">ImputationBeagle</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Imputation_Pipeline/README">Imputation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Imputation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/JointGenotyping_Pipeline/README">JointGenotyping</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Multiome_Pipeline/README">Multiome scATAC and GEX</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus</a><button aria-label="Toggle the collapsible sidebar category &#x27;Optimus&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/PairedTag_Pipeline/README">Paired-Tag</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README">RNA with UMIs</a><button aria-label="Toggle the collapsible sidebar category &#x27;RNA with UMIs&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/SlideSeq_Pipeline/README">Slide-seq</a><button aria-label="Toggle the collapsible sidebar category &#x27;Slide-seq&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/SlideTags_Pipeline/README">Slide-tags</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README">Smart-seq2 Single Nucleus Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Nucleus Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/snm3C/README">Single Nucleus Methyl-Seq and Chromatin Capture</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Nucleus Methyl-Seq and Chromatin Capture&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Ultima_Genomics_Whole_Genome_Germline_Pipeline/README">Ultima Genomics Whole Genome Germline</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/README">Whole Genome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Whole Genome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">WARP Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Slide-tags</span><meta itemprop="position" content="2"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Slide-tags Overview</h1><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center">v1.0.0</td><td align="center">May, 2025</td><td align="center">WARP Pipelines</td><td align="center">Please <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">file an issue in WARP</a></td></tr></tbody></table><p><img loading="lazy" alt="Slide-tags_diagram" src="/warp/assets/images/Slide-tags-5be1909efda58ae24d0b8b7e44310720.png" width="1289" height="1135" class="img_ev3q"></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="introduction-to-the-slide-tags-pipeline">Introduction to the Slide-tags Pipeline<a href="#introduction-to-the-slide-tags-pipeline" class="hash-link" aria-label="Direct link to Introduction to the Slide-tags Pipeline" title="Direct link to Introduction to the Slide-tags Pipeline">​</a></h2><p>The <strong>Slide-tags Pipeline</strong> is an open-source, cloud-optimized workflow for processing spatial transcriptomics data. It supports data derived from spatially barcoded sequencing technologies, including Slide-tags-based single-molecule profiling. The pipeline processes raw sequencing data into spatially resolved gene expression matrices, ensuring accurate alignment, spatial positioning, and quantification.</p><p>This pipeline integrates multiple processing steps, including barcode extraction, spatial alignment, transcript counting, and output generation in formats compatible with community tools. The pipeline calls three subworkflows: the Optimus workflow for gene expression data, the SpatialCount workflow for spatial barcode processing, and the Positioning workflow for integrating both data types.</p><p>The Optimus workflow (GEX) corrects cell barcodes and Unique Molecular Identifiers (UMIs), aligns reads to the genome, calculates per-barcode and per-gene quality metrics, and produces a cell-by-gene count matrix.</p><p>The SpatialCount workflow processes spatial barcoding data from FASTQ files to generate spatial barcode counts that represent transcript locations within the tissue.</p><p>The Positioning workflow integrates the gene expression data with spatial information to generate coordinate files and visualizations, producing a Seurat object for downstream spatial analysis.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="quickstart-table">Quickstart Table<a href="#quickstart-table" class="hash-link" aria-label="Direct link to Quickstart Table" title="Direct link to Quickstart Table">​</a></h2><table><thead><tr><th>Pipeline Features</th><th>Description</th><th>Source</th></tr></thead><tbody><tr><td>Assay type</td><td>Spatial transcriptomics using Slide-tags</td><td><a href="https://macoskolab.com/" target="_blank" rel="noopener noreferrer">Macosko Lab</a></td></tr><tr><td>Overall workflow</td><td>Barcode extraction, spatial positioning, transcript quantification</td><td>Original code available from <a href="https://github.com/MacoskoLab/Macosko-Pipelines/tree/main/slide-tags" target="_blank" rel="noopener noreferrer">GitHub</a>; WDL workflow available in WARP.</td></tr><tr><td>Workflow language</td><td>WDL</td><td><a href="https://github.com/openwdl/wdl" target="_blank" rel="noopener noreferrer">openWDL</a></td></tr><tr><td>Sub-workflows</td><td>Optimus (gene expression), SpatialCount (spatial barcoding), Positioning (integration)</td><td>Imported from separate WDL scripts</td></tr><tr><td>Genomic Reference Sequence</td><td>STAR reference genome provided as tar file</td><td>Referenced as input parameter</td></tr><tr><td>Gene annotation reference</td><td>GTF file containing gene annotations</td><td>Referenced as input parameter</td></tr><tr><td>Aligners</td><td>STARsolo</td><td><a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">STAR aligner</a></td></tr><tr><td>Data input file format</td><td>File format in which sequencing data is provided</td><td><a href="https://academic.oup.com/nar/article/38/6/1767/3112533" target="_blank" rel="noopener noreferrer">FASTQ</a> and <a href="https://www.loc.gov/preservation/digital/formats/fdd/fdd000323.shtml" target="_blank" rel="noopener noreferrer">CSV</a></td></tr><tr><td>Data output file format</td><td>Output formats for downstream analysis</td><td><a href="https://www.hdfgroup.org/" target="_blank" rel="noopener noreferrer">HDF5</a>, <a href="https://satijalab.org/seurat/" target="_blank" rel="noopener noreferrer">Seurat</a> and <a href="https://www.loc.gov/preservation/digital/formats/fdd/fdd000323.shtml" target="_blank" rel="noopener noreferrer">CSV</a></td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="set-up">Set-up<a href="#set-up" class="hash-link" aria-label="Direct link to Set-up" title="Direct link to Set-up">​</a></h2><h3 class="anchor anchorWithStickyNavbar_LWe7" id="installation">Installation<a href="#installation" class="hash-link" aria-label="Direct link to Installation" title="Direct link to Installation">​</a></h3><p>To download the latest Slide-tags release, see the release tags prefixed with &quot;Slide-tags&quot; on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a>. All Slide-tags pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/slidetags/SlideTags.changelog.md" target="_blank" rel="noopener noreferrer">Slide-tags changelog</a>.</p><p>The pipeline can be deployed using <a href="https://cromwell.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH-compliant workflow manager. Additionally, it can be run in cloud-based analysis platforms such as <a href="https://app.terra.bio" target="_blank" rel="noopener noreferrer">Terra</a>.</p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="inputs">Inputs<a href="#inputs" class="hash-link" aria-label="Direct link to Inputs" title="Direct link to Inputs">​</a></h3><p>The pipeline requires JSON-formatted configuration files detailing input parameters. Required inputs include:</p><ul><li><strong>Raw paired-end GEX FASTQ files</strong> containing sequencing GEX reads</li><li><strong>Raw paired-end Spatial FASTQ files</strong> containing spatial reads</li><li><strong>Pucks files</strong> contains spatial coordinates of bead centroids</li></ul><p><strong><em>General input variables used by both the Optimus and spatial/positioning components of the pipeline.</em></strong></p><table><thead><tr><th>Input Variables</th><th>Description</th><th>Format</th></tr></thead><tbody><tr><td>input_id</td><td>Unique input identifier</td><td>String</td></tr><tr><td>docker</td><td>Docker image used for the workflow</td><td>String</td></tr></tbody></table><p><strong><em>Input variables for the spatial and positioning components of the Slide-Tags pipeline can be found below.</em></strong> </p><table><thead><tr><th>Input Variables</th><th>Description</th><th>Format</th></tr></thead><tbody><tr><td>spatial_fastq</td><td>Array of paths to spatial FASTQ files. Requires at least one complete R1 and R2 pair. Each filename must include <em>R1</em> or <em>R2</em> to distinguish read pairs. The full directory is scanned, matching R1 and R2 files; an error is raised if any pair is incomplete</td><td>Array<!-- -->[String]</td></tr><tr><td>pucks</td><td>Array of paths to puck files</td><td>Array<!-- -->[String]</td></tr></tbody></table><p><strong><em>Optimus input variables can be found below.</em></strong></p><table><thead><tr><th>Input Variables</th><th>Description</th><th>Format</th></tr></thead><tbody><tr><td>gex_r1_fastq</td><td>Array of FASTQ files for R1 reads</td><td>Array<!-- -->[File]</td></tr><tr><td>gex_r2_fastq</td><td>Array of FASTQ files for R2 reads</td><td>Array<!-- -->[File]</td></tr><tr><td>gex_i1_fastq</td><td>Optional FASTQ files for I1 index reads</td><td>Array<!-- -->[File]<!-- -->?</td></tr><tr><td>tar_star_reference</td><td>Reference genome in a TAR format for STAR align</td><td>File</td></tr><tr><td>annotations_gtf</td><td>Gene annotation file in GTF format</td><td>File</td></tr><tr><td>gex_whitelist</td><td>Whitelist file for cell barcodes</td><td>File</td></tr><tr><td>cloud_provider</td><td>Cloud provider for computing resources</td><td>String</td></tr><tr><td>expected_cells</td><td>Expected number of cells in the dataset</td><td>Int</td></tr><tr><td>counting_mode</td><td>Counting mode (e.g., snRNA)</td><td>String</td></tr><tr><td>tenx_chemistry_version</td><td>Version of 10X chemistry used</td><td>Int</td></tr><tr><td>emptydrops_lower</td><td>Lower threshold for EmptyDrops filtering</td><td>Int</td></tr><tr><td>force_no_check</td><td>Flag to disable sanity checks</td><td>Boolean</td></tr><tr><td>ignore_r1_read_length</td><td>Ignore length check for R1 reads</td><td>Boolean</td></tr><tr><td>star_strand_mode</td><td>Strand mode setting for STAR alignment</td><td>String</td></tr><tr><td>count_exons</td><td>Flag to enable exon counting</td><td>Boolean</td></tr><tr><td>soloMultiMappers</td><td>Optional setting for handling multi-mapped reads</td><td>String?</td></tr><tr><td>gex_nhash_id</td><td>Optional NHash identifier for gene expression</td><td>String?</td></tr><tr><td>mt_genes</td><td>Optional file listing mitochondrial genes</td><td>File?</td></tr></tbody></table><p>Example input configurations can be found in the <code>test_inputs</code> folder of the GitHub repository.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="slide-tags-pipeline-tasks-and-tools">Slide-tags Pipeline Tasks and Tools<a href="#slide-tags-pipeline-tasks-and-tools" class="hash-link" aria-label="Direct link to Slide-tags Pipeline Tasks and Tools" title="Direct link to Slide-tags Pipeline Tasks and Tools">​</a></h2><p>The workflow is composed of several key steps, implemented in separate WDL tasks:</p><table><thead><tr><th>Task</th><th>Tool</th><th>Description</th></tr></thead><tbody><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/pipelines/skylab/optimus/Optimus.wdl" target="_blank" rel="noopener noreferrer">Optimus</a></td><td>STARsolo</td><td>Gene quantification subworkflow that aligns reads to a reference genome and produces a count matrix. Read more in the <a href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus Overview</a>.</td></tr><tr><td>spatial_count</td><td><a href="https://raw.githubusercontent.com/MacoskoLab/Macosko-Pipelines/d89176cf21e072fe8b5aad3a1454ad194fca7c9a/slide-tags/spatial-count.jl" target="_blank" rel="noopener noreferrer">Custom Julia script</a> developed by the Macosko lab</td><td>Extracts spatial barcodes, performs barcode sequencing error correction, maps reads to spatial barcodes and stores unique (cell, UMI, barcode) triplets in a count matrix, and calculates quality control metrics. Produces an h5 output.</td></tr><tr><td>positioning</td><td>Custom R scripts for developed by the Macosko lab; includes <a href="https://raw.githubusercontent.com/MacoskoLab/Macosko-Pipelines/d89176cf21e072fe8b5aad3a1454ad194fca7c9a/slide-tags/positioning.R" target="_blank" rel="noopener noreferrer">positioning.R</a>, <a href="https://raw.githubusercontent.com/MacoskoLab/Macosko-Pipelines/d89176cf21e072fe8b5aad3a1454ad194fca7c9a/slide-tags/helpers.R" target="_blank" rel="noopener noreferrer">helpers.R</a>, and <a href="https://raw.githubusercontent.com/MacoskoLab/Macosko-Pipelines/d89176cf21e072fe8b5aad3a1454ad194fca7c9a/slide-tags/run-positioning.R" target="_blank" rel="noopener noreferrer">run-positioning.R</a></td><td>Takes in the rna_paths (path to the filtered cell by gene count matrix, UMI counts, and the intronic metrics) to extract cell barcodes, calculates log-transformed UMI counts, and determines mitochondrial gene percentages. Performs data normalization, PCA, clustering, and UMAP embedding for visualization and produces quality metrics and graphs. Assigns cell barcodes to spatial barcode coordinates.</td></tr></tbody></table><p>Each of these tasks utilizes scripts from the <a href="https://github.com/MacoskoLab/Macosko-Pipelines/tree/main/slide-tags" target="_blank" rel="noopener noreferrer">Macosko Lab Pipelines</a> repository, modified for streamlined output handling. Dockers for running these scripts are maintained in the warp-tools repository under <a href="https://github.com/broadinstitute/warp-tools/tree/develop/3rd-party-tools/slide-tags" target="_blank" rel="noopener noreferrer">slide-tags</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="outputs">Outputs<a href="#outputs" class="hash-link" aria-label="Direct link to Outputs" title="Direct link to Outputs">​</a></h2><h3 class="anchor anchorWithStickyNavbar_LWe7" id="optimus-outputs">Optimus outputs<a href="#optimus-outputs" class="hash-link" aria-label="Direct link to Optimus outputs" title="Direct link to Optimus outputs">​</a></h3><table><thead><tr><th>Output Variable</th><th>File Name</th><th>Description</th><th>Format</th></tr></thead><tbody><tr><td>optimus_genomic_reference_version</td><td><code>&lt;reference_version&gt;.txt</code></td><td>File containing the Genome build, source and GTF annotation version.</td><td>TXT</td></tr><tr><td>optimus_bam</td><td><code>&lt;input_id&gt;.bam</code></td><td>BAM file containing aligned reads from Optimus workflow.</td><td>BAM</td></tr><tr><td>optimus_matrix</td><td><code>&lt;input_id&gt;_gex_sparse_counts.npz</code></td><td>NPZ file containing raw gene by cell counts.</td><td>NPZ</td></tr><tr><td>optimus_matrix_row_index</td><td><code>&lt;input_id&gt;_gex_sparse_counts_row_index.npy</code></td><td>NPY file containing the row indices.</td><td>NPY</td></tr><tr><td>optimus_matrix_col_index</td><td><code>&lt;input_id&gt;_gex_sparse_counts_col_index.npy</code></td><td>NPY file containing the column indices.</td><td>NPY</td></tr><tr><td>optimus_cell_metrics</td><td><code>&lt;input_id&gt;_gex.cell_metrics.csv.gz</code></td><td>CSV file containing the per-cell (barcode) metrics.</td><td>Compressed CSV</td></tr><tr><td>optimus_gene_metrics</td><td><code>&lt;input_id&gt;.gene_metrics.csv.gz</code></td><td>CSV file containing the per-gene metrics.</td><td>Compressed CSV</td></tr><tr><td>optimus_cell_calls</td><td><code>&lt;input_id&gt;.emptyDrops.csv</code></td><td>TSV file containing the EmptyDrops results when the Optimus workflow is run in sc_rna mode.</td><td>CSV</td></tr><tr><td>optimus_h5ad_output_file</td><td><code>&lt;input_id&gt;.h5ad</code></td><td>h5ad (Anndata) file containing the raw cell-by-gene count matrix, gene metrics, cell metrics, and global attributes. See the <a href="/warp/docs/Pipelines/Optimus_Pipeline/Loom_schema">Optimus Count Matrix Overview</a> for more details.</td><td>H5AD</td></tr><tr><td>optimus_multimappers_EM_matrix</td><td><code>UniqueAndMult-EM.mtx</code></td><td>Optional output produced when <code>soloMultiMappers</code> is &quot;EM&quot;; see STARsolo <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md#multi-gene-reads" target="_blank" rel="noopener noreferrer">documentation</a> for more information.</td><td>MTX</td></tr><tr><td>optimus_multimappers_Uniform_matrix</td><td><code>UniqueAndMult-Uniform.mtx</code></td><td>Optional output produced when <code>soloMultiMappers</code> is &quot;Uniform&quot;; see STARsolo <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md#multi-gene-reads" target="_blank" rel="noopener noreferrer">documentation</a> for more information.</td><td>MTX</td></tr><tr><td>optimus_multimappers_Rescue_matrix</td><td><code>UniqueAndMult-Rescue.mtx</code></td><td>Optional output produced when <code>soloMultiMappers</code> is &quot;Rescue&quot;; see STARsolo <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md#multi-gene-reads" target="_blank" rel="noopener noreferrer">documentation</a> for more information.</td><td>MTX</td></tr><tr><td>optimus_multimappers_PropUnique_matrix</td><td><code>UniqueAndMult-PropUnique.mtx</code></td><td>Optional output produced when <code>soloMultiMappers</code> is &quot;PropUnique&quot;; see STARsolo <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md#multi-gene-reads" target="_blank" rel="noopener noreferrer">documentation</a> for more information.</td><td></td></tr><tr><td>optimus_aligner_metrics</td><td><code>&lt;input_id&gt;.star_metrics.tar</code></td><td>Text file containing per barcode metrics (<code>CellReads.stats</code>) produced by the GEX pipeline STARsolo aligner.</td><td>TAR</td></tr><tr><td>optimus_library_metrics</td><td><code>&lt;input_id&gt;_gex_&lt;gex_nhash_id&gt;_library_metrics.csv</code></td><td>Optional CSV file containing all library-level metrics calculated with STARsolo for gene expression data.</td><td>CSV</td></tr><tr><td>optimus_mtx_files</td><td><code>&lt;input_id&gt;_gex.mtx_files.tar</code></td><td>TAR file with STARsolo matrix market files (barcodes.tsv, features.tsv, and matrix.mtx)</td><td>TAR</td></tr><tr><td>cb_cell_barcodes_csv</td><td><code>&lt;cell_csv&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td><td></td></tr><tr><td>cb_checkpoint_file</td><td><code>&lt;ckpt_file&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td><td></td></tr><tr><td>cb_h5_array</td><td><code>&lt;h5_array&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td><td></td></tr><tr><td>cb_html_report_array</td><td><code>&lt;report_array&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td><td></td></tr><tr><td>cb_log</td><td><code>&lt;log&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td><td></td></tr><tr><td>cb_metrics_csv_array</td><td><code>&lt;metrics_array&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td><td></td></tr><tr><td>cb_output_directory</td><td><code>&lt;output_dir&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td><td></td></tr><tr><td>cb_summary_pdf</td><td><code>&lt;pdf&gt;</code></td><td>Optional output produced when <code>run_cellbender</code> is &quot;true&quot;; see CellBender <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html" target="_blank" rel="noopener noreferrer">documentation</a> and <a href="https://github.com/broadinstitute/CellBender/tree/master" target="_blank" rel="noopener noreferrer">GitHub repository</a> for more information.</td><td></td></tr></tbody></table><h3 class="anchor anchorWithStickyNavbar_LWe7" id="output-variables-for-the-spatial-and-positioning-components-of-the-slide-tags-pipeline">Output variables for the spatial and positioning components of the Slide-Tags pipeline<a href="#output-variables-for-the-spatial-and-positioning-components-of-the-slide-tags-pipeline" class="hash-link" aria-label="Direct link to Output variables for the spatial and positioning components of the Slide-Tags pipeline" title="Direct link to Output variables for the spatial and positioning components of the Slide-Tags pipeline">​</a></h3><p>For more details regarding the output variables, please refer to the <a href="https://github.com/MacoskoLab/Macosko-Pipelines/tree/main/slide-tags" target="_blank" rel="noopener noreferrer">README</a> in the Slide-Tags directoty in the Macosko Lab repository.</p><table><thead><tr><th>Output Variable</th><th>File Name</th><th>Description</th><th>Format</th></tr></thead><tbody><tr><td>spatial_output_h5</td><td><code>&lt;input_id&gt;_SBcounts.h5</code></td><td>h5 file containing cell by bead matrix and spatial barcode information.</td><td>H5</td></tr><tr><td>spatial_output_log</td><td><code>&lt;input_id&gt;_spatial-count.log</code></td><td>Standard output of the spatial barcodes task.</td><td>TXT</td></tr><tr><td>positioning_seurat_qs</td><td><code>&lt;input_id&gt;_seurat.qs</code></td><td>Seurat object with processed spatial transcriptomics data.</td><td>SEURAT</td></tr><tr><td>positioning_coords_csv</td><td><code>&lt;input_id&gt;_coords.csv</code></td><td>Spatial coordinates for called cells.</td><td>CSV</td></tr><tr><td>positioning_coords2_csv</td><td><code>&lt;input_id&gt;_coords2.csv</code></td><td>Alternate or refined spatial coordinates.</td><td>CSV</td></tr><tr><td>positioning_summary_pdf</td><td><code>&lt;input_id&gt;_summary.pdf</code></td><td>QC summary report with plots and metrics.</td><td>PDF</td></tr><tr><td>positioning_intermediates</td><td><code>&lt;input_id&gt;_intermediates.tar.gz</code></td><td>Contains spatial barcodes matrix, cell barcode whitelist, and spatial metadata.</td><td>TAR</td></tr><tr><td>positioning_log</td><td><code>&lt;input_id&gt;_positioning.log</code></td><td>Standard output of the positioning task.</td><td>TXT</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="versioning">Versioning<a href="#versioning" class="hash-link" aria-label="Direct link to Versioning" title="Direct link to Versioning">​</a></h2><p>All releases of the pipeline are documented in the repository’s changelog.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="citing-the-slide-tags-pipeline">Citing the Slide-tags Pipeline<a href="#citing-the-slide-tags-pipeline" class="hash-link" aria-label="Direct link to Citing the Slide-tags Pipeline" title="Direct link to Citing the Slide-tags Pipeline">​</a></h2><p>If you use the Slide-tags Pipeline in your research, please cite the original sources:</p><ul><li><strong>Macosko Lab Pipelines:</strong> <a href="https://github.com/MacoskoLab/Macosko-Pipelines/tree/main/slide-tags" target="_blank" rel="noopener noreferrer">https://github.com/MacoskoLab/Macosko-Pipelines/tree/main/slide-tags</a></li></ul><p>Please also consider citing our preprint:</p><p>Degatano, K.; Awdeh, A.; Dingman, W.; Grant, G.; Khajouei, F.; Kiernan, E.; Konwar, K.; Mathews, K.; Palis, K.; Petrillo, N.; Van der Auwera, G.; Wang, C.; Way, J.; Pipelines, W. WDL Analysis Research Pipelines: Cloud-Optimized Workflows for Biological Data Processing and Reproducible Analysis. Preprints 2024, 2024012131. <a href="https://doi.org/10.20944/preprints202401.2131.v1" target="_blank" rel="noopener noreferrer">https://doi.org/10.20944/preprints202401.2131.v1</a></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="acknowledgements">Acknowledgements<a href="#acknowledgements" class="hash-link" aria-label="Direct link to Acknowledgements" title="Direct link to Acknowledgements">​</a></h2><p>We are immensely grateful Matthew Shabet and the Macosko Lab for development of these analsyes, for their generous time making these scripts FAIR, and for the many hours working with the WARP team to incoporate the scripts into WDL. </p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="feedback">Feedback<a href="#feedback" class="hash-link" aria-label="Direct link to Feedback" title="Direct link to Feedback">​</a></h2><p>Please help us make our tools better by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>; we welcome pipeline-related suggestions or questions.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/SlideTags_Pipeline/README.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2025-10-01T18:42:30.000Z">Oct 1, 2025</time></b> by <b>jsotobroad</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/Pipelines/SlideSeq_Pipeline/count-matrix-overview"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Slide-seq Count Matrix Overview</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Smart-seq2 Single Nucleus Multi-Sample Overview</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-slide-tags-pipeline" class="table-of-contents__link toc-highlight">Introduction to the Slide-tags Pipeline</a></li><li><a href="#quickstart-table" class="table-of-contents__link toc-highlight">Quickstart Table</a></li><li><a href="#set-up" class="table-of-contents__link toc-highlight">Set-up</a><ul><li><a href="#installation" class="table-of-contents__link toc-highlight">Installation</a></li><li><a href="#inputs" class="table-of-contents__link toc-highlight">Inputs</a></li></ul></li><li><a href="#slide-tags-pipeline-tasks-and-tools" class="table-of-contents__link toc-highlight">Slide-tags Pipeline Tasks and Tools</a></li><li><a href="#outputs" class="table-of-contents__link toc-highlight">Outputs</a><ul><li><a href="#optimus-outputs" class="table-of-contents__link toc-highlight">Optimus outputs</a></li><li><a href="#output-variables-for-the-spatial-and-positioning-components-of-the-slide-tags-pipeline" class="table-of-contents__link toc-highlight">Output variables for the spatial and positioning components of the Slide-Tags pipeline</a></li></ul></li><li><a href="#versioning" class="table-of-contents__link toc-highlight">Versioning</a></li><li><a href="#citing-the-slide-tags-pipeline" class="table-of-contents__link toc-highlight">Citing the Slide-tags Pipeline</a></li><li><a href="#acknowledgements" class="table-of-contents__link toc-highlight">Acknowledgements</a></li><li><a href="#feedback" class="table-of-contents__link toc-highlight">Feedback</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.f756d810.js"></script>
<script src="/warp/assets/js/main.affc3f00.js"></script>
</body>
</html>