#!/usr/bin/env bash

###########################################################
######### Wdl definitions for Pull approve Groups #########
# Add or remove files from these definitions to change 
#   which files a group is responsible for
# All dependencies of the files listed here 
#   (and dependencies of those dependencies and ...) 
#   will be automatically added by this script
###########################################################
# scientific_owners_gdc_pipeline
gdc_wgs_wdls=("pipelines/broad/dna_seq/somatic/single_sample/wgs/gdc_genome/GDCWholeGenomeSomaticSingleSample.wdl")
gdc_wgs_files=("pipelines/broad/dna_seq/somatic/single_sample/wgs/gdc_genome/**")

# scientific_owners_cram_to_unmapped_bam
cram_to_unmapped_bams_wdls=("pipelines/broad/reprocessing/cram_to_unmapped_bams/CramToUnmappedBams.wdl")
cram_to_unmapped_bams_files=("pipelines/broad/reprocessing/cram_to_unmapped_bams/**")

# scientific_owners_annotation_filtration:
annotation_filtration_wdls=("pipelines/broad/annotation_filtration/AnnotationFiltration.wdl")
annotation_filtration_files=("pipelines/broad/annotation_filtration/**")

# scientific_owners_arrays:
arrays_wdls=("pipelines/broad/arrays/validate_chip/ValidateChip.wdl"
             "pipelines/broad/arrays/multi_sample/MultiSampleArrays.wdl"
             "pipelines/broad/arrays/single_sample/Arrays.wdl"
             "verification/VerifyArrays.wdl"
             "verification/VerifyMultiSampleArrays.wdl"
             "verification/VerifyValidateChip.wdl")
arrays_files=("pipelines/broad/arrays/**")

# scientific_owners_germline_single_sample:
germline_single_sample_wdls=("pipelines/broad/dna_seq/germline/single_sample/exome/ExomeGermlineSingleSample.wdl"
                             "pipelines/broad/dna_seq/germline/single_sample/wgs/WholeGenomeGermlineSingleSample.wdl"
                             "pipelines/broad/reprocessing/exome/ExomeReprocessing.wdl"
                             "pipelines/broad/reprocessing/external/exome/ExternalExomeReprocessing.wdl"
                             "pipelines/broad/reprocessing/wgs/WholeGenomeReprocessing.wdl"
                             "verification/VerifyGermlineSingleSample.wdl"
                             "verification/VerifyReprocessing.wdl")
germline_single_sample_files=("pipelines/broad/dna_seq/germline/single_sample/**"
                              "pipelines/broad/reprocessing/**")

# scientific_owners_joint_genotyping:
joint_genotyping_wdls=("pipelines/broad/dna_seq/germline/joint_genotyping/JointGenotyping.wdl"
                       "pipelines/broad/dna_seq/germline/joint_genotyping/by_chromosome/JointGenotypingByChromosomePartOne.wdl"
                       "pipelines/broad/dna_seq/germline/joint_genotyping/by_chromosome/JointGenotypingByChromosomePartTwo.wdl"
                       "pipelines/broad/dna_seq/germline/joint_genotyping/reblocking/ReblockGVCF.wdl")
joint_genotyping_files=("pipelines/broad/dna_seq/germline/joint_genotyping/**")

# scientific_owners_somatic_single_sample:
somatic_single_sample_wdls=("beta-pipelines/broad/somatic/single_sample/targeted/TargetedSomaticSingleSample.wdl")
somatic_single_sample_files=("beta-pipelines/broad/somatic/single_sample/targeted/**")

# clinical_owners:
clinical_wdls=("pipelines/broad/dna_seq/germline/single_sample/wgs/WholeGenomeGermlineSingleSample.wdl")


# Gets all dependencies for a list of wdls and creates a string specific to a pullapprove config condition
function get_string() {
  local -a file_list=("$@")
  local -a string_array=()
  for f in ${file_list[@]}; do
    string_array+=("'${f}' in files or")
  done
  declare -i last_index=${#string_array[@]}-1
  string_array[${last_index}]=$(echo ${string_array[${last_index}]} | rev | cut -c4- | rev)
  printf -v str "%s\n        " "${string_array[@]}"
  echo "${str}"
}


function get_dependencies_for_wdls() {
  local -a wdls=("$@")
  local -a all_dependencies=()
  for wdl in ${wdls[@]}; do
    all_dependencies=("${all_dependencies[@]}" $(get_dependencies_for_wdl ${wdl}))
  done
  all_wdls=(${wdls[@]} ${all_dependencies[@]})
  unique_wdls=($(echo ${all_wdls[@]} | xargs -n1 | sort -u | xargs))
  echo ${unique_wdls[@]}
}


function get_dependencies_for_wdl() {
  local -r wdl=${1}
  local -a wdlImports=($(grep '^import ".*$' ${wdl} | cut -d ' ' -f 2 | sed 's|\.\.\/||g' | xargs -n1))
  local -a subWorkflowImports=()
  for import in ${wdlImports[@]}; do
    subWorkflowImports=("${subWorkflowImports[@]}" $(get_dependencies_for_wdl ${import}))
  done
  echo ${wdlImports[@]} ${subWorkflowImports[@]}
}

cd ..

# Edit instructions to be added to the .pullapprove.yml file generated by this script
edit_instructions='# DO NOT EDIT THIS FILE DIRECTLY!!!
# To edit the pullapprove.yml file, edit the pullapprove_template.yml file or the scripts/process.sh script instead.
# Top level wdls for each pipeline are defined in the scripts/process.sh script.
# All other content in this file can be found in pullapprove_template.yml.
# After editing, run the process.sh script from within the scripts directory to update this file based on your changes.
# Changing this file directly will result in your changes being overwritten the next time process.sh is run.'

export EDIT_INSTRUCTIONS="${edit_instructions}"

gdc_wgs_all_wdls=$(get_dependencies_for_wdls ${gdc_wgs_wdls[@]})
cram_to_unmapped_bams_all_wdls=$(get_dependencies_for_wdls ${cram_to_unmapped_bams_wdls[@]})
annotation_filtration_all_wdls=$(get_dependencies_for_wdls ${annotation_filtration_wdls[@]})
arrays_all_wdls=$(get_dependencies_for_wdls ${arrays_wdls[@]})
germline_single_sample_all_wdls=$(get_dependencies_for_wdls ${germline_single_sample_wdls[@]})
joint_genotyping_all_wdls=$(get_dependencies_for_wdls ${joint_genotyping_wdls[@]})
somatic_single_sample_all_wdls=$(get_dependencies_for_wdls ${somatic_single_sample_wdls[@]})
clinical_all_wdls=$(get_dependencies_for_wdls ${clinical_wdls[@]})

export GDC_WGS_FILES="$(get_string ${gdc_wgs_all_wdls[@]} ${gdc_wgs_files[@]})"
export CRAM_TO_UNMAPPED_BAMS="$(get_string ${cram_to_unmapped_bams_all_wdls[@]} ${cram_to_unmapped_bams_files[@]})"
export ANNOTATION_FILTRATION_FILES="$(get_string ${annotation_filtration_all_wdls[@]} ${annotation_filtration_files[@]})"
export ARRAYS_FILES="$(get_string ${arrays_all_wdls[@]} ${arrays_files[@]})"
export GERMLINE_SINGLE_SAMPLE_FILES="$(get_string ${germline_single_sample_all_wdls[@]} ${germline_single_sample_files[@]})"
export JOINT_GENOTYPING_FILES="$(get_string ${joint_genotyping_all_wdls[@]} ${joint_genotyping_files[@]})"
export SOMATIC_SINGLE_SAMPLE_FILES="$(get_string ${somatic_single_sample_all_wdls[@]} ${somatic_single_sample_files[@]})"
export CLINICAL_FILES="$(get_string ${clinical_all_wdls[@]})"


rm .pullapprove.yml
source pullapprove_template.yml
(cat pullapprove_template.yml;) >pullapprove_temp.yml
. pullapprove_temp.yml

rm pullapprove_temp.yml


