<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.5">
<link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Blog Atom Feed">
<link rel="preconnect" href="https://www.google-analytics.com">
<script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-181048107-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview")</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><title data-react-helmet="true">CEMBA Overview | WARP</title><meta data-react-helmet="true" name="twitter:card" content="summary_large_image"><meta data-react-helmet="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README"><meta data-react-helmet="true" name="docsearch:language" content="en"><meta data-react-helmet="true" name="docsearch:version" content="current"><meta data-react-helmet="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="CEMBA Overview | WARP"><meta data-react-helmet="true" name="description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><meta data-react-helmet="true" property="og:description" content="| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |"><link data-react-helmet="true" rel="shortcut icon" href="/warp/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README" hreflang="x-default"><link data-react-helmet="true" rel="preconnect" href="https://BH4D9OD16A-dsn.algolia.net" crossorigin="anonymous"><link rel="stylesheet" href="/warp/assets/css/styles.57c5b218.css">
<link rel="preload" href="/warp/assets/js/runtime~main.2d9b7af3.js" as="script">
<link rel="preload" href="/warp/assets/js/main.2b70a255.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#" class="skipToContent_1oUP">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle toggle_3Zt9 react-toggle--disabled"><div class="react-toggle-track" role="button" tabindex="-1"><div class="react-toggle-track-check"><span class="toggle_71bT">üåú</span></div><div class="react-toggle-track-x"><span class="toggle_71bT">üåû</span></div><div class="react-toggle-thumb"></div></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div><div class="searchBox_1Doo"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docs-wrapper docs-doc-page"><div class="docPage_31aa"><button class="clean-btn backToTopButton_35hR" type="button"><svg viewBox="0 0 24 24" width="28"><path d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z" fill="currentColor"></path></svg></button><aside class="docSidebarContainer_3Kbt"><div class="sidebar_15mo"><nav class="menu thin-scrollbar menu_Bmed menuWithAnnouncementBar_2WvA"><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">About WARP</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#">WARP Pipelines</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#" tabindex="0">CEMBA</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">CEMBA Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/CEMBA.methods">CEMBA_v1.1.0 Publication Methods</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Exome Germline Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">GDC Whole Genome Somatic Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Illumina Genotyping Array</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Imputation</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Optimus</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">RNA with UMIs</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Single Cell ATAC</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Slide-seq</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Multi-Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Single Nucleus Multi-Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Ultima Genomics Whole Genome Germline</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Whole Genome Germline Single Sample</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">WDL Best Practices</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">Contribution Guide</a></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_1CGd"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_3E-R"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></aside><main class="docMainContainer_3ufF"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_3FnS"><div class="docItemContainer_33ec"><article><div class="tocCollapsible_1PrD tocMobile_3Hoh"><button type="button" class="clean-btn tocCollapsibleButton_2O1e">On this page</button></div><div class="markdown"><header><h1>CEMBA Overview</h1></header><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">CEMBA_v1.1.0</a></td><td align="center">February, 2021</td><td align="center"><a href="mailto:ekiernan@broadinstitute.org" target="_blank" rel="noopener noreferrer">Elizabeth Kiernan</a></td><td align="center">Please file GitHub issues in warp or contact <a href="mailto:warp-pipelines-help@broadinstitute.org" target="_blank" rel="noopener noreferrer">the WARP team</a></td></tr></tbody></table><p><img alt="CEMBA" src="/warp/assets/images/CEMBA-d92c4880c99cd432f81ba94fcd614461.png"></p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="introduction-to-the-cemba-workflow"></a>Introduction to the CEMBA Workflow<a class="hash-link" href="#introduction-to-the-cemba-workflow" title="Direct link to heading">#</a></h2><p>CEMBA is a pipeline developed by the <a href="https://braininitiative.nih.gov/" target="_blank" rel="noopener noreferrer">BRAIN Initiative</a> that supports the processing of multiplexed single-nuclei bisulfite sequencing data. It is an alignment and methylated base calling pipeline that trims adaptors, attaches cell barcodes, aligns reads to the genome, filters reads based on quality, and creates both a VCF and ALLC file with methylation-site coverage.</p><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>tip</h5></div><div class="admonition-content"><p>Interested in using the pipeline for your publication? See the <a href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/CEMBA.methods">‚ÄúCEMBA publication methods‚Äù</a> for a generic &quot;methods&quot; style description of the pipeline.</p></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="quick-start-table"></a>Quick Start Table<a class="hash-link" href="#quick-start-table" title="Direct link to heading">#</a></h2><table><thead><tr><th>Pipeline Features</th><th>Description</th><th>Source</th></tr></thead><tbody><tr><td>Assay Type</td><td>Single-nucleus methylcytosine sequencing (snmC-seq)</td><td><a href="https://science.sciencemag.org/content/357/6351/600/tab-pdf" target="_blank" rel="noopener noreferrer">Luo et al. 2017</a></td></tr><tr><td>Overall Workflow</td><td>Aligns reads and identifies methylated bases</td><td>Code available from <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/cemba/cemba_methylcseq/CEMBA.wdl" target="_blank" rel="noopener noreferrer">Github</a></td></tr><tr><td>Workflow Language</td><td>WDL 1.0</td><td><a href="https://github.com/openwdl/wdl" target="_blank" rel="noopener noreferrer">openWDL</a></td></tr><tr><td>Genomic Reference Sequence</td><td>GRCH38 and GRCM38</td><td><a href="https://www.gencodegenes.org/" target="_blank" rel="noopener noreferrer">GENCODE</a></td></tr><tr><td>Aligner</td><td>BISMARK v0.21.0 with  --bowtie2</td><td><a href="https://www.bioinformatics.babraham.ac.uk/projects/bismark/" target="_blank" rel="noopener noreferrer">Bismark</a></td></tr><tr><td>Variant Caller</td><td>GATK 4.1.2.0</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK 4.1.2.0</a></td></tr><tr><td>Data Input File Format</td><td>File format in which sequencing data is provided</td><td><a href="https://support.illumina.com/bulletins/2016/04/fastq-files-explained.html" target="_blank" rel="noopener noreferrer">Zipped FASTQs (.fastq.gz)</a></td></tr><tr><td>Data Output File Format</td><td>File formats in which CEMBA output is provided</td><td><a href="http://samtools.github.io/hts-specs/" target="_blank" rel="noopener noreferrer">BAM</a>, <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf" target="_blank" rel="noopener noreferrer">VCF</a>, <a href="https://github.com/yupenghe/methylpy#output-format" target="_blank" rel="noopener noreferrer">ALLC</a></td></tr></tbody></table><header><h1>Set-up</h1></header><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="cemba-installation-and-requirements"></a>CEMBA Installation and Requirements<a class="hash-link" href="#cemba-installation-and-requirements" title="Direct link to heading">#</a></h2><p>The CEMBA pipeline code is written in the <a href="https://openwdl.org/" target="_blank" rel="noopener noreferrer">Workflow Description Language (WDL)</a> and can be downloaded by cloning the GitHub repository <a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer">WARP</a>. For the latest release of CEMBA, please see the release tags prefixed with &quot;CEMBA&quot; <a href="https://github.com/broadinstitute/warp/tags" target="_blank" rel="noopener noreferrer">here</a>. CEMBA can be deployed using <a href="https://github.com/broadinstitute/cromwell" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH compliant, flexible workflow management system that supports multiple computing platforms.</p><p>The workflow can also be run on Terra using the <a href="https://app.terra.bio/#workspaces/brain-initiative-bcdc/Methyl-c-seq_Pipeline" target="_blank" rel="noopener noreferrer">Methyl-c-seq_Pipeline workspace</a>. This workspace contains the CEMBA workflow, workflow configurations, required reference data and other inputs, and demultiplexed example testing data.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="inputs"></a>Inputs<a class="hash-link" href="#inputs" title="Direct link to heading">#</a></h2><p>CEMBA pipeline inputs are detailed in the example human configuration file (<a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/cemba/cemba_methylcseq/example_inputs/CEMBA.inputs.json" target="_blank" rel="noopener noreferrer">CEMBA.inputs.json</a>). Genomic reference files were built using the <a href="https://github.com/BICCN/CEMBA/blob/master/pipelines/build_cemba_references/BuildCembaReferences.wdl" target="_blank" rel="noopener noreferrer">BuildCembaReferencesWDL script (private repository)</a>. See descriptions of all inputs in the tables below.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="sample-data-input"></a>Sample data input<a class="hash-link" href="#sample-data-input" title="Direct link to heading">#</a></h3><p>The pipeline accepts paired-end reads in the form of two compressed FASTQ files (fastq.gz). FASTQ files may represent a single cell sample or in the case of multiplexed samples, multiple cells.</p><table><thead><tr><th>Parameter Name</th><th>Description</th></tr></thead><tbody><tr><td>fastq_r1_gzipped_input</td><td>Compressed FASTQ (.gz) for R1</td></tr><tr><td>fastq_r2_gzipped_input</td><td>Compressed FASTQ (.gz) for R2</td></tr></tbody></table><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="additional-inputs"></a>Additional inputs<a class="hash-link" href="#additional-inputs" title="Direct link to heading">#</a></h3><table><thead><tr><th>Parameter Name</th><th>Description</th></tr></thead><tbody><tr><td>barcode_white_list</td><td>List of known cell barcodes</td></tr><tr><td>output_base_sample_name</td><td>Prefix for all pipeline output files (final and intermediate)</td></tr><tr><td>barcode_start_pos</td><td>Base location of barcode start</td></tr><tr><td>barcode_length</td><td>Length of cell barcode (bp)</td></tr><tr><td>reference_fasta</td><td>Reference FASTA</td></tr><tr><td>reference_fasta_index</td><td>Reference FASTA index</td></tr><tr><td>fwd_converted_reference_fasta</td><td>Bisulfite-converted forward reference genome reads for Bismark alignment</td></tr><tr><td>rev_converted_reference_fasta</td><td>Bisulfite-converted reverse reference genome reads for Bismark alignment</td></tr><tr><td>reference_dictionary</td><td>Reference genome dictionary</td></tr><tr><td>fwd_bowtie2_index_files</td><td>Forward bowtie2 index files</td></tr><tr><td>rev_bowtie2_index_files</td><td>Reverse bowtie2 index files</td></tr><tr><td>quality_cutoff</td><td>Interval representing the number of base pairs to remove from 5‚Äô and 3‚Äô end in order to trim low quality reads</td></tr><tr><td>min_length_paired_end_trim</td><td>An interval to specify a minimum read length to avoid empty reads in paired-end mode</td></tr><tr><td>min_length_single_end_trim</td><td>An interval to specify a minimum read length to avoid empty reads in single-end mode</td></tr><tr><td>read1_adapter_seq</td><td>The R1 adaptor sequence</td></tr><tr><td>read2_adapter_seq</td><td>The R2 adaptor sequence</td></tr><tr><td>cut_length</td><td>Interval provided to trim degenerate bases, random primer indexes, and Adaptase C/T tail</td></tr><tr><td>paired_end_run</td><td>Boolean; if true, workflow will run in paired-end mode</td></tr><tr><td>remove_duplicates</td><td>Boolean; if true Picard will remove duplicates and report duplication removal metrics</td></tr><tr><td>extract_and_attach_barcodes_in_single_end_run</td><td>Boolean; if true, workflow will create an unaligned BAM and extract barcodes</td></tr><tr><td>min_map_quality</td><td>Numerical value that represents minimum map quality; if provided Samtools will filter reads and produce a BAM for reads above value and reads below value</td></tr><tr><td>read_group_library_name</td><td>Library preparation type used for read group; default is &quot;Methylation&quot;</td></tr><tr><td>read_group_platform_name</td><td>Sequencing platform used for read group; default is &quot;Illumina&quot;</td></tr><tr><td>read_group_platform_unit_name</td><td>Platform unit for the read group (i.e. run barcode); default is &quot;smmC-Seq&quot;</td></tr></tbody></table><header><h1>CEMBA Tasks and Tools</h1></header><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/cemba/cemba_methylcseq/CEMBA.wdl" target="_blank" rel="noopener noreferrer">CEMBA.wdl</a> implements the workflow by importing individual &quot;tasks&quot; written in the <a href="https://openwdl.org/" target="_blank" rel="noopener noreferrer">WDL script</a>.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="cemba-task-summary"></a>CEMBA Task Summary<a class="hash-link" href="#cemba-task-summary" title="Direct link to heading">#</a></h2><p>The table and summary sections below detail the tasks and tools of the CEMBA pipeline; <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/cemba/cemba_methylcseq/CEMBA.wdl" target="_blank" rel="noopener noreferrer">the code</a> is available through GitHub. Each task can be found in the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/cemba/cemba_methylcseq/CEMBA.wdl" target="_blank" rel="noopener noreferrer">CEMBA WDL</a> If you are looking for the specific parameters of each task/tool, please see the  <code>command {}</code> section of the WDL script.</p><table><thead><tr><th align="left">Task</th><th align="left">Tool(s)</th><th align="left">Purpose</th><th align="left">Docker</th></tr></thead><tbody><tr><td align="left">Trim</td><td align="left"><a href="https://cutadapt.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cutadapt v1.18</a></td><td align="left">Trim adaptors</td><td align="left">quay.io/broadinstitute/cutadapt:1.18</td></tr><tr><td align="left">CreateUnmappedBam</td><td align="left"><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard v2.18.23</a></td><td align="left">Create uBAM for attaching barcodes</td><td align="left">quay.io/broadinstitute/picard:2.18.23</td></tr><tr><td align="left">ExtractCellBarcodes</td><td align="left"><a href="https://sctools.readthedocs.io/en/latest/sctools.html" target="_blank" rel="noopener noreferrer">sctools v0.3.4</a></td><td align="left">Use whitelist to extract barcodes and tag to uBAM</td><td align="left">quay.io/humancellatlas/secondary-analysis-sctools:v0.3.4</td></tr><tr><td align="left">Trim</td><td align="left"><a href="https://cutadapt.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cutadapt v1.18</a></td><td align="left">Trim degenerate bases, primer index, C/T Adaptase tail of R1</td><td align="left">quay.io/broadinstitute/cutadapt:1.18</td></tr><tr><td align="left">Trim</td><td align="left"><a href="https://cutadapt.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cutadapt v1.18</a></td><td align="left">Trim bases, primer index, C/T Adaptase tail of R2</td><td align="left">quay.io/broadinstitute/cutadapt:1.18</td></tr><tr><td align="left">Align</td><td align="left"><a href="https://www.bioinformatics.babraham.ac.uk/projects/bismark/" target="_blank" rel="noopener noreferrer">Bismark v0.21.0</a></td><td align="left">Map multiplexed samples as single-end with --bowtie2</td><td align="left">quay.io/broadinstitute/bismark:0.21.0</td></tr><tr><td align="left">Sort</td><td align="left"><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard v2.18.23</a></td><td align="left">Sort BAM(s) in coordinate order</td><td align="left">quay.io/broadinstitute/picard:2.18.23</td></tr><tr><td align="left">FilterDuplicates</td><td align="left"><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard v2.18.23</a></td><td align="left">Removes duplicate reads from BAM</td><td align="left">quay.io/broadinstitute/picard:2.18.23</td></tr><tr><td align="left">Get MethylationReport</td><td align="left"><a href="https://www.bioinformatics.babraham.ac.uk/projects/bismark/" target="_blank" rel="noopener noreferrer">Bismark v0.21.0</a></td><td align="left">Produce methylation report for duplicates-filtered BAM</td><td align="left">quay.io/broadinstitute/bismark:0.21.0</td></tr><tr><td align="left">FilterMapQuality</td><td align="left"><a href="http://www.htslib.org/" target="_blank" rel="noopener noreferrer">Samtools v1.9</a></td><td align="left">Further filter duplicate-removed BAM by map quality</td><td align="left">quay.io/broadinstitute/samtools:1.9</td></tr><tr><td align="left">GetMethylationReport</td><td align="left"><a href="https://www.bioinformatics.babraham.ac.uk/projects/bismark/" target="_blank" rel="noopener noreferrer">Bismark v0.21.0</a></td><td align="left">Produce methylation report for reads above map quality and below map quality</td><td align="left">quay.io/broadinstitute/bismark:0.21.0</td></tr><tr><td align="left">AttachBarcodes</td><td align="left"><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard v2.18.23</a></td><td align="left">Add barcodes from the tagged uBAM to the aligned BAM</td><td align="left">quay.io/broadinstitute/picard:2.18.23</td></tr><tr><td align="left">MergeBams</td><td align="left"><a href="http://www.htslib.org/" target="_blank" rel="noopener noreferrer">Samtools v.19</a></td><td align="left">Merge R1 and R2 BAM files into single BAM</td><td align="left">quay.io/broadinstitute/samtools:1.9</td></tr><tr><td align="left">AddReadGroup</td><td align="left"><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK v4.1.2.0</a></td><td align="left">Add read groups to the merged BAM</td><td align="left">us.gcr.io/broad-gatk/gatk:4.3.0.0</td></tr><tr><td align="left">Sort</td><td align="left"><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard v2.18.23</a></td><td align="left">Sort in coordinate order after adding read group</td><td align="left">quay.io/broadinstitute/picard:2.18.23</td></tr><tr><td align="left">IndexBam</td><td align="left"><a href="http://www.htslib.org/" target="_blank" rel="noopener noreferrer">Samtools v1.9</a></td><td align="left">Index the output BAM</td><td align="left">quay.io/broadinstitute/samtools:1.9</td></tr><tr><td align="left">MethylationTypeCaller</td><td align="left"><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK v4.1.2.0</a></td><td align="left">Produce a  VCF with locus-specific methylation information</td><td align="left">us.gcr.io/broad-gatk/gatk:4.3.0.0</td></tr><tr><td align="left">VCFtoALLC</td><td align="left">Python</td><td align="left">Creates an <a href="https://github.com/yupenghe/methylpy#output-format" target="_blank" rel="noopener noreferrer">ALLC</a> file from the VCF produced with MethylationTypeCaller</td><td align="left">quay.io/cemba/vcftoallc:v0.0.1</td></tr><tr><td align="left">ComputeCoverageDepth</td><td align="left"><a href="http://www.htslib.org/" target="_blank" rel="noopener noreferrer">Samtools v1.9</a></td><td align="left">Compute number of sites with coverage greater than 1</td><td align="left">quay.io/broadinstitute/samtools:1.9</td></tr></tbody></table><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="prior-to-running-set-up-the-workflow-for-using-multiplexed-samples"></a>Prior to running: set-up the workflow for using multiplexed samples<a class="hash-link" href="#prior-to-running-set-up-the-workflow-for-using-multiplexed-samples" title="Direct link to heading">#</a></h3><p>The pipeline uses paired-end reads, but it can only perform multiplexing when running in single-end mode. If you have multiplexed samples and want to attach cell barcodes, you must run the pipeline in single-end mode. If you do not wish to attach cell barcodes, you may run in paired-end mode (even if your samples are multiplexed). You can specify single-end mode or paired-end mode using the paired_end_run boolean in the configuration file. You will also need to adjust the extract_and_attach_barcodes_in_single_end_run boolean to true if you want to attach barcodes.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="1-trim-adaptors"></a>1. Trim adaptors<a class="hash-link" href="#1-trim-adaptors" title="Direct link to heading">#</a></h3><p>The CEMBA workflow Trim task uses Cutadapt software to remove the Read1 (R1) and Read2 (R2) adaptor sequences specified in the input configuration from the zipped R1 and R2 FASTQ files. Low-quality reads are trimmed from the 5‚Äô and 3‚Äô ends using the interval specified in the quality_cutoff input parameter.  To avoid empty reads, a threshold for read length is set using the min_length_paired_end_trim option.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="2-extract-cell-barcodes"></a>2. Extract cell barcodes<a class="hash-link" href="#2-extract-cell-barcodes" title="Direct link to heading">#</a></h3><p>CEMBA can extract cell barcodes from multiplexed samples if the extract_and_attach_barcodes_in_single_end_run boolean is true and the samples are run in single-end mode. To do this, the workflow uses the CreateUnmappedBam and ExtractCellBarcodes tasks to first make an unaligned BAM (uBAM) for the trimmed R1 FASTQ and then tag barcodes identified with the barcode_white_list input to the uBAM.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="3-trim-degenerate-bases-random-primer-indexes-and-adaptase-ct-tail"></a>3. Trim degenerate bases, random primer indexes, and Adaptase C/T tail<a class="hash-link" href="#3-trim-degenerate-bases-random-primer-indexes-and-adaptase-ct-tail" title="Direct link to heading">#</a></h3><p>After barcode extraction, the Trim task is used a second time to remove additional bases resulting from R1 random primer indexes (often used as barcodes) and the R2 C/T tail introduced by the Adaptase enzyme. Reads are trimmed using the cut_length input. The read length threshold is set by the min_length_single_end input.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="4-align-to-a-reference-genome"></a>4. Align to a reference genome<a class="hash-link" href="#4-align-to-a-reference-genome" title="Direct link to heading">#</a></h3><p>The resulting trimmed FASTQ files can be aligned to a reference either in single-end mode for multiplexed samples or paired-end mode. For all modes, the workflow aligns with Bismark with --bowtie2 option. For paired-end or the single-end mode for R1, the workflow uses a directional option with --pbat parameter. For R2, the directional option is turned off.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="5-sort-remove-duplicates-and-filter"></a>5. Sort, remove duplicates, and filter<a class="hash-link" href="#5-sort-remove-duplicates-and-filter" title="Direct link to heading">#</a></h3><p>The aligned BAM(s) are scattered and sorted in coordinate order using Picard. Duplicate reads are then removed from the sorted BAM. If a min_map_quality is provided in the input, reads will be filtered accordingly and a BAM produced for all reads above the min_map_quality and a BAM for reads below the min_map_quality.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="6-generate-methylation-reports"></a>6. Generate methylation reports<a class="hash-link" href="#6-generate-methylation-reports" title="Direct link to heading">#</a></h3><p>Methylation reports are generated using the Bismark at two steps in the workflow: after the removal of duplicates and again after filtering on min_map_quality. The bismark_methylation_extraction function with the -- comprehensive --merge_non_CpG --report options outputs multiple reports which are detailed in the <a href="https://www.bioinformatics.babraham.ac.uk/projects/bismark/Bismark_User_Guide.pdf" target="_blank" rel="noopener noreferrer">Bismark documentation</a>. These outputs include mbias, splitting, CpG context, and non-CpG context reports.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="7-attach-barcodes-merge-bams-add-read-groups-sort-and-index-bams"></a>7. Attach barcodes, merge BAMs, add read groups, sort and index BAMs<a class="hash-link" href="#7-attach-barcodes-merge-bams-add-read-groups-sort-and-index-bams" title="Direct link to heading">#</a></h3><p>In the AttachBarcodes task, Picard attaches the barcodes in the R1 uBAM to the aligned, duplicate-removed, and if applicable, filtered, R1 BAM. This produces a tagged_mapped.bam file. Once the barcodes are attached, the MergeBams task uses Samtools to merge the (barcoded if applicable) R1 BAM with the aligned and filtered R2 BAM. Read groups are then attached to the merged BAM file with GATK4 and the BAM is sorted with Picard. The BAM is indexed with Samtools.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="8-call-methylated-bases"></a>8. Call methylated bases<a class="hash-link" href="#8-call-methylated-bases" title="Direct link to heading">#</a></h3><p>Methylated bases are identified using the MethylationTypeCaller task which calls the GATK4 function MethylationTypeCaller. This produces a VCF with methylation calls.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="9-create-allc-file"></a>9. Create ALLC file<a class="hash-link" href="#9-create-allc-file" title="Direct link to heading">#</a></h3><p>The VCF containing methylation calls is used to create an additional <a href="https://github.com/yupenghe/methylpy#output-format" target="_blank" rel="noopener noreferrer">ALLC (&quot;all-cytosine&quot;) file</a> which can be used for downstream differential methylated regions in downstream analyses.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="10-compute-coverage-depth"></a>10. Compute coverage depth<a class="hash-link" href="#10-compute-coverage-depth" title="Direct link to heading">#</a></h3><p>The ComputeCoverageDepth task uses Samtools to calculate any region in the filtered, sorted BAM with a coverage depth greater than 1. This interval is read in the stdout of the workflow.</p><header><h1>Outputs</h1></header><p>The table below details the pipeline outputs. <strong>If using multiplexed samples, the final files will represent reads from multiple cells and the output is not yet split by cell barcode.</strong></p><table><thead><tr><th>Workflow Output Name</th><th>Description</th><th>File Type (when applicable)</th></tr></thead><tbody><tr><td>bam_sort_output</td><td>Final aligned, filtered, (barcoded), and sorted BAM</td><td>BAM</td></tr><tr><td>bam_index_output</td><td>Index file for the final BAM</td><td>INDEX</td></tr><tr><td>methylation_vcf</td><td>VCF file demarcating methylation sites</td><td>VCF</td></tr><tr><td>methylation_vcf_index</td><td>Index file for the VCF</td><td>INDEX</td></tr><tr><td>total_depth_count</td><td>Interval demarcated in the stdout file for reads with coverage greater than 1</td><td>NA</td></tr><tr><td>mapping_reports</td><td>Mapping reports generated by Bismark alignment</td><td>Array</td></tr><tr><td>metric_remove_dup_output</td><td>Duplication metrics generated by Picard</td><td>TXT</td></tr><tr><td>methylation_mbias_report_output</td><td>Mbias reports generated by Bismark for the sorted duplicate-removed BAM and the two map quality-filtered BAMs (reads above map quality and reads below map quality)</td><td>TXT</td></tr><tr><td>methylation_splitting_report_output</td><td>Splitting reports generated by Bismark for the sorted duplicate-removed BAM and the two map quality-filtered BAMs (reads above map quality and reads below map quality)</td><td>TXT</td></tr><tr><td>methylation_CpG_context_report_output</td><td>CpG context reports generated by Bismark for the sorted duplicate-removed BAM and the two map quality-filtered BAMs (reads above map quality and reads below map quality)</td><td>TXT</td></tr><tr><td>methylation_non_CpG_context_report_output</td><td>Non-CpG context reports generated by Bismark for the sorted duplicate-removed BAM and the two map quality-filtered BAMs (reads above map quality and reads below map quality)</td><td>TXT</td></tr></tbody></table><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="versioning"></a>Versioning<a class="hash-link" href="#versioning" title="Direct link to heading">#</a></h2><p>All CEMBA pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/cemba/cemba_methylcseq/CEMBA.changelog.md" target="_blank" rel="noopener noreferrer">CEMBA changelog</a>.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="citing-the-cemba-pipeline"></a>Citing the CEMBA Pipeline<a class="hash-link" href="#citing-the-cemba-pipeline" title="Direct link to heading">#</a></h2><p>Please identify the pipeline in your methods section using the CEMBA Pipeline&#x27;s <a href="https://scicrunch.org/scicrunch/Resources/record/nlx_144509-1/SCR_021219/resolver?q=CEMBA&amp;l=CEMBA" target="_blank" rel="noopener noreferrer">SciCrunch resource identifier</a>.</p><ul><li>Ex: <em>CEMBA MethylC Seq Pipeline (RRID:SCR_021219)</em></li></ul><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="consortia-support"></a>Consortia Support<a class="hash-link" href="#consortia-support" title="Direct link to heading">#</a></h2><p>This pipeline is supported and used by the <a href="https://biccn.org/" target="_blank" rel="noopener noreferrer">BRAIN Initiative Cell Census Network</a> (BICCN). </p><p>If your organization also uses this pipeline, we would love to list you! Please reach out to us by contacting <a href="mailto:warp-pipelines-help@broadinstitute.org" target="_blank" rel="noopener noreferrer">the WARP team</a>.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="have-suggestions"></a>Have Suggestions?<a class="hash-link" href="#have-suggestions" title="Direct link to heading">#</a></h2><p>Please help us make our tools better by contacting <a href="mailto:warp-pipelines-help@broadinstitute.org" target="_blank" rel="noopener noreferrer">the WARP team</a> for pipeline-related suggestions or questions.</p></div><footer class="docusaurus-mt-lg"><div class="row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_2_ui" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_13-_">Last updated on <b><time datetime="2023-06-01T16:02:20.000Z">6/1/2023</time></b> by <b>dependabot[bot]</b></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/warp/docs/About_WARP/VersionAndReleasePipelines"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">¬´ Version and Release Pipelines</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/CEMBA.methods"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">CEMBA_v1.1.0 Publication Methods ¬ª</div></a></div></nav></div></div><div class="col col--3"><div class="tableOfContents_35-E thin-scrollbar"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-cemba-workflow" class="table-of-contents__link">Introduction to the CEMBA Workflow</a></li><li><a href="#quick-start-table" class="table-of-contents__link">Quick Start Table</a></li><li><a href="#cemba-installation-and-requirements" class="table-of-contents__link">CEMBA Installation and Requirements</a></li><li><a href="#inputs" class="table-of-contents__link">Inputs</a><ul><li><a href="#sample-data-input" class="table-of-contents__link">Sample data input</a></li><li><a href="#additional-inputs" class="table-of-contents__link">Additional inputs</a></li></ul></li><li><a href="#cemba-task-summary" class="table-of-contents__link">CEMBA Task Summary</a><ul><li><a href="#prior-to-running-set-up-the-workflow-for-using-multiplexed-samples" class="table-of-contents__link">Prior to running: set-up the workflow for using multiplexed samples</a></li><li><a href="#1-trim-adaptors" class="table-of-contents__link">1. Trim adaptors</a></li><li><a href="#2-extract-cell-barcodes" class="table-of-contents__link">2. Extract cell barcodes</a></li><li><a href="#3-trim-degenerate-bases-random-primer-indexes-and-adaptase-ct-tail" class="table-of-contents__link">3. Trim degenerate bases, random primer indexes, and Adaptase C/T tail</a></li><li><a href="#4-align-to-a-reference-genome" class="table-of-contents__link">4. Align to a reference genome</a></li><li><a href="#5-sort-remove-duplicates-and-filter" class="table-of-contents__link">5. Sort, remove duplicates, and filter</a></li><li><a href="#6-generate-methylation-reports" class="table-of-contents__link">6. Generate methylation reports</a></li><li><a href="#7-attach-barcodes-merge-bams-add-read-groups-sort-and-index-bams" class="table-of-contents__link">7. Attach barcodes, merge BAMs, add read groups, sort and index BAMs</a></li><li><a href="#8-call-methylated-bases" class="table-of-contents__link">8. Call methylated bases</a></li><li><a href="#9-create-allc-file" class="table-of-contents__link">9. Create ALLC file</a></li><li><a href="#10-compute-coverage-depth" class="table-of-contents__link">10. Compute coverage depth</a></li></ul></li><li><a href="#versioning" class="table-of-contents__link">Versioning</a></li><li><a href="#citing-the-cemba-pipeline" class="table-of-contents__link">Citing the CEMBA Pipeline</a></li><li><a href="#consortia-support" class="table-of-contents__link">Consortia Support</a></li><li><a href="#have-suggestions" class="table-of-contents__link">Have Suggestions?</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item"><span>GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_3J9K"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></span></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright ¬© 2023 Copyright ¬© Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.2d9b7af3.js"></script>
<script src="/warp/assets/js/main.2b70a255.js"></script>
</body>
</html>