<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Deprecated_Pipelines/Single_Cell_ATAC_Seq_Pipeline/README">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">scATAC Overview | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/Single_Cell_ATAC_Seq_Pipeline/README"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="scATAC Overview | WARP"><meta data-rh="true" name="description" content="9/12/2024"><meta data-rh="true" property="og:description" content="9/12/2024"><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/Single_Cell_ATAC_Seq_Pipeline/README"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Single_Cell_ATAC_Seq_Pipeline/README" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/Single_Cell_ATAC_Seq_Pipeline/README" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.fe8d495e.js" as="script">
<link rel="preload" href="/warp/assets/js/main.affc3f00.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">CEMBA</a><button aria-label="Toggle the collapsible sidebar category &#x27;CEMBA&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/README">GDC Whole Genome Somatic Single Sample</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible menu__list-item-collapsible--active"><a class="menu__link menu__link--sublist menu__link--active" aria-current="page" aria-expanded="true" tabindex="0" href="/warp/docs/Pipelines/Single_Cell_ATAC_Seq_Pipeline/README">Single Cell ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Cell ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Deprecated_Pipelines/Single_Cell_ATAC_Seq_Pipeline/scatac.methods">scATAC v1.3.2 Methods</a></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Multi_Sample_Pipeline/README">Smart-seq2 Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Sample_Pipeline/README">Smart-seq2 Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">Deprecated Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Single Cell ATAC</span><meta itemprop="position" content="2"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>scATAC Overview</h1><div class="theme-admonition theme-admonition-warning alert alert--danger admonition_LlT9"><div class="admonitionHeading_tbUL"><span class="admonitionIcon_kALy"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M5.05.31c.81 2.17.41 3.38-.52 4.31C3.55 5.67 1.98 6.45.9 7.98c-1.45 2.05-1.7 6.53 3.53 7.7-2.2-1.16-2.67-4.52-.3-6.61-.61 2.03.53 3.33 1.94 2.86 1.39-.47 2.3.53 2.27 1.67-.02.78-.31 1.44-1.13 1.81 3.42-.59 4.78-3.42 4.78-5.56 0-2.84-2.53-3.22-1.25-5.61-1.52.13-2.03 1.13-1.89 2.75.09 1.08-1.02 1.8-1.86 1.33-.67-.41-.66-1.19-.06-1.78C8.18 5.31 8.68 2.45 5.05.32L5.03.3l.02.01z"></path></svg></span>danger</div><div class="admonitionContent_S0QG"><p>9/12/2024</p><p>We are deprecating the Single-cell ATAC Seq Pipeline. Although the code will continue to be available, we are no longer supporting it. This workflow has been replaced with the <a href="/warp/docs/Pipelines/ATAC/README">ATAC workflow</a>.</p></div></div><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Author</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">scATAC 1.3.2</a></td><td align="center">May 2024</td><td align="center">Elizabeth Kiernan</td><td align="center">Please <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">file an issue in WARP</a>.</td></tr></tbody></table><p><img loading="lazy" alt="scATAC_diagram" src="/warp/assets/images/scATAC_diagram-8a049158447e04f48529d186ea9da5f5.png" width="676" height="1030" class="img_ev3q"></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="introduction">Introduction<a href="#introduction" class="hash-link" aria-label="Direct link to Introduction" title="Direct link to Introduction">​</a></h2><p>The scATAC Pipeline was developed by the Broad DSP Pipelines team to process single cell/nucleus ATAC-seq datasets. The pipeline is based on the <a href="https://github.com/r3fang/SnapATAC" target="_blank" rel="noopener noreferrer">SnapATAC pipeline</a> described by <a href="https://www.biorxiv.org/content/10.1101/615179v2.full" target="_blank" rel="noopener noreferrer">Fang et al. (2019)</a>. Overall, the pipeline uses the python module <a href="https://github.com/r3fang/SnapTools" target="_blank" rel="noopener noreferrer">SnapTools</a> to align and process paired reads in the form of FASTQ files. It produces an hdf5-structured Snap file that includes a cell-by-bin count matrix. In addition to the Snap file, the final outputs include a GA4GH compliant aligned BAM and QC metrics.</p><div class="theme-admonition theme-admonition-tip alert alert--success admonition_LlT9"><div class="admonitionHeading_tbUL"><span class="admonitionIcon_kALy"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Want to use the scATAC Pipeline for your publication?</div><div class="admonitionContent_S0QG"><p>Check out the <a href="/warp/docs/Deprecated_Pipelines/Single_Cell_ATAC_Seq_Pipeline/scatac.methods">scATAC Publication Methods</a> to get started!</p></div></div><h3 class="anchor anchorWithStickyNavbar_LWe7" id="quick-start-table">Quick Start Table<a href="#quick-start-table" class="hash-link" aria-label="Direct link to Quick Start Table" title="Direct link to Quick Start Table">​</a></h3><table><thead><tr><th>Pipeline Features</th><th>Description</th><th>Source</th></tr></thead><tbody><tr><td>Assay Type</td><td>Single nucleus ATAC-seq</td><td><a href="https://www.biorxiv.org/content/biorxiv/early/2019/05/13/615179.full.pdf" target="_blank" rel="noopener noreferrer">Preprint here </a></td></tr><tr><td>Overall Workflow</td><td>Generates Snap file with cell-by-bin matrix</td><td>Code available from <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/scATAC/scATAC.wdl" target="_blank" rel="noopener noreferrer">GitHub</a></td></tr><tr><td>Workflow Language</td><td>WDL 1.0</td><td><a href="https://github.com/openwdl/wdl" target="_blank" rel="noopener noreferrer">openWDL</a></td></tr><tr><td>Aligner</td><td>BWA</td><td><a href="https://pubmed.ncbi.nlm.nih.gov/19451168/" target="_blank" rel="noopener noreferrer">Li H. and Durbin R., 2009</a></td></tr><tr><td>Data Input File Format</td><td>File format in which sequencing data is provided</td><td>Paired-end FASTQs with cell barcodes appended to read names (read barcode demultiplexing section <a href="https://github.com/r3fang/SnapATAC/wiki/FAQs#whatissnap" target="_blank" rel="noopener noreferrer">here</a>)</td></tr><tr><td>Data Output File Format</td><td>File formats in which scATAC output is provided</td><td><a href="http://samtools.github.io/hts-specs/" target="_blank" rel="noopener noreferrer">BAM</a>, <a href="https://github.com/r3fang/SnapATAC/wiki/FAQs#whatissnap" target="_blank" rel="noopener noreferrer">Snap</a></td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="set-up">Set-up<a href="#set-up" class="hash-link" aria-label="Direct link to Set-up" title="Direct link to Set-up">​</a></h2><h3 class="anchor anchorWithStickyNavbar_LWe7" id="workflow-installation-and-requirements">Workflow Installation and Requirements<a href="#workflow-installation-and-requirements" class="hash-link" aria-label="Direct link to Workflow Installation and Requirements" title="Direct link to Workflow Installation and Requirements">​</a></h3><p>The <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/scATAC/scATAC.wdl" target="_blank" rel="noopener noreferrer">scATAC workflow</a> is written in the Workflow Description Language WDL and can be downloaded by cloning the GitHub <a href="https://github.com/broadinstitute/warp/" target="_blank" rel="noopener noreferrer">WARP repository</a>. The workflow can be deployed using <a href="https://cromwell.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH compliant, flexible workflow management system that supports multiple computing platforms. For the latest workflow version and release notes, please see the scATAC <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/scATAC/scATAC.changelog.md" target="_blank" rel="noopener noreferrer">changelog</a>.</p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="pipeline-inputs">Pipeline Inputs<a href="#pipeline-inputs" class="hash-link" aria-label="Direct link to Pipeline Inputs" title="Direct link to Pipeline Inputs">​</a></h3><p>The pipeline inputs are detailed in the table below. You can test the workflow by using the <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/wdl/scATAC/example_inputs/human_example.json" target="_blank" rel="noopener noreferrer">human_example.json</a> example configuration file.</p><table><thead><tr><th>Input name</th><th>Input type</th><th>Description</th></tr></thead><tbody><tr><td>input_fastq1</td><td>File</td><td>FASTQ file of the first reads (R1)</td></tr><tr><td>input_fastq2</td><td>File</td><td>FASTQ file of the second reads (R2)</td></tr><tr><td>input_id</td><td>String</td><td>Unique identifier for the sample; will also be used to name the output files</td></tr><tr><td>input_reference</td><td>File</td><td>Reference bundle that is generated with bwa-mk-index-wdl found <a href="https://github.com/broadinstitute/warp/tree/develop/tasks/wdl/accessory_workflows/build_bwa_reference/bwa-mk-index.wdl" target="_blank" rel="noopener noreferrer">here</a></td></tr><tr><td>genome_name</td><td>String</td><td>Name of the genomic reference (name that precedes the “.tar” in the input_reference)</td></tr><tr><td>output_bam</td><td>String</td><td>Name for the output BAM; default is set to the <code>input_id</code> + &quot;<!-- -->_<!-- -->aligned_bam&quot;</td></tr><tr><td>bin_size_list</td><td>String</td><td>List of bin sizes used to generate cell-by-bin matrices; default is 10000 bp</td></tr></tbody></table><h3 class="anchor anchorWithStickyNavbar_LWe7" id="input-file-preparation">Input File Preparation<a href="#input-file-preparation" class="hash-link" aria-label="Direct link to Input File Preparation" title="Direct link to Input File Preparation">​</a></h3><h4 class="anchor anchorWithStickyNavbar_LWe7" id="r1-and-r2-fastq-preparation">R1 and R2 FASTQ Preparation<a href="#r1-and-r2-fastq-preparation" class="hash-link" aria-label="Direct link to R1 and R2 FASTQ Preparation" title="Direct link to R1 and R2 FASTQ Preparation">​</a></h4><p>The scATAC workflow requires paired reads in the form FASTQ files with the cell barcodes appended to the readnames. A description of the barcode demultiplexing can be found on the SnapATAC documentation (see barcode demultiplexing section <a href="https://github.com/r3fang/SnapATAC/wiki/FAQs#CEMBA_snap" target="_blank" rel="noopener noreferrer">here</a>). The full cell barcode must form the first part of the read name (for both R1 and R2 files) and be separated from the rest of the line by a colon. You can find an example python code to perform demultiplexing in the <a href="https://github.com/r3fang/SnapTools/blob/master/snaptools/dex_fastq.py" target="_blank" rel="noopener noreferrer">SnapTools documentation here</a>. The codeblock below demonstrates the correct format.</p><div class="codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#bfc7d5;--prism-background-color:#292d3e"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-text codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#bfc7d5"><span class="token plain">@CAGTTGCACGTATAGAACAAGGATAGGATAAC:7001113:915:HJ535BCX2:1:1106:1139:1926 1:N:0:0</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">ACCCTCCGTGTGCCAGGAGATACCATGAATATGCCATAGAACCTGTCTCT</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">+</span><br></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">DDDDDIIIIIIIIIIIIIIHHIIIIIIIIIIIIIIIIIIIIIIIIIIIII</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h4 class="anchor anchorWithStickyNavbar_LWe7" id="input_reference-preparation">Input_reference Preparation<a href="#input_reference-preparation" class="hash-link" aria-label="Direct link to Input_reference Preparation" title="Direct link to Input_reference Preparation">​</a></h4><p>The input_reference is a BWA compatible reference bundle in TAR file format. You can create this BWA reference using the accessory workflow  <a href="https://github.com/broadinstitute/warp/tree/develop/tasks/wdl/accessory_workflows/build_bwa_reference/bwa-mk-index.wdl" target="_blank" rel="noopener noreferrer">here</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="workflow-tasks-and-tools">Workflow Tasks and Tools<a href="#workflow-tasks-and-tools" class="hash-link" aria-label="Direct link to Workflow Tasks and Tools" title="Direct link to Workflow Tasks and Tools">​</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/wdl/scATAC/scATAC.wdl" target="_blank" rel="noopener noreferrer">scATAC workflow</a> is divided into multiple tasks which are described in the table below. The table also links to the Docker Image for each task and to the documentation or code for the relevant software tool parameters.</p><table><thead><tr><th>Task</th><th>Task Description</th><th>Tool Docker Image</th><th>Parameter Descriptions or Code</th></tr></thead><tbody><tr><td>AlignPairedEnd</td><td>Align the modified FASTQ files to the genome</td><td><a href="https://github.com/broadinstitute/warp-tools/blob/develop/3rd-party-tools/snaptools-bwa/Dockerfile" target="_blank" rel="noopener noreferrer">snaptools-bwa:1.0.0-1.4.8-0.7.17-1690310027</a></td><td><a href="https://github.com/r3fang/SnapTools" target="_blank" rel="noopener noreferrer">SnapTools documentation</a></td></tr><tr><td>SnapPre</td><td>Initial generation of snap file</td><td><a href="https://github.com/broadinstitute/warp-tools/blob/develop/3rd-party-tools/snaptools-bwa/Dockerfile" target="_blank" rel="noopener noreferrer">snaptools-bwa:1.0.0-1.4.8-0.7.17-1690310027</a></td><td><a href="https://github.com/r3fang/SnapTools" target="_blank" rel="noopener noreferrer">SnapTools documentation</a></td></tr><tr><td>SnapCellByBin</td><td>Binning of data by genomic bins</td><td><a href="https://github.com/broadinstitute/warp-tools/blob/develop/3rd-party-tools/snaptools-bwa/Dockerfile" target="_blank" rel="noopener noreferrer">snaptools-bwa:1.0.0-1.4.8-0.7.17-1690310027</a></td><td><a href="https://github.com/r3fang/SnapTools" target="_blank" rel="noopener noreferrer">SnapTools documentation</a></td></tr><tr><td>MakeCompliantBAM</td><td>Generation of a GA4GH compliant BAM</td><td><a href="https://github.com/broadinstitute/warp-tools/blob/develop/tools/Dockerfile" target="_blank" rel="noopener noreferrer">warp-tools:1.0.1-1690997141</a></td><td><a href="https://github.com/broadinstitute/warp-tools/blob/develop/tools/scripts/makeCompliantBAM.py" target="_blank" rel="noopener noreferrer">Code</a></td></tr><tr><td>BreakoutSnap</td><td>Extraction of tables from snap file into text format (for testing and user availability)</td><td><a href="https://github.com/broadinstitute/warp/tree/master/dockers/skylab/snap-breakout" target="_blank" rel="noopener noreferrer">snap-breakout:0.0.1</a></td><td><a href="https://github.com/broadinstitute/warp/tree/master/dockers/skylab/snap-breakout/breakoutSnap.py" target="_blank" rel="noopener noreferrer">Code</a></td></tr></tbody></table><h3 class="anchor anchorWithStickyNavbar_LWe7" id="task-summary">Task Summary<a href="#task-summary" class="hash-link" aria-label="Direct link to Task Summary" title="Direct link to Task Summary">​</a></h3><h4 class="anchor anchorWithStickyNavbar_LWe7" id="alignpairedend">AlignPairedEnd<a href="#alignpairedend" class="hash-link" aria-label="Direct link to AlignPairedEnd" title="Direct link to AlignPairedEnd">​</a></h4><p>The AlignPairedEnd task takes the barcode demultiplexed FASTQ files and aligns reads to the genome using the BWA aligner. It uses the SnapTools min_cov parameter to set the minimum number of barcodes a fragment requires to be included in the final output. This parameter is set to 0. The final task output is an aligned BAM file.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="snappre">SnapPre<a href="#snappre" class="hash-link" aria-label="Direct link to SnapPre" title="Direct link to SnapPre">​</a></h4><p>The SnapPre task uses SnapTools to perform preprocessing and filtering on the aligned BAM. The task outputs are a Snap file and QC metrics. The table below details the filtering parameters for the task.</p><h5 class="anchor anchorWithStickyNavbar_LWe7" id="filtering-parameters">Filtering Parameters<a href="#filtering-parameters" class="hash-link" aria-label="Direct link to Filtering Parameters" title="Direct link to Filtering Parameters">​</a></h5><table><thead><tr><th>Parameter</th><th>Description</th><th>Value</th></tr></thead><tbody><tr><td>--min-mapq</td><td>Fragments with mappability less than value will be filtered</td><td>30</td></tr><tr><td>--min-flen</td><td>Fragments of length shorter than min_flen will be filtered</td><td>0</td></tr><tr><td>--max-flen</td><td>Fragments of length bigger than min_flen will be filtered</td><td>1000</td></tr><tr><td>--keep-chrm</td><td>Boolean variable indicates whether to keep reads mapped to chrM</td><td>TRUE</td></tr><tr><td>--keep-single</td><td>Boolean variable indicates whether to keep single-end reads</td><td>TRUE</td></tr><tr><td>--keep-secondary</td><td>Boolean variable indicates whether to keep secondary alignments</td><td>FALSE</td></tr><tr><td>--max-num</td><td>Max number of barcodes to be stored. Based on the coverage, top max_barcode barcodes are selected and stored</td><td>1000000</td></tr><tr><td>--min-cov</td><td>Fragments with less than min-cov number of barcodes will be filtered</td><td>100</td></tr></tbody></table><h4 class="anchor anchorWithStickyNavbar_LWe7" id="snapcellbybin">SnapCellByBin<a href="#snapcellbybin" class="hash-link" aria-label="Direct link to SnapCellByBin" title="Direct link to SnapCellByBin">​</a></h4><p>The SnapCellByBin task uses the Snap file to create cell-by-bin count matrices in which a “1” represents a bin with an accessible region of the genome and a “0” represents an inaccessible region. The bin_size_list sets the bin size to 10,000 bp by default but can be changed by specifying the value in the inputs to the workflow.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="makecompliantbam">MakeCompliantBAM<a href="#makecompliantbam" class="hash-link" aria-label="Direct link to MakeCompliantBAM" title="Direct link to MakeCompliantBAM">​</a></h4><p>The MakeCompliantBAM task uses a <a href="https://github.com/broadinstitute/warp-tools/blob/develop/tools/scripts/makeCompliantBAM.py" target="_blank" rel="noopener noreferrer">custom python script</a> to make a GA4GH compliant BAM by moving the cellular barcodes in the read names to the CB tag.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="breakoutsnap">BreakoutSnap<a href="#breakoutsnap" class="hash-link" aria-label="Direct link to BreakoutSnap" title="Direct link to BreakoutSnap">​</a></h4><p>The BreakoutSnap task extracts data from the Snap file and exports it to individual CSVs. These CSV outputs are listed in the table in the Outputs section below. The code is available <a href="https://github.com/broadinstitute/warp/tree/master/dockers/skylab/snap-breakout/breakoutSnap.py" target="_blank" rel="noopener noreferrer">here</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="outputs">Outputs<a href="#outputs" class="hash-link" aria-label="Direct link to Outputs" title="Direct link to Outputs">​</a></h2><p>The main outputs of the scATAC workflow are the Snap file, Snap QC metrics, and the GA4GH compliant BAM file. All files with the prefix “breakout” are CSV files containing individual pieces of data from the Snap. The sessions for the Snap file are described in the <a href="https://github.com/r3fang/SnapTools" target="_blank" rel="noopener noreferrer">SnapTools documentation</a>. Additionally, you can read detailed information on the <a href="https://github.com/r3fang/SnapTools/blob/master/docs/snap_format.docx" target="_blank" rel="noopener noreferrer">Snap file fields for each session</a> (select &quot;View Raw&quot;).</p><table><thead><tr><th>Output File Name</th><th>Description</th></tr></thead><tbody><tr><td>output_snap_qc</td><td>Quality control file corresponding to the snap file</td></tr><tr><td>output_snap</td><td>Output snap file (in hdf5 container format)</td></tr><tr><td>output_aligned_bam</td><td>Output BAM file, compliant with GA4GH standards</td></tr><tr><td>breakout_barcodes</td><td>Text file containing the FM (&#x27;Fragment session&#x27;) barcodeLen and barcodePos fields</td></tr><tr><td>breakout_fragments</td><td>Text file containing the FM (&#x27;Fragments session&#x27;) fragChrom, fragLen, and fragStart fields</td></tr><tr><td>breakout_binCoordinates</td><td>Text file with the AM session (&#x27;Cell x bin accessibility&#x27; matrix) binChrom and binStart fields</td></tr><tr><td>breakout_binCounts</td><td>Text file with the AM session (&#x27;Cell x bin accessibility&#x27; matrix) idx, idy, and count fields</td></tr><tr><td>breakout_barcodesSection</td><td>Text file with the data from the BD session (&#x27;Barcode session&#x27; table)</td></tr><tr><td>output_pipeline_version</td><td>Strong describing the version of the scATAC pipeline used.</td></tr></tbody></table><h5 class="anchor anchorWithStickyNavbar_LWe7" id="snap-qc-metrics">Snap QC Metrics<a href="#snap-qc-metrics" class="hash-link" aria-label="Direct link to Snap QC Metrics" title="Direct link to Snap QC Metrics">​</a></h5><p>The following table details the metrics available in the output_snap_qc file.</p><table><thead><tr><th>QC Metric</th><th>Abbreviation</th></tr></thead><tbody><tr><td>Total number of unique barcodes</td><td>No abbreviation</td></tr><tr><td>Total number of fragments</td><td>TN</td></tr><tr><td>Total number of uniquely mapped</td><td>UM</td></tr><tr><td>Total number of single ends</td><td>SE</td></tr><tr><td>Total number of secondary alignments</td><td>SA</td></tr><tr><td>Total number of paired ends</td><td>PE</td></tr><tr><td>Total number of proper paired</td><td>PP</td></tr><tr><td>Total number of proper frag len</td><td>PL</td></tr><tr><td>Total number of usable fragments</td><td>US</td></tr><tr><td>Total number of unique fragments</td><td>UQ</td></tr><tr><td>Total number of chrM fragments</td><td>CM</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="running-on-terra">Running on Terra<a href="#running-on-terra" class="hash-link" aria-label="Direct link to Running on Terra" title="Direct link to Running on Terra">​</a></h2><p><a href="https://app.terra.bio/" target="_blank" rel="noopener noreferrer">Terra</a> is a public, cloud-based platform for biomedical research. If you would like to try the scATAC workflow (previously named &quot;snap-atac&quot;) in Terra, you can import the most recent version from <a href="https://dockstore.org/workflows/github.com/broadinstitute/warp/scATAC:scATAC_v1.1.0?tab=info" target="_blank" rel="noopener noreferrer">Dockstore</a>. Additionally, there is a public <a href="https://app.terra.bio/#workspaces/brain-initiative-bcdc/scATAC" target="_blank" rel="noopener noreferrer">scATAC workspace</a> preloaded with the scATAC workflow and downsampled data.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="versioning">Versioning<a href="#versioning" class="hash-link" aria-label="Direct link to Versioning" title="Direct link to Versioning">​</a></h2><p>All scATAC workflow releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/scATAC/scATAC.changelog.md" target="_blank" rel="noopener noreferrer">scATAC changelog</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="citing-the-scatac-pipeline">Citing the scATAC Pipeline<a href="#citing-the-scatac-pipeline" class="hash-link" aria-label="Direct link to Citing the scATAC Pipeline" title="Direct link to Citing the scATAC Pipeline">​</a></h2><p>If you use the scATAC Pipeline in your research, please identify the pipeline in your methods section using the <a href="https://scicrunch.org/resources/data/record/nlx_144509-1/SCR_018919/resolver?q=SCR_018919&amp;l=SCR_018919&amp;i=rrid:scr_018919" target="_blank" rel="noopener noreferrer">scATAC SciCrunch resource identifier</a>.</p><ul><li>Ex: <em>scATAC Pipeline (RRID:SCR_018919)</em></li></ul><p>Please also consider citing our preprint:</p><p>Degatano, K., Awdeh, A., Cox III, R.S., Dingman, W., Grant, G., Khajouei, F., Kiernan, E., Konwar, K., Mathews, K.L., Palis, K., et al. Warp Analysis Research Pipelines: Cloud-optimized workflows for biological data processing and reproducible analysis. Bioinformatics 2025; btaf494. <a href="https://doi.org/10.1093/bioinformatics/btaf494" target="_blank" rel="noopener noreferrer">https://doi.org/10.1093/bioinformatics/btaf494</a></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="consortia-support">Consortia Support<a href="#consortia-support" class="hash-link" aria-label="Direct link to Consortia Support" title="Direct link to Consortia Support">​</a></h2><p>This pipeline is supported and used by the <a href="https://biccn.org/" target="_blank" rel="noopener noreferrer">BRAIN Initiative Cell Census Network</a> (BICCN). </p><p>If your organization also uses this pipeline, we would love to list you! Please reach out to us by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="pipeline-improvements">Pipeline Improvements<a href="#pipeline-improvements" class="hash-link" aria-label="Direct link to Pipeline Improvements" title="Direct link to Pipeline Improvements">​</a></h2><p>Please help us make our tools better by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>; we welcome pipeline-related suggestions or questions.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Deprecated_Pipelines/Single_Cell_ATAC_Seq_Pipeline/README.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2025-10-03T19:23:16.000Z">Oct 3, 2025</time></b> by <b>Robert Sidney Cox</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/Pipelines/Genomic_Data_Commons_Whole_Genome_Somatic/README"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Genomic Data Commons (GDC) Whole Genome Somatic Single Sample Overview</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/Deprecated_Pipelines/Single_Cell_ATAC_Seq_Pipeline/scatac.methods"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">scATAC v1.3.2 Methods</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction" class="table-of-contents__link toc-highlight">Introduction</a><ul><li><a href="#quick-start-table" class="table-of-contents__link toc-highlight">Quick Start Table</a></li></ul></li><li><a href="#set-up" class="table-of-contents__link toc-highlight">Set-up</a><ul><li><a href="#workflow-installation-and-requirements" class="table-of-contents__link toc-highlight">Workflow Installation and Requirements</a></li><li><a href="#pipeline-inputs" class="table-of-contents__link toc-highlight">Pipeline Inputs</a></li><li><a href="#input-file-preparation" class="table-of-contents__link toc-highlight">Input File Preparation</a></li></ul></li><li><a href="#workflow-tasks-and-tools" class="table-of-contents__link toc-highlight">Workflow Tasks and Tools</a><ul><li><a href="#task-summary" class="table-of-contents__link toc-highlight">Task Summary</a></li></ul></li><li><a href="#outputs" class="table-of-contents__link toc-highlight">Outputs</a></li><li><a href="#running-on-terra" class="table-of-contents__link toc-highlight">Running on Terra</a></li><li><a href="#versioning" class="table-of-contents__link toc-highlight">Versioning</a></li><li><a href="#citing-the-scatac-pipeline" class="table-of-contents__link toc-highlight">Citing the scATAC Pipeline</a></li><li><a href="#consortia-support" class="table-of-contents__link toc-highlight">Consortia Support</a></li><li><a href="#pipeline-improvements" class="table-of-contents__link toc-highlight">Pipeline Improvements</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.fe8d495e.js"></script>
<script src="/warp/assets/js/main.affc3f00.js"></script>
</body>
</html>