<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-Pipelines/snM3C/README">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">Single Nucleus Methyl-Seq and Chromatin Capture (snm3C) Overview | WARP</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/snm3C/README"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Single Nucleus Methyl-Seq and Chromatin Capture (snm3C) Overview | WARP"><meta data-rh="true" name="description" content="| Pipeline Version | Date Updated | Documentation Authors | Questions or Feedback |"><meta data-rh="true" property="og:description" content="| Pipeline Version | Date Updated | Documentation Authors | Questions or Feedback |"><link data-rh="true" rel="icon" href="/warp/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/snm3C/README"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/snm3C/README" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/snm3C/README" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://BX0CVVQ6KQ-dsn.algolia.net" crossorigin="anonymous"><link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Atom Feed">

<link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-NZ4BLRHQYS"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-NZ4BLRHQYS",{anonymize_ip:!0})</script>


<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><link rel="stylesheet" href="/warp/assets/css/styles.ddc3fdb8.css">
<link rel="preload" href="/warp/assets/js/runtime~main.a2ee2c75.js" as="script">
<link rel="preload" href="/warp/assets/js/main.435ed8cb.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title text--truncate">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/About_WARP/BestPractices">About WARP</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/warp/docs/Pipelines/ATAC/README">WARP Pipelines</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/ATAC/README">ATAC</a><button aria-label="Toggle the collapsible sidebar category &#x27;ATAC&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/BuildIndices_Pipeline/README">BuildIndices</a><button aria-label="Toggle the collapsible sidebar category &#x27;BuildIndices&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Exome_Germline_Single_Sample_Pipeline/README">Exome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Exome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Illumina_Genotyping_Arrays_Pipeline/README">Illumina Genotyping Array</a><button aria-label="Toggle the collapsible sidebar category &#x27;Illumina Genotyping Array&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/ImputationBeagle_Pipeline/README">ImputationBeagle</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Imputation_Pipeline/README">Imputation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Imputation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/JointGenotyping_Pipeline/README">JointGenotyping</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Multiome_Pipeline/README">Multiome scATAC and GEX</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Optimus_Pipeline/README">Optimus</a><button aria-label="Toggle the collapsible sidebar category &#x27;Optimus&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/PairedTag_Pipeline/README">Paired-Tag</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README">RNA with UMIs</a><button aria-label="Toggle the collapsible sidebar category &#x27;RNA with UMIs&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/SlideSeq_Pipeline/README">Slide-seq</a><button aria-label="Toggle the collapsible sidebar category &#x27;Slide-seq&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/SlideTags_Pipeline/README">Slide-tags</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/README">Smart-seq2 Single Nucleus Multi-Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Smart-seq2 Single Nucleus Multi-Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible menu__list-item-collapsible--active"><a class="menu__link menu__link--sublist menu__link--active" aria-current="page" aria-expanded="true" tabindex="0" href="/warp/docs/Pipelines/snm3C/README">Single Nucleus Methyl-Seq and Chromatin Capture</a><button aria-label="Toggle the collapsible sidebar category &#x27;Single Nucleus Methyl-Seq and Chromatin Capture&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/snM3C/summary_metrics">snm3C Mapping Summary Metrics Overview</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/snM3C/snm3c.methods">snM3C v4.0.1 Methods</a></li></ul></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/Ultima_Genomics_Whole_Genome_Germline_Pipeline/README">Ultima Genomics Whole Genome Germline</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/warp/docs/Pipelines/Whole_Genome_Germline_Single_Sample_Pipeline/README">Whole Genome Germline Single Sample</a><button aria-label="Toggle the collapsible sidebar category &#x27;Whole Genome Germline Single Sample&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Best_practices/GC_cost_optimization">WDL Best Practices</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/warp/docs/Pipelines/CEMBA_MethylC_Seq_Pipeline/README">Deprecated Pipelines</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/warp/docs/contribution/README">Contribution Guide</a><button aria-label="Toggle the collapsible sidebar category &#x27;Contribution Guide&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/warp/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">WARP Pipelines</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Single Nucleus Methyl-Seq and Chromatin Capture</span><meta itemprop="position" content="2"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Single Nucleus Methyl-Seq and Chromatin Capture (snm3C) Overview</h1><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Authors</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">snm3C_v4.0.1</a></td><td align="center">October, 2024</td><td align="center">Kaylee Mathews</td><td align="center">Please <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">file an issue in WARP</a>.</td></tr></tbody></table><p><img loading="lazy" alt="snm3C_diagram" src="/warp/assets/images/snm3C_diagram-cad90f278de866a4b04174f5552803be.png" width="960" height="2240" class="img_ev3q"></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="introduction-to-snm3c">Introduction to snm3C<a href="#introduction-to-snm3c" class="hash-link" aria-label="Direct link to Introduction to snm3C" title="Direct link to Introduction to snm3C">​</a></h2><p>The Single Nucleus Methly-Seq and Chromatin Capture (snm3C) workflow is an open-source, cloud-optimized computational workflow for processing single-nucleus methylome and chromatin contact (snm3C) sequencing data. The workflow is designed to demultiplex and align raw sequencing reads, call chromatin contacts, and generate summary metrics. </p><p>The workflow is developed in collaboration with Hanqing Liu, Wei Tian, Wubin Ding, Huaming Chen, Chongyuan Luo, Jingtian Zhou, and the entire laboratory of Joseph Ecker. Please see the <a href="#acknowledgements">Acknowledgments</a> section below.</p><p>For more information about the snm3C tools and analysis, please see the <a href="https://hq-1.gitbook.io/mc/" target="_blank" rel="noopener noreferrer">YAP documentation</a> or the cemba_data GitHub repositories created by <a href="https://github.com/lhqing/cemba_data" target="_blank" rel="noopener noreferrer">Hanqing Liu</a> and <a href="https://github.com/DingWB/cemba_data" target="_blank" rel="noopener noreferrer">Wubin Ding</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="quickstart-table">Quickstart table<a href="#quickstart-table" class="hash-link" aria-label="Direct link to Quickstart table" title="Direct link to Quickstart table">​</a></h2><p>The following table provides a quick glance at the Multiome pipeline features:</p><table><thead><tr><th>Pipeline features</th><th>Description</th><th>Source</th></tr></thead><tbody><tr><td>Assay type</td><td>single-nucleus methylome and chromatin contact (snm3C) sequencing data</td><td><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6765423/" target="_blank" rel="noopener noreferrer">Lee et al. 2019</a></td></tr><tr><td>Overall workflow</td><td>Read alignment and chromatin contact calling</td><td></td></tr><tr><td>Workflow language</td><td>WDL 1.0</td><td><a href="https://github.com/openwdl/wdl" target="_blank" rel="noopener noreferrer">openWDL</a></td></tr><tr><td>Genomic Reference Sequence</td><td>GRCh38 human genome primary sequence</td><td>GENCODE <a href="https://www.gencodegenes.org/human/release_43.html" target="_blank" rel="noopener noreferrer">human reference files</a></td></tr><tr><td>Aligner</td><td>HISAT-3N</td><td><a href="https://genome.cshlp.org/content/31/7/1290" target="_blank" rel="noopener noreferrer">Zhang at al. 2021</a></td></tr><tr><td>Data input file format</td><td>File format in which sequencing data is provided</td><td><a href="https://academic.oup.com/nar/article/38/6/1767/3112533" target="_blank" rel="noopener noreferrer">FASTQ</a></td></tr><tr><td>Data output file format</td><td>File formats in which snm3C output is provided</td><td>CSV, <a href="https://academic.oup.com/nar/article/38/6/1767/3112533" target="_blank" rel="noopener noreferrer">FASTQ</a>, <a href="http://samtools.github.io/hts-specs/" target="_blank" rel="noopener noreferrer">BAM</a>, and <a href="https://lhqing.github.io/ALLCools/intro.html" target="_blank" rel="noopener noreferrer">ALLC</a></td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="set-up">Set-up<a href="#set-up" class="hash-link" aria-label="Direct link to Set-up" title="Direct link to Set-up">​</a></h2><h3 class="anchor anchorWithStickyNavbar_LWe7" id="snm3c-installation">snm3C installation<a href="#snm3c-installation" class="hash-link" aria-label="Direct link to snm3C installation" title="Direct link to snm3C installation">​</a></h3><p>To download the latest snm3C release, see the release tags prefixed with &quot;snm3C&quot; on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a>. All snm3C pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/snm3C/snm3C.changelog.md" target="_blank" rel="noopener noreferrer">snm3C changelog</a>. </p><p>To discover and search releases, use the WARP command-line tool <a href="https://github.com/broadinstitute/warp/tree/master/wreleaser" target="_blank" rel="noopener noreferrer">Wreleaser</a>.</p><p>If you’re running a version of the snm3C workflow prior to the latest release, the accompanying documentation for that release may be downloaded with the source code on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a> (see the source code folder <code>website/docs/Pipelines/snm3C</code>).</p><p>The snm3C workflow can be deployed using <a href="https://cromwell.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH-compliant, flexible workflow management system that supports multiple computing platforms. The workflow can also be run in <a href="https://app.terra.bio" target="_blank" rel="noopener noreferrer">Terra</a>, a cloud-based analysis platform. </p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="inputs">Inputs<a href="#inputs" class="hash-link" aria-label="Direct link to Inputs" title="Direct link to Inputs">​</a></h3><p>The snm3C workflow requires a JSON configuration file specifying the input files and parameters for the analysis. Example configuration files can be found in the snm3C <a href="https://github.com/broadinstitute/warp/tree/develop/pipelines/wdl/snm3C/test_inputs" target="_blank" rel="noopener noreferrer"><code>test_inputs</code></a> directory in the WARP repository.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="input-descriptions">Input descriptions<a href="#input-descriptions" class="hash-link" aria-label="Direct link to Input descriptions" title="Direct link to Input descriptions">​</a></h4><table><thead><tr><th>Parameter</th><th>Description</th></tr></thead><tbody><tr><td>fastq_input_read1</td><td>Array of multiplexed FASTQ files for read 1.</td></tr><tr><td>fastq_input_read2</td><td>Array of multiplexed FASTQ files for read 2.</td></tr><tr><td>random_primer_indexes</td><td>File containing random primer indexes.</td></tr><tr><td>plate_id</td><td>String specifying the plate ID.</td></tr><tr><td>tarred_index_files</td><td>File containing tarred index files for hisat-3 mapping.</td></tr><tr><td>genome_fa</td><td>File containing the reference genome in FASTA format.</td></tr><tr><td>chromosome_sizes</td><td>File containing the genome chromosome sizes.</td></tr><tr><td>r1_adapter</td><td>Optional string describing the adapter sequence for read 1 paired-end reads to be used during adapter trimming with Cutadapt; default is &quot;AGATCGGAAGAGCACACGTCTGAAC&quot;.</td></tr><tr><td>r2_adapter</td><td>Optional string describing the adapter sequence for read 2 paired-end reads to be used during adapter trimming with Cutadapt; default is  &quot;AGATCGGAAGAGCGTCGTGTAGGGA&quot;.</td></tr><tr><td>r1_left_cut</td><td>Optional integer describing the number of bases to be trimmed from the beginning of read 1 with Cutadapt; default is 10.</td></tr><tr><td>r1_right_cut</td><td>Optional integer describing the number of bases to be trimmed from the end of read 1 with Cutadapt; default is 10.</td></tr><tr><td>r2_left_cut</td><td>Optional integer describing the number of bases to be trimmed from the beginning of read 2 with Cutadapt; default is 10.</td></tr><tr><td>r2_right_cut</td><td>Optional integer describing the number of bases to be trimmed from the end of read 2 with Cutadapt; default is 10.</td></tr><tr><td>min_read_length</td><td>Optional integer; if a read length is smaller than <code>min_read_length</code>, both paired-end reads will be discarded; default is 30.</td></tr><tr><td>num_upstr_bases</td><td>Optional integer describing the number of bases upstream of the C base to include in ALLC file context column created using ALLCools; default is 0.</td></tr><tr><td>num_downstr_bases</td><td>Optional integer describing the number of bases downstream of the C base to include in ALLC file context column created using ALLCools; default is 2.</td></tr><tr><td>compress_level</td><td>Optional integer describing the compression level for the output ALLC file; default is 5.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="snm3c-tasks-and-tools">snm3C tasks and tools<a href="#snm3c-tasks-and-tools" class="hash-link" aria-label="Direct link to snm3C tasks and tools" title="Direct link to snm3C tasks and tools">​</a></h2><p>The workflow contains several tasks described below.</p><p>Overall, the snm3C workflow:</p><ol><li>Demultiplexes reads.</li><li>Sorts, filters, trims, and aligns paired-end reads, and separates unmapped, uniquely aligned, and multi-aligned reads.</li><li>Aligns unmapped, single-end reads and removes overlapping reads.</li><li>Merges mapped reads, calls chromatin contacts, and creates ALLC files.</li><li>Creates summary output file.</li></ol><p>The tools each snm3C task employs are detailed in the table below. </p><p>To see specific tool parameters, select the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/snm3C/snm3C.wdl" target="_blank" rel="noopener noreferrer">workflow WDL link</a>; then find the task and view the <code>command {}</code> section of the task in the WDL script. To view or use the exact tool software, see the task&#x27;s Docker image which is specified in the task WDL <code># runtime values</code> section as <code>docker: </code>. More details about these tools and parameters can be found in the <a href="https://hq-1.gitbook.io/mc/" target="_blank" rel="noopener noreferrer">YAP documentation</a>.</p><table><thead><tr><th>Task name</th><th>Tool</th><th>Software</th><th>Description</th></tr></thead><tbody><tr><td>Demultiplexing</td><td>Cutadapt</td><td><a href="https://cutadapt.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cutadapt</a></td><td>Performs demultiplexing to cell-level FASTQ files based on random primer indices.</td></tr><tr><td>Hisat-paired-end</td><td>Cutadapt, HISAT-3N, <a href="https://github.com/lhqing/cemba_data/blob/788e83cd66f3b556bdfacf3485bed9500d381f23/cemba_data/hisat3n/hisat3n_general.py" target="_blank" rel="noopener noreferrer">hisat3n_general.py</a>, <a href="https://github.com/lhqing/cemba_data/blob/bf6248239074d0423d45a67d83da99250a43e50c/cemba_data/hisat3n/hisat3n_m3c.py" target="_blank" rel="noopener noreferrer">hisat3n_m3c.py</a></td><td><a href="https://cutadapt.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cutadapt</a>, <a href="https://daehwankimlab.github.io/hisat2/hisat-3n/" target="_blank" rel="noopener noreferrer">HISAT-3N</a>, python3</td><td>Sorts, filters, and trims reads using the <code>r1_adapter</code>, <code>r2_adapter</code>, <code>r1_left_cut</code>, <code>r1_right_cut</code>, <code>r2_left_cut</code>, and <code>r2_right_cut</code> input parameters; performs paired-end read alignment; imports 2 custom python3 scripts developed by Hanqing Liu and calls the <code>separate_unique_and_multi_align_reads()</code> and <code>split_hisat3n_unmapped_reads()</code> functions to separate unmapped, uniquely aligned, multi-aligned reads from HISAT-3N BAM file, then splits the unmapped reads FASTQ file by all possible enzyme cut sites and output new R1 and R2 FASTQ files; unmapped reads are stored in unmapped FASTQ files and uniquely and multi-aligned reads are stored in separate BAM files.</td></tr><tr><td>Hisat_single_end</td><td>HISAT-3N, <a href="https://github.com/lhqing/cemba_data/blob/bf6248239074d0423d45a67d83da99250a43e50c/cemba_data/hisat3n/hisat3n_m3c.py" target="_blank" rel="noopener noreferrer">hisat3n_m3c.py</a></td><td><a href="https://daehwankimlab.github.io/hisat2/hisat-3n/" target="_blank" rel="noopener noreferrer">HISAT-3N</a>, python3</td><td>Performs single-end alignment of unmapped reads to maximize read mapping, imports a custom python3 script developed by Hanqing Liu, and calls the <code>remove_overlap_read_parts()</code> function to remove overlapping reads from the split alignment BAM file produced during single-end alignment.</td></tr><tr><td>Merge_sort_analyze</td><td>merge, sort, MarkDuplicates, <a href="https://github.com/lhqing/cemba_data/blob/bf6248239074d0423d45a67d83da99250a43e50c/cemba_data/hisat3n/hisat3n_m3c.py" target="_blank" rel="noopener noreferrer">hisat3n_m3c.py</a>, bam-to-allc, extract-allc</td><td><a href="https://www.htslib.org/" target="_blank" rel="noopener noreferrer">samtools</a>, <a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a>, python3, <a href="https://lhqing.github.io/ALLCools/intro.html" target="_blank" rel="noopener noreferrer">ALLCools</a></td><td>Merges and sorts all mapped reads from the paired-end and single-end alignments; creates a position-sorted BAM file and a name-sorted BAM file; removes duplicate reads from the position-sorted, merged BAM file; imports a custom python3 script developed by Hanqing Liu and calls the <code>call_chromatin_contacts()</code> function to call chromatin contacts from the name-sorted, merged BAM file; reads are considered chromatin contacts if they are greater than 2,500 base pairs apart; creates a first ALLC file with a list of methylation points and a second ALLC file containing methylation contexts.</td></tr><tr><td>Summary_PerCellOutput</td><td>Custom function</td><td>bash</td><td>Untar files needed at per cell level.</td></tr><tr><td>Summary</td><td><a href="https://github.com/lhqing/cemba_data/blob/788e83cd66f3b556bdfacf3485bed9500d381f23/cemba_data/hisat3n/summary.py" target="_blank" rel="noopener noreferrer">summary.py</a></td><td>python3</td><td>Imports a custom python3 script developed by Hanqing Liu and calls the <code>snm3c_summary()</code> function to generate a single, summary file for the pipeline in CSV format; contains trimming, mapping, deduplication, chromatin contact, and AllC site statistics.</td></tr></tbody></table><h4 class="anchor anchorWithStickyNavbar_LWe7" id="1-demultiplexes-reads">1. Demultiplexes reads<a href="#1-demultiplexes-reads" class="hash-link" aria-label="Direct link to 1. Demultiplexes reads" title="Direct link to 1. Demultiplexes reads">​</a></h4><p>In the first step of the pipeline (<code>Demultiplexing</code>), raw sequencing reads are demultiplexed by random primer index into cell-level FASTQ files using <a href="https://cutadapt.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cutadapt</a>. For more information on barcoding, see the <a href="https://hq-1.gitbook.io/mc/tech-background/barcoding#two-round-of-barcoding" target="_blank" rel="noopener noreferrer">YAP documentation</a>. </p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="2-sorts-filters-trims-and-aligns-paired-end-reads-and-separates-unmapped-uniquely-aligned-and-multi-aligned-reads">2. Sorts, filters, trims, and aligns paired-end reads, and separates unmapped, uniquely aligned, and multi-aligned reads<a href="#2-sorts-filters-trims-and-aligns-paired-end-reads-and-separates-unmapped-uniquely-aligned-and-multi-aligned-reads" class="hash-link" aria-label="Direct link to 2. Sorts, filters, trims, and aligns paired-end reads, and separates unmapped, uniquely aligned, and multi-aligned reads" title="Direct link to 2. Sorts, filters, trims, and aligns paired-end reads, and separates unmapped, uniquely aligned, and multi-aligned reads">​</a></h4><p><strong>Sorts, filters, and trims reads</strong>
After demultiplexing, the pipeline uses <a href="https://cutadapt.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cutadapt</a> to sort, filter, and trim reads in the <code>hisat-paired-end</code> task. The R1 and R2 adapter sequences are removed, along with the number of bases specified by the <code>r1_left_cut</code>, <code>r1_right_cut</code>, <code>r2_left_cut</code>, and <code>r2_right_cut</code> input parameters. Any reads shorter than the specified <code>min_read_length</code> are filtered out in this step.</p><p><strong>Aligns paired-end reads</strong>
Next, the task uses <a href="https://daehwankimlab.github.io/hisat2/hisat-3n/" target="_blank" rel="noopener noreferrer">HISAT-3N</a> to perform paired-end read alignment to a reference genome FASTA file (<code>genome_fa</code>) and outputs an aligned BAM file. Additionally, the task outputs a stats file and a text file containing the genomic reference version used.</p><p><strong>Separates unmapped, uniquely aligned, and multi-aligned reads</strong>
After paired-end alignment, the task imports a custom python3 script (<a href="https://github.com/lhqing/cemba_data/blob/788e83cd66f3b556bdfacf3485bed9500d381f23/cemba_data/hisat3n/hisat3n_general.py" target="_blank" rel="noopener noreferrer">hisat3n_general.py</a>) developed by Hanqing Liu. The task calls the script&#x27;s <code>separate_unique_and_multi_align_reads()</code> function to separate unmapped, uniquely aligned, and multi-aligned reads from the HISAT-3N BAM file. Three new files are output from this step of the pipeline: </p><ol><li>A FASTQ file that contains the unmapped reads (<code>unmapped_fastq_tar</code>)</li><li>A BAM file that contains the uniquely aligned reads (<code>unique_bam_tar</code>)</li><li>A BAM file that contains the multi-aligned reads (<code>multi_bam_tar</code>)</li></ol><p>After separating reads, the task imports a custom python3 script (<a href="https://github.com/lhqing/cemba_data/blob/bf6248239074d0423d45a67d83da99250a43e50c/cemba_data/hisat3n/hisat3n_m3c.py" target="_blank" rel="noopener noreferrer">hisat3n_m3c.py</a>) developed by Hanqing Liu and calls the script&#x27;s <code>split_hisat3n_unmapped_reads()</code> function. This splits the FASTQ file containing the unmapped reads by all possible enzyme cut sites and outputs new R1 and R2 files. </p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="3-aligns-unmapped-single-end-reads-and-removes-overlapping-reads">3. Aligns unmapped, single-end reads and removes overlapping reads<a href="#3-aligns-unmapped-single-end-reads-and-removes-overlapping-reads" class="hash-link" aria-label="Direct link to 3. Aligns unmapped, single-end reads and removes overlapping reads" title="Direct link to 3. Aligns unmapped, single-end reads and removes overlapping reads">​</a></h4><p>In the next step of the pipeline, the <code>Hisat_single_end </code> task uses <a href="https://daehwankimlab.github.io/hisat2/hisat-3n/" target="_blank" rel="noopener noreferrer">HISAT-3N</a> to perform single-end read alignment of the previously unmapped reads to maximize read mapping and outputs a single, aligned BAM file.</p><p>After the second alignment step, the task imports a custom python3 script (<a href="https://github.com/lhqing/cemba_data/blob/bf6248239074d0423d45a67d83da99250a43e50c/cemba_data/hisat3n/hisat3n_m3c.py" target="_blank" rel="noopener noreferrer">hisat3n_m3c.py</a>) developed by Hanqing Liu. The task calls the script&#x27;s <code>remove_overlap_read_parts()</code> function to remove overlapping reads from the BAM file produced during single-end alignment and output another BAM file.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="4-merges-mapped-reads-calls-chromatin-contacts-and-creates-allc-files">4. Merges mapped reads, calls chromatin contacts, and creates ALLC files<a href="#4-merges-mapped-reads-calls-chromatin-contacts-and-creates-allc-files" class="hash-link" aria-label="Direct link to 4. Merges mapped reads, calls chromatin contacts, and creates ALLC files" title="Direct link to 4. Merges mapped reads, calls chromatin contacts, and creates ALLC files">​</a></h4><p><strong>Merged mapped reads</strong>
The <code>Merge_sort_analyze</code> task uses <a href="https://www.htslib.org/" target="_blank" rel="noopener noreferrer">samtools</a> to merge and sort all of the mapped reads from the paired-end and single-end alignments into a single BAM file. The BAM file is output as both a position-sorted and a name-sorted BAM file.</p><p>After merging, the task uses Picard&#x27;s MarkDuplicates tool to remove duplicate reads from the position-sorted, merged BAM file and output a deduplicated BAM file.</p><p><strong>Calls chromatin contacts</strong>
Next, the pipeline imports a custom python3 script (<a href="https://github.com/lhqing/cemba_data/blob/bf6248239074d0423d45a67d83da99250a43e50c/cemba_data/hisat3n/hisat3n_m3c.py" target="_blank" rel="noopener noreferrer">hisat3n_m3c.py</a>) developed by Hanqing Liu. The task calls the script&#x27;s <code>call_chromatin_contacts()</code> function to call chromatin contacts from the name-sorted, merged BAM file. If reads are greater than 2,500 base pairs apart, they are considered chromatin contacts. If reads are less than 2,500 base pairs apart, they are considered the same fragment. </p><p><strong>Creates ALLC files</strong>
After calling chromatin contacts, the task uses the <a href="https://lhqing.github.io/ALLCools/intro.html" target="_blank" rel="noopener noreferrer">ALLCools</a> <code>bam-to-allc</code> function to create an ALLC file from the deduplicated BAM file that contains a list of methylation points. The <code>num_upstr_bases</code> and <code>num_downstr_bases</code> input parameters are used to define the number of bases upstream and downstream of the C base to include in the ALLC context column.</p><p>Next, the task uses the <a href="https://lhqing.github.io/ALLCools/intro.html" target="_blank" rel="noopener noreferrer">ALLCools</a> <code>extract-allc</code> function to extract methylation contexts from the input ALLC file and output a second ALLC file that can be used to generate an <a href="https://github.com/lhqing/allcools_doc/blob/master/tech-background/file-formats.md#mcds-file" target="_blank" rel="noopener noreferrer">MCDS file</a>.</p><h4 class="anchor anchorWithStickyNavbar_LWe7" id="6-creates-summary-output-file">6. Creates summary output file<a href="#6-creates-summary-output-file" class="hash-link" aria-label="Direct link to 6. Creates summary output file" title="Direct link to 6. Creates summary output file">​</a></h4><p>In the last step of the pipeline, the <code>summary</code> task imports a custom python3 script (<a href="https://github.com/lhqing/cemba_data/blob/788e83cd66f3b556bdfacf3485bed9500d381f23/cemba_data/hisat3n/summary.py" target="_blank" rel="noopener noreferrer">summary.py</a>) developed by Hanqing Liu. The task calls the script&#x27;s <code>snm3c_summary()</code> function to generate a single, summary file for the pipeline in CSV format; contains trimming, mapping, deduplication, chromatin contact, and AllC site statistics. This is the main output of the pipeline.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="outputs">Outputs<a href="#outputs" class="hash-link" aria-label="Direct link to Outputs" title="Direct link to Outputs">​</a></h2><p>The following table lists the output variables and files produced by the pipeline.</p><table><thead><tr><th>Output name</th><th>Filename, if applicable</th><th>Output format and description</th></tr></thead><tbody><tr><td>MappingSummary</td><td><code>&lt;plate_id&gt;_MappingSummary.csv.gz</code></td><td>Mapping summary file in CSV format.</td></tr><tr><td>name_sorted_bams</td><td><code>&lt;plate_id&gt;.hisat3n_dna.all_reads.name_sort.tar.gz</code></td><td>Array of tarred files containing name-sorted, merged BAM files.</td></tr><tr><td>unique_reads_cgn_extraction_allc</td><td><code>&lt;plate_id&gt;.allc.tsv.tar.gz</code></td><td>Array of tarred files containing list of methylation points.</td></tr><tr><td>unique_reads_cgn_extraction_tbi</td><td><code>&lt;plate_id&gt;.allc.tbi.tar.gz</code></td><td>Array of tarred files containing ALLC index files.</td></tr><tr><td>reference_version</td><td><code>&lt;plate_id&gt;.reference_version.txt</code></td><td>Array of tarred files containing the genomic reference version used.</td></tr><tr><td>all_reads_dedup_contacts</td><td><code>&lt;plate_id&gt;.hisat3n_dna.all_reads.dedup_contacts.tar.gz</code></td><td>Array of tarred TSV files containing deduplicated chromatin contacts.</td></tr><tr><td>all_reads_3C_contacts</td><td><code>&lt;plate_id&gt;.hisat3n_dna.all_reads.3C.contact.tar.gz</code></td><td>Array of tarred TSV files containing chromatin contacts in Hi-C format.</td></tr><tr><td>chromatin_contact_stats</td><td><code>&lt;plate_id&gt;.chromatin_contact_stats.tar.gz</code></td><td>Array of tarred files containing chromatin contact statistics.</td></tr><tr><td>unique_reads_cgn_extraction_allc_extract</td><td><code>&lt;plate_id&gt;.extract-allc.tar.gz</code></td><td>Array of tarred files containing CGN context-specific ALLC files that can be used to generate an <a href="https://github.com/lhqing/allcools_doc/blob/master/tech-background/file-formats.md#mcds-file" target="_blank" rel="noopener noreferrer">MCDS file</a>.</td></tr><tr><td>unique_reads_cgn_extraction_tbi_extract</td><td><code>&lt;plate_id&gt;.extract-allc_tbi.tar.gz</code></td><td>Array of tarred files containing ALLC index files.</td></tr></tbody></table><h2 class="anchor anchorWithStickyNavbar_LWe7" id="versioning">Versioning<a href="#versioning" class="hash-link" aria-label="Direct link to Versioning" title="Direct link to Versioning">​</a></h2><p>All snm3C pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/develop/pipelines/wdl/snm3C/snm3C.changelog.md" target="_blank" rel="noopener noreferrer">pipeline changelog</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="citing-the-snm3c-pipeline">Citing the snm3C Pipeline<a href="#citing-the-snm3c-pipeline" class="hash-link" aria-label="Direct link to Citing the snm3C Pipeline" title="Direct link to Citing the snm3C Pipeline">​</a></h2><p>If you use the snm3C Pipeline in your research, please identify the pipeline in your methods section using the <a href="https://scicrunch.org/resources/data/record/nlx_144509-1/SCR_025041/resolver?q=SCR_025041&amp;l=SCR_025041&amp;i=rrid:scr_025041" target="_blank" rel="noopener noreferrer">snm3C SciCrunch resource identifier</a>.</p><ul><li>Ex: <em>snm3C Pipeline (RRID:SCR_025041)</em></li></ul><p>Please cite the following publications for the snm3C pipeline: </p><p>Lee, DS., Luo, C., Zhou, J. et al. Simultaneous profiling of 3D genome structure and DNA methylation in single human cells. Nat Methods 16, 999–1006 (2019). <a href="https://doi.org/10.1038/s41592-019-0547-z" target="_blank" rel="noopener noreferrer">https://doi.org/10.1038/s41592-019-0547-z</a></p><p>Liu, H., Zhou, J., Tian, W. et al. DNA methylation atlas of the mouse brain at single-cell resolution. Nature 598, 120–128 (2021). <a href="https://doi.org/10.1038/s41586-020-03182-8" target="_blank" rel="noopener noreferrer">https://doi.org/10.1038/s41586-020-03182-8</a></p><p>Please cite the following preprint for the WARP repository and website:</p><p>Degatano, K., Awdeh, A., Cox III, R.S., Dingman, W., Grant, G., Khajouei, F., Kiernan, E., Konwar, K., Mathews, K.L., Palis, K., et al. Warp Analysis Research Pipelines: Cloud-optimized workflows for biological data processing and reproducible analysis. Bioinformatics 2025; btaf494. <a href="https://doi.org/10.1093/bioinformatics/btaf494" target="_blank" rel="noopener noreferrer">https://doi.org/10.1093/bioinformatics/btaf494</a></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="consortia-support">Consortia support<a href="#consortia-support" class="hash-link" aria-label="Direct link to Consortia support" title="Direct link to Consortia support">​</a></h2><p>This pipeline is supported by the <a href="https://braininitiative.nih.gov/" target="_blank" rel="noopener noreferrer">BRAIN Initiative</a> (BICCN and BICAN). </p><p>If your organization also uses this pipeline, we would like to list you! Please reach out to us by <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">filing an issue in WARP</a>.</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="acknowledgements">Acknowledgements<a href="#acknowledgements" class="hash-link" aria-label="Direct link to Acknowledgements" title="Direct link to Acknowledgements">​</a></h2><p>We are immensely grateful to the members of the BRAIN Initiative (<a href="https://brainblog.nih.gov/brain-blog/brain-issues-suite-funding-opportunities-advance-brain-cell-atlases-through-centers" target="_blank" rel="noopener noreferrer">BICAN</a> Sequencing Working Group) and <a href="https://nida.nih.gov/about-nida/organization/divisions/division-neuroscience-behavior-dnb/basic-research-hiv-substance-use-disorder/scorch-program" target="_blank" rel="noopener noreferrer">SCORCH</a> for their invaluable and exceptional contributions to this pipeline. Our heartfelt appreciation goes to our collaborators and the developers of these tools, Hanqing Liu, Wei Tian, Wubin Ding, Huaming Chen, Chongyuan Luo, Jingtian Zhou, and the entire laboratory of Joseph Ecker. </p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="feedback">Feedback<a href="#feedback" class="hash-link" aria-label="Direct link to Feedback" title="Direct link to Feedback">​</a></h2><p>For questions, suggestions, or feedback related to the snm3C pipeline, please <a href="https://github.com/broadinstitute/warp/issues" target="_blank" rel="noopener noreferrer">file an issue in WARP</a>. Your feedback is valuable for improving the pipeline and addressing any issues that may arise during its usage.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/snM3C/README.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2026-02-11T20:27:07.000Z">Feb 11, 2026</time></b> by <b>jsotobroad</b></span></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/warp/docs/Pipelines/Smart-seq2_Single_Nucleus_Multi_Sample_Pipeline/consortia-processing"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Consortia Data Processing</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/warp/docs/Pipelines/snM3C/summary_metrics"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">snm3C Mapping Summary Metrics Overview</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-snm3c" class="table-of-contents__link toc-highlight">Introduction to snm3C</a></li><li><a href="#quickstart-table" class="table-of-contents__link toc-highlight">Quickstart table</a></li><li><a href="#set-up" class="table-of-contents__link toc-highlight">Set-up</a><ul><li><a href="#snm3c-installation" class="table-of-contents__link toc-highlight">snm3C installation</a></li><li><a href="#inputs" class="table-of-contents__link toc-highlight">Inputs</a></li></ul></li><li><a href="#snm3c-tasks-and-tools" class="table-of-contents__link toc-highlight">snm3C tasks and tools</a></li><li><a href="#outputs" class="table-of-contents__link toc-highlight">Outputs</a></li><li><a href="#versioning" class="table-of-contents__link toc-highlight">Versioning</a></li><li><a href="#citing-the-snm3c-pipeline" class="table-of-contents__link toc-highlight">Citing the snm3C Pipeline</a></li><li><a href="#consortia-support" class="table-of-contents__link toc-highlight">Consortia support</a></li><li><a href="#acknowledgements" class="table-of-contents__link toc-highlight">Acknowledgements</a></li><li><a href="#feedback" class="table-of-contents__link toc-highlight">Feedback</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2026 Copyright © Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.a2ee2c75.js"></script>
<script src="/warp/assets/js/main.435ed8cb.js"></script>
</body>
</html>