<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.5">
<link rel="alternate" type="application/rss+xml" href="/warp/blog/rss.xml" title="WARP Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/warp/blog/atom.xml" title="WARP Blog Atom Feed">
<link rel="preconnect" href="https://www.google-analytics.com">
<script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-181048107-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview")</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
<link rel="search" type="application/opensearchdescription+xml" title="WARP" href="/warp/opensearch.xml">
<script>!function(t,h,e,j,s,n){t.hj=t.hj||function(){(t.hj.q=t.hj.q||[]).push(arguments)},t._hjSettings={hjid:2427684,hjsv:6},s=h.getElementsByTagName("head")[0],(n=h.createElement("script")).async=1,n.src="https://static.hotjar.com/c/hotjar-"+t._hjSettings.hjid+".js?sv="+t._hjSettings.hjsv,s.appendChild(n)}(window,document)</script><title data-react-helmet="true">RNA with UMIs Overview | WARP</title><meta data-react-helmet="true" name="twitter:card" content="summary_large_image"><meta data-react-helmet="true" property="og:url" content="https://broadinstitute.github.io/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README"><meta data-react-helmet="true" name="docsearch:language" content="en"><meta data-react-helmet="true" name="docsearch:version" content="current"><meta data-react-helmet="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="RNA with UMIs Overview | WARP"><meta data-react-helmet="true" name="description" content="| Pipeline Version | Date Updated | Documentation Authors | Questions or Feedback |"><meta data-react-helmet="true" property="og:description" content="| Pipeline Version | Date Updated | Documentation Authors | Questions or Feedback |"><link data-react-helmet="true" rel="shortcut icon" href="/warp/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://broadinstitute.github.io/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://broadinstitute.github.io/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README" hreflang="x-default"><link data-react-helmet="true" rel="preconnect" href="https://BH4D9OD16A-dsn.algolia.net" crossorigin="anonymous"><link rel="stylesheet" href="/warp/assets/css/styles.57c5b218.css">
<link rel="preload" href="/warp/assets/js/runtime~main.ba3180c6.js" as="script">
<link rel="preload" href="/warp/assets/js/main.1308a829.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#" class="skipToContent_1oUP">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/warp/"><b class="navbar__title">WARP Pipelines</b></a></div><div class="navbar__items navbar__items--right"><a class="navbar__item navbar__link navbar__link--active" href="/warp/docs/get-started">Documentation</a><a class="navbar__item navbar__link" href="/warp/blog">Blog</a><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle toggle_3Zt9 react-toggle--disabled"><div class="react-toggle-track" role="button" tabindex="-1"><div class="react-toggle-track-check"><span class="toggle_71bT">ðŸŒœ</span></div><div class="react-toggle-track-x"><span class="toggle_71bT">ðŸŒž</span></div><div class="react-toggle-thumb"></div></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div><div class="searchBox_1Doo"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docs-wrapper docs-doc-page"><div class="docPage_31aa"><button class="clean-btn backToTopButton_35hR" type="button"><svg viewBox="0 0 24 24" width="28"><path d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z" fill="currentColor"></path></svg></button><aside class="docSidebarContainer_3Kbt"><div class="sidebar_15mo"><nav class="menu thin-scrollbar menu_Bmed menuWithAnnouncementBar_2WvA"><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" href="/warp/docs/get-started">Welcome to WARP</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">About WARP</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#">WARP Pipelines</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">CEMBA</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Exome Germline Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">GDC Whole Genome Somatic Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Illumina Genotyping Array</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Imputation</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Optimus</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#" tabindex="0">RNA with UMIs</a><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/README">RNA with UMIs Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/rna-with-umis.methods">RNA with UMIs v1.0.6 Methods</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Single Cell ATAC</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Slide-seq</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Multi-Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Single Nucleus Multi-Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Smart-seq2 Single Sample</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Ultima Genomics Whole Genome Germline</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#" tabindex="0">Whole Genome Germline Single Sample</a></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">WDL Best Practices</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#">Contribution Guide</a></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_1CGd"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_3E-R"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></aside><main class="docMainContainer_3ufF"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_3FnS"><div class="docItemContainer_33ec"><article><div class="tocCollapsible_1PrD tocMobile_3Hoh"><button type="button" class="clean-btn tocCollapsibleButton_2O1e">On this page</button></div><div class="markdown"><header><h1>RNA with UMIs Overview</h1></header><table><thead><tr><th align="center">Pipeline Version</th><th align="center">Date Updated</th><th align="center">Documentation Authors</th><th align="center">Questions or Feedback</th></tr></thead><tbody><tr><td align="center"><a href="https://github.com/broadinstitute/warp/releases?q=RNAwithUMIs&amp;expanded=true" target="_blank" rel="noopener noreferrer">RNAWithUMIsPipeline_v1.0.6</a></td><td align="center">April, 2022</td><td align="center"><a href="mailto:ekiernan@broadinstitute.org" target="_blank" rel="noopener noreferrer">Elizabeth Kiernan</a> &amp; <a href="mailto:kmathews@broadinstitute.org" target="_blank" rel="noopener noreferrer">Kaylee Mathews</a></td><td align="center">Please file GitHub issues in warp or contact <a href="mailto:warp-pipelines-help@broadinstitute.org" target="_blank" rel="noopener noreferrer">the WARP team</a></td></tr></tbody></table><p><img alt="RNAWithUMIs_diagram" src="/warp/assets/images/rna-with-umis_diagram-bce76bb42e6044bea0ff32d184622560.png"></p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="introduction-to-the-rna-with-umis-workflow"></a>Introduction to the RNA with UMIs workflow<a class="hash-link" href="#introduction-to-the-rna-with-umis-workflow" title="Direct link to heading">#</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl" target="_blank" rel="noopener noreferrer">RNA with UMIs pipeline</a> is an open-source, cloud-optimized workflow for processing total RNA isolated with the Transcriptome Capture (TCap) method. TCap is a technique that hybridizes exome baits to cDNA library preparations to better facilitate RNA-sequencing in low-input or poor quality (degraded) samples. These libraries may additionally be prepared with Unique Molecular Identifiers (UMIs), which can help distinguish biological signal from noise resulting from PCR amplification.</p><p>Overall, the workflow performs UMI correction, aligns reads to the genome, quantifies gene counts, and calculates quality metrics. The workflow produces genome- and transcriptome-aligned BAMs with indices and quality metrics files. </p><p>While this workflow was created to be used with TCap RNA-seq data, it can be used to process any bulk RNA-seq data. </p><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Want to use the RNA with UMIs pipeline for your publication?</h5></div><div class="admonition-content"><p>Check out the <a href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/rna-with-umis.methods">RNA with UMIs Methods</a> section to get started!</p></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="quickstart-table"></a>Quickstart table<a class="hash-link" href="#quickstart-table" title="Direct link to heading">#</a></h2><p>The following table provides a quick glance at the RNA with UMIs pipeline features:</p><table><thead><tr><th>Pipeline features</th><th>Description</th><th>Source</th></tr></thead><tbody><tr><td>Assay type</td><td>TCap (or any bulk) RNA-seq data</td><td><a href="https://genome.cshlp.org/content/25/9/1372" target="_blank" rel="noopener noreferrer">Cieslik et al. 2015</a></td></tr><tr><td>Overall workflow</td><td>Read alignment and transcriptome quantification</td><td>Code available from <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl" target="_blank" rel="noopener noreferrer">GitHub</a></td></tr><tr><td>Workflow language</td><td>WDL 1.0</td><td><a href="https://github.com/openwdl/wdl" target="_blank" rel="noopener noreferrer">openWDL</a></td></tr><tr><td>Genomic reference sequence</td><td>GRCh38 (hg38) and GRCh37 (hg19) human genome primary sequence</td><td>Genome Reference Consortium <a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.39" target="_blank" rel="noopener noreferrer">GRCh38</a> and <a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.25/" target="_blank" rel="noopener noreferrer">GRCh37</a></td></tr><tr><td>Gene annotations</td><td>GENCODE v34 (hg38) and v19 (hg19) gene annotations</td><td>GENCODE <a href="https://www.gencodegenes.org/human/release_34.html" target="_blank" rel="noopener noreferrer">v34</a> and <a href="https://www.gencodegenes.org/human/release_19.html" target="_blank" rel="noopener noreferrer">v19</a></td></tr><tr><td>Aligner</td><td>STAR</td><td><a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">STAR</a></td></tr><tr><td>Transcript quantification and metric calculation</td><td>RNA-SeQC, Picard, and GATK</td><td><a href="https://github.com/getzlab/rnaseqc" target="_blank" rel="noopener noreferrer">RNA-SeQC</a>, <a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a>, and <a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td></tr><tr><td>Data input file format</td><td>File format in which sequencing data is provided</td><td><a href="http://samtools.github.io/hts-specs/" target="_blank" rel="noopener noreferrer">BAM</a> or <a href="https://academic.oup.com/nar/article/38/6/1767/3112533" target="_blank" rel="noopener noreferrer">FASTQ</a></td></tr><tr><td>Data output file formats</td><td>File formats in which RNA with UMIs outputs are provided</td><td><a href="http://samtools.github.io/hts-specs/" target="_blank" rel="noopener noreferrer">BAM</a>; GCT (counts); TXT, TSV, and PDF (metrics)</td></tr></tbody></table><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="set-up"></a>Set-up<a class="hash-link" href="#set-up" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="installation"></a>Installation<a class="hash-link" href="#installation" title="Direct link to heading">#</a></h3><p>To download the latest release of the RNA with UMIs pipeline, see the release tags prefixed with &quot;RNAwithUMIs&quot; on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a>. All releases of the RNA with UMIs pipeline are documented in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/broad/rna_seq/RNAWithUMIsPipeline.changelog.md" target="_blank" rel="noopener noreferrer">RNA with UMIs changelog</a>. </p><p>To search releases of this and other pipelines, use the WARP command-line tool <a href="https://github.com/broadinstitute/warp/tree/master/wreleaser" target="_blank" rel="noopener noreferrer">Wreleaser</a>.</p><p>If youâ€™re running an RNA with UMIs workflow version prior to the latest release, the accompanying documentation for that release may be downloaded with the source code on the WARP <a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer">releases page</a> (see the source code folder <code>website/docs/Pipelines/RNA_with_UMIs_Pipeline</code>).</p><p>The RNA with UMIs pipeline can be deployed using <a href="https://cromwell.readthedocs.io/en/stable/" target="_blank" rel="noopener noreferrer">Cromwell</a>, a GA4GH compliant, flexible workflow management system that supports multiple computing platforms. The workflow can also be run in <a href="https://app.terra.bio" target="_blank" rel="noopener noreferrer">Terra</a>, a cloud-based analysis platform. </p><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="inputs"></a>Inputs<a class="hash-link" href="#inputs" title="Direct link to heading">#</a></h3><p>The RNA with UMIs workflow inputs are specified in JSON configuration files. Example configuration files can be found in the <a href="https://github.com/broadinstitute/warp/tree/master/pipelines/broad/rna_seq/test_inputs" target="_blank" rel="noopener noreferrer">test_inputs</a> folder in the WARP repository.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="input-descriptions"></a>Input descriptions<a class="hash-link" href="#input-descriptions" title="Direct link to heading">#</a></h4><p>The workflow takes in either a set of paired-end FASTQ files or a read group unmapped BAM that has gone through base calling. </p><table><thead><tr><th>Input variable name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>bam</td><td>Read group-specific unmapped BAM file; alternatively, paired-end FASTQ files (<code>r1_fastq</code> and <code>r2_fastq</code>) may be used.</td><td>File</td></tr><tr><td>r1_fastq</td><td>Read 1 FASTQ file; alternatively, the unmapped bam file (<code>bam</code>) may be used as input.</td><td>File</td></tr><tr><td>r2_fastq</td><td>Read 2 FASTQ file; alternatively, the unmapped bam file (<code>bam</code>) may be used as input.</td><td>File</td></tr><tr><td>read1Structure</td><td>String describing how the bases in a sequencing run should be allocated into logical reads for read 1 by fgbio&#x27;s <a href="http://fulcrumgenomics.github.io/fgbio/tools/latest/ExtractUmisFromBam.html" target="_blank" rel="noopener noreferrer">ExtractUmisFromBam</a> tool; for more information about read structures, see the <a href="https://github.com/fulcrumgenomics/fgbio/wiki/Read-Structures" target="_blank" rel="noopener noreferrer">fgbio documentation</a>.</td><td>String</td></tr><tr><td>read2Structure</td><td>String describing how the bases in a sequencing run should be allocated into logical reads for read 2 by fgbio&#x27;s <a href="http://fulcrumgenomics.github.io/fgbio/tools/latest/ExtractUmisFromBam.html" target="_blank" rel="noopener noreferrer">ExtractUmisFromBam</a> tool; for more information about read structures, see the <a href="https://github.com/fulcrumgenomics/fgbio/wiki/Read-Structures" target="_blank" rel="noopener noreferrer">fgbio documentation</a>.</td><td>String</td></tr><tr><td>output_basename</td><td>String used as a prefix in workflow output files.</td><td>String</td></tr><tr><td>platform</td><td>String used to describe the sequencing platform.</td><td>String</td></tr><tr><td>library_name</td><td>String used to describe the library.</td><td>String</td></tr><tr><td>platform_unit</td><td>String used to describe the platform unit.</td><td>String</td></tr><tr><td>read_group_name</td><td>String used to describe the read group name.</td><td>String</td></tr><tr><td>sequencing_center</td><td>String used to describe the sequencing center; default is set to &quot;BI&quot;.</td><td>String</td></tr><tr><td>starIndex</td><td>TAR file containing genome indices used for the <a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">STAR aligner</a>.</td><td>File</td></tr><tr><td>gtf</td><td>Gene annotation file (GTF) used for the <a href="https://github.com/getzlab/rnaseqc" target="_blank" rel="noopener noreferrer">RNA-SeQC</a> tool.</td><td>File</td></tr><tr><td>ref</td><td>FASTA file used for metric collection with <a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a> tools.</td><td>File</td></tr><tr><td>refIndex</td><td>FASTA index file used for metric collection with <a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a> tools.</td><td>File</td></tr><tr><td>refDict</td><td>Dictionary file used for metric collection with <a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a> tools.</td><td>File</td></tr><tr><td>refFlat</td><td>refFlat file used for metric collection with <a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a> tools.</td><td>File</td></tr><tr><td>ribosomalIntervals</td><td>Intervals file used for RNA metric collection with <a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a> tools.</td><td>File</td></tr><tr><td>exonBedFile</td><td>Bed file used for fragment size calculations with the <a href="https://github.com/getzlab/rnaseqc" target="_blank" rel="noopener noreferrer">RNA-SeQC</a> tool; contains non-overlapping exons.</td><td>File</td></tr><tr><td>population_vcf</td><td>VCF file used for contamination estimation by GATK&#x27;s <a href="https://gatk.broadinstitute.org/hc/en-us/articles/4418051471643-GetPileupSummaries" target="_blank" rel="noopener noreferrer">GetPileupSummaries</a> and <a href="https://gatk.broadinstitute.org/hc/en-us/articles/4418054253211-CalculateContamination" target="_blank" rel="noopener noreferrer">CalculateContamination</a> tools; contains common SNP sites from population-wide studies.</td><td>File</td></tr><tr><td>population_vcf_index</td><td>Population VCF index file used for contamination estimation by GATK&#x27;s <a href="https://gatk.broadinstitute.org/hc/en-us/articles/4418051471643-GetPileupSummaries" target="_blank" rel="noopener noreferrer">GetPileupSummaries</a> and <a href="https://gatk.broadinstitute.org/hc/en-us/articles/4418054253211-CalculateContamination" target="_blank" rel="noopener noreferrer">CalculateContamination</a> tools.</td><td>File</td></tr></tbody></table><h3><a aria-hidden="true" tabindex="-1" class="anchor anchor__h3 anchorWithStickyNavbar_31ik" id="references"></a>References<a class="hash-link" href="#references" title="Direct link to heading">#</a></h3><p>The pipeline supports both hg19 and hg38 references. The reference set consists of:</p><ol><li>.fasta, .fai, and .dict files</li><li>STAR index</li><li>GTF file (RNASeQC)</li><li>ribosomal interval list and refFlat file (Picard)</li></ol><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="fasta-index-and-dictionary-files"></a>FASTA, index, and dictionary files<a class="hash-link" href="#fasta-index-and-dictionary-files" title="Direct link to heading">#</a></h4><p>When running the workflow with the hg38 reference, we recommend using a version without HLA, ALT, and decoy contigs. These non-primary assembly contigs lead to reduced sensitivity unless the mapper is ALT-aware (e.g., bwa-mem). STAR is not ALT-aware, so these contigs should be removed.</p><p>In contrast, the hg19 reference does not have nearly as many contigs as the hg38 reference, so the workflow can be run using the standard hg19 reference stored in Broad&#x27;s <a href="https://console.cloud.google.com/storage/browser/_details/gcp-public-data--broad-references/hg19/v0/Homo_sapiens_assembly19.fasta" target="_blank" rel="noopener noreferrer">public reference bucket</a>.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="gtf-file"></a>GTF file<a class="hash-link" href="#gtf-file" title="Direct link to heading">#</a></h4><p>Genome annotation files (GTFs) contain information about genes, such as the start and end coordinates of each exon, the name of the gene, and the type of the transcript (e.g., protein-coding, antisense). The workflow uses the GENCODE v34 GTF for hg38 and v19 for hg19.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="ribosomal-interval-list-and-refflat-file"></a>Ribosomal interval list and refFlat file<a class="hash-link" href="#ribosomal-interval-list-and-refflat-file" title="Direct link to heading">#</a></h4><p>The workflow ribosomal interval list and the refFlat file are used by Picard metrics calculation tools. The workflow uses a custom ribosomal interval list based on the public hg38 ribosomal interval list, which has been modified to include mitochondrial rRNA coding genes. </p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="additional-reference-resources"></a>Additional reference resources<a class="hash-link" href="#additional-reference-resources" title="Direct link to heading">#</a></h4><p>For more information about ALT contigs, HLA, decoys, and ALT-aware mapping, see the following resources:</p><ul><li><a href="https://lh3.github.io/2017/11/13/which-human-reference-genome-to-use" target="_blank" rel="noopener noreferrer">Heng Li&#x27;s blog</a> where he argues for removing ALT contigs to improve sensitivity (for both DNA and RNA). </li><li><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360041155232-Reference-Genome-Components" target="_blank" rel="noopener noreferrer">GATK blog post</a> describing the components of a reference genome. </li><li><a href="https://support.illumina.com/content/dam/illumina-support/help/Illumina_DRAGEN_Bio_IT_Platform_v3_7_1000000141465/Content/SW/Informatics/Dragen/GPipelineAltMap_fDG.htm" target="_blank" rel="noopener noreferrer">Illumina DRAGEN documentation</a> describing ALT-aware mapping. </li></ul><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="rna-with-umis-tasks-and-tools"></a>RNA with UMIs tasks and tools<a class="hash-link" href="#rna-with-umis-tasks-and-tools" title="Direct link to heading">#</a></h2><p>The <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/broad/rna_seq/RNAWithUMIsPipeline.wdl" target="_blank" rel="noopener noreferrer">RNA with UMIs workflow</a> imports two additional WDL scripts. The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/UMIAwareDuplicateMarking.wdl" target="_blank" rel="noopener noreferrer">UMIAwareDuplicateMarking.wdl</a> script is a nested workflow used to mark duplicate sequencing reads, while the <a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">RNAWithUMIsTasks.wdl</a> script contains individual &quot;tasks&quot; called by the workflow.</p><p>Overall, the RNA with UMIs workflow:</p><ol><li>Converts FASTQs to unmapped BAMs. </li><li>Extracts UMIs. </li><li>Converts unmapped BAMs to FASTQs and filters reads.</li><li>Trims adapters and poly(A) tails.</li><li>Converts FASTQs to unmapped BAMs. </li><li>Aligns reads. </li><li>Marks duplicate reads and sorts BAMs.</li><li>Quantifies gene counts. </li><li>Calculates RNA and genomic metrics.</li></ol><p>The tools each task employs in the RNA with UMIs workflow are detailed in the table below. </p><p>To see specific tool parameters, select the task WDL link in the table; then find the task and view the <code>command {}</code> section of the task in the WDL script. To view or use the exact tool software, see the task&#x27;s Docker image which is specified in the task WDL <code>input {}</code> section as <code>String docker =</code>.</p><table><thead><tr><th>Task name and WDL link</th><th>Tool</th><th>Software</th><th>Description</th></tr></thead><tbody><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.FastqToUbam</a></td><td>FastqToSam</td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a></td><td>Converts the paired-end FASTQ files to unmapped BAM.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.ExtractUMIs</a></td><td>ExtractUmisFromBam</td><td><a href="http://fulcrumgenomics.github.io/fgbio/" target="_blank" rel="noopener noreferrer">fgbio</a></td><td>Extracts UMIs from the unmapped BAM and stores them in the RX tag of output BAM.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.SamToFastq</a></td><td>FastqToSam</td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a></td><td>Converts the BAM file to paired-end FASTQ files for adapter clipping and removes reads that fail platform/vendor quality checks performed by the sequencing platform.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.Fastp</a></td><td>fastp</td><td><a href="https://github.com/OpenGene/fastp" target="_blank" rel="noopener noreferrer">fastp</a></td><td>Trims adapters and polyA tails from reads.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.FastqToUbam (alias = FastqToUbamAfterClipping)</a></td><td>FastqToSam</td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a></td><td>Converts the trimmed paired-end FASTQ files to unmapped BAM.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.FastQC</a></td><td>FastQC</td><td><a href="https://github.com/s-andrews/FastQC" target="_blank" rel="noopener noreferrer">FastQC</a></td><td>Collects overall quality control metrics before alignment and generates an HTML-formatted report.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.STAR</a></td><td>STAR</td><td><a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">STAR</a></td><td>Aligns reads to the genome (using the StarIndex file) and outputs aligned reads to BAM. The task additionally converts the resulting BAM file to transcriptome coordinates, producing a transcriptome-aligned BAM. Parameters are listed below.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.CopyReadGroupsToHeader</a></td><td>view, reheader</td><td><a href="http://www.htslib.org/" target="_blank" rel="noopener noreferrer">Samtools</a></td><td>Copies the read group information from the genome-aligned BAM to the transcriptome-aligned BAM.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/UMIAwareDuplicateMarking.wdl" target="_blank" rel="noopener noreferrer">UmiMD.UMIAwareDuplicateMarking</a></td><td>SortSam, MarkDuplicates, group</td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a>, <a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a>, <a href="https://umi-tools.readthedocs.io/en/latest/index.html" target="_blank" rel="noopener noreferrer">UMI-tools</a></td><td>Marks duplicates on the genome-aligned BAM and tags reads with error-corrected UMIs.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/UMIAwareDuplicateMarking.wdl" target="_blank" rel="noopener noreferrer">UmiMD.UMIAwareDuplicateMarking (alias = UMIAwareDuplicateMarkingTranscriptome)</a></td><td>SortSam, MarkDuplicates, group</td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a>, <a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a>, <a href="https://umi-tools.readthedocs.io/en/latest/index.html" target="_blank" rel="noopener noreferrer">UMI-tools</a></td><td>Marks duplicates on the transcriptome-aligned BAM and tags reads with error-corrected UMIs.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.PostprocessTranscriptomeForRSEM</a></td><td>PostProcessReadsForRSEM</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Sorts reads for RSEM compatibility.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.GetSampleName</a></td><td>GetSampleName</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Writes the sample name from the unmapped BAM header into a separate text file.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.rnaseqc2</a></td><td>rnaseqc</td><td><a href="https://github.com/getzlab/rnaseqc" target="_blank" rel="noopener noreferrer">RNA-SeQC</a></td><td>Uses the genome-aligned, duplicate-marked BAM file to calculate TPMs, gene counts, exon counts, fragment sizes, and additional metrics, each of which is outputted to an individual file.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.CollectRNASeqMetrics</a></td><td>CollectRNASeqMetrics</td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a></td><td>Calculates RNA metrics; strand specificity is set to SECOND_READ_TRANSCRIPTION_STRAND.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.CollectMultipleMetrics</a></td><td>CollectMultipleMetrics</td><td><a href="https://broadinstitute.github.io/picard/" target="_blank" rel="noopener noreferrer">Picard</a></td><td>Collects multiple classes of metrics; runs tools CollectInsertSizeMetrics and CollectAlignmentSummaryMetrics.</td></tr><tr><td><a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.CalculateContamination</a></td><td>GetPileupSummaries, Calculate Contamination</td><td><a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer">GATK</a></td><td>Uses the population VCF and index files to calculate pileup metrics and estimate cross-sample contamination.</td></tr></tbody></table><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="1-convert-fastq-to-ubam"></a>1. Convert FASTQ to uBAM<a class="hash-link" href="#1-convert-fastq-to-ubam" title="Direct link to heading">#</a></h4><p>If paired-end FASTQ files are used as inputs for the pipeline, the first task in the pipeline converts those FASTQ files into an unmapped BAM file using Picard&#x27;s <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360036510672" target="_blank" rel="noopener noreferrer">FastqToSam</a>.</p><p>This task is skipped if a read group unmapped BAM file is used as input for the pipeline. </p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="2-umi-extraction"></a>2. UMI extraction<a class="hash-link" href="#2-umi-extraction" title="Direct link to heading">#</a></h4><p>Unique molecular identifiers (UMIs) are DNA tags (bases) that identify each DNA molecule before PCR amplification. They enable us to detect whether two reads that align to the same position in the reference are sequences of the same DNA molecule (have the same UMIs; PCR duplicates) or are two separate copies of DNA that happen to have the same sequence (have different UMIs; biological or natural duplicates).</p><p>For example, the Illumina sequencers used by the Broad Genomics Platform read 151 bp per read, of which 5 bp are reserved for the UMIs.</p><p>The first step of the RNA with UMIs workflow is to use the fgbio&#x27;s <a href="http://fulcrumgenomics.github.io/fgbio/tools/latest/ExtractUmisFromBam.html" target="_blank" rel="noopener noreferrer">ExtractUMIsFromBam</a> to remove UMI bases from the reads and store them in the RX read tag. </p><p>The resulting RX tag may contain information like &quot;ACT-GCT.&quot; The &quot;ACT&quot; is the 3 bp for read 1 and the &quot;GCT&quot; is the 3 bp for read 2. Even though we reserve 5 bases in a read, we only use 3 bases for UMI; the other two bases are reserved for cell barcode, which the workflow doesn&#x27;t use.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="3-convert-ubam-to-fastqs-and-filter-reads"></a>3. Convert uBAM to FASTQs and filter reads<a class="hash-link" href="#3-convert-ubam-to-fastqs-and-filter-reads" title="Direct link to heading">#</a></h4><p>The <a href="https://github.com/broadinstitute/warp/blob/develop/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.SamToFastq</a> task uses Picard&#x27;s <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360036510672" target="_blank" rel="noopener noreferrer">FastqToSam</a> to convert the unmapped BAM to paired-end FASTQ files that can then be used for adapter clipping. This step also removes reads that fail to pass platform/vendor quality checks performed by the sequencing platform and flagged with the corresponding <a href="https://broadinstitute.github.io/picard/explain-flags.html" target="_blank" rel="noopener noreferrer">SAM flag value</a>.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="4-trim-adapters-and-polya-tails"></a>4. Trim adapters and poly(A) tails<a class="hash-link" href="#4-trim-adapters-and-polya-tails" title="Direct link to heading">#</a></h4><p>After converting the uBAM to FASTQs, the workflow uses <a href="https://github.com/OpenGene/fastp" target="_blank" rel="noopener noreferrer">fastp</a> to trim sequencing adapters and poly(A) tails from the reads. This task requires an <a href="https://console.cloud.google.com/storage/browser/_details/gcp-public-data--broad-references/RNA/resources/Illumina_adapters.fasta;tab=live_object" target="_blank" rel="noopener noreferrer">adapter FASTA file</a> containing the list of sequences to be trimmed and is publicly available in Broad&#x27;s <a href="https://console.cloud.google.com/storage/browser/gcp-public-data--broad-references;tab=objects?prefix=&amp;forceOnObjectsSortingFiltering=false" target="_blank" rel="noopener noreferrer">public reference bucket</a>.</p><p>The adapter FASTA file contains the sequences shown below.</p><div class="codeBlockContainer_K1bP"><div class="codeBlockContent_hGly"><pre tabindex="0" class="prism-code language-undefined codeBlock_23N8 thin-scrollbar" style="color:#bfc7d5;background-color:#292d3e"><code class="codeBlockLines_39YC"><span class="token-line" style="color:#bfc7d5"><span class="token plain">&gt;Illumina TruSeq Adapter Read 1</span></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">AGATCGGAAGAGCACACGTCTGAACTCCAGTCA</span></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">&gt;Illumina TruSeq Adapter Read 2</span></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT</span></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">&gt;polyA</span></span><span class="token-line" style="color:#bfc7d5"><span class="token plain">AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA</span></span></code></pre><button type="button" aria-label="Copy code to clipboard" class="copyButton_Ue-o clean-btn">Copy</button></div></div><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="5-convert-fastqs-to-ubam"></a>5. Convert FASTQs to uBAM<a class="hash-link" href="#5-convert-fastqs-to-ubam" title="Direct link to heading">#</a></h4><p>The <a href="https://github.com/broadinstitute/warp/blob/master/tasks/broad/RNAWithUMIsTasks.wdl" target="_blank" rel="noopener noreferrer">tasks.FastqToUbam (alias = FastqToUbamAfterClipping)</a> task converts trimmed FASTQs to an unmapped BAM using Picard&#x27;s <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360036510672" target="_blank" rel="noopener noreferrer">FastqToSam</a>.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="6-alignment-with-star"></a>6. Alignment with STAR<a class="hash-link" href="#6-alignment-with-star" title="Direct link to heading">#</a></h4><p>After UMI extraction, the workflow aligns the paired-end reads to the reference (hg38 or hg19) using the <a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener noreferrer">STAR aligner</a>, which is specifically designed for RNA-seq data and can align cDNA sequences with many &quot;gaps&quot; that correspond to introns. </p><p>The task uses the following parameters:</p><table><thead><tr><th>Parameter</th><th>Value</th><th>Notes</th></tr></thead><tbody><tr><td>outSAMunmapped</td><td>Within</td><td>Includes unmapped reads in the output file rather than dropping those reads to facilitate potential downstream analysis.</td></tr><tr><td>outFilterMismatchNoverLmax</td><td>0.1</td><td>Sets the maximum allowable ratio of mismatches to read length. Reads with a ratio larger than the set value are filtered. For example, for paired-end reads with length 146, the reads are filtered if the number of mismatches is greater than 29 (146 * 2 * 0.1 = 29).</td></tr><tr><td>alignEndsProtrude</td><td>20 ConcordantPair</td><td>Allows a maximum of 20 protruding bases at alignment ends and marks these alignments as concordant pairs to prevent reads from small cDNA fragments that were sequenced into adapters from being dropped. This parameter allows for the processing of data derived from low-quality or degraded tissue such as formalin-fixed paraffin-embedded (FFPE) samples.</td></tr></tbody></table><p>Additional parameters are used to match <a href="https://www.encodeproject.org/data-standards/rna-seq/long-rnas/" target="_blank" rel="noopener noreferrer">ENCODE bulk RNA-seq data standards</a>. To learn more about ENCODE options in STAR, see the <a href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf" target="_blank" rel="noopener noreferrer">STAR manual</a>.</p><p>After STAR alignment, the workflow outputs both a genome- and transcriptome-aligned BAM.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="7-mark-duplicates-and-sort-bams"></a>7. Mark duplicates and sort BAMs<a class="hash-link" href="#7-mark-duplicates-and-sort-bams" title="Direct link to heading">#</a></h4><p>As described in Step 2 (UMI extraction), UMIs are DNA tags that allow us to distinguish between PCR duplicates (duplicate reads with the same UMI) and biological duplicates (duplicate reads with different UMIs).</p><p>In this step, the workflow sorts the aligned BAMs coordinates using Picard&#x27;s <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360036510732-SortSam-Picard-" target="_blank" rel="noopener noreferrer">SortSam</a> tool and then groups the duplicates using UMI-tools&#x27; <a href="https://umi-tools.readthedocs.io/en/latest/reference/group.html#" target="_blank" rel="noopener noreferrer">group</a> function. Once the duplicates are grouped by UMI, the PCR duplicates are marked using Picard&#x27;s <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-" target="_blank" rel="noopener noreferrer">MarkDuplicates</a>. This step outputs new genome- and transcriptome-aligned BAM files with PCR duplicates tagged and a corresponding index file.</p><p>The transcriptome-aligned BAM is then sorted using GATKâ€™s <a href="https://gatk.broadinstitute.org/hc/en-us/articles/5358861221659-PostProcessReadsForRSEM-BETA-" target="_blank" rel="noopener noreferrer">PostProcessReadsForRSEM</a> for compatibility with <a href="https://github.com/deweylab/RSEM" target="_blank" rel="noopener noreferrer">RSEM</a>. While RSEM is not used in this workflow, it is an additional tool that can be used to quantify expression from RNA-seq data.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="8-gene-quantification"></a>8. Gene quantification<a class="hash-link" href="#8-gene-quantification" title="Direct link to heading">#</a></h4><p>After duplicate reads have been tagged, the workflow uses <a href="https://github.com/getzlab/rnaseqc" target="_blank" rel="noopener noreferrer">RNA-SeQC</a> to quantify the expression level of transcripts based on the number of reads that align to one or more exons of each gene (<code>rnaseqc2_gene_counts</code>). Exon-level expression is also quantified based on the number of reads that align to each exon (<code>rnaseqc2_exon_counts</code>). </p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="9-metric-calculation"></a>9. Metric calculation<a class="hash-link" href="#9-metric-calculation" title="Direct link to heading">#</a></h4><p>The pipeline uses <a href="https://github.com/s-andrews/FastQC" target="_blank" rel="noopener noreferrer">FastQC</a>,  <a href="https://github.com/getzlab/rnaseqc" target="_blank" rel="noopener noreferrer">RNA-SeQC</a>, Picard&#x27;s <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037057492-CollectRnaSeqMetrics-Picard-" target="_blank" rel="noopener noreferrer">CollectRNASeqMetrics</a> and <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037594031-CollectMultipleMetrics-Picard-" target="_blank" rel="noopener noreferrer">CollectMultipleMetrics</a> tools, and GATK&#x27;s <a href="https://gatk.broadinstitute.org/hc/en-us/articles/4418051471643-GetPileupSummaries" target="_blank" rel="noopener noreferrer">GetPileupSummaries</a> and <a href="https://gatk.broadinstitute.org/hc/en-us/articles/4418054253211-CalculateContamination" target="_blank" rel="noopener noreferrer">CalculateContamination</a> tools to calculate summary metrics that can be used to assess the quality of the data each time the pipeline is run. </p><p>If you are a member of the Broad Institute&#x27;s Genomics Platform using the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/broad/internal/rna_seq/BroadInternalRNAWithUMIs.wdl" target="_blank" rel="noopener noreferrer">internal RNA with UMIs pipeline</a>, there is an additional step that merges the individual metrics files to create the <code>MergeMetrics.unified_metrics</code> output file and prepare the data for use in the Terra Data Repository.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor anchor__h4 anchorWithStickyNavbar_31ik" id="10-outputs"></a>10. Outputs<a class="hash-link" href="#10-outputs" title="Direct link to heading">#</a></h4><p>Workflow outputs are described in the table below. </p><table><thead><tr><th>Output variable name</th><th>Description</th><th>Type</th></tr></thead><tbody><tr><td>sample_name</td><td>Sample name extracted from the input unmapped BAM file header.</td><td>String</td></tr><tr><td>transcriptome_bam</td><td>Duplicate-marked BAM file containing alignments from STAR translated into transcriptome coordinates.</td><td>BAM</td></tr><tr><td>transcriptome_bam_index</td><td>Index file for the transcriptome_bam output.</td><td>BAM Index</td></tr><tr><td>transcriptome_duplicate_metrics</td><td>File containing duplication metrics.</td><td>TXT</td></tr><tr><td>output_bam</td><td>Duplicate-marked BAM file containing alignments from STAR translated into genome coordinates.</td><td>BAM</td></tr><tr><td>output_bam_index</td><td>Index file for the output_bam output.</td><td>BAM Index</td></tr><tr><td>duplicate_metrics</td><td>Duplicate metrics file containing the number of reads marked as duplicates.</td><td>TXT</td></tr><tr><td>rnaseqc2_gene_tpm</td><td>File containing TPMs.</td><td>GCT</td></tr><tr><td>rnaseqc2_gene_counts</td><td>File containing gene counts.</td><td>GCT</td></tr><tr><td>rnaseqc2_exon_counts</td><td>File containing exon counts.</td><td>GCT</td></tr><tr><td>rnaseqc2_fragment_size_histogram</td><td>File containing counts of observed fragment size.</td><td>TXT</td></tr><tr><td>rnaseqc2_metrics</td><td>File containing RNA-SeQC metrics including strand specificity, 3â€™/5â€™ bias, rRNA reads, and others.</td><td>TSV</td></tr><tr><td>picard_rna_metrics</td><td>Metrics file containing the output of Picardâ€™s CollectRnaSeqMetrics tool.</td><td>TXT</td></tr><tr><td>picard_alignment_summary_metrics</td><td>Metrics file containing output of Picardâ€™s CollectAlignmentSummaryMetrics tool.</td><td>TXT</td></tr><tr><td>picard_insert_size_metrics</td><td>Metrics file containing output of Picardâ€™s CollectInsertSizeMetrics tool.</td><td>TXT</td></tr><tr><td>picard_insert_size_histogram</td><td>Histogram chart of insert size.</td><td>PDF</td></tr><tr><td>picard_base_distribution_by_cycle_metrics</td><td>Metrics file containing the output of Picardâ€™s CollectBaseDistributionByCycle tool.</td><td>TXT</td></tr><tr><td>picard_base_distribution_by_cycle_pdf</td><td>Chart of nucleotide distribution per cycle.</td><td>PDF</td></tr><tr><td>picard_quality_by_cycle_metrics</td><td>Metrics file containing the output of Picardâ€™s MeanQualityByCycle tool.</td><td>TXT</td></tr><tr><td>picard_quality_by_cycle_pdf</td><td>Chart of mean quality by cycle.</td><td>PDF</td></tr><tr><td>picard_quality_distribution_metrics</td><td>Metrics file containing the output of Picardâ€™s QualityScoreDistribution tool.</td><td>TXT</td></tr><tr><td>picard_quality_distribution_pdf</td><td>Chart of quality score distribution.</td><td>PDF</td></tr><tr><td>contamination</td><td>Float representing the calculated cross-sample contamination.</td><td>Float</td></tr><tr><td>contamination_error</td><td>Float representing the error associated with the contamination calculation.</td><td>Float</td></tr><tr><td>fastqc_html_report</td><td>HTML report containing general quality control metrics generated by FastQC.</td><td>File</td></tr><tr><td>fastqc_percent_reads_with_adapter</td><td>Float representing the percent of reads with adapter sequences present following the adapter clipping task.</td><td>Float</td></tr></tbody></table><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="versioning"></a>Versioning<a class="hash-link" href="#versioning" title="Direct link to heading">#</a></h2><p>All RNA with UMIs pipeline releases are documented in the <a href="https://github.com/broadinstitute/warp/blob/master/pipelines/broad/rna_seq/RNAWithUMIsPipeline.changelog.md" target="_blank" rel="noopener noreferrer">pipeline changelog</a>.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor anchor__h2 anchorWithStickyNavbar_31ik" id="feedback"></a>Feedback<a class="hash-link" href="#feedback" title="Direct link to heading">#</a></h2><p>Please help us make our tools better by contacting <a href="mailto:warp-pipelines-help@broadinstitute.org" target="_blank" rel="noopener noreferrer">the WARP team</a> for pipeline-related suggestions or questions.</p></div><footer class="docusaurus-mt-lg"><div class="row"><div class="col"><a href="https://github.com/broadinstitute/warp/edit/develop/website/docs/Pipelines/RNA_with_UMIs_Pipeline/README.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_2_ui" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_13-_">Last updated on <b><time datetime="2023-06-01T16:11:27.000Z">6/1/2023</time></b> by <b>dependabot[bot]</b></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/warp/docs/Pipelines/Optimus_Pipeline/consortia-processing"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Â« Consortia Data Processing</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/warp/docs/Pipelines/RNA_with_UMIs_Pipeline/rna-with-umis.methods"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">RNA with UMIs v1.0.6 Methods Â»</div></a></div></nav></div></div><div class="col col--3"><div class="tableOfContents_35-E thin-scrollbar"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction-to-the-rna-with-umis-workflow" class="table-of-contents__link">Introduction to the RNA with UMIs workflow</a></li><li><a href="#quickstart-table" class="table-of-contents__link">Quickstart table</a></li><li><a href="#set-up" class="table-of-contents__link">Set-up</a><ul><li><a href="#installation" class="table-of-contents__link">Installation</a></li><li><a href="#inputs" class="table-of-contents__link">Inputs</a></li><li><a href="#references" class="table-of-contents__link">References</a></li></ul></li><li><a href="#rna-with-umis-tasks-and-tools" class="table-of-contents__link">RNA with UMIs tasks and tools</a></li><li><a href="#versioning" class="table-of-contents__link">Versioning</a></li><li><a href="#feedback" class="table-of-contents__link">Feedback</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/get-started">Get Started</a></li></ul></div><div class="col footer__col"><div class="footer__title">Guide and Policy</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/docs/contribution/README">Contribution Guide</a></li><li class="footer__item"><a class="footer__link-item" href="/warp/privacy">Privacy</a></li></ul></div><div class="col footer__col"><div class="footer__title">Resources</div><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/warp/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp/releases" target="_blank" rel="noopener noreferrer" class="footer__link-item">Changelog</a></li><li class="footer__item"><a href="https://github.com/broadinstitute/warp" target="_blank" rel="noopener noreferrer" class="footer__link-item"><span>GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_3J9K"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></span></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright Â© 2023 Copyright Â© Data Sciences Platform, Broad Institute.</div></div></div></footer></div>
<script src="/warp/assets/js/runtime~main.ba3180c6.js"></script>
<script src="/warp/assets/js/main.1308a829.js"></script>
</body>
</html>